{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction of HaloGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import xlrd\n",
    "import time\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from scipy.io import loadmat\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import urllib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create draft model from template iML1515\n",
    "* 1. use Bi-directional Best Hit to identify the ortholog pairs between halomonas and E coli\n",
    "* 2. get all reactions associated with all these genes\n",
    "* 3. For those reactions with missing genes in the gr rule, try to find homologes in _H.TD01_ based on the same identity and coverage thresholds (Not BBH). The coverage threshold used in this study is 45%\n",
    "* 4. resolve gr rules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_ecoli_model(mat_file,sbml_file):\n",
    "    # iJO1366: load mat file: containssubsystem information. while sbml doesn't\n",
    "    # mat file doesn't contain annotation field for reaction and metabolites. smbl does\n",
    "    # \n",
    "    # add subsystem information into sbml model\n",
    "    # \n",
    "    mat_model = cobra.io.load_matlab_model(mat_file)\n",
    "    sbml_model = cobra.io.read_sbml_model(sbml_file)\n",
    "    \n",
    "    for rxn in mat_model.reactions:\n",
    "        rxn_xml = sbml_model.reactions.get_by_id(rxn.id)\n",
    "        rxn_xml.subsystem = rxn.subsystem\n",
    "    \n",
    "    return sbml_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Download genes in template model based on uniprot ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrive_from_uniprot(uniprot_id):\n",
    "    # get fasta content for given uniprot_id, return as string format\n",
    "    url = 'https://www.uniprot.org/uniprot/{0}.fasta'.format(uniprot_id)\n",
    "    response = urllib.request.urlopen(url)\n",
    "    rec = response.read()\n",
    "    return rec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_genes(model,outname):\n",
    "    # 1. load ids that have been done\n",
    "    done_ids = [line.split('|')[1] for line in open(outname) if line.startswith('>')]\n",
    "    \n",
    "    fhand = open(outname,'a')\n",
    "    for g in model.genes:\n",
    "        uniprot_id = g.annotation.get('uniprot')\n",
    "        if uniprot_id is None: \n",
    "            print(g.id,g.name,g.annotation)\n",
    "            continue\n",
    "        if uniprot_id in done_ids: continue\n",
    "            \n",
    "        try:\n",
    "            rec = retrive_from_uniprot(uniprot_id)\n",
    "            fhand.write(str(rec)[2:-1].replace('\\\\n','\\n'))\n",
    "        except: print('Failed to retrive sequence for',uniprot_id)\n",
    "\n",
    "    fhand.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_fasta_format(model,fastafile):\n",
    "    # Replace the uniprot gene ids in fastafile with gene ids in the model\n",
    "    \n",
    "    if '.fasta' in fastafile: outname = fastafile.replace('.fasta','_formated.fasta')\n",
    "    elif '.fa' in fastafile: outname = fastafile.replace('.fa','_formated.fasta')\n",
    "    else: outname = fastafile + '_formated.fasta'\n",
    "    fhand = open(outname,'w')\n",
    "    \n",
    "    # load uniprotid-geneid\n",
    "    uni2gene = dict()\n",
    "    for gene in model.genes:\n",
    "        uni = gene.annotation.get('uniprot')\n",
    "        if uni is not None: uni2gene[uni] = gene.id\n",
    "    \n",
    "    for rec in SeqIO.parse(fastafile,'fasta'):\n",
    "        uni = rec.id.split('|')[1]\n",
    "        fhand.write('>{0}\\n{1}\\n'.format(uni2gene[uni],rec.seq))\n",
    "    fhand.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Perform blastp to identify BBH "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_coverage(combined_fasta,blast_results):\n",
    "    # combined_fasta: combined fasta file, which contains all sequnces from both query and target proteome.\n",
    "    # blast_results: tsv file produced by blastp\n",
    "    # \n",
    "    # Output: add a new column: coverage at the end of the blast_results.\n",
    "    # Coverage is caculated as: length of aligned query sequence/length of query sequence * 100\n",
    "    # \n",
    "    \n",
    "    # Load sequence lenght\n",
    "    Lenghs = dict()\n",
    "    for rec in SeqIO.parse(combined_fasta,'fasta'): Lenghs[rec.id] = len(rec.seq)\n",
    "    \n",
    "    \n",
    "    # calculate coverage\n",
    "    outfile = blast_results.replace('.tab','_cov.tab')\n",
    "    if outfile == blast_results: outfile += '_cov.tab'\n",
    "    \n",
    "    fhand = open(outfile,'w')\n",
    "    for line in open(blast_results):\n",
    "        cont = line.split()\n",
    "        aln_lengh = float(cont[7])-float(cont[6])\n",
    "        cov = aln_lengh/Lenghs[cont[0]]*100\n",
    "        fhand.write(line.strip()+'\\t{0}\\n'.format(cov))\n",
    "    fhand.close()\n",
    "    \n",
    "    # replace the orginal blast results file with the new one.\n",
    "    os.system('mv {0} {1}'.format(outfile,blast_results))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_blast(query_fasta,target_fasta):\n",
    "    # query_fasta: fasta file for the organsims to be modelled\n",
    "    # target_fasta: fasata file that contains enzyme sequences in the template model\n",
    "    \n",
    "    # combine two fasta files, do blast. Use default evalue. The blast results will be filetered based on evalue\n",
    "    # in the BBH extraction step\n",
    "    \n",
    "    # build combined fasta file for blast\n",
    "    fhand = open('input.fa','w')\n",
    "    for rec in SeqIO.parse(query_fasta,'fasta'):\n",
    "        fhand.write('>query|{0}\\n{1}\\n'.format(rec.id,rec.seq))\n",
    "    \n",
    "    for rec in SeqIO.parse(target_fasta,'fasta'):\n",
    "        fhand.write('>target|{0}\\n{1}\\n'.format(rec.id,rec.seq))\n",
    "    \n",
    "    fhand.close()\n",
    "    \n",
    "    blast_cmd = '''mkdir blast_tmps\n",
    "    mv input.fa blast_tmps\n",
    "    makeblastdb -dbtype prot -in blast_tmps/input.fa  -out blast_tmps/DB\n",
    "    blastp -query blast_tmps/input.fa -db blast_tmps/DB -outfmt 6  -max_target_seqs 500 -max_hsps 1  -num_threads 1 -out ../Results/bidirectional_blast.tab\n",
    "    '''\n",
    "    os.system(blast_cmd)\n",
    "    add_coverage('blast_tmps/input.fa','../Results/bidirectional_blast.tab')\n",
    "    os.system('rm -r blast_tmps')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_BBHs(blastfile,BBH_evalue,report=False,coverage=45):\n",
    "    # BBH_evalue: like 1e-10\n",
    "    # report: if True, print out the nubmer of BBHs\n",
    "    # coverage: the lower bound of query coverage to determine a true hit. from 0-100\n",
    "    # \n",
    "    \n",
    "    target_query = dict()\n",
    "    query_target = dict()\n",
    "    for line in open(blastfile):\n",
    "        cont = line.split()\n",
    "        evalue = float(cont[10])\n",
    "        cov = float(cont[12])\n",
    "        # filter based on evalue and coverage\n",
    "        if evalue > BBH_evalue: continue\n",
    "        if cov < coverage: continue\n",
    "        \n",
    "        if cont[0].startswith('target|') and cont[1].startswith('query|'):\n",
    "            target_gene = cont[0].split('|')[1]\n",
    "            query_gene = cont[1].split('|')[1]\n",
    "            if target_query.get(target_gene,(None,1000))[1]>evalue: target_query[target_gene] = (query_gene,evalue)\n",
    "        elif cont[0].startswith('query|') and cont[1].startswith('target|'):\n",
    "            target_gene = cont[1].split('|')[1]\n",
    "            query_gene = cont[0].split('|')[1]\n",
    "            if query_target.get(query_gene,(None,1000))[1]>evalue: query_target[query_gene] = (target_gene,evalue)\n",
    "        else: None\n",
    "\n",
    "    # 3. find BBH, the overlaped gene pairs\n",
    "    BBH = list()\n",
    "    for target_gene, (query_gene,evalue) in target_query.items():\n",
    "        if query_target.get(query_gene,(None,1000))[0] == target_gene: BBH.append((query_gene,target_gene))\n",
    "    \n",
    "    if report:\n",
    "        print('Number of homologs for query genes (evalue = {0}):'.format(BBH_evalue),len(query_target))\n",
    "        print('Number of homologs for target genes (evalue = {0}):'.format(BBH_evalue),len(target_query))\n",
    "        print('Number of BBHs (evalue = {0}):'.format(BBH_evalue),len(BBH))\n",
    "    return BBH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get all reactions associated with all genes in BBH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_rxns_in_BBH(template_model, BBHs):\n",
    "    # model file: template model file\n",
    "    # BBH, best bidirectional hit\n",
    "\n",
    "    # get all reactions\n",
    "    candidate_rxn_ids = list()\n",
    "    candidate_rxns = list()\n",
    "    for query_gene,target_gene in BBHs:\n",
    "        gene = template_model.genes.get_by_id(target_gene)\n",
    "        for rxn in gene.reactions:\n",
    "            if rxn.id not in candidate_rxn_ids: \n",
    "                candidate_rxn_ids.append(rxn.id)\n",
    "                candidate_rxns.append(rxn)\n",
    "    print('Number of candiate reactions:',len(candidate_rxn_ids))\n",
    "    \n",
    "    return candidate_rxns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update gene-reaction rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gr(gr):\n",
    "    # return a list of genes in gr rule\n",
    "    genes = list(set([item for item in gr.replace('(','').replace(')','').split() if item not in ['and','or']]))\n",
    "    genes.sort()\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gr_parts(gr):\n",
    "    if ' and ' in gr and ' or ' not in gr: gene_parts = ['( {0} )'.format(gr)]\n",
    "    else:\n",
    "        complexes = re.findall('(\\(.*?\\))',gr)\n",
    "        gr1 = gr\n",
    "        for i in range(len(complexes)):\n",
    "            gr1 = gr1.replace(complexes[i],'comp_{0}'.format(i))\n",
    "        gene_parts = parse_gr(gr1)\n",
    "\n",
    "        for i in range(len(gene_parts)):\n",
    "            if 'comp_' in gene_parts[i]: gene_parts[i] = complexes[int(gene_parts[i].split('_')[1])]\n",
    "    return list(set(gene_parts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def missing_type(gene_part):\n",
    "    # determin if a gene/coplex is totally, partly missing or complete\n",
    "    assert ' or ' not in gene_part\n",
    "    genes = parse_gr(gene_part)\n",
    "    missing_parts = re.findall('_missing',gene_part)\n",
    "    \n",
    "    if len(missing_parts) == 0: tpe = 'complete' # complete\n",
    "    elif len(missing_parts) == len(genes): tpe = 'missing'\n",
    "    else: tpe = 'partial'\n",
    "    return tpe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_repeated_gene_in_and(gene_part):\n",
    "    # for case: A and A\n",
    "    assert ' or ' not in gene_part\n",
    "    genes = parse_gr(gene_part)\n",
    "    \n",
    "    new_part = ''\n",
    "    for gene in genes: new_part += gene + ' and '\n",
    "    new_part = new_part[:-5]\n",
    "    \n",
    "    # in case the final gr rule is (A and B), remove parathesis\n",
    "    if len(genes) > 1: new_part = '( {0} )'.format(new_part) \n",
    "    return new_part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_repeated_gene_parts(gr):\n",
    "    gene_parts = parse_gr_parts(gr)\n",
    "    \n",
    "    # remove repeated genes in each part \n",
    "    new_gene_parts = list()\n",
    "    for gene_part in gene_parts:\n",
    "        new_gene_part = remove_repeated_gene_in_and(gene_part)\n",
    "        if new_gene_part not in new_gene_parts: new_gene_parts.append(new_gene_part)\n",
    "    \n",
    "    new_gr = ''\n",
    "    for gene_part in new_gene_parts: \n",
    "        new_gr += gene_part + ' or '\n",
    "    new_gr = new_gr[:-4]\n",
    "    \n",
    "    # in case the final gr rule is (A and B), remove parathesis\n",
    "    if len(new_gene_parts) == 1 and '(' in new_gr: new_gr = new_gr[1:-1].strip()\n",
    "    return new_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_and_or_case(gr):\n",
    "    genes1 = parse_gr_parts(gr)\n",
    "    tpes = list()\n",
    "    for g in genes1: tpes.append(missing_type(g))\n",
    "\n",
    "    # case 1: if one of parts is complete, remove all parts with are missing or partial\n",
    "    if 'complete' in tpes:\n",
    "        new_gr = ''\n",
    "        for i in range(len(genes1)):\n",
    "            if tpes[i] == 'complete': new_gr += genes1[i] + ' or '\n",
    "    # case 2: if none of parts is complete, only remove those ones that are missing\n",
    "    else:\n",
    "        new_gr = ''\n",
    "        for i in range(len(genes1)):\n",
    "            if tpes[i] != 'missing': new_gr += genes1[i] + ' or '\n",
    "    new_gr = new_gr[:-4]\n",
    "    \n",
    "    # in case the final gr rule is (A and B), remove parathesis\n",
    "    if 'or' not in new_gr and '(' in new_gr: new_gr = new_gr[1:-1].strip()\n",
    "    \n",
    "    new_gr = remove_repeated_gene_parts(new_gr)\n",
    "    return new_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_gr(gr,BBHs):\n",
    "    # replace the targetli gene ids in gr rule with the ones in querymonas based on BBH. For those targetli genes\n",
    "    # wihout BBH in querymonas, add a tag '_missing' at the end of the gene id\n",
    "    \n",
    "    target2query_map = dict()\n",
    "    for query_gene,target_gene in BBHs: target2query_map[target_gene] = query_gene\n",
    "    \n",
    "    genes = parse_gr(gr)\n",
    "    new_gr = gr\n",
    "    for gene in genes:\n",
    "        query_gene = target2query_map.get(gene)\n",
    "        if query_gene is None: query_gene = gene+'_missing'\n",
    "        new_gr = new_gr.replace(gene,query_gene)\n",
    "    \n",
    "    # refine new gr rule. \n",
    "     \n",
    "    # case 1: if 'missing' not in gr rule, do not need to update\n",
    "    if '_missing' not in new_gr: refined_gr = new_gr\n",
    "    \n",
    "    # case 2: only 'or' relationships are in gr rule, remove all genes with '_missing' tag\n",
    "    elif ' and ' not in new_gr: \n",
    "        new_genes = [item for item in parse_gr(new_gr) if '_missing' not in item]\n",
    "        refined_gr = ''\n",
    "        for g in new_genes: refined_gr += g + ' or '\n",
    "        refined_gr = refined_gr[:-4]\n",
    "    \n",
    "    # case 3: if 'or' not in gr rule, keep missing genes\n",
    "    elif 'or' not in new_gr: refined_gr = new_gr\n",
    "    \n",
    "    # case 4: if (A and B) or (C and D_missing) or E. In such or structure, if one of the gene/gene complex is \n",
    "    #         complete, then remove all others with missing gene(s). if all parts are incompelte, only removes \n",
    "    #         those part which are totally missing.\n",
    "    elif 'and' in new_gr and 'or' in new_gr: refined_gr = update_and_or_case(new_gr)\n",
    "\n",
    "    # Ohter cases: export to a txt file and manully edit\n",
    "    else:  refined_gr = new_gr\n",
    "    return refined_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def report_model_status(model):\n",
    "    print('Number of reactions:',len(model.reactions))\n",
    "    print('Number of metabolits:',len(model.metabolites))\n",
    "    print('Number of compartments:',len(model.compartments))\n",
    "    k = 0\n",
    "    for gene in model.genes:\n",
    "        if 'missing' in gene.id: k += 1\n",
    "    print('Number of genes:',len(model.genes))\n",
    "    print('Number of missing genes:',k)\n",
    "    \n",
    "    k = 0\n",
    "    for rxn in model.reactions:\n",
    "        if '_missing' in rxn.gene_reaction_rule: k += 1\n",
    "    print('Number of reactions with missing genes:',k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_model_from_template(candidate_rxns,BBHs,report=False):\n",
    "    rxns_to_add = list()\n",
    "    for rxn in candidate_rxns:\n",
    "        new_gr = update_gr(rxn.gene_reaction_rule,BBHs)\n",
    "        new_rxn = rxn.copy()\n",
    "        new_rxn.gene_reaction_rule = new_gr\n",
    "        rxns_to_add.append(new_rxn)\n",
    "    model_from_template = cobra.Model('model_from_template')\n",
    "    model_from_template.add_reactions(rxns_to_add)\n",
    "    if report: report_model_status(model_from_template)\n",
    "    return model_from_template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_missing_genes(model,homolog_evalue,report=False,coverage=45):\n",
    "    # homolog_evalue: evalue used in blast to infer homology relationships\n",
    "    # model: tmeplate model from BBH, which in most cases contains missing genes in gene-reaction rules\n",
    "    \n",
    "    # load best hit based on homolog_evalue\n",
    "    best_hits = dict()\n",
    "    # best_hits = {eco_gene_id:(halo_gene,evalue)}\n",
    "    for line in open('../Results/bidirectional_blast.tab'):\n",
    "        cont = line.split()\n",
    "        evalue = float(cont[10])\n",
    "        cov = float(cont[12])\n",
    "        \n",
    "        # filter hits based on evalue and coverage\n",
    "        if evalue > homolog_evalue: continue\n",
    "        if cov < coverage: continue\n",
    "        \n",
    "        if line.startswith('target|') and cont[1].startswith('query'):\n",
    "            target = cont[0].split('|')[1]\n",
    "            query = cont[1].split('|')[1]\n",
    "            best_hit, best_evalue = best_hits.get(target,(None,0))\n",
    "            if best_hit is None or best_evalue > evalue: best_hits[target] = (query,evalue)\n",
    "    \n",
    "    # update missing genes in the model. Replace the missing genes with homologs found in above step\n",
    "    \n",
    "    updated_rxns = list()\n",
    "    for rxn in model.reactions:\n",
    "        rxn_copy = rxn.copy()\n",
    "        gr = rxn_copy.gene_reaction_rule\n",
    "\n",
    "        if '_missing' in gr:\n",
    "            genes = parse_gr(gr)\n",
    "            for gene in genes:\n",
    "                if '_missing' in gene: \n",
    "                    homolog = best_hits.get(gene.replace('_missing',''),(None,None))[0]\n",
    "                    if homolog is not None: gr = gr.replace(gene,homolog)\n",
    "        \n",
    "        gr = remove_repeated_gene_parts(gr)\n",
    "        gr = update_and_or_case(gr)\n",
    "        rxn_copy.gene_reaction_rule = gr\n",
    "\n",
    "        updated_rxns.append(rxn_copy)\n",
    "    \n",
    "    # create a new model\n",
    "    new_model = cobra.Model('model_from_template')\n",
    "    new_model.add_reactions(updated_rxns)\n",
    "    if report: report_model_status(new_model)\n",
    "    return new_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_rxns_with_missing_genes(model,report=True):\n",
    "    rxns = list()\n",
    "    for rxn in model.reactions:\n",
    "        if '_missing' in rxn.gene_reaction_rule: continue\n",
    "        rxns.append(rxn)\n",
    "    new_model = cobra.Model('model_from_template')\n",
    "    new_model.add_reactions(rxns)\n",
    "    if report: report_model_status(new_model)\n",
    "    return new_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run pipeline for reconstruction of draft model from template model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using license file /Users/gangl/gurobi.lic\n",
      "Academic license - for non-commercial use only\n"
     ]
    }
   ],
   "source": [
    "# template_model = cobra.io.read_sbml_model('../Dataset/iJO1366.xml')\n",
    "template_model = load_ecoli_model('../../../ComplementaryData/iML1515.mat',\n",
    "                                  '../../../ComplementaryData/iML1515.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# download enzyme sequences in the template model\n",
    "# download_genes(template_model,outname='../../../ComplementaryData/orf_iML1515.fasta')\n",
    "# update_fasta_format(template_model,'../../../ComplementaryData/orf_iML1515.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# blast\n",
    "#do_blast('../../../ComplementaryData/protein_sequences_v2.0.fasta',\n",
    "#         '../../../ComplementaryData/orf_iML1515_formated.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARgAAADQCAYAAADcQn7hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xd8VVW2wPHfSkISCIFQAkKChgwIOEiN9KJiRRRUVFQEGco4gmAbR9/MvHHGmdF5tqEoiBRxpAgIgooF0QEpAqFKE0KPKL33st4f50QiptzIPTk3N+v7+dzPPe2eu3IgK2fvffbeoqoYY4wXIvwOwBgTvizBGGM8YwnGGOMZSzDGGM9YgjHGeMYSjDHGM5ZgjDGesQRjjPGMpwlGRB4TkdUiskpExotIrIhUF5GFIrJBRN4VkWj32Bh3PcPdn+JlbMYY74lXT/KKSBIwF7hCVY+LyERgBtAemKKqE0RkGLBCVYeKyMNAPVV9SES6ALer6j15fUfFihU1JSXFk/iNMblbsmTJHlVNzO+4KI/jiAJKishpoBTwPXAtcJ+7fwzwLDAU6OguA0wGhoiIaB4ZMCUlhfT0dG8iN8bkSkS2BnKcZ0UkVf0OeAnYhpNYDgJLgAOqesY9LBNIcpeTgO3uZ8+4x1e48Lwi0kdE0kUkfffu3V6Fb4wJAs8SjIiUw7krqQ5UBeKAm3M4NOsORfLYd36D6nBVTVPVtMTEfO/QjDE+8rKS9zpgs6ruVtXTwBSgBZAgIllFs2Rgh7ucCVQDcPeXBfZ5GJ8xxmNeJphtQDMRKSUiArQD1gBfAp3dY7oD09zl6e467v4v8qp/CZQNR2GMf7ysg1mIU1m7FPjG/a7hwB+Ax0UkA6eOZaT7kZFABXf748DTwYjj3cXbeXD0IuZv3GPJxphC5mkrkqr+BfjLBZs3AU1yOPYEcFfQYwC+yTzIfW8upG5SGfq0+RXt615CVKQ9Y2iM1zx7DqYwpKWlaSDN1CdOn2XK0u8Y8dUmNu05SlJCSXq2qs49V1UjLsbrlnpjwo+ILFHVtHyPKw4JJsu5c8qsdbsYPmcji7fsp0xsFF2bXcaDLVKoVCbWw0iNCS+WYPKxdNt+3pyziU9W/0CJiAg6NqhKnzap1KwcH+QojQk/lmACtHXvUUbO3czE9O2cOH2Oa2tXonfrVJqllsdp/DLGXMgSTAHtO3qKd77eypj5W9h79BT1ksvSu3UqN1uFsDE/YwnmF7qwQji5nFMhfHeaVQgbk8USzEU6d075fO1O3vxqE4u37KdsyRJ0bXYp3VukUCneKoRN8WYJJogurBC+vWES/a6tQbXypTz/bmNCkSUYD2zZc75CWBW6t7iMvtfUIKFUdKHFYEwosATjoe8PHufVmeuZtCST+Jgo+l5Tg+4tUogtEVnosRjjh0ATjDWP/AJVypbk/zrX5+MBrWl8WTme/3gd7V6ezZSlmZw7V3QTtjHBZgnmItS+pAyjezRhXO+mlI+L5vGJK7hl8FzmrLeBsIwBSzBB0eJXFZnWtyWD7m3IkZOn6TZqEQ+MXMiq7w76HZoxvrIEEyQREcJt9avy+eNt+XOHK/jmu4PcOmQuj727nMz9x/wOzxhfWCWvRw4eP82w2RsZNXczqvBgyxT6Xl2DsqVK+B2aMRfNWpFCxI4Dx3ll5nreW+q0OPW7tgbdmluLkynarBUpRFRNKMlLd9VnRv/WNLqsHP+c4bQ4TV1mLU4m/FmCKSR1qpThrR5NGNurKeXiSvDYuyvoMHguX22wFicTvizBFLKWNSoyvW8rBnZpwKETp3lgpNPilLHriN+hGRN0lmB8EBEhdGyQxKwn2vKnW+qwMvMgtw6ey3tLMv0OzZigsgTjo5ioSHq1TuXTR9twZXJZnpi0gicmruDYqTP5f9iYIsASTAi4pGws43o1pX+7mkxZlsmtg+ey7odDfodlzEXzcurYWiKyPNvrkIg8KiLlRWSmiGxw38u5x4uIDBKRDBFZKSKNvIotFEVFRvD49ZcztmdTDp04Q8ch8xi/aJvN5WSKNC8nXvtWVRuoagOgMXAMmIozodosVa0JzOL8BGs3AzXdVx9gqFexhbIWNSoyo39rmlQvzzNTvqH/hOUcPnHa77CM+UUKq4jUDtioqluBjsAYd/sYoJO73BF4Wx1f48xhXaWQ4gspifExjOnRhN/fWIuPVu6gw+C51q/JFEmFlWC6AOPd5cqq+j2A+17J3Z4EbM/2mUx320+ISB8RSReR9N27w/cZkogIoe81NZjQpzknT5/jjtfn89a8zVZkMkWK5wlGRKKB24BJ+R2aw7af/Tap6nBVTVPVtMTExGCEGNKaVC/PjAGtaVWzIs9+sIaH3lnCwWNWZDJFQ2HcwdwMLFXVne76zqyij/u+y92eCVTL9rlkYEchxBfyysdFM6JbGn9sX4dZa3fRftBXLNu23++wjMlXYSSYezlfPAKYDnR3l7sD07Jt7+a2JjUDDmYVpYxTZOrdJpVJDzUH4K5hCxg+Z6P1ZzIhzdMEIyKlgOuBKdk2vwBcLyIb3H0vuNtnAJuADOBN4GEvYyuqGl5ajhn9W3Ndncr8c8Y6er2dzr6jp/wOy5gc2XANRZSq8vaCrfzjo7WUj4tm0L0NaVK9vN9hmWIiaMM1iMgAESnjFl1GishSEbkhOGGaX0pE6N4ihSkPtyC2RARdhi9gyBcbOGtFJhNCAiki/UZVDwE3AIlAD84Xa4zP6iaV5YNHWnFLvaq89Nl6uo9axO7DJ/0OyxggsAST1XzcHhitqivIuUnZ+CQ+tgSDujTghTuuZPGWfdw88CvmZ+zxOyxjAkowS0TkM5wE86mIxAPnvA3LFJSI0KXJpUzr15KEUiV4YNQiRtuDecZngSSYnjj9ha5S1WNANE4xyYSg2peU4f2+Lbm2diX++sEa/vDeSk6eOet3WKaYisptRw69mVNFrGRUFJSOieKNro159fP1DP4ig427jzK0ayMqxcf6HZopZnJNMMDL2ZYbA0uyrStwrScRmaCIiBCeuKEWtS6J58lJK+g4ZB5vdkujblJZv0MzxUiuCUZVr8laFpFl2ddN0dGhXlVSKsTR5+10Og+bz4ud63Nr/ap+h2WKiUCf5LWawiKsblJZpvVrRd2qZXlk/DJe/HSddTEwhcKGzCwmEuNjGNe7GfekVeO1LzfS5z/pNpCV8VxelbyDOX/nkiwig7LvV9X+XgZmgi86KoIX7rySOlXiee6jtdzx+nxGdE/jsgpxfodmwlRelbzZO/ksyfUoU6SICA+2rE7NyvH0HbeU24bM4/X7G9GyRkW/QzNhyDo7FmNb9x6l15h0Nu05yp9vqUP3FinYowgmEBfd2VFEKorIX0Skv4iUFpGhIrJKRKaJSI3ghmv8cFmFOKY83IJraiXy7AdreGbKN5w6Yw9pm+DJq5J3HBCDM8r/IpyxWjoDHwIjvA/NFIb42BIMfyCNftfUYMLi7dz35tfsOWKdJU1w5JVgKqvq/wD9gdKq+qKqrlPVN4GEwgnPFIaICOHJG2sx+N6GrNpxkNtsFgMTJHklmLMA6lTSXNg11+6jw9Ct9asy+aEWKNB52Hw+XGlDIpuLk1crUqqITMcZmiFrGXe9uueRGV/UTSrL9H6teOidJfQbt4xvfzjMY9ddTkSEVf6agsu1FUlE2ub1QVWd7UlEBWCtSN45eeYsf35/FRPTM7n+isq8ek8DSsfk9ffIFCeBtiLl1RfJ9wRi/BMTFcm/7qxHnSpl+PtHa7nj9XmM6HYVl1Yo5XdopgjJq5m6poiMFpFXRCRZRD4WkSMiskJE8s1cpugTEXq0rM6YHk3Yeegkt702l/kbbaQ8E7i8KnlHAwtwJj9bCIwCKgJPAq8FcnIRSRCRySKyTkTWikhzESkvIjNFZIP7Xs49VkRkkIhkiMjKHMajMT5pVbMi0/q2pGLpGB4YuYi3F2yxkfJMQPJKMKXdaVpfAo6r6iRVPaGqM3GejwnEQOATVa0N1AfW4oyON0tVawKz3HVwZoCs6b76AEML/uMYr6RUjGPqwy24+vJE/nfaav5nqj2UZ/KXV4LJ/r/nUB77ciQiZYA2wEgAVT2lqgeAjsAY97AxQCd3uSPwtjq+BhKyppg1oSE+tgTDu6Xx8NW/Yvyi7dw/wh7KM3nLK8HUdosq32RbzlqvFcC5U4HdwGgRWSYiI0QkDucBvu8B3PdK7vFJwPZsn890t5kQEhkhPHVTbQZ2acDKzIN0HDKP1TvsoTyTs7zaHesE4dyNgEdUdaGIDOR8cSgnOT1o8bOCvoj0wSlCcemll15kiOaX6tggieoV4+jz9hI6D13AS3fV55Z6dsNpfirXOxhV3ZrXK4BzZwKZqrrQXZ+Mk3B2ZhV93Pdd2Y6vlu3zyTgVzBfGNVxV01Q1LTExMYAwjFfqJScw/ZGW1KniDP3wysz1NlKe+QnPRrRT1R+A7SKSVZxqB6wBpgPd3W3dgWnu8nSgm9ua1Aw4mFWUMqGrUnws4/s0o3PjZAbN2sDvxi7h6MkzfodlQoTXj2Y+AowVkWic3tg9cJLaRBHpCWwD7nKPnYEzuVsGcAybe6nIiImK5MXOzkN5//hoDXcOnc+b3dKoVt4eyivu8uoqMEtV24nIv1T1D4UcV0Csq0DombN+N/3GLSUyQnj9/sY0/1UFv0MyHrjoAaeAKm5/pNtEpKGINMr+Cl6oJpy0uTyRaf1aUT4umgdGLuQ/XwdSXWfCVV5FpP/FafVJBl65YJ9NvGZyVb1iHFP7tuTRCcv58/urWPv9IZ699ddER9kkFsVNXp0dJwOTReTPqvpcIcZkwkCZ2BK82S2NFz/9lmGzN5Kx8whDuzaiQulAHwI34SDfPymq+pyI3CYiL7mvDoURmCn6IiOEp292HspbkXmA24bMY82OCx8KN+Es3wQjIs8DA3CamNcAA9xtxgSkY4MkJv62OWfOnePOofP5+Bt7+qC4CKRQfAtwvaqOUtVRwE3uNmMCVr9aAh/0a0XtKvH8buxSXrWH8oqFQGvdsg/yXdaLQEz4q1QmlvG9nYfyBs7awFPvreSsJZmwFsiDds8Dy0TkS5z+Qm2AZzyNyoSt2BLOQ3lJCSUZOGsDx0+f5d/3NKBEpLUwhaN8E4yqjheR/wJX4SSYP7jdAIz5RUSEx66/nFLRkTz/8TpOnj7LkPsaEVsi0u/QTJAF9GdDVb9X1emqOs2SiwmW37b9Fc91/DWfr91FrzHpHDtlfZjCjd2XGl890DyFl+6qz/yNe+g+ahGHT5z2OyQTRJZgjO86N05m0L0NWbbtAPePWMiBY6f8DskESZ4JRkQiRGRVYQVjiq8O9aryxgONWffDYboM/5rdh20oznCQZ4JR1XPAChGxoeOM59rVqcyo7lexde8x7nljAd8fPO53SOYiBVJEqgKsFpFZIjI96+V1YKZ4alWzIm/3bMLuwye5a9gCtu095ndI5iLkOh7MjwfkMoVsKMz8aOPBhK+VmQfoNmoRMVERjO3VjBqVSvsdkskmGOPBAD8mki1ACXd5MbD0oiM0Jg/1khOY0KcZZ8/BPW8ssE6SRVQgnR174wzY/Ya7KQl438ugjAGofUkZJv62GdFREXQZvoBl2/b7HZIpoEDqYPoCLXEnX1PVDZyfy8gYT6Umlmbib5uTUCqariMWsnDTXr9DMgUQSII5qao/PpggIlHkMF+RMV6pVr4Ukx5qTpWEknQfvYjZ63f7HZIJUCAJZraI/A9QUkSuByYBH3gbljE/VblMLO/2aUZqxdL0HpPOp6utx0pREEiCeRpnCthvgN/iTC/yJy+DMiYnFUrHML53M66oWoaHxy5l2vLv/A7J5COQ3tTnRGQMsBCnaPSt5te27RKRLcBh4CxwRlXTRKQ88C6QgtM6dbeq7hcRAQbizI10DHhQVa21yvxE2VIleKdXU3q+tZhH313OidNnuecqew40VAXSinQLsBEYBAwBMkTk5gJ8xzWq2iBbm/nTwCxVrQnM4vx81TcDNd1XH2BoAb7DFCOlY6J4q0cT2tRM5A/vfcPoeZv9DsnkIpAi0ss4SeJqVW0LXAO8ehHf2REY4y6PATpl2/62Or4GErLmsDbmQiWjIxnerTE3/royf/1gDa99meF3SCYHgSSYXaqa/V9vE+cnrM+PAp+JyBIR6eNuq5w157T7ntXknQRsz/bZTHfbT4hIHxFJF5H03butNaE4i4mK5LX7GtGpQVVe/PRbnp2+mpNnzvodlskm1zoYEbnDXVwtIjOAiTgJ4y6cp3kD0VJVd4hIJWCmiKzL41jJYdvP6npUdTgwHJyuAgHGYcJUVGQEL9/dgHJx0Yyet4X0rfsY1KUhqYnWtSAU5HUHc6v7igV2Am2Bq3FalMoFcnJV3eG+7wKmAk2AnVlFH/c9624oE6iW7ePJwI4Afw5TjEVGCH+59dcMf6AxmfuP02HwXN5bkkmAbRHGQ3nN7NjjYk4sInFAhKoedpdvAP4GTAe6Ay+479Pcj0wH+onIBKApcDCrKGVMIG749SVcmVyWRycs54lJK/hqw26e61SX+NgSfodWbOXbTC0i1YFHcJqVfzxeVW/L56OVgalO6zNRwDhV/UREFgMTRaQnsA2nyAXO8zXtgQycZuqLSnCmeKpStiTjejfjtS8z+Pfn61m2/QCDujSkfrWE/D9sgi6Q4RpWACNxHrQ7l7XdhmswoS59yz4GTFjOzkMn+P2NtejdOpWIiJyq+kxBBTpcQyDzIp1Q1UFBiMmYQpWWUp4Z/Vvz9JSVPP/xOuZm7OHlu+tTKT7W79CKjUCaqQeKyF9EpLmINMp6eR6ZMUFQtlQJXr+/Ef+8/UoWbd5H+4FfWWfJQhTIHcyVwAPAtZwvIqm7bkzIExHua3opaSnleGTcMrqPWkSfNqk8eUMtoqNsYg0vBZJgbgdSsw/ZYExRdHnleKb1a8k/PlrL8DmbWLBxL4PubUj1inF+hxa2AknfKwCrgjdhIbZEJM91qsuwro3Ztu8YHQZ9xZSlmX6HFbYCuYOpDKxzm5d/nKwmgGZqY0LWTXUvoZ77zMzjE1fw1YY9PNepLqVjAvmVMIEK5Gr+xfMojPFB1YSSjOvdlCFfZjBo1gaWbdvPoHsbUi/ZbtiDJd/nYEKZPQdjgmXR5n08OmEZu4+c5Kkba9OzVXV7ZiYPQZu2REQOi8gh93VCRM6KiM0hYcJKk+rlmTGgNdfWrsQ/ZqzlwbcW2/S1QRDIvEjxqlrGfcUCd+IMPGVMWEkoFc2wro35e6e6LNy0l+temc0bszdy4rQNAfFLFfghAFV9H3sGxoQpEaFrs8v48JFWNLw0gec/XkfbF79k/KJtnDl7Lv8TmJ8IpC/SHdlWI4A0oK2qNvcysEBYHYzx2teb9vJ/n6xj6bYDVK8YxxM3XE77ulWKff1MoHUwgSSY0dlWz+AM1P2mO8aLryzBmMKgqny+dhcvfrqO9TuPUDepDE/dWJvWNSvijhZQ7AQtwYQySzCmMJ09p0xb/h2vzFxP5v7jNE+twFM31aLhpQGNvxZWLjrBiMj/5vE5VdXnfmlwwWIJxvjh5JmzjF+4jcFfZLD36CluuKIyv7+xFjUrx/sdWqEJRoJ5IofNcUBPoIKq+j7oqSUY46cjJ88wau5mhs/ZxLFTZ7ijUTKPXleT5HKl/A7Nc0EtIolIPDAAJ7lMBF62OhhjHPuOnmLofzMYs2ArKHRtdhl9r/kVFUrH+B2aZ4KSYNxZGB8H7seZw2igqu4PWpQXyRKMCSU7Dhxn4OcbmLRkOyVLRNKrdSq9WlcPyzGBL/pJXhF5EWd6ksPAlar6bCglF2NCTdWEkvyrcz0+e6wNbS5PZOCsDbR98b+MnLu52D6sl1cdzDmc3tNn+On8RIJTyVvG+/DyZncwJpSt2H6A//t0HfMy9pKUUJLHrr+cOxslhUXT9kXfwahqhKqWvKCrQJms9eCGa0z4qV8tgbG9mvFOz6ZUKB3Nk5NW8Lt3lnLoxGm/Qys0Nl6gMR5rVbMi0/q25I/t6zBz7U46DpnHuh+KR39hzxOMiESKyDIR+dBdry4iC0Vkg4i8KyLR7vYYdz3D3Z/idWzGFBYRoXebVMb3bsaRk2fo9No8pi4L/5H0CuMOZgCwNtv6v4BXVbUmsB+n6Rv3fb+q1gBedY8zJqw0qV6ej/q3ol5yAo+9u4I/vf8NJ8+EbwWwpwlGRJKBW4AR7rrg9MSe7B4yBujkLnd013H3t5NwqA0z5gKV4mMZ16spv22Tyjtfb+PuYQv47sBxv8PyhNd3MP8GnuL8dCcVgAOqesZdzwSS3OUkYDuAu/+ge/xPiEgfEUkXkfTdu21+G1M0RUVG8Ez7Ogzr2oiNu4/SYVB4ztfkWYIRkQ7ALlVdkn1zDodqAPvOb1AdrqppqpqWmJgYhEiN8c9NdaswvV9LKsXH8uDoRQz8fAPnzhXdDsgX8vIOpiVwm4hsASbgFI3+DSSISNZg48nADnc5E6gG4O4vC+zzMD5jQkJqYmmm9m1BpwZJvPr5en4zZjH7j4bHNGSeJRhVfUZVk1U1BegCfKGq9wNfAp3dw7oD09zl6e467v4vtCiPJWFMAZSKjuKVu+vz9051mZ+xlw6D57Iy84DfYV00P56D+QPwuIhk4NSxjHS3jwQquNsfB572ITZjfJM1XOfEh5zBIjsPXcC4hdsoyn9nbcApY0LQvqOnGDBhGV9t2MOdjZL5e6e6lIyO9DusHwVt2hJjTOErHxfNWz2aMKBdTaYsy+T21+exZc9Rv8MqMEswxoSoyAjhsesvZ/SDV/HDoRPcOngun63+we+wCsQSjDEh7upalfjwkVZUT4yjz3+W8PzHa4vMFCqWYIwpApLLlWLSQ825v+mlvDF7E11HLmTX4RN+h5UvSzDGFBExUZH84/Yrefmu+izffoAOg+YyJ8Sf/rUEY0wRc2fjZKY+3JLSsVF0G7WI/uOXhezdjCUYY4qgOlXKMKN/ax69riafrPqBdi/P5p2vt4ZcNwNLMMYUUbElInn0usv55NHWXJlUlj+9v4o7h81n7fehM5iVJRhjirjUxNKM7dWUV+6uz9a9x+gweC7/nLGWY6fO5P9hj1mCMSYMiAh3NErmiyfaclfjZIbP2cT1r8xh1tqdvsZlCcaYMJJQKpoX7qzHpIeaExcTSc8x6Tz0nyV8f9CfAa0swRgThq5KKc+Hj7Tm9zfW4stvd3Hdy7MZNXczZwu5EtgSjDFhKjoqgr7X1GDmY21JSynP3z5cQ8fXCncYCEswxoS5SyuU4q0eVzHkvobsPHSSTq/N49npqzlcCPMzWYIxphgQETrUq8qsJ9rStdlljFmwhetemc3H33zv6XgzlmCMKUbKxJbgbx3rMvXhllSIi+F3Y5fSc0w62/cd8+T7LMEYUww1qJbA9H4t+dMtdfh6015ueHUOw2Zv5HSQe2lbgjGmmIqKjKBX61RmPt6WVjUr8sLH6+gxenFwvyOoZzPGFDlJCSV5s1san63+4efzBF0kSzDGGABu+PUlQT+nFZGMMZ6xBGOM8YyXU8fGisgiEVkhIqtF5K/u9uoislBENojIuyIS7W6Pcdcz3P0pXsVmjCkcXt7BnASuVdX6QAPgJhFpBvwLeFVVawL7gZ7u8T2B/apaA3jVPc4YU4R5OXWsquoRd7WE+1KcOaonu9vHAJ3c5Y7uOu7+diIiXsVnjPGep61IIhIJLAFqAK8BG4EDqpo1Ek4mkOQuJwHbAVT1jIgcxJlads8F5+wD9HFXj4jItwGEUvHC84SgUI8x1OOD0I8x1OODwGO8LJCTeZpgVPUs0EBEEoCpQJ2cDnPfc7pb+VmzvKoOB4YXJA4RSQ9kmks/hXqMoR4fhH6MoR4fBD/GQmlFUtUDwH+BZkCCiGQltmRgh7ucCVQDcPeXBfYVRnzGGG942YqU6N65ICIlgeuAtcCXQGf3sO7ANHd5uruOu/8L9bKbpzHGc14WkaoAY9x6mAhgoqp+KCJrgAki8ndgGTDSPX4k8B8RycC5c+kSxFgKVKTySajHGOrxQejHGOrxQZBjFLtJMMZ4xZ7kNcZ4xhKMMcYzYZ9gROQmEfnW7YLwdAjEU01EvhSRtW4XigHu9vIiMtPtQjFTRMr5HGekiCwTkQ/d9Ry7ePgYX4KITBaRde61bB6C1/Ax9994lYiMd7vP+HodRWSUiOwSkVXZtuV43cQxyP3dWSkijQr6fWGdYNwK5teAm4ErgHtF5Ap/o+IM8ISq1sFptu/rxvQ0MMvtQjHLXffTAJxWvyy5dfHwy0DgE1WtDdTHiTVkrqGIJAH9gTRVrQtE4jRc+H0d3wJuumBbbtftZqCm++oDDC3wt6lq2L6A5sCn2dafAZ7xO64LYpwGXA98C1Rxt1UBvvUxpmT3P9q1wIc4D0HuAaJyuq4+xFcG2IzbSJFteyhdw6wn08vjtNZ+CNwYCtcRSAFW5XfdgDeAe3M6LtBXWN/BkK37gSt71wTfuT3GGwILgcqq+j2A+17Jv8j4N/AUkDVAawVy7+Lhh1RgNzDaLcaNEJE4Qugaqup3wEvANuB74CBOt5lQuo5ZcrtuF/37E+4JJqDuB34QkdLAe8CjqnrI73iyiEgHYJeqLsm+OYdD/byOUUAjYKiqNgSO4n+R8ifceoyOQHWgKhCHU+S4UEj8f8zFRf+7h3uC+bH7gSt71wTfiEgJnOQyVlWnuJt3ikgVd38VYJdP4bUEbhORLcAEnGLSv8m9i4cfMoFMVV3ork/GSTihcg3BeXJ9s6ruVtXTwBSgBaF1HbPkdt0u+vcn3BPMYqCmW3MfjVPJNt3PgNwhKEYCa1X1lWy7sneVyN6FolCp6jOqmqyqKTjX6wtVvZ/cu3gUOlX9AdguIrXcTe2ANYTINXRtA5qJSCn33zwrxpC5jtnkdt2mA93c1qRmwMGsolTA/KoEK8QKrfbAepyhIv4YAvG0wrnNXAksd1/tceo5ZgEb3PfyIRDr1cCH7nIqsAjIACYBMT7H1gBId6/j+0C5ULuGwF+BdcAq4D9AjN/XERiPUyd0GucOpWdu1w2niJQ1zMo3OC1iBfo+6ypgjPFMuBeRjDE+sgRjjPGMJRjeDEbAAAADQUlEQVRjjGcswRhjPGMJxhjjGUswxhjPWIIxxnjGEkwxJyJH8j/qor+jpIjMdofPCOZ5f1HsIhItInOyPbJvPGIJxhSG3wBT1Jkny3eqegrnidV7/I4l3FmCMT8Skcfd0ddWicij2bb/2R05bqY7MtuTBTz1/WTrcyMiXUVkkYgsF5E33NHz/iUiD2c75lkRecJdfl9Elrijw/W58OQiknLBCG1PisizuX2Xe9j7blzGQ5ZgDAAi0hjoATTFGWmvt4g0FJE04E6ccWvuAAo065/byTRVVbe463Vw7hxaqmoD4CzOL/oEfnpHcTdOXx2A36hqY/e7+4tIhQC/O7fvAqd/0FUF+VlMwVkZ1GRpBUxV1aMAIjIFaI3zR2iaqh53t3/gvqcCfwTKqmpnd1sc8DpwCvivqo7Fmev4QLbvaQc0BhY7nYwpiTP+zNsiUklEqgKJwH5V3eZ+pr+I3O4uV8MZwnFvAD9Tjt8FzrTGInJKROJV9XABrpMpAEswxYyI9AV6u6vts+/K7SM5bVTVTUBPEZmcbfMdwGRV/UBE3gXGAseB2AvON0ZVn8nhtJNxhjK4BOeOBhG5GmdsleaqekxE/nvB+cAZ5zj73XjW/ry+C5zezSdy2WeCwIpIxYyqvqaqDdxX9sGD5gCd3PFL4oDbga+AucCt7oj4pYFb8jh9MueHWDzrft9+IFJEsn7pZwGdRaQS/Dii/WXuvgk4Y9B0xkk24MxRvt9NLrVxim8X2glUEpEKIhIDdMjvu9xiVtZgUMYjdgdjAFDVpSLyFs5YJQAjVHUZgIhMB1YAW3HGYDmYy2kycZLMcn76x+sznCLY56q6RkT+BHwmIhE445L0Bbaq6moRiQe+0/MDG30CPCQiK3EGnf46h9hPi8jfcMY23owzBgt5fRdwDTCjINfIFJyNB2PyJSKlVfWIiJTCudPpg/NL+g+cGRFGqOrz7p3PEJxix1y3DgYRaQg8rqoP+PMT/Jxbx/SMqn7rdyzhzBKMyZeIjMOZVyoWp07j+V9wjt+4n/X9WZis4VNV9W2/Ywl3lmCMMZ6xSl5jjGcswRhjPGMJxhjjGUswxhjPWIIxxnjGEowxxjOWYIwxnvl/zkk+jB8jIBQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# BBH, try different evalues\n",
    "evalues =  np.arange(0,110,10,dtype=float)\n",
    "BBH_nums = list()\n",
    "for evalue in evalues:\n",
    "    BBH = extract_BBHs('../Results/bidirectional_blast.tab',10**(-evalue))\n",
    "    BBH_nums.append(len(BBH))\n",
    "plt.figure(figsize=(4,3))\n",
    "plt.plot(evalues,BBH_nums)\n",
    "plt.xlabel('-log$_{10}$(evalue)')\n",
    "plt.ylabel('Number of BBHs')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of homologs for query genes (evalue = 1e-20): 1034\n",
      "Number of homologs for target genes (evalue = 1e-20): 944\n",
      "Number of BBHs (evalue = 1e-20): 718\n"
     ]
    }
   ],
   "source": [
    "BBHs = extract_BBHs('../Results/bidirectional_blast.tab',1e-20,True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of candiate reactions: 1235\n"
     ]
    }
   ],
   "source": [
    "# Build first template model\n",
    "candidate_rxns = get_all_rxns_in_BBH(template_model, BBHs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1235\n",
      "Number of metabolits: 1208\n",
      "Number of compartments: 3\n",
      "Number of genes: 857\n",
      "Number of missing genes: 144\n",
      "Number of reactions with missing genes: 136\n"
     ]
    }
   ],
   "source": [
    "model_from_template = build_model_from_template(candidate_rxns,BBHs,True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARgAAADQCAYAAADcQn7hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJztnXl4FUX2sN8DJASSsCWBAEFCkB1CIGET2VUEGXQYd0YRFXX0E9QZf+i4MzoyijMObqgwoqMCAo6ioKLIriwBwr6HAJEtYQlZyH6+P7oTAmRpkntzs9T7PP3ce6urq0537j05VafqHFFVDAaDwR3U8LQABoOh6mIUjMFgcBtGwRgMBrdhFIzBYHAbRsEYDAa3YRSMwWBwG0bBGAwGt+E2BSMi/xGREyKyrZBzfxERFZFA+7OIyFQR2SciW0Sku7vkMhgM5Yc7LZiZwPUXF4pIC+Ba4FCB4mFAG/t4AHjPjXIZDIZyopa7GlbVFSISWsipfwH/B3xdoOxG4BO1lhWvEZEGItJUVY8W10dgYKCGhhbWhcFgcCcbNmxIVNWgkuq5TcEUhoiMBH5T1c0iUvBUc+Bwgc/xdlmxCiY0NJTo6GiXy2kwGIpHRA46qVduCkZE6gLPANcVdrqQskI3SYnIA1jDKK644gqXyWcwGFxPeXqRWgOtgM0iEgeEABtFJBjLYmlRoG4IcKSwRlT1A1WNUtWooKASLTSDweBByk3BqOpWVW2sqqGqGoqlVLqr6jFgAXC37U3qDSSVNP9iMBgqPu50U88CfgXaiUi8iNxXTPVFQCywD/gQeNhdchkMhvLDnV6kO0o4H1rgvQKPuEsWg8EA6Vk5nErNJDElg5MpmSTYrydTMqyy1EzaB/vzzA0dXdZnuXqRDAaD61BVzp7LJjE1g8RkS0GcTMkgoaDSSMnkZGomickZJGdkF9pOXe+aBPh5E+hXm1o1XTuoMQrGYKgk5OYqu44l88v+RFbvS2TdgVOkZuZcUk8EGtX1JsDPmwDf2nRuXp8AX2+C/GsT4OtNgF9tAm2FEuDnTV1v96kBo2AMhgrMoZNprLYVyq/7T3IyNROAsEBfburWnFaBvrbiqJ1vhTSs6+VyS6S0GAVjMFQgElMy+GX/SVbvTWT1/kTiT58DoLF/bfq3DaLvlYFc1TqAZg3qeFhSZxgFYzC4m5wsiI+GozGQe+E8SEZ2LodOpnLgZBpxiamcSM4AINCrBhMa1SW0hS+tAn0J8PVGRCAD2OFGWRtcAR1vdFlzRsEYDK4mNxdObIfY5XBgOcSthqzUQqvW5vwuXwC8Cpw8ZR973CnsRbQeYhSMwVDhOHXAUiaxy+DACkg7CUBG/TCOhPyOmFpd+Sk1jDWH00jPyqWGQKdm9enduhG9WwUQ0aIBPl41PXsPAOJaGYyCMRhKQ8oJS5HELkMPLEfOWNFHUryD2OETyfI6Hfg6qQ3xxxvBcfCqKbQOqsvverSg75WB9AprRD0frxI6qfwYBWMwOCH9LBz8hfQ9P5Ozfxm+Z3YDkCJ+rMntwPLswfyS24n96c0IqVmX9sH+3BjhT9sm/rQPrkerQF+8a1UMz055UqKCERFf4Jyq5opIW6A98J2qZrldOoPBE+TmkplyksO7N5Cx52f8j6ymWeoOapIL6sXG3Hb8kns7W7y7khscTrumDekY7M9NTfxp28QP/2pgmTjFiQWzAugnIg2BJUA0cBsw2p2CGQwuJSsdUhMg9QSkJlrvU/Len4DUBDQ1gayk49RKP4U3ubQGclTYSmt+rXszJ5tchU9ob65sHsjYYH+C/GpzUVwjw0U4UTCiqmn2ZsW3VPU1EdnkbsEMBsdkpMChNXD6gK00Ei48UhIgM7nQS9XLl0yfABJy67E/zZffsrpwtkYDAoNb0LJ1B4I6D6JL02AiahhFUhocKRgR6YNlseTtiDZzNwbPkZ0Jv22wPTbLIX59gfUlAnUDwK8x+AZCs27ga7/3awy+QeAbxMEMXxbszeR/204Tm5CKV01hQNsgRkY056YOjd26fL464eQpPgY8DfxPVbeLSBiw1L1iGQwFyM2F49tsN/ByOPiLva5EoGlX6PMIhA2EJp0t5VKjcFfr0aRzfLv5KF9v/o1tvx1HBHq3CmBc/zCGdQ6mQV3v8ryrakGJCkZVlwPL7cleVDUWGO9uwQzVGFVruJO3UK3AuhIC2kDEHdBqAIReDXUbFdvU6dRMFm07yoKYI6yLO4UqhIfU59kbOjAivBnB9X3K4YaqL068SH2AGYAfcIWIdAUeVFUTFMrgOlJO2AplGcSugCQ7q41/U7jyWggbYCmV+s1LbCo1I5ufdh5nQcwRlu9JIDtXCQvy5bEhbRkZ0YxWgb7uvRdDPk6GSG8CQ7HCWmJnBOjvVqkMVZ/0s3Bw9Xkr5YS9wcanPoT2g77jLYUS2MaKP1ACmdm5rNiTwNebj/DTjuOcy8qhaX0f7r26FSO7NqNTs3rG4+MBHM1kqerhi/44lwahMBgKIycbTu6zFMiJHXB8h7VP5/RBQKGWD1zRG7rcYlkpTSOKnEMBK8jSkaR0dh87y65jyey2j9iEVDJzcmlY14tR3ZszsmszeoQ2oobx/ngUJwrmsIhcBaiIeGPNv+x0r1iGSocqnP3NViAFlEnibsixYpggNSDgSkuJdL0TWvaBkJ7gVfg8SFJaFruOnWX38eR8ZbLnWPIFkdma1vehXbA/A9oF0btVAFe3CcSrgsRCMThTMA8B/8ZKhBYPLMbEz63enDtjK5DttjLZab2mJ52v498MmnSE1oOgSSdo3AEC2xWqTNKzcth3IsWyRvKVyVmOn80435xPLWv5fbdmtAuuR/tgaxl+/Tpm1WxFxokXKRGzard6k7gXNn1quYpP7LQslTxq17cUSec/QOOO55VJnYZFNpeUlsWs9YfYEn+GXceSiUtMJddOs+ddswZXNvajb+tA2gb70y7Yn/bB/gTX8zFzKJUQJ16kIGAcEFqwvqre6z6xDBWCM4dh+WSI+dzaxh/U3nINN+5oK5OOUK+5o0lYsLw7M3+JY9ry/SSnZ9MyoC7tmvgzoktT2gXXo12wH6EBvhUm3KOh7DgZIn0NrAR+wkzuVg9STsDKf0L0DECg15/g6sfBr3SZNDOyc5i19hBvL91HYkom13Rowp+va0uHpvVcK7ehwuFEwdRV1Ylul8Tgec6dgV/egjXvQXY6dBsNAyZC/ZBSNZedk8uXm37j3z/t5bcz5+gd1oj372pPZMuih0+GqoUTBfOtiAxX1UVul8bgGTJTYe37sPpNa6K20ygY9AwEXlmq5nJzle+3H+ONxbvZn5BK15D6/OMP4fS9MsDMo1QznCiYCcBfRSQTyAQEKxmjsW8rO9mZsPFjWPE6pByHNkNh8LPQNLxUzakqy/ckMGXxbrb9dpY2jf2Y9sdIhnZqYhRLNcWJF8m/PAQxlCO5ObBlDix7Fc4cgpZ94dZPrAVvpWR93Cle/3436+JOEdKwDm/c0pWbujWnplnoVq1x4kUSLDd1K1X9m4i0AJqq6roSrvsPMAI4oaqd7bLXgd9hWUL7gbGqesY+9zRWOIgcYLyq/lD62zIUiirs/AZ+ftlaANe0K4z4lxVJvpQWxrbfknhj8W6W7k4gyL82f7uxE7f1uKJahoc0XIqTIdK7QC4wGPgbkAK8A/Qo4bqZwNvAJwXKfgSeVtVsEfkHVhiIiSLSEbgd6AQ0A34SkbaqarxWrkAVYpfCkklwZBMEtoVbPrbSU5RSscQmpPDGj3tYuOUo9et48dSw9ozpE0od7woQGd9QYXCiYHqpave8KHaqetreMlAsqrpCREIvKltc4OMa4Gb7/Y3AbFXNAA6IyD6gJ/CrA/kMxXF4naVY4lZC/RZw47sQfhvULF1Apd/OnGPqT3uZtzGe2rVq8OjgK7m/X5hZUWsoFCffsiwRqQko5C+8y3VB3/cCc+z3zbEUTh7xdtkliMgDwAMAV1xxhQvEqKIc22YNhfZ8Z0VxG/YaRN4DtWqXqrnElAzeXbqfT9ccBGBMn1AeHtSaQL/StWeoHjhRMFOB/wGNReQVLKvj2bJ0KiLPANnAZ3lFhVTTwq5V1Q+ADwCioqIKrVMtycm2UpPGLrOOuFXgUw+GPA+9HgLv0sdAWbT1KE/O3cy5rBxuiWzB+Gva0LyS5EY2eBYnXqTPRGQDMARLEdykqqXeTS0iY7Amf4eoap6CiAdaFKgWAhwpbR/VAlVI2FUgPekqyDhrnWvSBQb8H/T+U7F7gpzw5cZ4/jJ3M92uaMhrN4fTOsjPBcIbqgtOvEiNgBPArAJlXqXJiyQi1wMTgQGqmlbg1ALgcxH5J9YkbxugWC9VteTM4fNxaQ8st9auADQMhU6/t+KphPYv9ZL+i/l87SGe+WorV7UO4MO7o0wgbMNl4+QbsxHLujiNZcE0AI6KyAlgnKpuKOwiEZkFDAQCRSQeeAHLa1Qb+NFeeLVGVR+yg4l/AezAGjo9YjxIQOpJiFtxXqGcirXKfYOsaG95YSQbtnR51/9ZdYBJ3+5gcPvGvDu6e8XIm2yodMj5UUoRFUSmYWUU+MH+fB1wPfAF8G9V7eV2KYsgKipKo6OjPdW968lMhYO/2nFpl1kTtSh4+0No3/NKpXHHUruXnfDusn289v1uhnUO5t+3dzNrWgyXICIbVDWqpHpOLJgoVX0o74OqLhaRv6vqEyJiXAiuYMtciP6Pnd8nC2p6Q4te1n6gsAFWbp+a7ncDqyr/+nEPU3/ex40RzXjjlq4mdIKhTDhRMKdEZCIw2/58G3Dadl27wl1dvUlNhK8esuZR+jxsWSlX9AHvuuUqhqry90U7+XDlAW6LasHfR3Uxy/wNZcaJgrkTa/7kK/vzKrusJnCrm+SqPmybb2UlvPW/VgAnD5Cbq7ywYDv/XXOQe64K5fkRHU2wbINLcBoy89EiTu9zrTjVkJjPITjcY8olJ1d5av4W5m6I58EBYTx1fXuz89ngMswA25Oc2Gktjou40yPdZ+Xk8ticGOZuiOexa9oY5WJwOWZhgyfZPBtq1ILON5dc18VkZOfw6OebWLzjOE8Na89DA1qXuwyGqo9RMJ4iLybLlde6bGGcU9KzcnjwvxtYvieBl0Z2YsxVoeXav6H64GQl79RCipOAaFX92vUiVRMOLIfko3D95HLtNjUjm/s/jmbNgZNMHtWF23uaDaMG9+FkDsYHiAD22kc40Ai4T0TedKNsVZuYWVYe5nbDyq3Ls+lZ3P2fdayLO8W/bo0wysXgdpwMka4EBqtqNoCIvIeV3fFaYKsbZau6ZCRbkeUi7ih1+ITL5XRqJnf/Zx27jp3l7Tu6MaxL03Lp11C9caJgmgO+WMMi7PfNVDVHRDKKvsxQJDu+huxz0PWOcukuITmDu2asJTYxlffvimRw+ybl0q/B4ETBvAbEiMgyrM2O/YG/i4gvVjI2w+WyeTY0ag0hJUUdLTvHktK5c/oajp5J56N7etD3ykC392kw5OFkod0MEVmEFcJSgL+qal6slifdKVyV5PRBK3zloGfdumER4PCpNEZPX8up1Ew+vrcnPVs1cmt/BsPFOF1oVwNIAE4BV4pIf/eJVMXZ8oX12vU2t3ZzIDGVW9//lTNpmXx6fy+jXAwewYmb+h9YGxy3c35zowIr3ChX1UQVNs+C0H7QwH0enL3Hk7lz+lpycpVZD/SmU7P6buvLYCgOJ3MwNwHt7Ij/hrIQvx5O7Yd+f3ZbFwcSU7njw7WIwJwHetOmicmbZ/AcToZIsYDJSeEKNs8Cr7rQcaRbmo8/ncboD9eQq8qscb2McjF4HCcWTBqWF2kJkG/FqOp4t0lVFclKt0IzdPgd1Hb9D//E2XRGT19LSkY2sx7ozZWNjXIxeB4nCmaBfRjKwp7vIT0Jut7u8qZPpWYyevpaEpIz+PT+XmbOxVBhcOKm/rg8BKnybJ4F/s2siHUuJOlcFnfNWMuhU2nMHNuT7leULU2JweBKilQwIvKFqt4qIlu5MAmaAKqq4W6XrqqQkgB7f4SrHoUarovOn5qRzb0z17PneDIf3BVFn9YBLmvbYHAFxVkwE+zXEeUhSJVm61zQHJduDUjPymHcJ9FsOnSad+7szqD2jV3WtsHgKor0IqnqUfttInBYVQ9i5TTqism6eHlsnmVlBmjc3iXNZWbn8vBnG/ll/0mm3NLVbFw0VFicuKlXAD4i0hxYAowFZrpTqCrF8e1wbIvLrJecXOXxOTH8vOsEL9/UmVHdQ1zSrsHgDpwoGLHTvI4C3lLV3wOeiVBdGdk8y2VhMXNzlYnzt7Bw61GeGd6BP/Z2fUZHg8GVOFIwItIHGA0stMtMqE0n5GRbe4/aDAXfsk3AqiovfrOdeXaA7nH9w1wkpMHgPpwomMewckr/z84hHQYsLekiEfmPiJwQkW0FyhqJyI8istd+bWiXi4hMFZF9IrJFRLqX9oYqFLHLrAT1ZVz7oqpM/n4Xn/x6kAf6hzFhSBvXyGcwuJkSFYyqLlfVkar6DxGpASQ6XMU7EyuHdUGeApaoahus+Zyn7PJhQBv7eAB4z6H8FZvNs6BOQ2g7tEzNvP3zPt5fHsvoXlfw9DCTWsRQeShRwYjI5yJSzw4wtQPYLSIlxoFR1RVY4R0KciOQt3DvY6yNlHnln6jFGqCBiFRu10h6Euz6Fjr/oUxhMaevjOWNH/cwqntz/nZjZ6NcDJUKJ0Okjqp6FksZLAKuAO4qZX9N8tzf9mve4o3mwOEC9eLtssrLjq8hOx26lj6p2udrD/Hywp0M7xLMa38IN+lcDZUOJwrGS0S8sBTM16qaxYUre11BYb+cQvsQkQdEJFpEohMSElwshguJmQUBbaB56aaTvtr0G898tZVB7YJ487Zu1KppknAaKh9OvrXvA3FYwb5XiEhL4Gwp+zueN/SxX0/Y5fFAiwL1QihiMZ+qfqCqUaoaFRRUvgnLHHPqABz6xcoaUIohzffbjvHnuZvp3SqA9/4YiXcto1wMlRMnk7xTVbW5qg6350gOAoNK2d8CYIz9fgzwdYHyu21vUm8gqcBK4srHljmAQPjlh8VctvsEj87aSNeQ+kwfE4WPl+v2LhkM5U1xmx3/qKqfisgTRVT5Z3ENi8gsYCAQKCLxwAvAZOALEbkPOATcYldfBAwH9mHFnxl7OTdRocgLi9mqP9S/vFW2a2JP8uB/N9CmsT8fje2Jb22z3MhQuSnuG+xrv5YqcpGqFrU2fkghdRV4pDT9VDgOrYHTcTDgqRKrFiTm8Bnum7meFo3q8t/7elK/jgkiaKj8FKlgVPV9+/Wl8hOnCrB5Fnj5WpHrHLLjyFnunrGWAL/afHZ/LwL8yifbo8HgbpxkFWgFPAqEFqyvqu4JLFuZyToH2/9nxdyt7efokpMpGdzz0Tp8a9fis/t70aSej5uFNBjKDyeD/K+AGcA3nE9bYiiM3Ysg46zjrQGq1ubFM2lZfPVIX1o0qutmAQ2G8sWJgklX1alul6QqsHk21AuBUGd56T5de4ifdp7guREd6disnpuFMxjKHycK5t8i8gKwmAuzCmx0m1SVkeTjsG8J9J0ANUpet7L3eDIvf7uD/m2DGHtVqPvlMxg8gBMF0wVra8BgLszsONhdQlVKLiMsZkZ2DuNnx+BXuxZTbjFbAAxVFycK5vdAmKpmuluYSs3mWdA8EoLallj1te93s/PoWWaMiaKxv5nUNVRdnKxB3ww0cLcglZpjW+H4NkfWy4o9CcxYdYC7+7RkSIcm5SCcweA5nFgwTYBdIrKeC+dgjJs6j5hZUMPLCs1QDCdTMvjz3M20beLHX4d3KCfhDAbP4UTBvOB2KSozOdmw9QsrqFTdRkVWU1X+b94Wks5l8cm9Pc0eI0O1wElmx+XlIUilZf8SSE2AiOLjvny65iBLdp3g+REd6dDUuKQN1QMTB6CsbJ4FdRrBldcWWWXP8WReXriTAW2DGNs3tPxkMxg8jFEwZeHcadi1CLrcArW8C62SnpXD+FmbbJd0VxPy0lCtKFLBiMgS+/Uf5SdOJWP7V5CTUezWgNe+382uY8m8fks4Qf5mE6OhelHcHExTERkAjBSR2VwU1tKs5MXaGhDU3koLWwjLdp/gP6sPMKZPSwa3Ny5pQ/WjOAXzPFZakRAuDS5lVvKe3A+H18A1LxYaFjMxJYO/zN1Cuyb+PG1c0oZqSnHxYOYB80TkOVX9WznKVDkoJiymqjJx3hbOpmfx6f3GJW2ovjhxU/9NREYCeVuEl6nqt+4Vq4KTm2t5j8IGQr1ml5z+r+2SfuF3HWkfbFzShuqLk8RrrwITsJKu7QAm2GXVl0O/wplDhW4N2HM8mVcW7mRguyDuMbukDdUcJyt5bwAiVDUXQEQ+BjZh5auunmz+HLz9oMOIC4rzXNL+PrV4/WbjkjYYnK6DKbjZsb47BKk0ZKbC9q+h443g7XvBqX98v8tySd/c1bikDQacWTCvAptEZCmWq7o/1dl6WflPyEyGyAszqyzbfYKPVsdxz1WhDGrfuIiLDYbqhZNJ3lkisgzogaVgJqrqMXcLViFJ3Ae/TIXw26FFj/PFBVzSTw1r70EBDYaKhaPMXnaWxQVulqViowrfPQm1fODaSQWKlSfnbuZsehaf3d/LuKQNhgKYvUhO2bkA9v8Mg54B//Orcj/59SBLdyfw12HtaRdcqhx1BkOVxSgYJ2SmwvdPQ5Mu0OP+/OLdx5J5ZdFOBrULYoxxSRsMl1CsghGRGiKyrbyEqbCseB3O/gY3TIGa1qgyzyVdz6cWr5td0gZDoRSrYOy1L5tF5ApXdioij4vIdhHZJiKzRMRHRFqJyFoR2Ssic0Sk8PgH5U3CHvjlbYgYDVf0zi+e/N0udh9P5vVbuhJoUr0aDIXiZIjUFNguIktEZEHeUdoORaQ5MB6IUtXOQE3gduAfwL9UtQ1wGrivtH24jLyJXa+6cM35FN1Ld59g5i+2S7qdcUkbDEXhxIv0UslVStVvHRHJAuoCR7F2Z+fFnfwYeBF4zw19O2fHVxC7DIZPAb8gAJLSsnhy7mbjkjYYHFCiBWPH5I0DvOz364FSx4JR1d+AKcAhLMWSBGwAzqhqtl0tHmhe2PUi8oCIRItIdEJCQmnFKJmMZPj+rxAcDlH35hfPWBVLYkomb9za1bikDYYScLLZcRwwD3jfLmoOfFXaDkWkIXAj0ApoBvgCwwqpqoVdr6ofqGqUqkYFBQWVVoySWf4aJB+BG96AGpYiSUrL4qPVcQzrHEzn5tV7x4TB4AQnczCPAH2BswCquhcoy8TDNcABVU1Q1SzgS+AqoIGI5A3ZQoAjZeijbJzYBWvehW53QYue+cUzVh8gOSOb8UPaeEw0g6Ey4UTBZBRMG2srgUKtC4ccAnqLSF2xfLtDsMJALAVutuuMAb4uQx+lRxUW/cXaLX3Ni/nFSeey+Gj1Aa7vFGzSjhgMDnGiYJaLyF+xJmWvBeYC35S2Q1VdizXk2ghstWX4AJgIPCEi+4AAYEZp+ygT2+ZD3EoY8jz4BuYXf7T6AMnpxnoxGC4HJ16kp7BcxluBB4FFwPSydKqqL3BpxshYoGch1cuPjGT44RloGgGR9+QXJ53LYsaqA1zXsQkdmxnrxWBwipPd1Ll2kKm1WEOj3apaliFSxWXZZEg5Drd/nj+xCzBzdZyxXgyGUlCighGRG4BpwH6scA2tRORBVf3O3cKVK8d3wJr3oPvdEBKZX3w2PYsZq2K5tmMT4zkyGC4TJ0OkN4BBqroPQERaAwuBqqNg8iZ2ferBkAtHbjNXx3E2PZsJxnoxGC4bJwrmRJ5ysYkFTrhJHs+wdS4cXA0j3gTfgPzi5HRr7uWaDsZ6MRhKQ5EKRkRG2W+3i8gi4AusOZhbsFbzVg3Sk2Dxs9A8ErqPueDUx7/EkXQuq0JZL1lZWcTHx5Oenu5pUQzVAB8fH0JCQvDy8irV9cVZML8r8P44MMB+nwA0LFVvFZFlkyHlBNwxG2qc99onp2fx4coDDGnfmC4hFcd6iY+Px9/fn9DQUBMiwuBWVJWTJ08SHx9Pq1atStVGcZkdxxZ1rspwbBusfR+ixkLz7hec+uTXg5b1ck3FsV4A0tPTjXIxlAsiQkBAAGXZ8+fEi9QKeBQILVhfVUeWuteKQP7Ebn0Y/NwFp1IysvlwZSyD2zcmPKRBEQ14DqNcDOVFWb9rTiZ5v8JaVfsNkFum3ioSm2dbGRpHvgV1G11w6uNf4jiTVrHmXgyGyoiTrQLpqjpVVZeq6vK8w+2SuZNzZ+DH56B5FET88YJTqRnZTF8Zy6B2QXRtUfGsl8rCggULmDx58mVfd9VVV7lBGoOncGLB/FtEXgAWAxl5hapa6pgwHmfp3yHtJIyed8HELlhzL6fTsphwTVsPCVc1GDlyJCNHXv4o+pdffnGDNAZP4UTBdAHuwoo4lzdEUvtz5ePoFlj/IUTdB80iLjiVmpHNByv2M6BtEBGVwHp56Zvt7Dhy1qVtdmxWjxd+16nI83FxcVx//fVcffXVrFmzhq5duzJ27FheeOEFTpw4wWeffUbPnj2ZOXMm0dHRvP3228ydO5eXXnqJmjVrUr9+fVasWMH27dsZO3YsmZmZ5ObmMn/+fNq0aYOfnx8pKSksW7aMF198kcDAQLZt20ZkZCSffvopIsKiRYt44oknCAwMpHv37sTGxvLtt99eIGdaWhr33HMPu3btokOHDsTFxfHOO+8QFRXF4sWLeeGFF8jIyKB169Z89NFH+Pn5ERoaypgxY/jmm2/Iyspi7ty5tG/fntTUVB599FG2bt1KdnY2L774IjfeeGOR92A4j5Mh0u+BMFUdoKqD7KNyKpfcXGtit04jGPzMJaf/uybPejFfkuLYt28fEyZMYMuWLezatYvPP/+cVatWMWXKFP7+979fUn/SpEn88MMPbN68mQULrHDO06ZNY8KECcTExBAdHU1ISMgl123atIk333yTHTt2EBsby+rVq0lPT+fBBx/ku+++Y9WqVUV6ON59910aNmzIli1beO6559iwYQMAiYmJvPzyy/z0009s3LiRqKgo/vnPf+ZfFxgYyMaNG/nTn/7ElClTAHjllVcYPHjvefU2AAAT30lEQVQw69evZ+nSpTz55JOkpqY6uofqjhMLZjPQgKqwenfzLDi8Fm58F+pcuJQnLTObD1fE0r9tEN2vqBzLfIqzNNxJq1at6NKlCwCdOnViyJAhiAhdunQhLi7ukvp9+/blnnvu4dZbb2XUKGv9Zp8+fXjllVeIj49n1KhRhf7n79mzZ/6PNiIigri4OPz8/AgLC8tfl3HHHXfwwQcfXHLtqlWrmDBhAgCdO3cmPDwcgDVr1rBjxw769u0LQGZmJn369Mm/Lk++yMhIvvzySwAWL17MggUL8hVOeno6hw4dcnQP1R0nCqYJsEtE1nPhHEzlclOfOw0/Pg8tekHXOy45/emag5xMzTSeIwfUrn0+TUuNGjXyP9eoUYPs7OxL6k+bNo21a9eycOFCIiIiiImJ4c4776RXr14sXLiQoUOHMn36dAYPvtAwLthPzZo1yc7OxulG/qLqqSrXXnsts2bNKvbe8vrLu2b+/Pm0a9fugrodOnQo8R6qO06GSC9gDZP+jrXxMe+oXPz8Cpw7ZWUIuGhiNy0zm/eXx9KvTSCRLSuH9VKZ2L9/P7169WLSpEkEBgZy+PBhYmNjCQsLY/z48YwcOZItW7Y4aqt9+/bExsbmW0pz5swptN7VV1/NF198AcCOHTvYunUrAL1792b16tXs22dtr0tLS2PPnj3F9jl06FDeeuutfKW1adMmgFLfQ3XCSTyYyu2SBjgSA9EzoMc4aBp+yenP1hziZGomj5m5F7fw5JNPsnfvXlSVIUOG0LVrVyZPnsynn36Kl5cXwcHBPP/8847aqlOnDu+++y7XX389gYGB9OxZeIyyhx9+mDFjxhAeHk63bt0IDw+nfv36BAUFMXPmTO644w4yMiyD/OWXX6Zt26K9hs899xyPPfYY4eHhqCqhoaF8++23zJkzp1T3UJ2QkkxOEUnmfAxeb8ALSFVVj4d2i4qK0ujo6OIr5ebCjGvhzCH4f+uhzoXeoXOZOfR77WfaB9fj0/t7uVFa17Bz5046dOjgaTE8SkpKCn5+fqgqjzzyCG3atOHxxx+/oE5OTg5ZWVn4+Piwf/9+hgwZwp49e/D2rhgJQysThX3nRGSDqkaVdK0TC8b/ooZvwtOhLS+HmE/ht2j4/fuXKBeAz9YeJDEl03iOKhEffvghH3/8MZmZmXTr1o0HH3zwkjppaWkMGjSIrKwsVJX33nvPKBcP4GSS9wJU9SsRecodwriFkJ7Q5/9B+G2XnDqXmcO05bH0vTKAHqGNCrnYUBF5/PHHL7FYLsbf358SrVuD23Gy2XFUgY81gCjKlrakfGncHoa+Uuipz9cdIjElg3eHdC/0vMFgKBtOLJiCcWGysdLI3ugWacqR9Kwcpi3fz1WtA+jZylgvBoM7cDIHUyXjwny+9hAJyRm8fUc3T4tiMFRZiguZWZzPTVX1b26Qp1zIs156hzWiV1hAyRcYDIZSUdxCu9RCDrCSsE10s1xuZda6Q5xIzmDCELNjujJy5MgRbr7ZyjIcExPDokWL8s+9+OKL+Uv6y0p0dDTjx4+/7OuGDx/OmTNnXCJDZae4kJn5q3VFxB+YAIwFZlMZV/La5FkvvVo1ok9rY72UBVVFValRw8mCcNfRrFkz5s2bB5C/0XD48OEu7ycqKoqoqBKXelxCQYVX3Sn2myEijUTkZWALljLqrqoTVbVMGx9FpIGIzBORXSKyU0T62H39KCJ77Ve3rNmfs/4wx89mVI11L989BR/d4Nrju+JXIMTFxdGhQwcefvhhunfvzuHDh2nTpg2JiYnk5ubSr18/Fi9enF9v3LhxdOrUieuuu45z584BMHXqVDp27Eh4eDi33377JX0MHz48f9l9t27dmDRpEmCtqJ0+fTpxcXF07tyZzMxMnn/+eebMmUNERET+toEdO3YwcOBAwsLCmDp1aqH34efnx8SJE4mMjOSaa65h3bp1+dfk7fhetmwZI0aMAGD58uVEREQQERFBt27dSE5O5ujRo/Tv35+IiAg6d+7MypUrAQgNDSUxMbHYZ7B+/XrCw8Pp06cPTz75JJ07d75ExtzcXB5++GE6derEiBEjGD58eL5i3bBhAwMGDCAyMpKhQ4dy9OhRAAYOHMjEiRPp2bMnbdu2zZcpJyeHJ598kh49ehAeHs77778PUOQ9uIoiFYyIvI6VniQZ6KKqL6rqaRf1+2/ge1VtD3QFdmLlwF6iqm2AJfZnl5KelcO7y/bRs1Uj+pi5l1Kze/du7r77bjZt2kTLli2ZOHEiDz30EG+88QYdO3bkuuuuA2Dv3r088sgjbN++nQYNGjB//nwAJk+ezKZNm9iyZQvTpk27pP3+/fuzcuVKzp49S61atVi9ejVg7ZDu169ffj1vb28mTZrEbbfdRkxMDLfdZq112rVrFz/88APr1q3jpZdeIisr65I+UlNTGThwIBs2bMDf359nn32WH3/8kf/973+FLvmfMmUK77zzDjExMaxcuZI6derw+eefM3ToUGJiYti8eTMRERGXXFfUMxg7dizTpk3j119/pWbNmpdcB/Dll18SFxfH1q1bmT59Or/++itgpa559NFHmTdvHhs2bODee+/lmWfOhx/Jzs5m3bp1vPnmm7z00ksAzJgxg/r167N+/XrWr1/Phx9+yIEDBxzdQ1kozov0Z6zd088CzxQI/itYk7yl2iogIvWA/sA9WA1lApkiciMw0K72MbAMF8/1fBFtWS//ujWiagTOHnb5ISldQcuWLendu3f+5/vvv5+5c+cybdo0YmJi8stbtWqV/4WNjIzM36AYHh7O6NGjuemmm7jpppsuab9fv35MnTqVVq1accMNN/Djjz+SlpZGXFwc7dq1KzQkREFuuOEGateuTe3atWncuDHHjx+/JFaLt7c3119/PQBdunShdu3aeHl5FRty4oknnmD06NGMGjWKkJAQevTowb333ktWVhY33XRToT/Owp7BmTNnSE5Ozg8Peuedd14SMAsshXrLLbdQo0YNgoODGTRoEGAp+G3btnHttdcClnXStGnT/OsKhpzIu5fFixezZcuWfAsoKSmJvXv3OrqHslCkBaOqNVS1jqr6q2q9Aod/GfchhWHlVvpIRDaJyHQR8QWaqOpRu++jQOPCLhaRB0QkWkSiLyedQkZ2Du8t20+P0IZm7qWM+Pr6XvA5LS2N+Ph4wNonlEdh4RYAFi5cyCOPPMKGDRuIjIy8JMRDjx49iI6OZuXKlfTv359u3brx4YcfEhkZiROK6rcgXl5e+f9knISceOqpp5g+fTrnzp2jd+/e7Nq1i/79+7NixQqaN2/OXXfdxSeffOJIFleEnOjUqRMxMTHExMSwdetWFi9efEmfF4eceOutt/KvOXDgANddd52jeygL5Ts7Z1EL6A68p6rdsLxTjodDqvqBqkapalRQUJDjTr+IjudoUjqPXdO2algvFYiJEycyevRoJk2axLhx44qtm5uby+HDhxk0aBCvvfYaZ86cuUApgWVdtGjRgi+++ILevXvTr18/pkyZcsHwKA9/f3+Sk5Ndej+FsX//frp06cLEiROJiopi165dHDx4kMaNGzNu3Djuu+8+Nm50Fqa6YcOG+Pv7s2bNGgBmz55daL2rr76a+fPnk5uby/Hjx1m2bBkA7dq1IyEh4YIh0/bt24vtc+jQobz33nv5w8U9e/aQmppa6ntwymXvRXIB8UC8qq61P8/DUjDHRaSpqh4Vkaa4MIJeRnYO7y7dR1TLhlxlrBeXsnz5ctavX8/q1aupWbMm8+fP56OPPso35y8mJyeHP/7xjyQlJaGqPP744zRocOkm1H79+rFkyRLq1q1Lv379iI+PL1TBDBo0iMmTJxMREcHTTz/t8vvL480332Tp0qXUrFmTjh07MmzYMGbPns3rr7+Ol5cXfn5+l/Xff8aMGYwbNw5fX18GDhxI/fqXZg/9wx/+wJIlS+jcuTNt27alV69e1K9fH29vb+bNm8f48eNJSkoiOzubxx57jE6dio5weP/99xMXF0f37t1RVYKCgvjqq69YtmxZqe/BEXmuxvI8gJVAO/v9i8Dr9vGUXfYU8FpJ7URGRqoT/vtrnLac+K2u2HPCUf2KzI4dOzwtgsEFJCcn579/9dVXdfz48cXWS0xM1LCwMD169Gi5yFeQwr5zQLQ6+K17woIBK1PkZyLiDcRira+pAXwhIvcBh4BbXNFRZnYu7y7dR2TLhlx9ZaArmjQYyszChQt59dVXyc7OpmXLlsycObPQeiNGjODMmTNkZmby3HPPERwcXL6ClhGPKBhVjcHalX0xQ1zd19wNhzmSlM7kP4SbuRdDheG2227Ld6sXR968S2XFE5O85UqfsAAeu6YN/dpUHetFHXohDIayUtbvWpVXMGFBflXKc+Tj48PJkyeNkjG4HVXl5MmT+Pj4lLoNT83BGEpJSEgI8fHxRSYcMxhciY+PT5kSyhkFU8nw8vLKTzpmMFR0qvwQyWAweA6jYAwGg9swCsZgMLiNEhOvVWREJAE46KBqIJDoZnHKSkWXsaLLBxVfxoouHziXsaWqlrgZsFIrGKeISLQ6yELnSSq6jBVdPqj4MlZ0+cD1MpohksFgcBtGwRgMBrdRXRTMB54WwAEVXcaKLh9UfBkrunzgYhmrxRyMwWDwDNXFgjEYDB7AKBiDweA2qryCEZHrRWS3iOwTEZenQimFPC1EZKmdD2q7iEywy8slL9RlylrTDsz+rf25lYistWWcYwcM85RsHsutdRkyPm7/jbeJyCwR8fH0MxSR/4jICRHZVqCs0OcmFlPt384WEel+uf1VaQUjIjWBd4BhQEfgDhHp6FmpyAb+rKodgN7AI7ZMbs8LVQomYOWsyuMfwL9sGU9jpRH2FB7JreUUEWkOjAeiVLUzUBO4Hc8/w5nA9ReVFfXchgFt7OMB4L3L7s1JXM3KegB9gB8KfH4aeNrTcl0k49fAtcBuoKld1hTY7WG5Quwv22DgW6x8WIlArcKebTnLVg84gO2kKFBeYZ4h0Bw4DDTCilrwLTC0IjxDIBTYVtJzA94H7iisntOjSlswnP8j5xFvl1UIRCQU6AasxWFeqHLkTeD/gFz7cwBwRlXzkgZ58lmWKbdWeaCqvwFTsOJLHwWSgA1UnGdYkKKeW5l/P1VdwRQWxq5C+OVFxA+YDzymqmc9LU9BRGQEcEJVNxQsLqSqp55lmXJrlQf2PMaNQCugGeCLNeS4mArxfSyCMv/Nq7qCiQdaFPgcAhzxkCz5iIgXlnL5TFW/tIuP2/mgcHVeqFLQFxgpInHAbKxh0ptAAxHJC1LmyWdZWG6t7lSsZ3gNcEBVE1Q1C/gSuIqK8wwLUtRzK/Pvp6ormPVAG3vm3htrkm2BJwUSKzjwDGCnqv6zwKkFwBj7/RisuRmPoKpPq2qIqoZiPbOfVXU0sBS42a7mMRlV9RhwWETa2UVDgB1UoGeINTTqLSJ17b95nowV4hleRFHPbQFwt+1N6g0k5Q2lHOOpSbBynNAaDuwB9gPPVAB5rsYyM7cAMfYxHGuOYwmw135t5GlZbXkHAt/a78OAdcA+YC5Q24NyRQDR9nP8CmhY0Z4h8BKwC9gG/Beo7elnCMzCmhPKwrJQ7ivquWENkd6xfztbsTxil9Wf2SpgMBjcRlUfIhkMBg9iFIzBYHAbRsEYDAa3YRSMwWBwG0bBGAwGt2EUjMFgcBtGwRgMBrdhFEw1R0RSyqGPOiKy3A6f4cp2SyW7iHiLyIoCS/YNbsIoGEN5cC/wparmeFoQAFXNxFqxepunZanqGAVjyEdEnrCjr20TkccKlD9nR4770Y7M9pfLbHo0BfbciMgfRWSdiMSIyPt25Lx/iMjDBeq8KCJ/tt9/JSIb7OhwDxQid+hFEdr+IiIvFtWXXe0rWy6DGzEKxgCAiEQCY4FeWJH2xolINxGJAv6AFbdmFHBZWf/sTaZhqhpnf+6AZTn0VdUIIAfrhz6bCy2KW7H26gDcq6qRdt/jRSTAYd9F9QXW/qAel3MvhsvHjEENeVwN/E9VUwFE5EugH9Y/oa9V9Zxd/o39GgY8A9RX1ZvtMl/gXSATWKaqn2HlOj5ToJ8hQCSw3tpkTB2s2DOfiEhjEWkGBAGnVfWQfc14Efm9/b4FVgjHkw7uqdC+AFQ1R0QyRcRfVZMv4zkZLgOjYKoZIvIIMM7+OLzgqaIuKaxQVWOB+0RkXoHiUcA8Vf1GROYAnwHnAJ+L2vtYVZ8upNl5WKEMgrEsGkRkIFZslT6qmiYiyy5qD6w4xwWt8bzzxfUF1u7m9CLOGVyAGSJVM1T1HVWNsI+CwYNWADfZ8Ut8gd8DK4FVwO/siPh+wA3FNB/C+RCLOXZ/p4GaIpL3o18C3CwijSE/on1L+9xsrPgzN2MpG4D6WNZMmoi0xxq+XcxxoLGIBIhIbWBESX3Zw6y8YFAGN2EsGAMAqrpRRGZixSoBmK6qmwBEZAGwGTiIFYMlqYhm4rGUTAwX/vNajDUE+0lVd4jIs8BiEamBFZfkEeCgqm4XEX/gNz0f2Oh74CER2YIVdHpNIbJnicgkrNjGB7BisFBcX8AgYNHlPCPD5WPiwRhKRET8VDVFROpiWToPYP1IX8HKiDBdVV+1LZ+3sYYdq+w5GESkG/CEqt7lmTu4FHuO6WlV3e1pWaoyRsEYSkREPsfKK+WDNafxainauNe+1uNrYfLCp6rqJ56WpapjFIzBYHAbZpLXYDC4DaNgDAaD2zAKxmAwuA2jYAwGg9swCsZgMLgNo2AMBoPbMArGYDC4jf8PEWl6Rf4gcx8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test different evalues\n",
    "evalues =  np.arange(0,110,10,dtype=float)\n",
    "missing_genes = list()\n",
    "rxns_with_missing_genes = list()\n",
    "for evalue in evalues:\n",
    "    model_from_template2 = update_missing_genes(model_from_template,10**(-evalue))\n",
    "    missing_genes.append(len([gene for gene in model_from_template2.genes if 'missing' in gene.id]))\n",
    "    rxns_with_missing_genes.append(len([rxn for rxn in model_from_template2.reactions if 'missing' in rxn.gene_reaction_rule]))\n",
    "plt.figure(figsize=(4,3))\n",
    "plt.plot(evalues,missing_genes,label='missing genes')\n",
    "plt.plot(evalues,rxns_with_missing_genes,label='rxns with missing genes')\n",
    "plt.xlabel('-log$_{10}$(evalue)')\n",
    "plt.ylabel('Number of missing genes')\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1235\n",
      "Number of metabolits: 1208\n",
      "Number of compartments: 3\n",
      "Number of genes: 825\n",
      "Number of missing genes: 96\n",
      "Number of reactions with missing genes: 116\n"
     ]
    }
   ],
   "source": [
    "model_from_template2 = update_missing_genes(model_from_template,1e-20,True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions with MetaCyc id: 541 / 1235\n"
     ]
    }
   ],
   "source": [
    "k = 0\n",
    "for rxn in model_from_template2.reactions:\n",
    "    meta_id = rxn.annotation.get('biocyc')\n",
    "    if meta_id is not None:\n",
    "        k += 1\n",
    "print('Number of reactions with MetaCyc id:',k,'/',len(model_from_template2.reactions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "and or rxns: 28/1235\n",
      "reactions with missing genes: 116/1235\n"
     ]
    }
   ],
   "source": [
    "and_or_rxns = 0\n",
    "rxns_with_missing_genes = 0\n",
    "for rxn in model_from_template2.reactions:\n",
    "    gr = rxn.gene_reaction_rule\n",
    "    if ' and ' in gr and ' or ' in gr: \n",
    "        and_or_rxns += 1\n",
    "    if '_missing' in gr: rxns_with_missing_genes += 1\n",
    "        \n",
    "print('and or rxns: {0}/{1}'.format(and_or_rxns,len(model_from_template2.reactions)))\n",
    "print('reactions with missing genes: {0}/{1}'.format(rxns_with_missing_genes,len(model_from_template2.reactions)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1119\n",
      "Number of metabolits: 1099\n",
      "Number of compartments: 3\n",
      "Number of genes: 662\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "# remove reactions with missing genes from the draft model\n",
    "model_from_template3 = remove_rxns_with_missing_genes(model_from_template2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_json_model(model_from_template2,\n",
    "                        '../Results/halo_iML1515_template_with_missing_genes.json')\n",
    "cobra.io.save_json_model(model_from_template3,\n",
    "                    '../Results/halo_iML1515_template_without_missing_genes.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All models and templates we have now\n",
    "* draft model reconstructed from MetaCyc by RAVEN\n",
    "* draft model reconstructed from KEGG by RAVEN\n",
    "* draft model reconstructed from template iML1515<br>\n",
    "\n",
    "<br>Next step, we are going to merge the model from MetaCyc and iML1515. We decided not using the draft model from KEGG due to following reasons:\n",
    "* 1. MetaCyc contains experimentally elucidated pathways only;, while KEGG contains ortholog groups based annotations\n",
    "* 2. The reactions from MetaCyc is charge balanced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "halo_meta = '../Results/halo_metacycmodel_100.mat'\n",
    "halo_kegg = '../Results/halo_keggmodel_50.mat'\n",
    "halo_iJO = '../Results/halo_iML1515_template_without_missing_genes.json'\n",
    "ecoli_sbml = '../../../ComplementaryData/iML1515.xml'\n",
    "ecoli_mat = '../../../ComplementaryData/iML1515.mat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_metacyc_mat(model_file):\n",
    "    # model_file: a mat file of the model. This model is the model that is reconstructed from MetaCyc with RAVEN.\n",
    "    # return a dictionary, with filed name as key, field values as value\n",
    "    # like data['id'] = 'metacycmodel'\n",
    "    #\n",
    "    \n",
    "    ###### \n",
    "    # load mat model with cobrapy\n",
    "    model = cobra.io.load_matlab_model(model_file)\n",
    "    \n",
    "    \n",
    "    ###### \n",
    "    # load mnx of metabolites\n",
    "    meta2mnx = dict()\n",
    "    dfmnx = pd.read_csv('../../../ComplementaryData/MNX2metacyc.tsv',index_col=1,sep='\\t')\n",
    "\n",
    "    for ind in dfmnx.index:meta2mnx[ind] = meta2mnx.get(ind,[]) + [dfmnx.loc[ind,'MNX']]\n",
    "    \n",
    "    \n",
    "    ######\n",
    "    # load annotation fields\n",
    "    data = dict()\n",
    "    \n",
    "    mat = mat = loadmat(model_file)\n",
    "    meta_vars = {\"__globals__\", \"__header__\", \"__version__\"}\n",
    "    for item in mat.keys(): \n",
    "        if item not in meta_vars: model_id = item\n",
    "    \n",
    "    # field names\n",
    "    names = mat[model_id].dtype.names\n",
    "    \n",
    "    # save data into a dictionary\n",
    "    for i,item in enumerate(mat[model_id][0][0]):\n",
    "        if names[i] in ['rxns','rxnNames','eccodes','rxnReferences','mets','metNames','metFormulas','inchis']:\n",
    "            data[names[i]] = [i[0][0] if len(i[0])!=0 else '' for i in item]\n",
    "        \n",
    "        elif names[i] in ['subSystems']:\n",
    "            lst = []\n",
    "            for sub_item in item:\n",
    "                try: lst.append([k[0] for k in sub_item[0][0]])\n",
    "                except: lst.append([])\n",
    "            data[names[i]] = lst\n",
    "        elif names[i] in ['rxnMiriams','metMiriams']:\n",
    "            lst = []\n",
    "            for sub_item in item:\n",
    "                anno = dict()\n",
    "                try:\n",
    "                    sub_item = sub_item[0][0][0]\n",
    "                    for j in range(len(sub_item['name'])):\n",
    "                        name = sub_item['name'][j][0][0]\n",
    "                        value = sub_item['value'][j][0][0]\n",
    "                        anno[name] = str(value)\n",
    "                except:pass\n",
    "                lst.append(anno)\n",
    "            data[names[i]] = lst\n",
    "        elif names[i] in ['rxnConfidenceScores','metCharges']:\n",
    "            data[names[i]] = [k[0] for k in item]\n",
    "        else: pass\n",
    "    \n",
    "    print('Loaded fields:',data.keys())\n",
    "    \n",
    "    ######\n",
    "    # there are some metabolites formula cannot be correctly parsed by cobrapy. like ZN1 from RAVEN, which will\n",
    "    # be treated as Z1N1. Most of such metabolites are ions\n",
    "    for i,item in enumerate(data['metFormulas']):\n",
    "        for element in ['ZN','MN','CU','MG','FE','CL','BR','AS','CA','CD','MO','TE','SE','NA']: \n",
    "            if element in item: item = item.replace(element,element.lower().title())\n",
    "            data['metFormulas'][i] = item\n",
    "    \n",
    "    ######\n",
    "    # assign annotation to the cobrapy model\n",
    "    ## 1. assign rxn annotation and reaction subsystem\n",
    "    \n",
    "    for i in range(len(data['rxns'])):\n",
    "        rxn_id = data['rxns'][i]\n",
    "        anno = data['rxnMiriams'][i]\n",
    "        \n",
    "        # ec number\n",
    "        ec = data['eccodes'][i]\n",
    "        ec = ec.replace('ec-code/','')\n",
    "\n",
    "        if len(ec) == 0: pass\n",
    "        elif len(ec.split(';')) == 1: anno['ec-code'] = ec\n",
    "        else: anno['ec-code'] = ec.split(';')\n",
    "        \n",
    "        # metacyc id\n",
    "        anno['biocyc'] = 'META:'+rxn_id\n",
    "        \n",
    "        rxn = model.reactions.get_by_id(rxn_id)\n",
    "        rxn.annotation = anno\n",
    "        \n",
    "        # assign reaction subsystem\n",
    "        subss = ''\n",
    "        for item in data['subSystems'][i]: \n",
    "            subss += item +';'\n",
    "        if subss.endswith(';'): subss = subss[:-1]\n",
    "        rxn.subsystem = subss\n",
    "    \n",
    "    ## assign metabolites annotation\n",
    "    for i in range(len(data['mets'])):\n",
    "        met_id = data['mets'][i]\n",
    "        anno = data['metMiriams'][i]\n",
    "        \n",
    "        # assign metacyc id\n",
    "        anno['biocyc'] = 'META:'+met_id\n",
    "        \n",
    "        # assign mnx id\n",
    "        mnx_id = meta2mnx.get(met_id)\n",
    "        if mnx_id is not None: anno['metanetx.chemical'] = mnx_id\n",
    "        \n",
    "        met = model.metabolites.get_by_id(met_id)\n",
    "        met.annotation = anno\n",
    "        \n",
    "        # assign charge\n",
    "        met.charge = data['metCharges'][i]\n",
    "        \n",
    "        # assign formula\n",
    "        met.formula = data['metFormulas'][i]\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using license file /Users/gangl/gurobi.lic\n",
      "Academic license - for non-commercial use only\n",
      "Loaded fields: dict_keys(['rxns', 'rxnNames', 'eccodes', 'subSystems', 'rxnMiriams', 'rxnReferences', 'rxnConfidenceScores', 'mets', 'metNames', 'metFormulas', 'metCharges', 'inchis', 'metMiriams'])\n"
     ]
    }
   ],
   "source": [
    "halo_meta = load_metacyc_mat(halo_meta)\n",
    "halo_kegg = cobra.io.load_matlab_model(halo_kegg)\n",
    "halo_iJO = cobra.io.load_json_model(halo_iJO)\n",
    "eco = load_ecoli_model(ecoli_mat,ecoli_sbml)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for rxn in halo_iJO.reactions: print(rxn.annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "halo_meta.solver = 'gurobi'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_keggrxn(rxn):\n",
    "    eq = rxn.reaction\n",
    "    for met in rxn.metabolites:\n",
    "        eq = eq.replace(met.id,met.name)\n",
    "    print('KEGG:',eq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def report_model_status(model):\n",
    "    print('Number of reactions:',len(model.reactions))\n",
    "    print('Number of metabolits:',len(model.metabolites))\n",
    "    print('Number of compartments:',len(model.compartments))\n",
    "    k = 0\n",
    "    for gene in model.genes:\n",
    "        if 'missing' in gene.id: k += 1\n",
    "    print('Number of genes:',len(model.genes))\n",
    "    print('Number of missing genes:',k)\n",
    "    \n",
    "    k = 0\n",
    "    for rxn in model.reactions:\n",
    "        if '_missing' in rxn.gene_reaction_rule: k += 1\n",
    "    print('Number of reactions with missing genes:',k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1462\n",
      "Number of metabolits: 1639\n",
      "Number of compartments: 1\n",
      "Number of genes: 852\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_kegg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1271\n",
      "Number of metabolits: 1520\n",
      "Number of compartments: 1\n",
      "Number of genes: 1230\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1119\n",
      "Number of metabolits: 1099\n",
      "Number of compartments: 3\n",
      "Number of genes: 662\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_iJO)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Check all metabolites ends with -GLU-N. Re-associate these metabolites to non-GLU-N format\n",
    "We identified that some metabolites are in the form of -GLU-N(poly-glutamine). After carefully check the MetaCyc website and cross-referenced to KEGG database, N in those metabolites should be 0. Then we replaced those ones in -GLU-N format with non-GLU-N format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5747, 1) Index(['kegg'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# load all metacyc-kegg reaction pair\n",
    "dfrxnpair = pd.read_csv('../../../ComplementaryData/metacyc2keggrxns.tsv',sep='\\t',index_col=0)\n",
    "print(dfrxnpair.shape,dfrxnpair.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5401, 1) Index(['metacyc'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# load metacyc-kegg metabolites pairs\n",
    "dfmetpair = pd.read_csv('../../../ComplementaryData/metaCycMets_KEGG_pair.tsv',sep='\\t',index_col=1)\n",
    "print(dfmetpair.shape,dfmetpair.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5-10-METHENYL-THF-GLU-N\n",
      "5-10-METHENYL-THF-GLU-N + WATER <=> FORMYL-THF-GLU-N + PROTON\n",
      "KEGG: H2O + 5,10-Methenyltetrahydrofolate <=> H+ + 10-Formyltetrahydrofolate\n",
      "C00001 + C00445 <=> C00080 + C00234\n",
      "\n",
      "5-METHYL-THF-GLU-N\n",
      "5-METHYL-THF-GLU-N + HOMO-CYS --> MET + THF-GLU-N\n",
      "KEGG: L-Homocysteine + 5-Methyltetrahydrofolate <=> L-Methionine + Tetrahydrofolate\n",
      "C00155 + C00440 <=> C00073 + C00101\n",
      "\n",
      "DIHYDROFOLATE-GLU-N\n",
      "DIHYDROFOLATE-GLU-N + NADPH + PROTON --> NADP + THF-GLU-N\n",
      "KEGG: NADP+ + Tetrahydrofolate <=> NADPH + H+ + Dihydrofolate\n",
      "C00006 + C00101 <=> C00005 + C00080 + C00415\n",
      "\n",
      "FORMYL-THF-GLU-N\n",
      "5-PHOSPHO-RIBOSYL-GLYCINEAMIDE + FORMYL-THF-GLU-N <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE + PROTON + THF-GLU-N\n",
      "KEGG: 10-Formyltetrahydrofolate + 5'-Phosphoribosylglycinamide <=> Tetrahydrofolate + 5'-Phosphoribosyl-N-formylglycinamide\n",
      "C00234 + C03838 <=> C00101 + C04376\n",
      "\n",
      "METHYLENE-THF-GLU-N\n",
      "AMINOMETHYLDIHYDROLIPOYL-GCVH + THF-GLU-N <=> AMMONIUM + DIHYDROLIPOYL-GCVH + METHYLENE-THF-GLU-N\n",
      "KEGG: Tetrahydrofolate + S-Aminomethyldihydrolipoylprotein <=> Ammonia + 5,10-Methylenetetrahydrofolate + Dihydrolipoylprotein\n",
      "C00101 + C01242 <=> C00014 + C00143 + C02972\n",
      "\n",
      "N5-Formyl-THF-Glu-N\n",
      "5-10-METHENYL-THF-GLU-N + WATER --> N5-Formyl-THF-Glu-N + PROTON\n",
      "KEGG: H2O + 5,10-Methenyltetrahydrofolate --> Folinic acid\n",
      "C00001 + C00445 --> C03479\n",
      "\n",
      "THF-GLU-N\n",
      "AMINOMETHYLDIHYDROLIPOYL-GCVH + THF-GLU-N <=> AMMONIUM + DIHYDROLIPOYL-GCVH + METHYLENE-THF-GLU-N\n",
      "KEGG: Tetrahydrofolate + S-Aminomethyldihydrolipoylprotein <=> Ammonia + 5,10-Methylenetetrahydrofolate + Dihydrolipoylprotein\n",
      "C00101 + C01242 <=> C00014 + C00143 + C02972\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# For each of metabolites that endswith '-GLU-N', find one reaction that are from metacyc model \n",
    "# which has been associated with a kegg reaction\n",
    "for met in halo_meta.metabolites:\n",
    "    if met.id.upper().endswith('-GLU-N'): \n",
    "        print(met.id)\n",
    "        for rxn in met.reactions:\n",
    "            try: \n",
    "                kegg_id = dfrxnpair.loc[rxn.id,'kegg']\n",
    "                kegg_rxn = halo_kegg.reactions.get_by_id(kegg_id)\n",
    "                break\n",
    "            except: continue\n",
    "        print(rxn.reaction)\n",
    "        print_keggrxn(kegg_rxn)\n",
    "        print(kegg_rxn.reaction)\n",
    "        print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually check above reactions, associate each of '*-GLU-N' metabolites to a kegg id \n",
    "# based on metacyc-kegg reaction pairs\n",
    "GluNs = {\n",
    "    '5-10-METHENYL-THF-GLU-N': 'C00445',\n",
    "    '5-METHYL-THF-GLU-N': 'C00440',\n",
    "    'DIHYDROFOLATE-GLU-N':'C00415',\n",
    "    'FORMYL-THF-GLU-N':'C00234',\n",
    "    'METHYLENE-THF-GLU-N': 'C00143',\n",
    "    'N5-Formyl-THF-Glu-N':'C03479',\n",
    "    'THF-GLU-N': 'C00101'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5-10-METHENYL-THF-GLU-N C00445 5-10-METHENYL-THF\n",
      "5-METHYL-THF-GLU-N C00440 5-METHYL-THF\n",
      "DIHYDROFOLATE-GLU-N C00415 DIHYDROFOLATE\n",
      "FORMYL-THF-GLU-N C00234 10-FORMYL-THF\n",
      "METHYLENE-THF-GLU-N C00143 METHYLENE-THF\n",
      "N5-Formyl-THF-Glu-N C03479 5-FORMYL-THF\n",
      "THF-GLU-N C00101 THF\n"
     ]
    }
   ],
   "source": [
    "# Reassociate those GLU-N to non-GLU metabolites in metacyc\n",
    "for metg,kegg_id in GluNs.items():\n",
    "    print(metg,kegg_id,dfmetpair.loc[kegg_id,'metacyc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_rxn(model,old_met_id,new_met_id):\n",
    "    old_met = model.metabolites.get_by_id(old_met_id)\n",
    "    for old_rxn in old_met.reactions:\n",
    "        \n",
    "        # create a new reaction with replaced metabolite\n",
    "        new_rxn = cobra.Reaction(old_rxn.id)\n",
    "        new_rxn.name = old_rxn.name\n",
    "        new_rxn.subsystem = old_rxn.subsystem\n",
    "        new_rxn.lower_bound = old_rxn.lower_bound\n",
    "        new_rxn.upper_bound = old_rxn.upper_bound\n",
    "        \n",
    "        new_met = cobra.Metabolite(new_met_id,name=new_met_id,compartment=old_met.compartment)\n",
    "        \n",
    "        coeffs = dict()\n",
    "        for met in old_rxn.metabolites:\n",
    "            if met.id == old_met.id: coeffs[new_met] = old_rxn.get_coefficient(met)\n",
    "            else: coeffs[met] = old_rxn.get_coefficient(met)\n",
    "        new_rxn.add_metabolites(coeffs)\n",
    "        new_rxn.gene_reaction_rule = old_rxn.gene_reaction_rule \n",
    "        \n",
    "        # remove old reactions and add new reaction\n",
    "        model.remove_reactions([old_rxn])\n",
    "        model.add_reaction(new_rxn)\n",
    "    \n",
    "    # findally, remove old metabolites\n",
    "    model.remove_metabolites([old_met])\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace the all metabolites that endswith -GLU-N with non-GLU-N format\n",
    "for met in halo_meta.metabolites:\n",
    "    if met.id.upper().endswith('-GLU-N'):\n",
    "        kegg_id = GluNs[met.id]\n",
    "        new_id = dfmetpair.loc[kegg_id,'metacyc']\n",
    "        halo_meta = update_rxn(halo_meta,met.id,new_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "METHYLENE-THF + NADP <=> 5-10-METHENYL-THF + NADPH\n",
      "5-10-METHENYL-THF + WATER --> 5-FORMYL-THF + PROTON\n",
      "5-10-METHENYL-THF + WATER <=> 10-FORMYL-THF + PROTON\n",
      "5-FORMYL-THF + ATP --> 5-10-METHENYL-THF + ADP + Pi\n",
      "SS-DIMETHYL-BETA-PROPIOTHETIN + THF --> 5-METHYL-THF + CPD-7672 + PROTON\n",
      "5-METHYL-THF + HOMO-CYS --> MET + THF\n",
      "METHYLENE-THF + NADH + PROTON --> 5-METHYL-THF + NAD\n",
      "DUMP + METHYLENE-THF --> DIHYDROFOLATE + TMP\n",
      "DIHYDROFOLATE + NADPH + PROTON --> NADP + THF\n",
      "Acceptor + GLY + GTP + METHYLENE-THF + WATER + tRNA-uridine34 --> DIHYDROFOLATE + Donor-H2 + GDP + 2.0 PROTON + Pi + tRNA-containing-5-CarbMeAminome-uridine\n",
      "7-8-DIHYDROPTEROATE + ATP + GLT --> ADP + DIHYDROFOLATE + PROTON + Pi\n",
      "Acceptor + GLY + GTP + METHYLENE-THF + WATER + tRNA-2-thiouridine34 --> DIHYDROFOLATE + Donor-H2 + GDP + 2.0 PROTON + Pi + tRNA-containing-5-CarbMeAminome-2-ThioU\n",
      "10-FORMYL-THF + L-methionyl-tRNAfmet --> N-formyl-L-methionyl-tRNAfmet + PROTON + THF\n",
      "10-FORMYL-THF + AICAR <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE + THF\n",
      "5-10-METHENYL-THF + WATER <=> 10-FORMYL-THF + PROTON\n",
      "10-FORMYL-THF + WATER --> FORMATE + PROTON + THF\n",
      "10-FORMYL-THF + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE + PROTON + THF\n",
      "Acceptor + GLY + GTP + METHYLENE-THF + WATER + tRNA-uridine34 --> DIHYDROFOLATE + Donor-H2 + GDP + 2.0 PROTON + Pi + tRNA-containing-5-CarbMeAminome-uridine\n",
      "METHYLENE-THF + NADH + PROTON --> 5-METHYL-THF + NAD\n",
      "METHYLENE-THF + NADP <=> 5-10-METHENYL-THF + NADPH\n",
      "GLY + NAD + THF <=> AMMONIUM + CARBON-DIOXIDE + METHYLENE-THF + NADH\n",
      "Acceptor + GLY + GTP + METHYLENE-THF + WATER + tRNA-2-thiouridine34 --> DIHYDROFOLATE + Donor-H2 + GDP + 2.0 PROTON + Pi + tRNA-containing-5-CarbMeAminome-2-ThioU\n",
      "SER + THF <=> GLY + METHYLENE-THF + WATER\n",
      "DUMP + METHYLENE-THF --> DIHYDROFOLATE + TMP\n",
      "2-KETO-ISOVALERATE + METHYLENE-THF + WATER <=> 2-DEHYDROPANTOATE + THF\n",
      "AMINOMETHYLDIHYDROLIPOYL-GCVH + THF <=> AMMONIUM + DIHYDROLIPOYL-GCVH + METHYLENE-THF\n",
      "5-10-METHENYL-THF + WATER --> 5-FORMYL-THF + PROTON\n",
      "5-FORMYL-THF + ATP --> 5-10-METHENYL-THF + ADP + Pi\n",
      "10-FORMYL-THF + L-methionyl-tRNAfmet --> N-formyl-L-methionyl-tRNAfmet + PROTON + THF\n",
      "DIHYDROFOLATE + NADPH + PROTON --> NADP + THF\n",
      "10-FORMYL-THF + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE + PROTON + THF\n",
      "GLY + NAD + THF <=> AMMONIUM + CARBON-DIOXIDE + METHYLENE-THF + NADH\n",
      "SER + THF <=> GLY + METHYLENE-THF + WATER\n",
      "2-KETO-ISOVALERATE + METHYLENE-THF + WATER <=> 2-DEHYDROPANTOATE + THF\n",
      "10-FORMYL-THF + AICAR <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE + THF\n",
      "AMINOMETHYLDIHYDROLIPOYL-GCVH + THF <=> AMMONIUM + DIHYDROLIPOYL-GCVH + METHYLENE-THF\n",
      "5-METHYL-THF + HOMO-CYS --> MET + THF\n",
      "10-FORMYL-THF + WATER --> FORMATE + PROTON + THF\n",
      "SS-DIMETHYL-BETA-PROPIOTHETIN + THF --> 5-METHYL-THF + CPD-7672 + PROTON\n"
     ]
    }
   ],
   "source": [
    "# print all changed reactions to have a final manual check\n",
    "for metg,kegg_id in GluNs.items():\n",
    "    new_id = dfmetpair.loc[kegg_id,'metacyc']\n",
    "    met = halo_meta.metabolites.get_by_id(new_id)\n",
    "    for rxn in met.reactions: print(rxn.reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manually curate metabolites ids\n",
    "Since some metabolites have multiple metacyc ids. In the case that one metabolites has more than one metacyc id existing in the model, remove one of them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fe4S4-Cluster-Protein + Octanoylated-domains + 6.0 PROTON + 2.0 Reduced-ferredoxins + 2.0 S-ADENOSYLMETHIONINE --> 2.0 CH33ADO + Dihydro-Lipoyl-Proteins + 4.0 FE+2 + 2.0 HS + Iron-Sulfur-Cluster-Scaffold-Proteins + 2.0 MET + 2.0 Oxidized-ferredoxins\n"
     ]
    }
   ],
   "source": [
    "# 'HS' and 'CPD-7046' are both in the halo_meta model. \n",
    "# They represent the same metabolit with mnx id: 'MNXM89582'\n",
    "# Replace 'CPD-7046' by 'HS'\n",
    "rxns = list()\n",
    "old_met = halo_meta.metabolites.get_by_id('CPD-7046' )\n",
    "for rxn in old_met.reactions:\n",
    "    rxn1 = rxn.copy()\n",
    "    rxn1.subtract_metabolites(rxn1.metabolites)\n",
    "    coeffs = dict()\n",
    "    for met in rxn.metabolites:\n",
    "        coeff = rxn.get_coefficient(met)\n",
    "        if met.id == 'CPD-7046':\n",
    "            met = halo_meta.metabolites.get_by_id('HS')\n",
    "        coeffs[met] = coeff\n",
    "    \n",
    "    rxn1.add_metabolites(coeffs)\n",
    "    print(rxn1.reaction)\n",
    "    rxns.append(rxn1)\n",
    "halo_meta.remove_reactions(old_met.reactions)\n",
    "halo_meta.add_reactions(rxns)\n",
    "halo_meta.repair()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find all transport reactions in metacyc model. Save all these reactions into a new model. Remove those reactions from metacyc model\n",
    "Sincd there is no compartment information in the reaction, the metabolites transported are missing in the reaction because they exit in both side of the reaction. We will manually check the metacyc website for those transporters and then add them back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 RXN0-21 glutathione exporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "2 RXN-9615 ammonium transporter :  --> \n",
      "3 TRANS-RXN0-479 selenite ABC transporter;selenite  ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "4 ABC-70-RXN sulfate  ABC transporter;sulfate ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "5 ABC-27-RXN transport of phosphate : ATP + WATER --> ADP + PROTON + Pi\n",
      "6 TRANS-RXN0-286 lipid II flippase :  --> \n",
      "7 TRANS-RXN0-268 Transport of L-Ala-D-Glu-meso-A2pm : ATP + WATER --> ADP + PROTON + Pi\n",
      "8 TRANS-RXN-355 n-hexane efflux pump :  --> \n",
      "9 RXN0-11 transport of glutathione : ATP + WATER --> ADP + PROTON + Pi\n",
      "10 TRANS-RXN-167 N-acetyl-D-galactosamine PTS;N-acetylglucosamine PTS permease;N-acetyl-D-glucosamine PTS permease : Hpr-pi-phospho-L-histidines + N-acetyl-D-glucosamine --> Hpr-Histidine + N-ACETYL-D-GLUCOSAMINE-6-P\n",
      "11 ABC-24-RXN putrescine/spermidine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "12 TRANS-RXN-352 deoxycholate efflux pump :  --> \n",
      "13 TRANS-RXN-296 cobinamide ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "14 RXN-19739 glutaconyl-CoA decarboxylase : GLUTACONYL-COA + PROTON --> CARBON-DIOXIDE + CROTONYL-COA\n",
      "15 TRANS-RXN0-513 lanthionine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "16 TRANS-RXN0-244 Cd2+:H+ antiporter :  <=> \n",
      "17 TRANS-RXN-42 K+ : H+ antiporter KefB;K+ : H+ antiporter KefC;K+:H+ antiporter;Na+/K+:H+ antiporter ChaA :  <=> \n",
      "18 TRANS-RXN-365 deoxycholate efflux pump :  --> \n",
      "19 RXN0-2501 chloride:H+ antiporter;chloride:H+ antiporter ClcA :  <=> \n",
      "20 TRANS-RXN-273 nitroimidazole export :  --> \n",
      "21 TRANS-RXN-127 L-tartrate:succinate antiporter :  <=> \n",
      "22 TRANS-RXN0-510 N-acetyl-DL-methionine transport : ATP + WATER --> ADP + PROTON + Pi\n",
      "23 TRANS-RXN0-492 transport of beta-D-galactofuranose : ATP + WATER --> ADP + PROTON + Pi\n",
      "24 ABC-36-RXN transport of L-valine : ATP + WATER --> ADP + PROTON + Pi\n",
      "25 TRANS-RXN0-447 transport of adenine :  <=> \n",
      "26 ABC-16-RXN maltose ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "27 ATPSYN-RXN ATP synthase;H+-translocating V-ATPase;ATP hydrolase : ATP + WATER <=> ADP + PROTON + Pi\n",
      "28 TRANS-RXN-347 norfloxacin:Na+ antiport :  <=> \n",
      "29 RXN0-3 L-cysteine exporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "30 TRANS-RXN-312 L-phenylalanine transport;transport of L-phenylalanine : ATP + WATER --> ADP + PROTON + Pi\n",
      "31 RXN0-18 inosine:H+ antiporter :  <=> \n",
      "32 TRANS-RXN-242 L-homoserine:H+ antiporter :  <=> \n",
      "33 ABC-13-RXN transport of L-glutamate : ATP + WATER --> ADP + PROTON + Pi\n",
      "34 TRANS-RXN0-200 Zn2+:H+ antiporter :  <=> \n",
      "35 RXN0-5111 glycolate:cation symporter :  <=> \n",
      "36 TRANS-RXN-359 ethidium efflux pump :  --> \n",
      "37 TRANS-RXN-325 canavanine exporter :  --> \n",
      "38 RXN0-1702 transport of ferric coprogen :  --> \n",
      "39 TRANS-RXN0-593 L-cystine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "40 TRANS-RXN-122A transport of L-aspartate;L-aspartate:H+ symporter;aspartate transport :  <=> \n",
      "41 TRANS-RXN0-503 maltotriose ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "42 ABC-8-RXN Transport of a dipeptide : ATP + WATER --> ADP + PROTON + Pi\n",
      "43 ABC-10-RXN transport of ferric enterobactin : ATP + WATER --> ADP + PROTON + Pi\n",
      "44 TRANS-RXN0-269 L-valine exporter :  <=> \n",
      "45 TRANS-RXN0-222 transport of L-aspartate : ATP + WATER --> ADP + PROTON + Pi\n",
      "46 TRANS-RXN0-0244 L-threonine:H+ antiport;L-threonine:H+ antiporter :  <=> \n",
      "47 TRANS-RXN-356 decanoate efflux pump :  --> \n",
      "48 TRANS-RXN-340 bicozamycin:H+ antiporter :  <=> \n",
      "49 RXN0-1981 acetate:cation symporter :  <=> \n",
      "50 TRANS-RXN-167A D-glucosamine PTS permease : GLUCOSAMINE + Hpr-pi-phospho-L-histidines --> D-GLUCOSAMINE-6-P + Hpr-Histidine\n",
      "51 RXN-5076 xanthine:H+ symporter XanQ;xanthine:H+ symporter XanP :  <=> \n",
      "52 ABC-25-RXN putrescine/spermidine ABC transporter;transport of putresince : ATP + WATER --> ADP + PROTON + Pi\n",
      "53 RXN0-4083 serine/threonine:Na+ symporter :  <=> \n",
      "54 TRANS-RXN-353 novobiocin efflux pump :  --> \n",
      "55 TRANS-RXN-297 ferric coprogen ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "56 RXN0-22 guanosine:H+ antiporter :  <=> \n",
      "57 RXN0-1923 O-acetyl-L-serine exporter;export of O-acetyl-L-serine :  --> \n",
      "58 TRANS-RXN-162 L-glutamate:H+ symporter :  <=> \n",
      "59 TRANS-RXN-132 uracil:H+ symporter;uracil:H+ symporter UraA :  <=> \n",
      "60 RXN0-6 Fe2+:H+ antiporter :  <=> \n",
      "61 ABC-5-RXN vitamin B12 ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "62 TRANS-RXN0-452 p-aminobenzoyl glutamate:H+ symporter :  <=> \n",
      "63 ABC-20-RXN nickel ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "64 TRANS-RXN0-258 transport of N-acetyl-beta-D-glucosamine(anhydrous)-N-acetylmuramate :  <=> \n",
      "65 TRANS-RXN0-207 Cu+ exporting P-type ATPase : ATP + WATER --> ADP + PROTON + Pi\n",
      "66 RXN0-4522 methionine transport : ATP + WATER --> ADP + PROTON + Pi\n",
      "67 TRANS-RXN-366 cefoperazone ABC exporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "68 TRANS-RXN0-549 transport of poly-beta-1,6-N-acetyl-D-glucosamine :  --> \n",
      "69 TRANS-RXN0-511 L-methionine S-oxide transport : ATP + WATER --> ADP + PROTON + Pi\n",
      "70 TRANS-RXN0-239 nitrate:nitrite antiporter :  --> \n",
      "71 TRANS-RXN-99 choline:H+ symporter :  <=> \n",
      "72 TRANS-RXN-40 L-arabinose exporter :  <=> \n",
      "73 TRANS-RXN-363 nalidixate efflux pump :  --> \n",
      "74 TRANS-RXN-348 tetraphenylphosphonium:Na+ antiport :  <=> \n",
      "75 TRANS-RXN-290 D-cystine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "76 RXN0-1802 long-chain fatty acid outer membrane porin bacteriophage T2 binding :  --> \n",
      "77 TRANS-RXN-242A L-homoserine lactone:H+ antiport :  <=> \n",
      "78 TRANS-RXN0-617 homocystine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "79 TRANS-RXN0-569 L-amino acid exporter :  --> \n",
      "80 TRANS-RXN0-508 djenkolate ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "81 ABC-34-RXN transport of glycerol-3-phosphate : ATP + WATER --> ADP + PROTON + Pi\n",
      "82 RXN-8638 transport of a quarternary amine : ATP + WATER --> ADP + PROTON + Pi\n",
      "83 TRANS-RXN-360 gentamicin efflux pump :  --> \n",
      "84 RXN0-1721 ferric dihyroxybenzoylserine outer membrane transporter :  --> \n",
      "85 TRANS-RXN0-595 flavin mononucleotide (FMN) export :  --> \n",
      "86 TRANS-RXN0-504 maltotetraose ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "87 ABC-9-RXN transport of ferric dicitrate : ATP + WATER --> ADP + PROTON + Pi\n",
      "88 TRANS-RXN0-226 transport of GlcNAc-1,6-anhMurNAc-L-Ala-gamma-D-Glu-DAP-D-Ala :  <=> \n",
      "89 TRANS-RXN0-162 ABC complex for formation and release of holoCcmE : ATP + WATER --> ADP + PROTON + Pi\n",
      "90 TRANS-RXN-357 enterobactin efflux pump :  --> \n",
      "91 TRANS-RXN-3 K+ transporter Kup;K+ transporter TrkH;K+ transporter TrkG :  <=> \n",
      "92 TRANS-RXN0-478 selenate ABC transporter;selenate  ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "93 ABC-7-RXN thiosulfate ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "94 ABC-26-RXN transport of proline : ATP + WATER --> ADP + PROTON + Pi\n",
      "95 TRANS-RXN-354 acriflavine efflux pump;acriflavin efflux pump :  --> \n",
      "96 TRANS-RXN-298 ferrichrome ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "97 RXN0-1924 cysteine exporter;export of L-cysteine :  --> \n",
      "98 TRANS-RXN-194 calmodulin-stimulated calcium-ATPase;plastid envelope ATPase;calcium-transporting ATPase;plasma membrane calcium-transporting ATPase;plasma membrane Ca2+-ATPase : ATP + WATER --> ADP + PROTON + Pi\n",
      "99 TRANS-RXN-137 transport of nitrite;nitrite transporter NirC :  --> \n",
      "100 TRANS-RXN0-576 transport of tellurite :  <=> \n",
      "101 TRANS-RXN-118 proline:Na+ symporter :  <=> \n",
      "102 TRANS-RXN0-453 autoinducer 2 exporter :  <=> \n",
      "103 TRANS-RXN0-281 acetoacetate transporter :  --> \n",
      "104 TRANS-RXN-351 sulfadiazine efflux pump :  --> \n",
      "105 TRANS-RXN0-573 transport of a glycerophosphodiester : ATP + WATER --> ADP + PROTON + Pi\n",
      "106 RXN66-448 Sodium independent arginine transport;L-arginine exporter;sodium independent arginine plasma membrane transport :  <=> \n",
      "107 ABC-4-RXN L-arginine transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "108 ABC-19-RXN transport of molybdate;molybdate ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "109 TRANS-RXN0-240 biotin transporter :  --> \n",
      "110 TRANS-RXN-364 CCCP efflux pump :  --> \n",
      "111 TRANS-RXN-349 deoxycholate:Na+ antiport :  <=> \n",
      "112 TRANS-RXN-291 diaminopimelate ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "113 TRANS-RXN-272 flavin adenine dinucleotide (FAD) export :  --> \n",
      "114 TRANS-RXN0-491 transport of alpha-D-galactofuranose : ATP + WATER --> ADP + PROTON + Pi\n",
      "115 ABC-35-RXN transport of L-leucine;L-leucine transport : ATP + WATER --> ADP + PROTON + Pi\n",
      "116 TRANS-RXN0-445 chitobiose transport;transport of N,N'-diacetyl-chitobiose :  --> \n",
      "117 ABC-15-RXN transport of L-isoleucine : ATP + WATER --> ADP + PROTON + Pi\n",
      "118 TRANS-RXN0-235 2,3-diketo-L-gulonate:Na+ symporter :  --> \n",
      "119 RXN-18029 teichoic acid transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "120 TRANS-RXN0-202 methionine transport : ATP + WATER --> ADP + PROTON + Pi\n",
      "121 TRANS-RXN-220 glucosyl-oleandomycin-exporting ATPase : ATP + WATER --> ADP + PROTON + Pi\n",
      "122 TRANS-RXN-145 water channel AqpZ :  <=> \n",
      "123 TRANS-RXN0-505 propionate transporter :  --> \n",
      "124 ABC-12-RXN transport of L-glutamine : ATP + WATER --> ADP + PROTON + Pi\n",
      "125 TRANS-RXN0-181 TatABCE protein export complex :  --> \n",
      "126 TRANS-RXN-283 glycine betaine ABC transporter : ATP + WATER --> ADP + PROTON + Pi\n",
      "127 RXN0-1701 transport of ferrichrome :  --> \n",
      "128 TRANS-RXN-2 K+ transporting ATPase : ATP + WATER --> ADP + PROTON + Pi\n",
      "129 TRANS-RXN0-592 chenodeoxycholate efflux pump :  --> \n",
      "130 TRANS-RXN0-277 NAD(P) transhydrogenase;pyridine nucleotide transhydrogenase : NADH + NADP --> NAD + NADPH\n",
      "131 TRANS-RXN-237 transport of lipopolysaccharide : ATP + WATER --> ADP + PROTON + Pi\n",
      "132 TRANS-RXN-214 Na(+)-translocating NADH-quinone reductase : NADH + PROTON + Ubiquinones --> NAD + Ubiquinols\n"
     ]
    }
   ],
   "source": [
    "# check transport reaction in halo_meta\n",
    "# load transport reaction ids in metacyc database\n",
    "# The transport reactions were identified based on the transport reaction ids from RAVEN.\n",
    "is_transport = dict()\n",
    "for item in open('../../../ComplementaryData/metaCyc_transport_rxns.txt').read().split(): is_transport[item] = True\n",
    "fhand = open('../Results/transport_rxns.tsv','w')\n",
    "transmodel = cobra.Model('transmodel')\n",
    "non_trans_rxns = list()\n",
    "trans_rxns = list()\n",
    "k = 0\n",
    "for rxn in halo_meta.reactions:\n",
    "    if is_transport.get(rxn.id,False): \n",
    "        k += 1\n",
    "        print(k,rxn.id,rxn.name,':',rxn.reaction)\n",
    "        fhand.write('{0}\\t{1}\\t{2}\\t{3}\\n'.format(k,rxn.id,rxn.name,rxn.reaction))\n",
    "        trans_rxns.append(rxn)\n",
    "    else: non_trans_rxns.append(rxn)\n",
    "transmodel.add_reactions(trans_rxns)\n",
    "for met in transmodel.metabolites: met.charge = int(met.charge)\n",
    "halo_meta.remove_reactions(trans_rxns)\n",
    "cobra.io.write_sbml_model(transmodel,'../Results/transport_rxns.xml')\n",
    "fhand.close()\n",
    "\n",
    "halo_meta = cobra.Model(halo_meta.id)\n",
    "halo_meta.add_reactions(non_trans_rxns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add _c for all metabolite identifiers. Change the compartment to 'c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "halo_meta.repair()\n",
    "for met in halo_meta.metabolites:\n",
    "    met.id = met.id+'_c'\n",
    "    met.compartment = 'c'\n",
    "halo_meta.repair()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add transport reactions back manully"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No objective coefficients in model. Unclear what should be optimized\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 132\n",
      "Number of metabolits: 20\n",
      "Number of compartments: 1\n",
      "Number of genes: 236\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "transmodel = cobra.io.read_sbml_model('../Results/transport_rxns.xml')\n",
    "report_model_status(transmodel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_transported_met(rxn_id,transmodel,metacyc_model,direction,met_ids,weights):\n",
    "    # rxn_id: rxn_id in transmodel\n",
    "    # transmodel: the model containing all transport reactions\n",
    "    # metacyc model: model reconstructed from MetaCyc\n",
    "    # direction: 'in' or 'out'. 'in': e to c, out: c to e\n",
    "    # met_ids: transported metabolite ids\n",
    "    # weights: coeffs of tranported metabolite\n",
    "    # \n",
    "    # Assume all metbolites in transport reactions before adding transported metabolit as in cyctosal.\n",
    "    # \n",
    "    rxn = transmodel.reactions.get_by_id(rxn_id)\n",
    "    new_rxn = rxn.copy()\n",
    "    \n",
    "    # 1. add _c to exsiting metaboites in transport reaction\n",
    "    coeffs = dict()\n",
    "    for met in rxn.metabolites:\n",
    "        new_id = met.id+'_c'\n",
    "        try: new_met = metacyc_model.metabolites.get(new_id)\n",
    "        except: \n",
    "            new_met = met.copy()\n",
    "            new_met.id = new_met.id +'_c'\n",
    "            new_met.compartment = 'c'\n",
    "        coeffs[new_met] = rxn.get_coefficient(met)\n",
    "    \n",
    "    # 2. add transported metaboites\n",
    "    def get_met_ec(met_id):\n",
    "        try: met_e = metacyc_model.metabolites.get_by_id(met_id+'_e')\n",
    "        except: \n",
    "            met_e = cobra.Metabolite(met_id+'_e')\n",
    "            met_e.annotation = {'biocyc':'META:'+met_id}\n",
    "            #met_e.name = met_id\n",
    "            met_e.compartment = 'e'\n",
    "            \n",
    "    \n",
    "        try: met_c = metacyc_model.metabolites.get_by_id(met_id+'_c')\n",
    "        except: \n",
    "            met_c = cobra.Metabolite(met_id+'_c')\n",
    "            #met_c.name = met_id\n",
    "            met_e.annotation = {'biocyc':'META:'+met_id}\n",
    "            met_c.compartment = 'c'\n",
    "    \n",
    "        return met_e,met_c\n",
    "    \n",
    "    for i in range(len(met_ids)):\n",
    "        met_id = met_ids[i]\n",
    "        number_of_molecular = weights[i]\n",
    "        met_e,met_c = get_met_ec(met_id)\n",
    "        if direction == 'in': \n",
    "            coeffs[met_e] = -number_of_molecular\n",
    "            coeffs[met_c] = number_of_molecular\n",
    "        elif direction == 'out':\n",
    "            coeffs[met_c] = -number_of_molecular\n",
    "            coeffs[met_e] = number_of_molecular\n",
    "        else: print('Error: check direction. ')\n",
    "    new_rxn = rxn.copy()\n",
    "    new_rxn.subtract_metabolites(new_rxn.metabolites)\n",
    "    new_rxn.add_metabolites(coeffs)\n",
    "    print(rxn.id,new_rxn.reaction)\n",
    "    return new_rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1139\n",
      "Number of metabolits: 1460\n",
      "Number of compartments: 1\n",
      "Number of genes: 994\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RXN0-21 ATP_c + GLUTATHIONE_c + WATER_c --> ADP_c + GLUTATHIONE_e + PROTON_c + Pi_c\n",
      "RXN-9615 AMMONIUM_e --> AMMONIUM_c\n",
      "TRANS-RXN0-479 ATP_c + SELENITE_e + WATER_c --> ADP_c + PROTON_c + Pi_c + SELENITE_c\n",
      "ABC-70-RXN ATP_c + SULFATE_e + WATER_c --> ADP_c + PROTON_c + Pi_c + SULFATE_c\n",
      "ABC-27-RXN ATP_c + Pi_e + WATER_c --> ADP_c + PROTON_c + Pi_c + Pi_c\n",
      "TRANS-RXN0-286 C6_c --> C6_e\n",
      "TRANS-RXN0-268 ATP_c + L-ALA-GAMMA-D-GLU-DAP_e + WATER_c --> ADP_c + L-ALA-GAMMA-D-GLU-DAP_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-355 CPD-9288_c --> CPD-9288_e\n",
      "RXN0-11 ATP_c + GLUTATHIONE_e + WATER_c --> ADP_c + GLUTATHIONE_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-167 Hpr-pi-phospho-L-histidines_c + N-acetyl-D-glucosamine_c + N-acetyl-D-glucosamine_e --> Hpr-Histidine_c + N-ACETYL-D-GLUCOSAMINE-6-P_c + N-acetyl-D-glucosamine_c\n",
      "ABC-24-RXN ATP_c + SPERMIDINE_e + WATER_c --> ADP_c + PROTON_c + Pi_c + SPERMIDINE_c\n",
      "TRANS-RXN-352 DEOXYCHOLATE_c --> DEOXYCHOLATE_e\n",
      "TRANS-RXN-296 ATP_c + COBINAMIDE_e + WATER_c --> ADP_c + COBINAMIDE_c + PROTON_c + Pi_c\n",
      "RXN-19739 GLUTACONYL-COA_c + NA+_c + PROTON_c --> CARBON-DIOXIDE_c + CROTONYL-COA_c + NA+_e\n",
      "TRANS-RXN0-513 ATP_c + CPD-3736_e + WATER_c --> ADP_c + CPD-3736_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-244 CD+2_e + PROTON_e <=> CD+2_c + PROTON_c\n",
      "TRANS-RXN-42 K+_e + PROTON_e <=> K+_c + PROTON_c\n",
      "TRANS-RXN-365 DEOXYCHOLATE_c + PROTON_c --> DEOXYCHOLATE_e + PROTON_e\n",
      "RXN0-2501 2 CL-_e + PROTON_e <=> 2 CL-_c + PROTON_c\n",
      "TRANS-RXN-273 CPD-14545_c --> CPD-14545_e\n",
      "TRANS-RXN-127 SUC_e + TARTRATE_e <=> SUC_c + TARTRATE_c\n",
      "TRANS-RXN0-510 ATP_c + CPD-3738_e + WATER_c --> ADP_c + CPD-3738_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-492 ATP_c + CPD0-2485_e + WATER_c --> ADP_c + CPD0-2485_c + PROTON_c + Pi_c\n",
      "ABC-36-RXN ATP_c + VAL_e + WATER_c --> ADP_c + PROTON_c + Pi_c + VAL_c\n",
      "TRANS-RXN0-447 ADENINE_e + PROTON_e <=> ADENINE_c + PROTON_c\n",
      "ABC-16-RXN ATP_c + MALTOSE_e + WATER_c --> ADP_c + MALTOSE_c + PROTON_c + Pi_c\n",
      "ATPSYN-RXN ATP_c + 4 PROTON_e + WATER_c <=> ADP_c + PROTON_c + 4 PROTON_c + Pi_c\n",
      "TRANS-RXN-347 CPD-21015_e + NA+_e <=> CPD-21015_c + NA+_c\n",
      "RXN0-3 ATP_c + CYS_c + WATER_c --> ADP_c + CYS_e + PROTON_c + Pi_c\n",
      "TRANS-RXN-312 ATP_c + PHE_e + WATER_c --> ADP_c + PHE_c + PROTON_c + Pi_c\n",
      "RXN0-18 INOSINE_e + PROTON_e <=> INOSINE_c + PROTON_c\n",
      "TRANS-RXN-242 HOMO-SER_e + PROTON_e <=> HOMO-SER_c + PROTON_c\n",
      "ABC-13-RXN ATP_c + GLT_e + WATER_c --> ADP_c + GLT_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-200 PROTON_e + ZN+2_e <=> PROTON_c + ZN+2_c\n",
      "RXN0-5111 GLYCOLLATE_e + NA+_e <=> GLYCOLLATE_c + NA+_c\n",
      "TRANS-RXN-359 CPD0-1938_e + PROTON_e --> CPD0-1938_c + PROTON_c\n",
      "TRANS-RXN-325 CANAVANINE_c --> CANAVANINE_e\n",
      "RXN0-1702 CPD0-621_e + PROTON_e --> CPD0-621_c + PROTON_c\n",
      "TRANS-RXN0-593 ATP_c + CYSTINE_e + WATER_c --> ADP_c + CYSTINE_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-122A L-ASPARTATE_e + 2 PROTON_e <=> L-ASPARTATE_c + 2 PROTON_c\n",
      "TRANS-RXN0-503 ATP_c + MALTOTRIOSE_e + WATER_c --> ADP_c + MALTOTRIOSE_c + PROTON_c + Pi_c\n",
      "ABC-8-RXN ATP_c + DIPEPTIDES_e + WATER_c --> ADP_c + DIPEPTIDES_c + PROTON_c + Pi_c\n",
      "ABC-10-RXN ATP_c + FERRIC-ENTEROBACTIN-COMPLEX_e + WATER_c --> ADP_c + FERRIC-ENTEROBACTIN-COMPLEX_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-269 VAL_e <=> VAL_c\n",
      "TRANS-RXN0-222 ATP_c + TRANS-RXN0-222_e + WATER_c --> ADP_c + PROTON_c + Pi_c + TRANS-RXN0-222_c\n",
      "TRANS-RXN0-0244 PROTON_e + THR_e <=> PROTON_c + THR_c\n",
      "TRANS-RXN-356 CPD-3617_c + PROTON_c --> CPD-3617_e + PROTON_e\n",
      "TRANS-RXN-340 CPD-20921_e + PROTON_e <=> CPD-20921_c + PROTON_c\n",
      "RXN0-1981 ACET_e + NA+_e <=> ACET_c + NA+_c\n",
      "TRANS-RXN-167A GLUCOSAMINE_c + GLUCOSAMINE_e + Hpr-pi-phospho-L-histidines_c --> D-GLUCOSAMINE-6-P_c + GLUCOSAMINE_c + Hpr-Histidine_c\n",
      "RXN-5076 PROTON_e + XANTHINE_e <=> PROTON_c + XANTHINE_c\n",
      "ABC-25-RXN ATP_c + PUTRESCINE_e + WATER_c --> ADP_c + PROTON_c + PUTRESCINE_c + Pi_c\n",
      "RXN0-4083 NA+_e + SER_e <=> NA+_c + SER_c\n",
      "TRANS-RXN-353 CPD-15417_c + PROTON_c --> CPD-15417_e + PROTON_e\n",
      "TRANS-RXN-297 ATP_c + CPD0-621_e + WATER_c --> ADP_c + CPD0-621_c + PROTON_c + Pi_c\n",
      "RXN0-22 GUANOSINE_e + PROTON_e <=> GUANOSINE_c + PROTON_c\n",
      "RXN0-1923 ACETYLSERINE_c --> ACETYLSERINE_e\n",
      "TRANS-RXN-162 GLT_e + 2 PROTON_e <=> GLT_c + 2 PROTON_c\n",
      "TRANS-RXN-132 PROTON_e + URACIL_e <=> PROTON_c + URACIL_c\n",
      "RXN0-6 FE+2_e + PROTON_e <=> FE+2_c + PROTON_c\n",
      "ABC-5-RXN ATP_c + COB-I-ALAMIN_e + WATER_c --> ADP_c + COB-I-ALAMIN_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-452 CPD0-889_e + PROTON_e <=> CPD0-889_c + PROTON_c\n",
      "ABC-20-RXN ATP_c + NI+2_e + WATER_c --> ADP_c + NI+2_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-258 CPD0-1081_e + PROTON_e <=> CPD0-1081_c + PROTON_c\n",
      "TRANS-RXN0-207 ATP_c + CU+_c + WATER_c --> ADP_c + CU+_e + PROTON_c + Pi_c\n",
      "RXN0-4522 ATP_c + MET_e + WATER_c --> ADP_c + MET_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-366 ATP_c + CPD-19953_c + WATER_c --> ADP_c + CPD-19953_e + PROTON_c + Pi_c\n",
      "TRANS-RXN0-549 CPD0-1192_c --> CPD0-1192_e\n",
      "TRANS-RXN0-511 ATP_c + L-Methionine-sulfoxides_e + WATER_c --> ADP_c + L-Methionine-sulfoxides_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-239 NITRATE_e + NITRITE_c --> NITRATE_c + NITRITE_e\n",
      "TRANS-RXN-99 CHOLINE_e + PROTON_e <=> CHOLINE_c + PROTON_c\n",
      "TRANS-RXN-40 L-ARABINOSE_e + PROTON_e <=> L-ARABINOSE_c + PROTON_c\n",
      "TRANS-RXN-363 CPD-21025_c + PROTON_c --> CPD-21025_e + PROTON_e\n",
      "TRANS-RXN-348 CPD-20890_e + NA+_e <=> CPD-20890_c + NA+_c\n",
      "TRANS-RXN-290 ATP_c + CPD0-1564_e + WATER_c --> ADP_c + CPD0-1564_c + PROTON_c + Pi_c\n",
      "RXN0-1802 Long-Chain-Fatty-Acids_e --> Long-Chain-Fatty-Acids_c\n",
      "TRANS-RXN-242A CPD-15554_e + PROTON_e <=> CPD-15554_c + PROTON_c\n",
      "TRANS-RXN0-617 ATP_c + HOMOCYSTINE_e + WATER_c --> ADP_c + HOMOCYSTINE_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-569 LYS_c --> LYS_e\n",
      "TRANS-RXN0-508 ATP_c + CPD-3740_e + WATER_c --> ADP_c + CPD-3740_c + PROTON_c + Pi_c\n",
      "ABC-34-RXN ATP_c + GLYCEROL-3P_e + WATER_c --> ADP_c + GLYCEROL-3P_c + PROTON_c + Pi_c\n",
      "RXN-8638 ATP_c + Quaternary-Amines_e + WATER_c --> ADP_c + PROTON_c + Pi_c + Quaternary-Amines_c\n",
      "TRANS-RXN-360 Gentamycins_c + PROTON_c --> Gentamycins_e + PROTON_e\n",
      "RXN0-1721 CPD0-2482_e --> CPD0-2482_c\n",
      "TRANS-RXN0-595 FMN_c --> FMN_e\n",
      "TRANS-RXN0-504 ATP_c + MALTOTETRAOSE_e + WATER_c --> ADP_c + MALTOTETRAOSE_c + PROTON_c + Pi_c\n",
      "ABC-9-RXN ATP_c + FERRIC-CITRATE-COMPLEX_e + WATER_c --> ADP_c + FERRIC-CITRATE-COMPLEX_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-226 CPD0-1080_e + PROTON_e <=> CPD0-1080_c + PROTON_c\n",
      "TRANS-RXN0-162 ATP_c + TRANS-RXN0-162_c + WATER_c --> ADP_c + PROTON_c + Pi_c + TRANS-RXN0-162_e\n",
      "TRANS-RXN-357 ENTEROBACTIN_c + PROTON_c --> ENTEROBACTIN_e + PROTON_e\n",
      "TRANS-RXN-3 K+_e + PROTON_e <=> K+_c + PROTON_c\n",
      "TRANS-RXN0-478 ATP_c + SELENATE_e + WATER_c --> ADP_c + PROTON_c + Pi_c + SELENATE_c\n",
      "ABC-7-RXN ABC-7-RXN_e + ATP_c + WATER_c --> ABC-7-RXN_c + ADP_c + PROTON_c + Pi_c\n",
      "ABC-26-RXN ATP_c + PRO_e + WATER_c --> ADP_c + PROTON_c + PRO_c + Pi_c\n",
      "TRANS-RXN-354 CPD-21070_e + PROTON_e --> CPD-21070_c + PROTON_c\n",
      "TRANS-RXN-298 ATP_c + CPD0-2241_e + WATER_c --> ADP_c + CPD0-2241_c + PROTON_c + Pi_c\n",
      "RXN0-1924 CYS_c --> CYS_e\n",
      "TRANS-RXN-194 ATP_c + CA+2_c + WATER_c --> ADP_c + CA+2_e + PROTON_c + Pi_c\n",
      "TRANS-RXN-137 NITRITE_c --> NITRITE_e\n",
      "TRANS-RXN0-576 CPD-4544_e + NA+_e <=> CPD-4544_c + NA+_c\n",
      "TRANS-RXN-118 NA+_e + PRO_e <=> NA+_c + PRO_c\n",
      "TRANS-RXN0-453 TRANS-RXN0-453_e <=> TRANS-RXN0-453_c\n",
      "TRANS-RXN0-281 3-KETOBUTYRATE_e --> 3-KETOBUTYRATE_c\n",
      "TRANS-RXN-351 CPD-20940_c --> CPD-20940_e\n",
      "TRANS-RXN0-573 ATP_c + Glycerophosphodiesters_e + WATER_c --> ADP_c + Glycerophosphodiesters_c + PROTON_c + Pi_c\n",
      "RXN66-448 ARG_e <=> ARG_c\n",
      "ABC-4-RXN ARG_e + ATP_c + WATER_c --> ADP_c + ARG_c + PROTON_c + Pi_c\n",
      "ABC-19-RXN ATP_c + CPD-3_e + WATER_c --> ADP_c + CPD-3_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-240 BIOTIN_e --> BIOTIN_c\n",
      "TRANS-RXN-364 CPD-7970_c + PROTON_c --> CPD-7970_e + PROTON_e\n",
      "TRANS-RXN-349 DEOXYCHOLATE_e + NA+_e <=> DEOXYCHOLATE_c + NA+_c\n",
      "TRANS-RXN-291 ATP_c + MESO-DIAMINOPIMELATE_e + WATER_c --> ADP_c + MESO-DIAMINOPIMELATE_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-272 FAD_c --> FAD_e\n",
      "TRANS-RXN0-491 ATP_c + CPD0-2486_e + WATER_c --> ADP_c + CPD0-2486_c + PROTON_c + Pi_c\n",
      "ABC-35-RXN ATP_c + LEU_e + WATER_c --> ADP_c + LEU_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-445 CHITOBIOSE_e --> CHITOBIOSE_c\n",
      "ABC-15-RXN ATP_c + ILE_e + WATER_c --> ADP_c + ILE_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-235 CPD-334_e + NA+_e --> CPD-334_c + NA+_c\n",
      "RXN-18029 ATP_c + CPD-19309_c + WATER_c --> ADP_c + CPD-19309_e + PROTON_c + Pi_c\n",
      "TRANS-RXN0-202 ATP_c + CPD-218_e + WATER_c --> ADP_c + CPD-218_c + PROTON_c + Pi_c\n",
      "TRANS-RXN-220 ATP_c + CPD-13949_c + WATER_c --> ADP_c + CPD-13949_e + PROTON_c + Pi_c\n",
      "TRANS-RXN-145 WATER_e <=> WATER_c\n",
      "TRANS-RXN0-505 TRANS-RXN0-505_e --> TRANS-RXN0-505_c\n",
      "ABC-12-RXN ATP_c + GLN_e + WATER_c --> ADP_c + GLN_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-181 General-Protein-Substrates_c --> General-Protein-Substrates_e\n",
      "TRANS-RXN-283 ATP_c + BETAINE_e + WATER_c --> ADP_c + BETAINE_c + PROTON_c + Pi_c\n",
      "RXN0-1701 CPD0-2241_e + PROTON_e --> CPD0-2241_c + PROTON_c\n",
      "TRANS-RXN-2 ATP_c + K+_e + WATER_c --> ADP_c + K+_c + PROTON_c + Pi_c\n",
      "TRANS-RXN0-592 CPD-15189_c + PROTON_c --> CPD-15189_e + PROTON_e\n",
      "TRANS-RXN0-277 NADH_c + NADP_c + PROTON_e --> NADPH_c + NAD_c + PROTON_c\n",
      "TRANS-RXN-237 ATP_c + Lipopolysaccharides_c + WATER_c --> ADP_c + Lipopolysaccharides_e + PROTON_c + Pi_c\n",
      "TRANS-RXN-214 NA+_c + NADH_c + PROTON_c + Ubiquinones_c --> NA+_e + NAD_c + Ubiquinols_c\n"
     ]
    }
   ],
   "source": [
    "trans_rxns_formated = list()\n",
    "for line in open('../../../ComplementaryData/transport_rxns_mannual.tsv'):\n",
    "    cont = line.strip().split('\\t')\n",
    "    rxn_id = cont[1]\n",
    "    direction = cont[5]\n",
    "    met_ids = cont[4].split(',')\n",
    "    weights = [int(item) for item in cont[6].split(',')]\n",
    "        \n",
    "    trans_rxns_formated.append(format_transported_met(rxn_id,transmodel,halo_meta,direction,\n",
    "                                                  met_ids,weights)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1271\n",
      "Number of metabolits: 1641\n",
      "Number of compartments: 2\n",
      "Number of genes: 1230\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "halo_meta.add_reactions(trans_rxns_formated)\n",
    "halo_meta.repair()\n",
    "report_model_status(halo_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'c': '', 'e': ''}\n"
     ]
    }
   ],
   "source": [
    "print(halo_meta.compartments)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add annotation field to rxn and met in metacyc model. Add 'biocyc' to both"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for rxn in halo_meta.reactions:\n",
    "    rxn.annotation['biocyc'] = rxn.id\n",
    "    print(rxn.id,rxn.annotation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for met in halo_meta.metabolites:\n",
    "    met.annotation['biocyc'] = met.id.replace('_c','').replace('_e','')\n",
    "    print(met.id,met.annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iML1515</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01325818c88</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1877</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2712</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>38</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_Ec_iML1515_core_75p37M - 1.0*BIOMASS_Ec_iML1515_core_75p37M_reverse_35685</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iML1515 at 0x1325818c88>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eco"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine draft models based on metacyc and iML1515\n",
    "* For template model on iML1515\n",
    "    * replace _p with _c\n",
    "    * use metacyc id for metabolites and reactions if available\n",
    "    * if multiple metacyc ids exist, use the one in halomodel if available\n",
    "    * ignore transport reaction between c and e \n",
    "* Add reactions from metacyc into above model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_universal_met(eco_met,halomodel,universal,mnx2meta):\n",
    "    is_in_halo = dict()\n",
    "    for met in halomodel.metabolites: is_in_halo[met.id] = True\n",
    "    mnx_id = eco_met.annotation.get('metanetx.chemical')\n",
    "    \n",
    "    comp = eco_met.compartment\n",
    "    if comp == 'p': comp = 'c'\n",
    "    met_id = None\n",
    "    \n",
    "    if mnx_id is None or mnx2meta.get(mnx_id,None) is None: \n",
    "        # if there is no mnx id or no metacyc id, use bigg id instead\n",
    "        met_id = eco_met.annotation['bigg.metabolite'] + '_' + comp\n",
    "    else:\n",
    "        meta_ids = mnx2meta[mnx_id]\n",
    "        for item in meta_ids:\n",
    "            if is_in_halo.get(item+ '_' + comp,False): \n",
    "                met_id = item+ '_' + comp\n",
    "                break\n",
    "        if met_id is None: met_id = meta_ids[0] + '_' + comp\n",
    "    try: \n",
    "        new_met = halomodel.metabolites.get_by_id(met_id)\n",
    "    except:\n",
    "        try:\n",
    "            new_met = universal.metabolites.get_by_id(met_id)\n",
    "        except:\n",
    "            new_met = eco_met.copy()\n",
    "            new_met.id = met_id\n",
    "            new_met.compartment = comp\n",
    "    return new_met"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_rxnid(new_rxn,halomodel,universal):\n",
    "    # metacyc reaction ids in halomodel:\n",
    "    halo_rxn_ids = dict()\n",
    "    halo_rxn_mets = dict()\n",
    "    mets_rxn = dict()\n",
    "    for rxn in halomodel.reactions: \n",
    "        halo_rxn_ids[rxn.id] = True\n",
    "        mets = [met.id for met in rxn.metabolites]\n",
    "        mets.sort()\n",
    "        mets = tuple(mets)\n",
    "        halo_rxn_mets[mets] = True\n",
    "        mets_rxn[mets] = rxn\n",
    "        \n",
    "    # 1. check metacyc id\n",
    "    meta_ids = new_rxn.annotation.get('biocyc')\n",
    "    new_rxn_id = new_rxn.id\n",
    "    # case 1: there is metacyc ids, map metacyc id\n",
    "    if meta_ids is not None:\n",
    "        if type(meta_ids) == str: meta_ids = [meta_ids]\n",
    "        for meta_id in meta_ids:\n",
    "            meta_id = meta_id.replace('META:','')\n",
    "            if halo_rxn_ids.get(meta_id,False): new_rxn_id = meta_id\n",
    "    \n",
    "    # case 2: if case 1 failed, check reaction equations\n",
    "    else:\n",
    "        new_mets = [met.id for met in new_rxn.metabolites]\n",
    "        new_mets.sort()\n",
    "        new_mets = tuple(new_mets)\n",
    "        if halo_rxn_mets.get(new_mets,False): \n",
    "            # the two reactions contain the same metabolites. Now compare the equations, including direction\n",
    "            eq1 = new_rxn.reaction\n",
    "            eq2 = mets_rxn[new_mets].reaction\n",
    "            eq1 = eq1.split()\n",
    "            eq2 = eq2.split()\n",
    "            \n",
    "            eq1.sort()\n",
    "            eq2.sort()\n",
    "            if eq1 == eq2:\n",
    "                # print(new_rxn.id,new_rxn.reaction)\n",
    "                # print(mets_rxn[new_mets].id,mets_rxn[new_mets].reaction)\n",
    "                # print('')\n",
    "                new_rxn_id = mets_rxn[new_mets].id\n",
    "            \n",
    "    new_rxn.id = new_rxn_id\n",
    "    \n",
    "    # if new_rxn_id is in universal model, use the reactions in univeral model\n",
    "    try: \n",
    "        rxn_universal = universal.reactions.get_by_id(new_rxn_id)\n",
    "        gr_rule = new_rxn.gene_reaction_rule \n",
    "        new_rxn = rxn_universal.copy()\n",
    "        new_rxn.gene_reaction_rule = gr_rule\n",
    "    except: None\n",
    "    \n",
    "    \n",
    "    return new_rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_universal_rxn(eco_rxn,halomodel,universal,mnx2meta):\n",
    "    coeffs = dict()\n",
    "    is_transport = False\n",
    "    \n",
    "    # convert metabolite ids\n",
    "    mets = list()\n",
    "    for eco_met in eco_rxn.metabolites:\n",
    "        new_met = create_universal_met(eco_met,halomodel,universal,mnx2meta)\n",
    "        if new_met.id in mets:\n",
    "            is_transport = True\n",
    "\n",
    "        mets.append(new_met.id)\n",
    "        coeffs[new_met] = coeffs.get(new_met,0) + eco_rxn.get_coefficient(eco_met)\n",
    "\n",
    "    if is_transport and 'BIOMASS' not in eco_rxn.id: return None\n",
    "    else:\n",
    "        # convert reaction ids\n",
    "        new_rxn = eco_rxn.copy()\n",
    "        new_rxn.subtract_metabolites(new_rxn.metabolites)\n",
    "        new_rxn.add_metabolites(coeffs)\n",
    "        \n",
    "        new_rxn = update_rxnid(new_rxn,halomodel,universal)\n",
    "        return new_rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_transport_rxn_from_model_on_template(model):\n",
    "    rxns = list()\n",
    "    \n",
    "    for rxn in model.reactions:\n",
    "        is_transport = False\n",
    "        mets = dict()\n",
    "        for met in rxn.metabolites: \n",
    "            comp = met.compartment\n",
    "            met_id = met.id.replace('_{0}'.format(comp),'')\n",
    "            if mets.get(met_id) is not None: is_transport = True\n",
    "            else: mets[met_id] = 0\n",
    "\n",
    "        if not is_transport: rxns.append(rxn)\n",
    "    \n",
    "    new_model = cobra.Model('model_from_template_without_transport')\n",
    "    new_model.add_reactions(rxns)\n",
    "    print('{0} transport reactions were removed'.format(len(model.reactions)-len(rxns)))\n",
    "    return new_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# break point\n",
    "# print(gang)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. covert iML1515 to a universal model\n",
    "* replace _p with _c\n",
    "* use metacyc id for metabolites and reactions if available\n",
    "* if multiple metacyc ids exist, use the one in halomodel if available\n",
    "* ignore transport reaction between c and e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "universal = cobra.Model('universal')\n",
    "mnx2meta = dict()\n",
    "for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "    cont = line.split()\n",
    "    mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "univeral_rxns = dict()\n",
    "for eco_rxn in eco.reactions:\n",
    "    new_rxn = create_universal_rxn(eco_rxn,halo_meta,universal,mnx2meta)\n",
    "    if new_rxn is None: continue\n",
    "    \n",
    "    univeral_rxns[new_rxn.id] = new_rxn\n",
    "universal.add_reactions(univeral_rxns.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 2225\n",
      "Number of metabolits: 1503\n",
      "Number of compartments: 2\n",
      "Number of genes: 1093\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(universal)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "def block_amino_acid_update_eco(model):\n",
    "    amino_acids = ['CYS_c','L-ASPARTATE_c','GLT_c','PHE_c','GLY_c',\n",
    "                  'HIS_c','ILE_c','LYS_c','LEU_c','MET_c','ASN_c',\n",
    "                  'PRO_c','GLN_c','ARG_c','SER_c','THR_c','VAL_c',\n",
    "                  'TRP_c','L-ALPHA-ALANINE_c','TYR_c']\n",
    "    for met_id in amino_acids: \n",
    "        met = model.metabolites.get_by_id(met_id)\n",
    "        for rxn in met.reactions:\n",
    "            if 'BIOMASS' in rxn.name: continue\n",
    "            if rxn.id.endswith('tex'):\n",
    "                rxn.upper_bound = 0\n",
    "                print(rxn.id,rxn.reaction)\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 0.344 at 0x1321995160>\n",
      "<Solution 0.877 at 0x13219950f0>\n"
     ]
    }
   ],
   "source": [
    "universal.objective = 'BIOMASS_Ec_iML1515_core_75p37M'\n",
    "eco.objective = 'BIOMASS_Ec_iML1515_core_75p37M'\n",
    "\n",
    "# check why? make sure growth rate is higher after removal of compartment p. \n",
    "print(universal.optimize())\n",
    "print(eco.optimize())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for rxn in eco.reactions:\n",
    "    try:\n",
    "        if 'biomass' in rxn.name.lower():print(rxn.id,rxn.name)\n",
    "    except: pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. convert  the draft model based template to a small universal model. Remove all transport reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 transport reactions were removed\n"
     ]
    }
   ],
   "source": [
    "# change the ids in model halo_iJO with metacyc ids.\n",
    "halo_iJO_meta = cobra.Model('halo_iML1515_metacycids')\n",
    "template_rxns = dict()\n",
    "\n",
    "for rxn in halo_iJO.reactions:\n",
    "    new_rxn = create_universal_rxn(rxn,halo_meta,universal,mnx2meta)\n",
    "    if new_rxn is None: continue\n",
    "    template_rxns[new_rxn.id] = new_rxn\n",
    "halo_iJO_meta.add_reactions(template_rxns.values())\n",
    "halo_iJO_meta = remove_transport_rxn_from_model_on_template(halo_iJO_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1119\n",
      "Number of metabolits: 1099\n",
      "Number of compartments: 3\n",
      "Number of genes: 662\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_iJO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 933\n",
      "Number of metabolits: 837\n",
      "Number of compartments: 1\n",
      "Number of genes: 533\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_iJO_meta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. compare draft models from template and metacyc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 DEOXYRIBOSE-P-ALD-RXN DEOXY-RIBOSE-5P_c --> ACETALD_c + GAP_c\n",
      "1 DEOXYRIBOSE-P-ALD-RXN DEOXY-RIBOSE-5P_c <=> ACETALD_c + GAP_c\n",
      "\n",
      "2 RXN-6321 5-FORMYL-THF_c + PROTON_c --> 5-10-METHENYL-THF_c + WATER_c\n",
      "2 RXN-6321 5-10-METHENYL-THF_c + WATER_c --> 5-FORMYL-THF_c + PROTON_c\n",
      "\n",
      "3 GLYOHMETRANS-RXN SER_c + THF_c <=> GLY_c + METHYLENE-THF_c + WATER_c\n",
      "3 GLYOHMETRANS-RXN SER_c + THF_c <=> GLY_c + METHYLENE-THF_c + WATER_c\n",
      "\n",
      "4 LTAA-RXN L-ALLO-THREONINE_c --> ACETALD_c + GLY_c\n",
      "4 LTAA-RXN L-ALLO-THREONINE_c --> ACETALD_c + GLY_c\n",
      "\n",
      "5 RXN0-5240 D-ALANINE_c + PYRIDOXAL_PHOSPHATE_c --> PYRIDOXAMINE-5P_c + PYRUVATE_c\n",
      "5 RXN0-5240 D-ALANINE_c + PYRIDOXAL_PHOSPHATE_c <=> PYRIDOXAMINE-5P_c + PYRUVATE_c\n",
      "\n",
      "6 THREONINE-ALDOLASE-RXN THR_c --> ACETALD_c + GLY_c\n",
      "6 THREONINE-ALDOLASE-RXN THR_c --> ACETALD_c + GLY_c\n",
      "\n",
      "7 RXN0-6563 CPD0-2189_c <=> GLYCOLALDEHYDE_c + GLY_c\n",
      "7 RXN0-6563 CPD0-2189_c <=> GLYCOLALDEHYDE_c + GLY_c\n",
      "\n",
      "8 DIMETHUROPORDEHYDROG-RXN DIHYDROSIROHYDROCHLORIN_c + NAD_c --> NADH_c + PROTON_c + SIROHYDROCHLORIN_c\n",
      "8 DIMETHUROPORDEHYDROG-RXN DIHYDROSIROHYDROCHLORIN_c + NAD_c --> NADH_c + 2.0 PROTON_c + SIROHYDROCHLORIN_c\n",
      "\n",
      "9 RXN-13403 2.0 S-ADENOSYLMETHIONINE_c + UROPORPHYRINOGEN-III_c --> 2.0 ADENOSYL-HOMO-CYS_c + DIHYDROSIROHYDROCHLORIN_c + PROTON_c\n",
      "9 RXN-13403 2.0 S-ADENOSYLMETHIONINE_c + UROPORPHYRINOGEN-III_c --> 2.0 ADENOSYL-HOMO-CYS_c + DIHYDROSIROHYDROCHLORIN_c + PROTON_c\n",
      "\n",
      "10 SIROHEME-FERROCHELAT-RXN FE+2_c + SIROHYDROCHLORIN_c --> 3.0 PROTON_c + SIROHEME_c\n",
      "10 SIROHEME-FERROCHELAT-RXN FE+2_c + SIROHYDROCHLORIN_c --> 2.0 PROTON_c + SIROHEME_c\n",
      "\n",
      "11 RXN0-1461 COPROPORPHYRINOGEN_III_c + OXYGEN-MOLECULE_c + 2.0 PROTON_c --> 2.0 CARBON-DIOXIDE_c + PROTOPORPHYRINOGEN_c + 2.0 WATER_c\n",
      "11 RXN0-1461 COPROPORPHYRINOGEN_III_c + OXYGEN-MOLECULE_c + 2.0 PROTON_c --> 2.0 CARBON-DIOXIDE_c + PROTOPORPHYRINOGEN_c + 2.0 WATER_c\n",
      "\n",
      "12 GLUTATHIONE-REDUCT-NADPH-RXN NADPH_c + OXIDIZED-GLUTATHIONE_c + PROTON_c <=> 2.0 GLUTATHIONE_c + NADP_c\n",
      "12 GLUTATHIONE-REDUCT-NADPH-RXN NADPH_c + OXIDIZED-GLUTATHIONE_c + PROTON_c --> 2.0 GLUTATHIONE_c + NADP_c\n",
      "\n",
      "13 RXN0-2382 INDOLE_c + SER_c --> TRP_c + WATER_c\n",
      "13 RXN0-2382 INDOLE_c + SER_c --> TRP_c + WATER_c\n",
      "\n",
      "14 TRYPSYN-RXN INDOLE-3-GLYCEROL-P_c + SER_c --> GAP_c + TRP_c + WATER_c\n",
      "14 TRYPSYN-RXN INDOLE-3-GLYCEROL-P_c + SER_c --> GAP_c + TRP_c + WATER_c\n",
      "\n",
      "15 RXN0-2381 INDOLE-3-GLYCEROL-P_c --> GAP_c + INDOLE_c\n",
      "15 RXN0-2381 INDOLE-3-GLYCEROL-P_c <=> GAP_c + INDOLE_c\n",
      "\n",
      "16 BRANCHED-CHAINAMINOTRANSFERILEU-RXN 2-KETOGLUTARATE_c + ILE_c <=> 2-KETO-3-METHYL-VALERATE_c + GLT_c\n",
      "16 BRANCHED-CHAINAMINOTRANSFERILEU-RXN 2-KETOGLUTARATE_c + ILE_c <=> 2-KETO-3-METHYL-VALERATE_c + GLT_c\n",
      "\n",
      "17 BRANCHED-CHAINAMINOTRANSFERVAL-RXN 2-KETOGLUTARATE_c + VAL_c <=> 2-KETO-ISOVALERATE_c + GLT_c\n",
      "17 BRANCHED-CHAINAMINOTRANSFERVAL-RXN 2-KETOGLUTARATE_c + VAL_c <=> 2-KETO-ISOVALERATE_c + GLT_c\n",
      "\n",
      "18 BRANCHED-CHAINAMINOTRANSFERLEU-RXN 2K-4CH3-PENTANOATE_c + GLT_c --> 2-KETOGLUTARATE_c + LEU_c\n",
      "18 BRANCHED-CHAINAMINOTRANSFERLEU-RXN 2-KETOGLUTARATE_c + LEU_c <=> 2K-4CH3-PENTANOATE_c + GLT_c\n",
      "\n",
      "19 TYROSINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + TYR_c <=> GLT_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "19 TYROSINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + TYR_c <=> GLT_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "\n",
      "20 SHIKIMATE-5-DEHYDROGENASE-RXN 3-DEHYDRO-SHIKIMATE_c + NADPH_c + PROTON_c <=> NADP_c + SHIKIMATE_c\n",
      "20 SHIKIMATE-5-DEHYDROGENASE-RXN 3-DEHYDRO-SHIKIMATE_c + NADPH_c + PROTON_c --> NADP_c + SHIKIMATE_c\n",
      "\n",
      "21 GLUTSEMIALDEHYDROG-RXN L-GLUTAMATE-5-P_c + NADPH_c + PROTON_c --> L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c + NADP_c + Pi_c\n",
      "21 GLUTSEMIALDEHYDROG-RXN L-GLUTAMATE-5-P_c + NADPH_c + PROTON_c --> L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c + NADP_c + Pi_c\n",
      "\n",
      "22 RIBOFLAVINSYNDEAM-RXN DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + PROTON_c + WATER_c --> AMMONIUM_c + CPD-602_c\n",
      "22 RIBOFLAVINSYNDEAM-RXN DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + PROTON_c + WATER_c --> AMMONIUM_c + CPD-602_c\n",
      "\n",
      "23 RIBOFLAVINSYNREDUC-RXN CPD-602_c + NADPH_c + PROTON_c --> CPD-1086_c + NADP_c\n",
      "23 RIBOFLAVINSYNREDUC-RXN CPD-602_c + NADPH_c + PROTON_c --> CPD-1086_c + NADP_c\n",
      "\n",
      "24 RIBOFLAVINKIN-RXN ATP_c + RIBOFLAVIN_c --> ADP_c + FMN_c + PROTON_c\n",
      "24 RIBOFLAVINKIN-RXN ATP_c + RIBOFLAVIN_c --> ADP_c + FMN_c + PROTON_c\n",
      "\n",
      "25 FADSYN-RXN ATP_c + FMN_c + PROTON_c --> FAD_c + PPI_c\n",
      "25 FADSYN-RXN ATP_c + FMN_c + PROTON_c --> FAD_c + PPI_c\n",
      "\n",
      "26 RIBOPHOSPHAT-RXN CPD-1086_c + WATER_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + Pi_c\n",
      "26 RIBOPHOSPHAT-RXN CPD-1086_c + WATER_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + Pi_c\n",
      "\n",
      "27 RXN0-5187 FMN_c + WATER_c --> Pi_c + RIBOFLAVIN_c\n",
      "27 RXN0-5187 FMN_c + WATER_c --> Pi_c + RIBOFLAVIN_c\n",
      "\n",
      "28 RIBOFLAVIN-SYN-RXN 2.0 DIMETHYL-D-RIBITYL-LUMAZINE_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + RIBOFLAVIN_c\n",
      "28 RIBOFLAVIN-SYN-RXN 2.0 DIMETHYL-D-RIBITYL-LUMAZINE_c + PROTON_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + RIBOFLAVIN_c\n",
      "\n",
      "29 FPPSYN-RXN DELTA3-ISOPENTENYL-PP_c + GERANYL-PP_c --> FARNESYL-PP_c + PPI_c\n",
      "29 FPPSYN-RXN DELTA3-ISOPENTENYL-PP_c + GERANYL-PP_c --> FARNESYL-PP_c + PPI_c\n",
      "\n",
      "30 GPPSYN-RXN CPD-4211_c + DELTA3-ISOPENTENYL-PP_c --> GERANYL-PP_c + PPI_c\n",
      "30 GPPSYN-RXN CPD-4211_c + DELTA3-ISOPENTENYL-PP_c --> GERANYL-PP_c + PPI_c\n",
      "\n",
      "31 UROGENIIISYN-RXN HYDROXYMETHYLBILANE_c --> UROPORPHYRINOGEN-III_c + WATER_c\n",
      "31 UROGENIIISYN-RXN HYDROXYMETHYLBILANE_c --> UROPORPHYRINOGEN-III_c + WATER_c\n",
      "\n",
      "32 UROGENDECARBOX-RXN 4.0 PROTON_c + UROPORPHYRINOGEN-III_c --> 4.0 CARBON-DIOXIDE_c + COPROPORPHYRINOGEN_III_c\n",
      "32 UROGENDECARBOX-RXN 4.0 PROTON_c + UROPORPHYRINOGEN-III_c --> 4.0 CARBON-DIOXIDE_c + COPROPORPHYRINOGEN_III_c\n",
      "\n",
      "33 THYM-PHOSPH-RXN Pi_c + THYMIDINE_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + THYMINE_c\n",
      "33 THYM-PHOSPH-RXN Pi_c + THYMIDINE_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + THYMINE_c\n",
      "\n",
      "34 URA-PHOSPH-RXN DEOXYURIDINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + URACIL_c\n",
      "34 URA-PHOSPH-RXN DEOXYURIDINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + URACIL_c\n",
      "\n",
      "35 CARBODEHYDRAT-RXN CARBON-DIOXIDE_c + WATER_c <=> H2CO3_c + PROTON_c\n",
      "35 CARBODEHYDRAT-RXN H2CO3_c <=> CARBON-DIOXIDE_c + WATER_c\n",
      "\n",
      "36 AICARTRANSFORM-RXN 10-FORMYL-THF_c + AICAR_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c + THF_c\n",
      "36 AICARTRANSFORM-RXN 10-FORMYL-THF_c + AICAR_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c + THF_c\n",
      "\n",
      "37 IMPCYCLOHYDROLASE-RXN IMP_c + WATER_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c\n",
      "37 IMPCYCLOHYDROLASE-RXN IMP_c + WATER_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c\n",
      "\n",
      "38 IMP-DEHYDROG-RXN IMP_c + NAD_c + WATER_c --> NADH_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "38 IMP-DEHYDROG-RXN IMP_c + NAD_c + WATER_c <=> NADH_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "39 RXN-11667 CPD-12173_c <=> CROTONYL-COA_c + WATER_c\n",
      "39 RXN-11667 S-3-HYDROXYBUTANOYL-COA_c <=> CROTONYL-COA_c + WATER_c\n",
      "\n",
      "40 RXN-12567 OH-HEXANOYL-COA_c <=> CPD0-2121_c + WATER_c\n",
      "40 RXN-12567 OH-HEXANOYL-COA_c --> CPD0-2121_c + WATER_c\n",
      "\n",
      "41 RXN-11662 ACETOACETYL-COA_c + NADH_c + PROTON_c <=> CPD-12173_c + NAD_c\n",
      "41 RXN-11662 NAD_c + S-3-HYDROXYBUTANOYL-COA_c <=> ACETOACETYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "42 RXN-12490 CPD0-2123_c + NADH_c + PROTON_c <=> CPD0-2244_c + NAD_c\n",
      "42 RXN-12490 CPD0-2244_c + NAD_c --> CPD0-2123_c + NADH_c + PROTON_c\n",
      "\n",
      "43 RXN-13616 CPD0-2244_c <=> T2-DECENOYL-COA_c + WATER_c\n",
      "43 RXN-13616 T2-DECENOYL-COA_c + WATER_c --> CPD0-2244_c\n",
      "\n",
      "44 HYPOXANPRIBOSYLTRAN-RXN HYPOXANTHINE_c + PRPP_c --> IMP_c + PPI_c\n",
      "44 HYPOXANPRIBOSYLTRAN-RXN HYPOXANTHINE_c + PRPP_c --> IMP_c + PPI_c\n",
      "\n",
      "45 GUANPRIBOSYLTRAN-RXN GUANINE_c + PRPP_c --> GMP_c + PPI_c\n",
      "45 GUANPRIBOSYLTRAN-RXN GUANINE_c + PRPP_c --> GMP_c + PPI_c\n",
      "\n",
      "46 XANPRIBOSYLTRAN-RXN PRPP_c + XANTHINE_c --> PPI_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "46 XANPRIBOSYLTRAN-RXN PRPP_c + XANTHINE_c --> PPI_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "47 ADENYL-KIN-RXN AMP_c + ATP_c <=> 2.0 ADP_c\n",
      "47 ADENYL-KIN-RXN AMP_c + ATP_c <=> 2.0 ADP_c\n",
      "\n",
      "48 DTDPKIN-RXN ATP_c + TDP_c <=> ADP_c + TTP_c\n",
      "48 DTDPKIN-RXN ATP_c + TDP_c --> ADP_c + TTP_c\n",
      "\n",
      "49 GDPKIN-RXN ATP_c + GDP_c <=> ADP_c + GTP_c\n",
      "49 GDPKIN-RXN ATP_c + GDP_c --> ADP_c + GTP_c\n",
      "\n",
      "50 RXN-14074 AMP_c + GTP_c <=> ADP_c + GDP_c\n",
      "50 RXN-14074 AMP_c + GTP_c --> ADP_c + GDP_c\n",
      "\n",
      "51 DGDPKIN-RXN ATP_c + DGDP_c <=> ADP_c + DGTP_c\n",
      "51 DGDPKIN-RXN ATP_c + DGDP_c --> ADP_c + DGTP_c\n",
      "\n",
      "52 DADPKIN-RXN ATP_c + DADP_c <=> ADP_c + DATP_c\n",
      "52 DADPKIN-RXN ATP_c + DADP_c --> ADP_c + DATP_c\n",
      "\n",
      "53 UDPKIN-RXN ATP_c + UDP_c <=> ADP_c + UTP_c\n",
      "53 UDPKIN-RXN ATP_c + UDP_c --> ADP_c + UTP_c\n",
      "\n",
      "54 CDPKIN-RXN ATP_c + CDP_c <=> ADP_c + CTP_c\n",
      "54 CDPKIN-RXN ATP_c + CDP_c --> ADP_c + CTP_c\n",
      "\n",
      "55 DUDPKIN-RXN ATP_c + DUDP_c <=> ADP_c + DUTP_c\n",
      "55 DUDPKIN-RXN ATP_c + DUDP_c --> ADP_c + DUTP_c\n",
      "\n",
      "56 DCDPKIN-RXN ATP_c + DCDP_c <=> ADP_c + DCTP_c\n",
      "56 DCDPKIN-RXN ATP_c + DCDP_c --> ADP_c + DCTP_c\n",
      "\n",
      "57 DIHYDROOROT-RXN DI-H-OROTATE_c + WATER_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c\n",
      "57 DIHYDROOROT-RXN DI-H-OROTATE_c + WATER_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c\n",
      "\n",
      "58 OROTPDECARB-RXN OROTIDINE-5-PHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + UMP_c\n",
      "58 OROTPDECARB-RXN OROTIDINE-5-PHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + UMP_c\n",
      "\n",
      "59 CITSYN-RXN ACETYL-COA_c + OXALACETIC_ACID_c + WATER_c --> CIT_c + CO-A_c + PROTON_c\n",
      "59 CITSYN-RXN ACETYL-COA_c + OXALACETIC_ACID_c + WATER_c --> CIT_c + CO-A_c + PROTON_c\n",
      "\n",
      "60 ACETALD-DEHYDROG-RXN ACETALD_c + CO-A_c + NAD_c <=> ACETYL-COA_c + NADH_c + PROTON_c\n",
      "60 ACETALD-DEHYDROG-RXN ACETALD_c + CO-A_c + NAD_c <=> ACETYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "61 INORGPYROPHOSPHAT-RXN PPI_c + WATER_c --> PROTON_c + 2.0 Pi_c\n",
      "61 INORGPYROPHOSPHAT-RXN PPI_c + WATER_c --> PROTON_c + 2.0 Pi_c\n",
      "\n",
      "62 TRIPHOSPHATASE-RXN P3I_c + WATER_c --> PPI_c + PROTON_c + Pi_c\n",
      "62 TRIPHOSPHATASE-RXN P3I_c + WATER_c --> PPI_c + Pi_c\n",
      "\n",
      "63 RXN-7607 IMP_c + WATER_c --> INOSINE_c + Pi_c\n",
      "63 RXN-7607 IMP_c + WATER_c --> INOSINE_c + Pi_c\n",
      "\n",
      "64 RXN-14142 DGMP_c + WATER_c --> DEOXYGUANOSINE_c + Pi_c\n",
      "64 RXN-14142 DGMP_c + WATER_c --> DEOXYGUANOSINE_c + Pi_c\n",
      "\n",
      "65 XMPXAN-RXN WATER_c + XANTHOSINE-5-PHOSPHATE_c --> Pi_c + XANTHOSINE_c\n",
      "65 XMPXAN-RXN WATER_c + XANTHOSINE-5-PHOSPHATE_c --> Pi_c + XANTHOSINE_c\n",
      "\n",
      "66 AMP-DEPHOSPHORYLATION-RXN AMP_c + WATER_c --> ADENOSINE_c + Pi_c\n",
      "66 AMP-DEPHOSPHORYLATION-RXN AMP_c + WATER_c --> ADENOSINE_c + Pi_c\n",
      "\n",
      "67 RXN-7609 GMP_c + WATER_c --> GUANOSINE_c + Pi_c\n",
      "67 RXN-7609 GMP_c + WATER_c --> GUANOSINE_c + Pi_c\n",
      "\n",
      "68 PEPCARBOXYKIN-RXN ATP_c + OXALACETIC_ACID_c --> ADP_c + CARBON-DIOXIDE_c + PHOSPHO-ENOL-PYRUVATE_c\n",
      "68 PEPCARBOXYKIN-RXN ATP_c + OXALACETIC_ACID_c --> ADP_c + CARBON-DIOXIDE_c + PHOSPHO-ENOL-PYRUVATE_c\n",
      "\n",
      "69 1.1.1.39-RXN MAL_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "69 1.1.1.39-RXN MAL_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "\n",
      "70 ALARACECAT-RXN L-ALPHA-ALANINE_c <=> D-ALANINE_c\n",
      "70 ALARACECAT-RXN L-ALPHA-ALANINE_c <=> D-ALANINE_c\n",
      "\n",
      "71 ALANINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + L-ALPHA-ALANINE_c <=> GLT_c + PYRUVATE_c\n",
      "71 ALANINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + L-ALPHA-ALANINE_c <=> GLT_c + PYRUVATE_c\n",
      "\n",
      "72 XYLULOKIN-RXN ATP_c + D-XYLULOSE_c --> ADP_c + PROTON_c + XYLULOSE-5-PHOSPHATE_c\n",
      "72 XYLULOKIN-RXN ATP_c + D-XYLULOSE_c --> ADP_c + PROTON_c + XYLULOSE-5-PHOSPHATE_c\n",
      "\n",
      "73 GLYCEROL-KIN-RXN ATP_c + GLYCEROL_c --> ADP_c + GLYCEROL-3P_c + PROTON_c\n",
      "73 GLYCEROL-KIN-RXN ATP_c + GLYCEROL_c <=> ADP_c + GLYCEROL-3P_c + PROTON_c\n",
      "\n",
      "74 D-PPENTOMUT-RXN DEOXY-D-RIBOSE-1-PHOSPHATE_c <=> DEOXY-RIBOSE-5P_c\n",
      "74 D-PPENTOMUT-RXN DEOXY-D-RIBOSE-1-PHOSPHATE_c <=> DEOXY-RIBOSE-5P_c\n",
      "\n",
      "75 PEPDEPHOS-RXN ADP_c + PHOSPHO-ENOL-PYRUVATE_c + PROTON_c --> ATP_c + PYRUVATE_c\n",
      "75 PEPDEPHOS-RXN ADP_c + PHOSPHO-ENOL-PYRUVATE_c + PROTON_c --> ATP_c + PYRUVATE_c\n",
      "\n",
      "76 DARAB5PISOM-RXN RIBULOSE-5P_c <=> ARABINOSE-5P_c\n",
      "76 DARAB5PISOM-RXN ARABINOSE-5P_c <=> RIBULOSE-5P_c\n",
      "\n",
      "77 RXN-8991 2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE_c <=> CPD-9451_c + WATER_c\n",
      "77 RXN-8991 2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE_c <=> CPD-9451_c + WATER_c\n",
      "\n",
      "78 3-ISOPROPYLMALISOM-RXN CPD-9451_c + WATER_c <=> 3-CARBOXY-3-HYDROXY-ISOCAPROATE_c\n",
      "78 3-ISOPROPYLMALISOM-RXN CPD-9451_c + WATER_c <=> 3-CARBOXY-3-HYDROXY-ISOCAPROATE_c\n",
      "\n",
      "79 ACETOLACTSYN-RXN PROTON_c + 2.0 PYRUVATE_c --> 2-ACETO-LACTATE_c + CARBON-DIOXIDE_c\n",
      "79 ACETOLACTSYN-RXN PROTON_c + 2.0 PYRUVATE_c --> 2-ACETO-LACTATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "80 ACETOOHBUTSYN-RXN 2-OXOBUTANOATE_c + PROTON_c + PYRUVATE_c --> 2-ACETO-2-HYDROXY-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "80 ACETOOHBUTSYN-RXN 2-OXOBUTANOATE_c + PROTON_c + PYRUVATE_c --> 2-ACETO-2-HYDROXY-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "81 2.5.1.19-RXN PHOSPHO-ENOL-PYRUVATE_c + SHIKIMATE-5P_c <=> 3-ENOLPYRUVYL-SHIKIMATE-5P_c + Pi_c\n",
      "81 2.5.1.19-RXN PHOSPHO-ENOL-PYRUVATE_c + SHIKIMATE-5P_c <=> 3-ENOLPYRUVYL-SHIKIMATE-5P_c + Pi_c\n",
      "\n",
      "82 PRTRANS-RXN ANTHRANILATE_c + PRPP_c --> N-5-PHOSPHORIBOSYL-ANTHRANILATE_c + PPI_c\n",
      "82 PRTRANS-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c + PPI_c <=> ANTHRANILATE_c + PRPP_c\n",
      "\n",
      "83 ANTHRANSYN-RXN CHORISMATE_c + GLN_c --> ANTHRANILATE_c + GLT_c + PROTON_c + PYRUVATE_c\n",
      "83 ANTHRANSYN-RXN CHORISMATE_c + GLN_c <=> ANTHRANILATE_c + GLT_c + PROTON_c + PYRUVATE_c\n",
      "\n",
      "84 PREPHENATEDEHYDRAT-RXN PREPHENATE_c + PROTON_c --> CARBON-DIOXIDE_c + PHENYL-PYRUVATE_c + WATER_c\n",
      "84 PREPHENATEDEHYDRAT-RXN PREPHENATE_c + PROTON_c <=> CARBON-DIOXIDE_c + PHENYL-PYRUVATE_c + WATER_c\n",
      "\n",
      "85 CHORISMATEMUT-RXN CHORISMATE_c --> PREPHENATE_c\n",
      "85 CHORISMATEMUT-RXN CHORISMATE_c <=> PREPHENATE_c\n",
      "\n",
      "86 CHORISMATE-SYNTHASE-RXN 3-ENOLPYRUVYL-SHIKIMATE-5P_c --> CHORISMATE_c + Pi_c\n",
      "86 CHORISMATE-SYNTHASE-RXN 3-ENOLPYRUVYL-SHIKIMATE-5P_c --> CHORISMATE_c + Pi_c\n",
      "\n",
      "87 IGPSYN-RXN CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c + PROTON_c --> CARBON-DIOXIDE_c + INDOLE-3-GLYCEROL-P_c + WATER_c\n",
      "87 IGPSYN-RXN CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c + PROTON_c --> CARBON-DIOXIDE_c + INDOLE-3-GLYCEROL-P_c + WATER_c\n",
      "\n",
      "88 PRAISOM-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c --> CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c\n",
      "88 PRAISOM-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c --> CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c\n",
      "\n",
      "89 2PGADEHYDRAT-RXN 2-PG_c <=> PHOSPHO-ENOL-PYRUVATE_c + WATER_c\n",
      "89 2PGADEHYDRAT-RXN 2-PG_c <=> PHOSPHO-ENOL-PYRUVATE_c + WATER_c\n",
      "\n",
      "90 METHGLYSYN-RXN DIHYDROXY-ACETONE-PHOSPHATE_c --> METHYL-GLYOXAL_c + Pi_c\n",
      "90 METHGLYSYN-RXN DIHYDROXY-ACETONE-PHOSPHATE_c --> METHYL-GLYOXAL_c + Pi_c\n",
      "\n",
      "91 PEPSYNTH-RXN ATP_c + PYRUVATE_c + WATER_c --> AMP_c + PHOSPHO-ENOL-PYRUVATE_c + 2.0 PROTON_c + Pi_c\n",
      "91 PEPSYNTH-RXN ATP_c + PYRUVATE_c + WATER_c <=> AMP_c + PHOSPHO-ENOL-PYRUVATE_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "92 PGLUCISOM-RXN D-glucopyranose-6-phosphate_c <=> FRUCTOSE-6P_c\n",
      "92 PGLUCISOM-RXN D-glucopyranose-6-phosphate_c <=> FRUCTOSE-6P_c\n",
      "\n",
      "93 PHOSGLYPHOS-RXN ATP_c + G3P_c <=> ADP_c + DPG_c\n",
      "93 PHOSGLYPHOS-RXN ATP_c + G3P_c <=> ADP_c + DPG_c\n",
      "\n",
      "94 KDPGALDOL-RXN 2-KETO-3-DEOXY-6-P-GLUCONATE_c --> GAP_c + PYRUVATE_c\n",
      "94 KDPGALDOL-RXN 2-KETO-3-DEOXY-6-P-GLUCONATE_c --> GAP_c + PYRUVATE_c\n",
      "\n",
      "95 OXALODECARB-RXN OXALACETIC_ACID_c + PROTON_c --> CARBON-DIOXIDE_c + PYRUVATE_c\n",
      "95 OXALODECARB-RXN OXALACETIC_ACID_c + PROTON_c --> CARBON-DIOXIDE_c + PYRUVATE_c\n",
      "\n",
      "96 RIBULP3EPIM-RXN RIBULOSE-5P_c <=> XYLULOSE-5-PHOSPHATE_c\n",
      "96 RIBULP3EPIM-RXN RIBULOSE-5P_c <=> XYLULOSE-5-PHOSPHATE_c\n",
      "\n",
      "97 TRANSALDOL-RXN D-SEDOHEPTULOSE-7-P_c + GAP_c <=> ERYTHROSE-4P_c + FRUCTOSE-6P_c\n",
      "97 TRANSALDOL-RXN D-SEDOHEPTULOSE-7-P_c + GAP_c <=> ERYTHROSE-4P_c + FRUCTOSE-6P_c\n",
      "\n",
      "98 2TRANSKETO-RXN ERYTHROSE-4P_c + XYLULOSE-5-PHOSPHATE_c <=> FRUCTOSE-6P_c + GAP_c\n",
      "98 2TRANSKETO-RXN ERYTHROSE-4P_c + XYLULOSE-5-PHOSPHATE_c <=> FRUCTOSE-6P_c + GAP_c\n",
      "\n",
      "99 ALCOHOL-DEHYDROG-RXN ETOH_c + NAD_c <=> ACETALD_c + NADH_c + PROTON_c\n",
      "99 ALCOHOL-DEHYDROG-RXN ETOH_c + NAD_c <=> ACETALD_c + NADH_c + PROTON_c\n",
      "\n",
      "100 2.7.7.1-RXN ATP_c + NICOTINAMIDE_NUCLEOTIDE_c + PROTON_c --> NAD_c + PPI_c\n",
      "100 2.7.7.1-RXN ATP_c + NICOTINAMIDE_NUCLEOTIDE_c + PROTON_c --> NAD_c + PPI_c\n",
      "\n",
      "101 NICONUCADENYLYLTRAN-RXN ATP_c + NICOTINATE_NUCLEOTIDE_c + PROTON_c <=> DEAMIDO-NAD_c + PPI_c\n",
      "101 NICONUCADENYLYLTRAN-RXN ATP_c + NICOTINATE_NUCLEOTIDE_c + PROTON_c --> DEAMIDO-NAD_c + PPI_c\n",
      "\n",
      "102 PYRIMSYN3-RXN AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + ATP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c\n",
      "102 PYRIMSYN3-RXN AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + ATP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c\n",
      "\n",
      "103 OHMETPYRKIN-RXN ATP_c + HMP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + PROTON_c\n",
      "103 OHMETPYRKIN-RXN ATP_c + HMP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + PROTON_c\n",
      "\n",
      "104 DEOXYINOPHOSPHOR-RXN DEOXYINOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + HYPOXANTHINE_c\n",
      "104 DEOXYINOPHOSPHOR-RXN DEOXYINOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + HYPOXANTHINE_c\n",
      "\n",
      "105 ADENPHOSPHOR-RXN ADENOSINE_c + Pi_c <=> ADENINE_c + RIBOSE-1P_c\n",
      "105 ADENPHOSPHOR-RXN ADENOSINE_c + Pi_c <=> ADENINE_c + RIBOSE-1P_c\n",
      "\n",
      "106 INOPHOSPHOR-RXN INOSINE_c + Pi_c <=> HYPOXANTHINE_c + RIBOSE-1P_c\n",
      "106 INOPHOSPHOR-RXN INOSINE_c + Pi_c <=> HYPOXANTHINE_c + RIBOSE-1P_c\n",
      "\n",
      "107 DEOXYGUANPHOSPHOR-RXN DEOXYGUANOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + GUANINE_c\n",
      "107 DEOXYGUANPHOSPHOR-RXN DEOXYGUANOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + GUANINE_c\n",
      "\n",
      "108 RXN0-5199 GUANOSINE_c + Pi_c <=> GUANINE_c + RIBOSE-1P_c\n",
      "108 RXN0-5199 GUANOSINE_c + Pi_c <=> GUANINE_c + RIBOSE-1P_c\n",
      "\n",
      "109 DEOXYADENPHOSPHOR-RXN DEOXYADENOSINE_c + Pi_c <=> ADENINE_c + DEOXY-D-RIBOSE-1-PHOSPHATE_c\n",
      "109 DEOXYADENPHOSPHOR-RXN DEOXYADENOSINE_c + Pi_c <=> ADENINE_c + DEOXY-D-RIBOSE-1-PHOSPHATE_c\n",
      "\n",
      "110 OROPRIBTRANS-RXN OROTIDINE-5-PHOSPHATE_c + PPI_c <=> OROTATE_c + PRPP_c\n",
      "110 OROPRIBTRANS-RXN OROTIDINE-5-PHOSPHATE_c + PPI_c <=> OROTATE_c + PRPP_c\n",
      "\n",
      "111 DIHYDROXYISOVALDEHYDRAT-RXN CPD-13357_c --> 2-KETO-ISOVALERATE_c + WATER_c\n",
      "111 DIHYDROXYISOVALDEHYDRAT-RXN CPD-13357_c --> 2-KETO-ISOVALERATE_c + WATER_c\n",
      "\n",
      "112 DIHYDROXYMETVALDEHYDRAT-RXN 1-KETO-2-METHYLVALERATE_c --> 2-KETO-3-METHYL-VALERATE_c + WATER_c\n",
      "112 DIHYDROXYMETVALDEHYDRAT-RXN 1-KETO-2-METHYLVALERATE_c --> 2-KETO-3-METHYL-VALERATE_c + WATER_c\n",
      "\n",
      "113 PHOSACETYLTRANS-RXN ACETYL-COA_c + Pi_c <=> ACETYL-P_c + CO-A_c\n",
      "113 PHOSACETYLTRANS-RXN ACETYL-COA_c + Pi_c <=> ACETYL-P_c + CO-A_c\n",
      "\n",
      "114 PTAALT-RXN PROPIONYL-COA_c + Pi_c --> CO-A_c + PROPIONYL-P_c\n",
      "114 PTAALT-RXN PROPIONYL-COA_c + Pi_c <=> CO-A_c + PROPIONYL-P_c\n",
      "\n",
      "115 ACETATEKIN-RXN ACET_c + ATP_c <=> ACETYL-P_c + ADP_c\n",
      "115 ACETATEKIN-RXN ACET_c + ATP_c <=> ACETYL-P_c + ADP_c\n",
      "\n",
      "116 KDOTRANS-RXN CMP-KDO_c + LIPID-IV-A_c --> CMP_c + KDO-LIPID-IVA_c + PROTON_c\n",
      "116 KDOTRANS-RXN CMP-KDO_c + LIPID-IV-A_c --> CMP_c + KDO-LIPID-IVA_c + PROTON_c\n",
      "\n",
      "117 KDOTRANS2-RXN CMP-KDO_c + KDO-LIPID-IVA_c --> CMP_c + KDO2-LIPID-IVA_c + PROTON_c\n",
      "117 KDOTRANS2-RXN CMP-KDO_c + KDO-LIPID-IVA_c --> CMP_c + KDO2-LIPID-IVA_c + PROTON_c\n",
      "\n",
      "118 LIPIDXSYNTHESIS-RXN OH-MYRISTOYL_c + WATER_c --> BISOHMYR-GLUCOSAMINYL-1P_c + 2.0 PROTON_c + UMP_c\n",
      "118 LIPIDXSYNTHESIS-RXN OH-MYRISTOYL_c + WATER_c --> BISOHMYR-GLUCOSAMINYL-1P_c + 2.0 PROTON_c + UMP_c\n",
      "\n",
      "119 LIPIDADISACCHARIDESYNTH-RXN BISOHMYR-GLUCOSAMINYL-1P_c + OH-MYRISTOYL_c --> BISOHMYR-GLC_c + PROTON_c + UDP_c\n",
      "119 LIPIDADISACCHARIDESYNTH-RXN BISOHMYR-GLUCOSAMINYL-1P_c + OH-MYRISTOYL_c --> BISOHMYR-GLC_c + PROTON_c + UDP_c\n",
      "\n",
      "120 TETRAACYLDISACC4KIN-RXN ATP_c + BISOHMYR-GLC_c --> ADP_c + LIPID-IV-A_c + PROTON_c\n",
      "120 TETRAACYLDISACC4KIN-RXN ATP_c + BISOHMYR-GLC_c --> ADP_c + LIPID-IV-A_c + PROTON_c\n",
      "\n",
      "121 RXN0-5107 TTP_c + WATER_c --> PPI_c + PROTON_c + TMP_c\n",
      "121 RXN0-5107 TTP_c + WATER_c --> PPI_c + PROTON_c + TMP_c\n",
      "\n",
      "122 RXN-14139 UTP_c + WATER_c --> PPI_c + PROTON_c + UMP_c\n",
      "122 RXN-14139 UTP_c + WATER_c --> PPI_c + PROTON_c + UMP_c\n",
      "\n",
      "123 ATP-PYROPHOSPHATASE-RXN ATP_c + WATER_c --> AMP_c + PPI_c + PROTON_c\n",
      "123 ATP-PYROPHOSPHATASE-RXN ATP_c + WATER_c --> AMP_c + PPI_c + PROTON_c\n",
      "\n",
      "124 1.1.1.83-RXN CPD-660_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "124 1.1.1.83-RXN CPD-660_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "\n",
      "125 RXN0-743 PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c <=> CPD0-181_c\n",
      "125 RXN0-743 CPD0-181_c --> PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c\n",
      "\n",
      "126 AMPSYN-RXN ADENYLOSUCC_c <=> AMP_c + FUM_c\n",
      "126 AMPSYN-RXN ADENYLOSUCC_c <=> AMP_c + FUM_c\n",
      "\n",
      "127 AICARSYN-RXN P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c <=> AICAR_c + FUM_c\n",
      "127 AICARSYN-RXN P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c <=> AICAR_c + FUM_c\n",
      "\n",
      "128 1.8.4.14-RXN CPD-8989_c + Red-Thioredoxin_c --> MET_c + Ox-Thioredoxin_c + WATER_c\n",
      "128 1.8.4.14-RXN CPD-8990_c + Red-Thioredoxin_c --> MET_c + Ox-Thioredoxin_c + WATER_c\n",
      "\n",
      "129 THIOREDOXIN-REDUCT-NADPH-RXN NADPH_c + Ox-Thioredoxin_c + PROTON_c --> NADP_c + Red-Thioredoxin_c\n",
      "129 THIOREDOXIN-REDUCT-NADPH-RXN NADPH_c + Ox-Thioredoxin_c + PROTON_c --> NADP_c + Red-Thioredoxin_c\n",
      "\n",
      "130 1.8.4.8-RXN PAPS_c + Red-Thioredoxin_c --> 3-5-ADP_c + Ox-Thioredoxin_c + 2.0 PROTON_c + SO3_c\n",
      "130 1.8.4.8-RXN 3-5-ADP_c + Ox-Thioredoxin_c + 2.0 PROTON_c + SO3_c <=> PAPS_c + Red-Thioredoxin_c\n",
      "\n",
      "131 THYMIDYLATESYN-RXN DUMP_c + METHYLENE-THF_c --> DIHYDROFOLATE_c + TMP_c\n",
      "131 THYMIDYLATESYN-RXN DUMP_c + METHYLENE-THF_c --> DIHYDROFOLATE_c + TMP_c\n",
      "\n",
      "132 P-PANTOCYSLIG-RXN 4-P-PANTOTHENATE_c + CTP_c + CYS_c --> CMP_c + PPI_c + PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c\n",
      "132 P-PANTOCYSLIG-RXN 4-P-PANTOTHENATE_c + CTP_c + CYS_c --> CMP_c + PPI_c + PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c\n",
      "\n",
      "133 P-PANTOCYSDECARB-RXN PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c --> CARBON-DIOXIDE_c + PANTETHEINE-P_c\n",
      "133 P-PANTOCYSDECARB-RXN PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c --> CARBON-DIOXIDE_c + PANTETHEINE-P_c\n",
      "\n",
      "134 FORMYLTHFDEFORMYL-RXN 10-FORMYL-THF_c + WATER_c --> FORMATE_c + PROTON_c + THF_c\n",
      "134 FORMYLTHFDEFORMYL-RXN 10-FORMYL-THF_c + WATER_c --> FORMATE_c + PROTON_c + THF_c\n",
      "\n",
      "135 PGLUCONDEHYDRAT-RXN CPD-2961_c --> 2-KETO-3-DEOXY-6-P-GLUCONATE_c + WATER_c\n",
      "135 PGLUCONDEHYDRAT-RXN CPD-2961_c --> 2-KETO-3-DEOXY-6-P-GLUCONATE_c + WATER_c\n",
      "\n",
      "136 FUMHYDR-RXN FUM_c + WATER_c <=> MAL_c\n",
      "136 FUMHYDR-RXN MAL_c <=> FUM_c + WATER_c\n",
      "\n",
      "137 D--TARTRATE-DEHYDRATASE-RXN D-TARTRATE_c --> OXALACETIC_ACID_c + WATER_c\n",
      "137 D--TARTRATE-DEHYDRATASE-RXN D-TARTRATE_c <=> OXALACETIC_ACID_c + WATER_c\n",
      "\n",
      "138 PMPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXAMINE-5P_c + WATER_c --> AMMONIUM_c + HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "138 PMPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXAMINE-5P_c + WATER_c --> AMMONIUM_c + HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "\n",
      "139 PNPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXINE-5P_c --> HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "139 PNPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXINE-5P_c --> HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "\n",
      "140 PSERTRANSAMPYR-RXN 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + GLT_c <=> 2-KETOGLUTARATE_c + 4-PHOSPHONOOXY-THREONINE_c\n",
      "140 PSERTRANSAMPYR-RXN 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + GLT_c <=> 2-KETOGLUTARATE_c + 4-PHOSPHONOOXY-THREONINE_c\n",
      "\n",
      "141 PSERTRANSAM-RXN 3-P-HYDROXYPYRUVATE_c + GLT_c --> 2-KETOGLUTARATE_c + 3-P-SERINE_c\n",
      "141 PSERTRANSAM-RXN 2-KETOGLUTARATE_c + 3-P-SERINE_c <=> 3-P-HYDROXYPYRUVATE_c + GLT_c\n",
      "\n",
      "142 SUCCINYLDIAMINOPIMTRANS-RXN 2-KETOGLUTARATE_c + N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c <=> GLT_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c\n",
      "142 SUCCINYLDIAMINOPIMTRANS-RXN 2-KETOGLUTARATE_c + N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c <=> GLT_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c\n",
      "\n",
      "143 HEMEOSYN-RXN FARNESYL-PP_c + PROTOHEME_c + WATER_c --> CPD-17063_c + PPI_c\n",
      "143 HEMEOSYN-RXN FARNESYL-PP_c + PROTOHEME_c + WATER_c --> CPD-17063_c + PPI_c\n",
      "\n",
      "144 AGMATIN-RXN AGMATHINE_c + WATER_c --> PUTRESCINE_c + UREA_c\n",
      "144 AGMATIN-RXN AGMATHINE_c + WATER_c --> PUTRESCINE_c + UREA_c\n",
      "\n",
      "145 ARGDECARBOX-RXN ARG_c + PROTON_c --> AGMATHINE_c + CARBON-DIOXIDE_c\n",
      "145 ARGDECARBOX-RXN ARG_c + PROTON_c <=> AGMATHINE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "146 ARGSUCCINSYN-RXN ATP_c + L-ASPARTATE_c + L-CITRULLINE_c --> AMP_c + L-ARGININO-SUCCINATE_c + PPI_c + PROTON_c\n",
      "146 ARGSUCCINSYN-RXN ATP_c + L-ASPARTATE_c + L-CITRULLINE_c --> AMP_c + L-ARGININO-SUCCINATE_c + PPI_c + PROTON_c\n",
      "\n",
      "147 SPERMIDINESYN-RXN PUTRESCINE_c + S-ADENOSYLMETHIONINAMINE_c --> 5-METHYLTHIOADENOSINE_c + PROTON_c + SPERMIDINE_c\n",
      "147 SPERMIDINESYN-RXN PUTRESCINE_c + S-ADENOSYLMETHIONINAMINE_c <=> 5-METHYLTHIOADENOSINE_c + PROTON_c + SPERMIDINE_c\n",
      "\n",
      "148 GSAAMINOTRANS-RXN GLUTAMATE-1-SEMIALDEHYDE_c <=> 5-AMINO-LEVULINATE_c\n",
      "148 GSAAMINOTRANS-RXN GLUTAMATE-1-SEMIALDEHYDE_c --> 5-AMINO-LEVULINATE_c\n",
      "\n",
      "149 CPM-KDOSYNTH-RXN CTP_c + KDO_c --> CMP-KDO_c + PPI_c\n",
      "149 CPM-KDOSYNTH-RXN CTP_c + KDO_c --> CMP-KDO_c + PPI_c\n",
      "\n",
      "150 5.4.2.10-RXN GLUCOSAMINE-1P_c <=> D-GLUCOSAMINE-6-P_c\n",
      "150 5.4.2.10-RXN D-GLUCOSAMINE-6-P_c <=> GLUCOSAMINE-1P_c\n",
      "\n",
      "151 UDPNACETYLGLUCOSAMENOLPYRTRANS-RXN PHOSPHO-ENOL-PYRUVATE_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> Pi_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c\n",
      "151 UDPNACETYLGLUCOSAMENOLPYRTRANS-RXN PHOSPHO-ENOL-PYRUVATE_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> Pi_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c\n",
      "\n",
      "152 GLUTRACE-RXN D-GLT_c <=> GLT_c\n",
      "152 GLUTRACE-RXN GLT_c <=> D-GLT_c\n",
      "\n",
      "153 UDPGLCNACEPIM-RXN UDP-N-ACETYL-D-GLUCOSAMINE_c <=> UDP-MANNAC_c\n",
      "153 UDPGLCNACEPIM-RXN UDP-N-ACETYL-D-GLUCOSAMINE_c <=> UDP-MANNAC_c\n",
      "\n",
      "154 MALONYL-ACPDECARBOX-RXN MALONYL-ACP_c + PROTON_c --> ACETYL-ACP_c + CARBON-DIOXIDE_c\n",
      "154 MALONYL-ACPDECARBOX-RXN MALONYL-ACP_c + PROTON_c --> ACETYL-ACP_c + CARBON-DIOXIDE_c\n",
      "\n",
      "155 3-OXOACYL-ACP-SYNTH-BASE-RXN ACETYL-ACP_c + MALONYL-ACP_c + PROTON_c --> ACP_c + Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c\n",
      "155 3-OXOACYL-ACP-SYNTH-BASE-RXN ACETYL-ACP_c + MALONYL-ACP_c + PROTON_c --> ACP_c + Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c\n",
      "\n",
      "156 DHHB-METHYLTRANSFER-RXN OCTAPRENYL-METHYL-OH-METHOXY-BENZQ_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + CPD-9956_c + PROTON_c\n",
      "156 DHHB-METHYLTRANSFER-RXN OCTAPRENYL-METHYL-OH-METHOXY-BENZQ_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + CPD-9956_c + PROTON_c\n",
      "\n",
      "157 2-OCTAPRENYL-6-OHPHENOL-METHY-RXN 2-OCTAPRENYL-6-HYDROXYPHENOL_c + S-ADENOSYLMETHIONINE_c --> 2-OCTAPRENYL-6-METHOXYPHENOL_c + ADENOSYL-HOMO-CYS_c + PROTON_c\n",
      "157 2-OCTAPRENYL-6-OHPHENOL-METHY-RXN 2-OCTAPRENYL-6-HYDROXYPHENOL_c + S-ADENOSYLMETHIONINE_c --> 2-OCTAPRENYL-6-METHOXYPHENOL_c + ADENOSYL-HOMO-CYS_c + PROTON_c\n",
      "\n",
      "158 H2PTERIDINEPYROPHOSPHOKIN-RXN AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE_c + ATP_c --> AMP_c + DIHYDROPTERIN-CH2OH-PP_c + PROTON_c\n",
      "158 H2PTERIDINEPYROPHOSPHOKIN-RXN AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE_c + ATP_c --> AMP_c + DIHYDROPTERIN-CH2OH-PP_c + PROTON_c\n",
      "\n",
      "159 PANTEPADENYLYLTRAN-RXN ATP_c + PANTETHEINE-P_c + PROTON_c --> DEPHOSPHO-COA_c + PPI_c\n",
      "159 PANTEPADENYLYLTRAN-RXN ATP_c + PANTETHEINE-P_c + PROTON_c --> DEPHOSPHO-COA_c + PPI_c\n",
      "\n",
      "160 DXPREDISOM-RXN DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c --> 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + NADP_c\n",
      "160 DXPREDISOM-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + NADP_c <=> DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c\n",
      "\n",
      "161 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PROTON_c --> 2-OCTAPRENYLPHENOL_c + CARBON-DIOXIDE_c\n",
      "161 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PROTON_c --> 2-OCTAPRENYLPHENOL_c + CARBON-DIOXIDE_c\n",
      "\n",
      "162 GABATRANSAM-RXN 2-KETOGLUTARATE_c + 4-AMINO-BUTYRATE_c --> GLT_c + SUCC-S-ALD_c\n",
      "162 GABATRANSAM-RXN 2-KETOGLUTARATE_c + 4-AMINO-BUTYRATE_c <=> GLT_c + SUCC-S-ALD_c\n",
      "\n",
      "163 ACETYLORNTRANSAM-RXN 2-KETOGLUTARATE_c + N-ALPHA-ACETYLORNITHINE_c <=> CPD-469_c + GLT_c\n",
      "163 ACETYLORNTRANSAM-RXN 2-KETOGLUTARATE_c + N-ALPHA-ACETYLORNITHINE_c <=> CPD-469_c + GLT_c\n",
      "\n",
      "164 SUCCSEMIALDDEHYDROG-RXN NADP_c + SUCC-S-ALD_c + WATER_c --> NADPH_c + 2.0 PROTON_c + SUC_c\n",
      "164 SUCCSEMIALDDEHYDROG-RXN NADP_c + SUCC-S-ALD_c + WATER_c --> NADPH_c + 2.0 PROTON_c + SUC_c\n",
      "\n",
      "165 SUCCINATE-SEMIALDEHYDE-DEHYDROGENASE-RXN NAD_c + SUCC-S-ALD_c + WATER_c --> NADH_c + 2.0 PROTON_c + SUC_c\n",
      "165 SUCCINATE-SEMIALDEHYDE-DEHYDROGENASE-RXN NAD_c + SUCC-S-ALD_c + WATER_c --> NADH_c + 2.0 PROTON_c + SUC_c\n",
      "\n",
      "166 4.3.1.17-RXN SER_c --> AMMONIUM_c + PYRUVATE_c\n",
      "166 4.3.1.17-RXN SER_c --> AMMONIUM_c + PYRUVATE_c\n",
      "\n",
      "167 ACETOLACTREDUCTOISOM-RXN CPD-13357_c + NADP_c <=> 2-ACETO-LACTATE_c + NADPH_c + PROTON_c\n",
      "167 ACETOLACTREDUCTOISOM-RXN 2-ACETO-LACTATE_c + NADPH_c + PROTON_c --> CPD-13357_c + NADP_c\n",
      "\n",
      "168 ACETOOHBUTREDUCTOISOM-RXN 2-ACETO-2-HYDROXY-BUTYRATE_c + NADPH_c + PROTON_c <=> 1-KETO-2-METHYLVALERATE_c + NADP_c\n",
      "168 ACETOOHBUTREDUCTOISOM-RXN 2-ACETO-2-HYDROXY-BUTYRATE_c + NADPH_c + PROTON_c --> 1-KETO-2-METHYLVALERATE_c + NADP_c\n",
      "\n",
      "169 NAD-KIN-RXN ATP_c + NAD_c --> ADP_c + NADP_c + PROTON_c\n",
      "169 NAD-KIN-RXN ATP_c + NAD_c --> ADP_c + NADP_c + PROTON_c\n",
      "\n",
      "170 GLYOXI-RXN GLUTATHIONE_c + METHYL-GLYOXAL_c --> S-LACTOYL-GLUTATHIONE_c\n",
      "170 GLYOXI-RXN S-LACTOYL-GLUTATHIONE_c <=> GLUTATHIONE_c + METHYL-GLYOXAL_c\n",
      "\n",
      "171 3.1.3.74-RXN PYRIDOXAL_PHOSPHATE_c + WATER_c --> PYRIDOXAL_c + Pi_c\n",
      "171 3.1.3.74-RXN PYRIDOXAL_PHOSPHATE_c + WATER_c --> PYRIDOXAL_c + Pi_c\n",
      "\n",
      "172 DXS-RXN GAP_c + PROTON_c + PYRUVATE_c --> CARBON-DIOXIDE_c + DEOXYXYLULOSE-5P_c\n",
      "172 DXS-RXN GAP_c + PROTON_c + PYRUVATE_c --> CARBON-DIOXIDE_c + DEOXYXYLULOSE-5P_c\n",
      "\n",
      "173 2.7.1.148-RXN 4-CYTIDINE-5-DIPHOSPHO-2-C_c + ATP_c --> 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c + ADP_c + PROTON_c\n",
      "173 2.7.1.148-RXN 4-CYTIDINE-5-DIPHOSPHO-2-C_c + ATP_c --> 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c + ADP_c + PROTON_c\n",
      "\n",
      "174 RXN0-302 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c --> 2C-METH-D-ERYTHRITOL-CYCLODIPHOSPHATE_c + CMP_c\n",
      "174 RXN0-302 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c --> 2C-METH-D-ERYTHRITOL-CYCLODIPHOSPHATE_c + CMP_c\n",
      "\n",
      "175 PHOSMANMUT-RXN MANNOSE-1P_c <=> CPD-15979_c\n",
      "175 PHOSMANMUT-RXN MANNOSE-1P_c <=> CPD-15979_c\n",
      "\n",
      "176 PYRUVDEH-RXN CO-A_c + NAD_c + PYRUVATE_c --> ACETYL-COA_c + CARBON-DIOXIDE_c + NADH_c\n",
      "176 PYRUVDEH-RXN CO-A_c + NAD_c + PYRUVATE_c --> ACETYL-COA_c + CARBON-DIOXIDE_c + NADH_c\n",
      "\n",
      "177 2OXOGLUTARATEDEH-RXN 2-KETOGLUTARATE_c + CO-A_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + SUC-COA_c\n",
      "177 2OXOGLUTARATEDEH-RXN 2-KETOGLUTARATE_c + CO-A_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + SUC-COA_c\n",
      "\n",
      "178 MMUM-RXN CPD-397_c + HOMO-CYS_c --> 2.0 MET_c + PROTON_c\n",
      "178 MMUM-RXN CPD-397_c + HOMO-CYS_c --> 2.0 MET_c + PROTON_c\n",
      "\n",
      "179 TRE6PHYDRO-RXN TREHALOSE-6P_c + WATER_c --> D-glucopyranose-6-phosphate_c + Glucopyranose_c\n",
      "179 TRE6PHYDRO-RXN TREHALOSE-6P_c + WATER_c --> D-glucopyranose-6-phosphate_c + Glucopyranose_c\n",
      "\n",
      "180 GTP-CYCLOHYDRO-II-RXN GTP_c + 3.0 WATER_c --> DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + FORMATE_c + PPI_c + 2.0 PROTON_c\n",
      "180 GTP-CYCLOHYDRO-II-RXN GTP_c + 3.0 WATER_c --> DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + FORMATE_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "181 MANNONOXIDOREDUCT-RXN D-MANNONATE_c + NAD_c <=> FRUCTURONATE_c + NADH_c + PROTON_c\n",
      "181 MANNONOXIDOREDUCT-RXN D-MANNONATE_c + NAD_c <=> FRUCTURONATE_c + NADH_c + PROTON_c\n",
      "\n",
      "182 GLYOCARBOLIG-RXN 2.0 GLYOX_c + PROTON_c --> CARBON-DIOXIDE_c + TARTRONATE-S-ALD_c\n",
      "182 GLYOCARBOLIG-RXN 2.0 GLYOX_c + PROTON_c --> CARBON-DIOXIDE_c + TARTRONATE-S-ALD_c\n",
      "\n",
      "183 QUINOPRIBOTRANS-RXN 2.0 PROTON_c + PRPP_c + QUINOLINATE_c --> CARBON-DIOXIDE_c + NICOTINATE_NUCLEOTIDE_c + PPI_c\n",
      "183 QUINOPRIBOTRANS-RXN 2.0 PROTON_c + PRPP_c + QUINOLINATE_c --> CARBON-DIOXIDE_c + NICOTINATE_NUCLEOTIDE_c + PPI_c\n",
      "\n",
      "184 MALONYL-COA-ACP-TRANSACYL-RXN ACP_c + MALONYL-COA_c <=> CO-A_c + MALONYL-ACP_c\n",
      "184 MALONYL-COA-ACP-TRANSACYL-RXN ACP_c + MALONYL-COA_c --> CO-A_c + MALONYL-ACP_c\n",
      "\n",
      "185 UDPNACETYLGLUCOSAMACYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-N-ACETYL-D-GLUCOSAMINE_c <=> ACP_c + UDP-OHMYR-ACETYLGLUCOSAMINE_c\n",
      "185 UDPNACETYLGLUCOSAMACYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> ACP_c + UDP-OHMYR-ACETYLGLUCOSAMINE_c\n",
      "\n",
      "186 SUCCDIAMINOPIMDESUCC-RXN N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c + WATER_c --> LL-DIAMINOPIMELATE_c + SUC_c\n",
      "186 SUCCDIAMINOPIMDESUCC-RXN N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c + WATER_c --> LL-DIAMINOPIMELATE_c + SUC_c\n",
      "\n",
      "187 ADENPRIBOSYLTRAN-RXN ADENINE_c + PRPP_c --> AMP_c + PPI_c\n",
      "187 ADENPRIBOSYLTRAN-RXN ADENINE_c + PRPP_c --> AMP_c + PPI_c\n",
      "\n",
      "188 RXN-8992 5.0 DELTA3-ISOPENTENYL-PP_c + FARNESYL-PP_c --> OCTAPRENYL-DIPHOSPHATE_c + 5.0 PPI_c\n",
      "188 RXN-8992 5.0 DELTA3-ISOPENTENYL-PP_c + FARNESYL-PP_c --> OCTAPRENYL-DIPHOSPHATE_c + 5.0 PPI_c\n",
      "\n",
      "189 CATAL-RXN 2.0 HYDROGEN-PEROXIDE_c --> OXYGEN-MOLECULE_c + 2.0 WATER_c\n",
      "189 CATAL-RXN 2.0 HYDROGEN-PEROXIDE_c --> OXYGEN-MOLECULE_c + 2.0 WATER_c\n",
      "\n",
      "190 RXN-10819 ATP_c + CO-A_c + PHENYLACETATE_c --> AMP_c + CPD-207_c + PPI_c\n",
      "190 RXN-10819 ATP_c + CO-A_c + PHENYLACETATE_c --> AMP_c + CPD-207_c + PPI_c\n",
      "\n",
      "191 3.6.1.41-RXN ADENOSYL-P4_c + WATER_c --> 2.0 ADP_c + 2.0 PROTON_c\n",
      "191 3.6.1.41-RXN ADENOSYL-P4_c + WATER_c --> 2.0 ADP_c + 2.0 PROTON_c\n",
      "\n",
      "192 RXN0-4342 ATP_c + D-BETA-D-HEPTOSE-1-P_c + PROTON_c --> ADP-D-GLYCERO-D-MANNO-HEPTOSE_c + PPI_c\n",
      "192 RXN0-4342 ATP_c + D-BETA-D-HEPTOSE-1-P_c + PROTON_c --> ADP-D-GLYCERO-D-MANNO-HEPTOSE_c + PPI_c\n",
      "\n",
      "193 RXN-7913 ATP_c + DCMP_c <=> ADP_c + DCDP_c\n",
      "193 RXN-7913 ATP_c + DCMP_c --> ADP_c + DCDP_c\n",
      "\n",
      "194 RXN-11832 ATP_c + CMP_c <=> ADP_c + CDP_c\n",
      "194 RXN-11832 ATP_c + CMP_c <=> ADP_c + CDP_c\n",
      "\n",
      "195 RXN-12002 ATP_c + UMP_c <=> ADP_c + UDP_c\n",
      "195 RXN-12002 ATP_c + UMP_c --> ADP_c + UDP_c\n",
      "\n",
      "196 GUANYL-KIN-RXN ATP_c + GMP_c <=> ADP_c + GDP_c\n",
      "196 GUANYL-KIN-RXN ATP_c + GMP_c <=> ADP_c + GDP_c\n",
      "\n",
      "197 GMKALT-RXN ATP_c + DGMP_c <=> ADP_c + DGDP_c\n",
      "197 GMKALT-RXN ATP_c + DGMP_c <=> ADP_c + DGDP_c\n",
      "\n",
      "198 ACETOACETYL-COA-TRANSFER-RXN 3-KETOBUTYRATE_c + ACETYL-COA_c --> ACETOACETYL-COA_c + ACET_c\n",
      "198 ACETOACETYL-COA-TRANSFER-RXN 3-KETOBUTYRATE_c + ACETYL-COA_c <=> ACETOACETYL-COA_c + ACET_c\n",
      "\n",
      "199 DTDPGLUCOSEPP-RXN GLC-1-P_c + PROTON_c + TTP_c --> DTDP-D-GLUCOSE_c + PPI_c\n",
      "199 DTDPGLUCOSEPP-RXN GLC-1-P_c + PROTON_c + TTP_c --> DTDP-D-GLUCOSE_c + PPI_c\n",
      "\n",
      "200 DTDPGLUCDEHYDRAT-RXN DTDP-D-GLUCOSE_c --> DTDP-DEOH-DEOXY-GLUCOSE_c + WATER_c\n",
      "200 DTDPGLUCDEHYDRAT-RXN DTDP-D-GLUCOSE_c --> DTDP-DEOH-DEOXY-GLUCOSE_c + WATER_c\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "201 DTDPDEHYDRHAMEPIM-RXN DTDP-DEOH-DEOXY-GLUCOSE_c --> DTDP-DEOH-DEOXY-MANNOSE_c\n",
      "201 DTDPDEHYDRHAMEPIM-RXN DTDP-DEOH-DEOXY-GLUCOSE_c --> DTDP-DEOH-DEOXY-MANNOSE_c\n",
      "\n",
      "202 DTDPDEHYRHAMREDUCT-RXN DTDP-DEOH-DEOXY-MANNOSE_c + NADPH_c + PROTON_c --> DTDP-RHAMNOSE_c + NADP_c\n",
      "202 DTDPDEHYRHAMREDUCT-RXN DTDP-RHAMNOSE_c + NADP_c <=> DTDP-DEOH-DEOXY-MANNOSE_c + NADPH_c + PROTON_c\n",
      "\n",
      "203 UDPNACETYLMURAMATEDEHYDROG-RXN NADPH_c + PROTON_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c --> NADP_c + UDP-N-ACETYLMURAMATE_c\n",
      "203 UDPNACETYLMURAMATEDEHYDROG-RXN NADPH_c + PROTON_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c --> NADP_c + UDP-N-ACETYLMURAMATE_c\n",
      "\n",
      "204 RXN0-4301 D-SEDOHEPTULOSE-7-P_c --> D-ALPHABETA-D-HEPTOSE-7-PHOSPHATE_c\n",
      "204 RXN0-4301 D-SEDOHEPTULOSE-7-P_c --> D-ALPHABETA-D-HEPTOSE-7-PHOSPHATE_c\n",
      "\n",
      "205 RXN-9510 FMN_c + NADH_c + PROTON_c --> FMNH2_c + NAD_c\n",
      "205 RXN-9510 FMNH2_c + NAD_c <=> FMN_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "206 RXN-12444 FMN_c + NADPH_c + PROTON_c --> FMNH2_c + NADP_c\n",
      "206 RXN-12444 FMN_c + NADPH_c + 2.0 PROTON_c --> FMNH2_c + NADP_c\n",
      "\n",
      "207 GDPPYPHOSKIN-RXN ATP_c + GDP_c --> AMP_c + GUANOSINE-5DP-3DP_c + PROTON_c\n",
      "207 GDPPYPHOSKIN-RXN ATP_c + GDP_c --> AMP_c + GUANOSINE-5DP-3DP_c\n",
      "\n",
      "208 GTPPYPHOSKIN-RXN ATP_c + GTP_c --> AMP_c + GDP-TP_c + PROTON_c\n",
      "208 GTPPYPHOSKIN-RXN ATP_c + GTP_c --> AMP_c + GDP-TP_c\n",
      "\n",
      "209 PROTOHEMEFERROCHELAT-RXN FE+2_c + PROTOPORPHYRIN_IX_c --> PROTOHEME_c + 2.0 PROTON_c\n",
      "209 PROTOHEMEFERROCHELAT-RXN PROTOHEME_c + 2.0 PROTON_c <=> FE+2_c + PROTOPORPHYRIN_IX_c\n",
      "\n",
      "210 DEPHOSPHOCOAKIN-RXN ATP_c + DEPHOSPHO-COA_c --> ADP_c + CO-A_c + PROTON_c\n",
      "210 DEPHOSPHOCOAKIN-RXN ATP_c + DEPHOSPHO-COA_c --> ADP_c + CO-A_c + PROTON_c\n",
      "\n",
      "211 THIAZOLSYN3-RXN ATP_c + THZ_c --> ADP_c + PROTON_c + THZ-P_c\n",
      "211 THIAZOLSYN3-RXN ATP_c + THZ_c --> ADP_c + PROTON_c + THZ-P_c\n",
      "\n",
      "212 KDO-8PPHOSPHAT-RXN KDO-8P_c + WATER_c --> KDO_c + Pi_c\n",
      "212 KDO-8PPHOSPHAT-RXN KDO-8P_c + WATER_c --> KDO_c + Pi_c\n",
      "\n",
      "213 ARGININE-N-SUCCINYLTRANSFERASE-RXN ARG_c + SUC-COA_c --> CO-A_c + CPD-421_c + PROTON_c\n",
      "213 ARGININE-N-SUCCINYLTRANSFERASE-RXN ARG_c + SUC-COA_c --> CO-A_c + CPD-421_c + PROTON_c\n",
      "\n",
      "214 SUCCGLUDESUCC-RXN N2-SUCCINYLGLUTAMATE_c + WATER_c --> GLT_c + SUC_c\n",
      "214 SUCCGLUDESUCC-RXN N2-SUCCINYLGLUTAMATE_c + WATER_c --> GLT_c + SUC_c\n",
      "\n",
      "215 GUANINE-DEAMINASE-RXN GUANINE_c + PROTON_c + WATER_c --> AMMONIUM_c + XANTHINE_c\n",
      "215 GUANINE-DEAMINASE-RXN GUANINE_c + PROTON_c + WATER_c --> AMMONIUM_c + XANTHINE_c\n",
      "\n",
      "216 DTMPKI-RXN ATP_c + TMP_c <=> ADP_c + TDP_c\n",
      "216 DTMPKI-RXN ATP_c + TMP_c --> ADP_c + TDP_c\n",
      "\n",
      "217 PHOSPHOGLUCMUT-RXN GLC-1-P_c <=> D-glucopyranose-6-phosphate_c\n",
      "217 PHOSPHOGLUCMUT-RXN GLC-1-P_c <=> D-glucopyranose-6-phosphate_c\n",
      "\n",
      "218 RXN0-5185 CPD-2961_c + WATER_c --> GLUCONATE_c + Pi_c\n",
      "218 RXN0-5185 CPD-2961_c + WATER_c --> GLUCONATE_c + Pi_c\n",
      "\n",
      "219 RIBOSYLHOMOCYSTEINASE-RXN CPD-564_c --> DIHYDROXYPENTANEDIONE_c + HOMO-CYS_c\n",
      "219 RIBOSYLHOMOCYSTEINASE-RXN CPD-564_c --> DIHYDROXYPENTANEDIONE_c + HOMO-CYS_c\n",
      "\n",
      "220 BADH-RXN BETAINE_ALDEHYDE_c + NAD_c + WATER_c --> BETAINE_c + NADH_c + 2.0 PROTON_c\n",
      "220 BADH-RXN BETAINE_ALDEHYDE_c + NAD_c + WATER_c <=> BETAINE_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "221 PYRNUTRANSHYDROGEN-RXN NADPH_c + NAD_c --> NADH_c + NADP_c\n",
      "221 PYRNUTRANSHYDROGEN-RXN NADPH_c + NAD_c --> NADH_c + NADP_c\n",
      "\n",
      "222 GMP-SYN-GLUT-RXN ATP_c + GLN_c + WATER_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GLT_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "222 GMP-SYN-GLUT-RXN ATP_c + GLN_c + WATER_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GLT_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "223 GMP-SYN-NH3-RXN AMMONIUM_c + ATP_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "223 GMP-SYN-NH3-RXN AMMONIUM_c + ATP_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "224 PABASYN-RXN CHORISMATE_c + GLN_c --> 4-AMINO-4-DEOXYCHORISMATE_c + GLT_c\n",
      "224 PABASYN-RXN CHORISMATE_c + GLN_c <=> 4-AMINO-4-DEOXYCHORISMATE_c + GLT_c\n",
      "\n",
      "225 ADCLY-RXN 4-AMINO-4-DEOXYCHORISMATE_c --> P-AMINO-BENZOATE_c + PROTON_c + PYRUVATE_c\n",
      "225 ADCLY-RXN 4-AMINO-4-DEOXYCHORISMATE_c --> P-AMINO-BENZOATE_c + PROTON_c + PYRUVATE_c\n",
      "\n",
      "226 URACIL-PRIBOSYLTRANS-RXN PRPP_c + URACIL_c --> PPI_c + UMP_c\n",
      "226 URACIL-PRIBOSYLTRANS-RXN PRPP_c + URACIL_c --> PPI_c + UMP_c\n",
      "\n",
      "227 CTPSYN-RXN ATP_c + GLN_c + UTP_c + WATER_c --> ADP_c + CTP_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "227 CTPSYN-RXN ATP_c + GLN_c + UTP_c + WATER_c --> ADP_c + CTP_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "228 5.1.3.20-RXN ADP-D-GLYCERO-D-MANNO-HEPTOSE_c --> ADP-L-GLYCERO-D-MANNO-HEPTOSE_c\n",
      "228 5.1.3.20-RXN ADP-D-GLYCERO-D-MANNO-HEPTOSE_c --> ADP-L-GLYCERO-D-MANNO-HEPTOSE_c\n",
      "\n",
      "229 GLYOXYLATE-REDUCTASE-NADP+-RXN GLYOX_c + NADPH_c + PROTON_c --> GLYCOLLATE_c + NADP_c\n",
      "229 GLYOXYLATE-REDUCTASE-NADP+-RXN GLYOX_c + NADPH_c + PROTON_c --> GLYCOLLATE_c + NADP_c\n",
      "\n",
      "230 RXN0-300 NADPH_c + OH-PYR_c + PROTON_c --> GLYCERATE_c + NADP_c\n",
      "230 RXN0-300 NADPH_c + OH-PYR_c + PROTON_c --> GLYCERATE_c + NADP_c\n",
      "\n",
      "231 ACETYLORNDEACET-RXN N-ALPHA-ACETYLORNITHINE_c + WATER_c --> ACET_c + L-ORNITHINE_c\n",
      "231 ACETYLORNDEACET-RXN N-ALPHA-ACETYLORNITHINE_c + WATER_c <=> ACET_c + L-ORNITHINE_c\n",
      "\n",
      "232 ADDALT-RXN DEOXYADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + DEOXYINOSINE_c\n",
      "232 ADDALT-RXN DEOXYADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + DEOXYINOSINE_c\n",
      "\n",
      "233 ADENODEAMIN-RXN ADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + INOSINE_c\n",
      "233 ADENODEAMIN-RXN ADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + INOSINE_c\n",
      "\n",
      "234 ADENYLOSUCCINATE-SYNTHASE-RXN GTP_c + IMP_c + L-ASPARTATE_c --> ADENYLOSUCC_c + GDP_c + 2.0 PROTON_c + Pi_c\n",
      "234 ADENYLOSUCCINATE-SYNTHASE-RXN GTP_c + IMP_c + L-ASPARTATE_c --> ADENYLOSUCC_c + GDP_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "235 ASPARAGHYD-RXN ASN_c + WATER_c --> AMMONIUM_c + L-ASPARTATE_c\n",
      "235 ASPARAGHYD-RXN ASN_c + WATER_c --> AMMONIUM_c + L-ASPARTATE_c\n",
      "\n",
      "236 ASPCARBTRANS-RXN CARBAMOYL-P_c + L-ASPARTATE_c --> CARBAMYUL-L-ASPARTATE_c + PROTON_c + Pi_c\n",
      "236 ASPCARBTRANS-RXN CARBAMOYL-P_c + L-ASPARTATE_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c + Pi_c\n",
      "\n",
      "237 DIAMINOPIMEPIM-RXN LL-DIAMINOPIMELATE_c <=> MESO-DIAMINOPIMELATE_c\n",
      "237 DIAMINOPIMEPIM-RXN LL-DIAMINOPIMELATE_c <=> MESO-DIAMINOPIMELATE_c\n",
      "\n",
      "238 SUCCORNTRANSAM-RXN 2-KETOGLUTARATE_c + N2-SUCCINYLORNITHINE_c --> CPD-822_c + GLT_c\n",
      "238 SUCCORNTRANSAM-RXN 2-KETOGLUTARATE_c + N2-SUCCINYLORNITHINE_c <=> CPD-822_c + GLT_c\n",
      "\n",
      "239 THRESYN-RXN O-PHOSPHO-L-HOMOSERINE_c + WATER_c --> Pi_c + THR_c\n",
      "239 THRESYN-RXN O-PHOSPHO-L-HOMOSERINE_c + WATER_c --> Pi_c + THR_c\n",
      "\n",
      "240 DAHPSYN-RXN ERYTHROSE-4P_c + PHOSPHO-ENOL-PYRUVATE_c + WATER_c --> 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c + Pi_c\n",
      "240 DAHPSYN-RXN ERYTHROSE-4P_c + PHOSPHO-ENOL-PYRUVATE_c + WATER_c <=> 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c + Pi_c\n",
      "\n",
      "241 NMNAMIDOHYDRO-RXN NICOTINAMIDE_NUCLEOTIDE_c + WATER_c --> AMMONIUM_c + NICOTINATE_NUCLEOTIDE_c\n",
      "241 NMNAMIDOHYDRO-RXN NICOTINAMIDE_NUCLEOTIDE_c + WATER_c --> AMMONIUM_c + NICOTINATE_NUCLEOTIDE_c\n",
      "\n",
      "242 SPERMACTRAN-RXN ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-568_c + PROTON_c\n",
      "242 SPERMACTRAN-RXN ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-568_c + PROTON_c\n",
      "\n",
      "243 RXN0-7165 ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-3462_c + PROTON_c\n",
      "243 RXN0-7165 ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-3462_c + PROTON_c\n",
      "\n",
      "244 MALSYN-RXN ACETYL-COA_c + GLYOX_c + WATER_c --> CO-A_c + MAL_c + PROTON_c\n",
      "244 MALSYN-RXN ACETYL-COA_c + GLYOX_c + WATER_c --> CO-A_c + MAL_c + PROTON_c\n",
      "\n",
      "245 GLUC1PURIDYLTRANS-RXN GLC-1-P_c + PROTON_c + UTP_c --> CPD-12575_c + PPI_c\n",
      "245 GLUC1PURIDYLTRANS-RXN GLC-1-P_c + PROTON_c + UTP_c <=> CPD-12575_c + PPI_c\n",
      "\n",
      "246 UDPHYDROXYMYRGLUCOSAMNACETYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-OHMYR-GLUCOSAMINE_c --> ACP_c + OH-MYRISTOYL_c + PROTON_c\n",
      "246 UDPHYDROXYMYRGLUCOSAMNACETYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-OHMYR-GLUCOSAMINE_c --> ACP_c + OH-MYRISTOYL_c + PROTON_c\n",
      "\n",
      "247 UDP-NACMUR-ALA-LIG-RXN ATP_c + L-ALPHA-ALANINE_c + UDP-N-ACETYLMURAMATE_c --> ADP_c + CPD0-1456_c + PROTON_c + Pi_c\n",
      "247 UDP-NACMUR-ALA-LIG-RXN ATP_c + L-ALPHA-ALANINE_c + UDP-N-ACETYLMURAMATE_c --> ADP_c + CPD0-1456_c + PROTON_c + Pi_c\n",
      "\n",
      "248 UDP-NACMURALA-GLU-LIG-RXN ATP_c + CPD0-1456_c + D-GLT_c --> ADP_c + PROTON_c + Pi_c + UDP-AA-GLUTAMATE_c\n",
      "248 UDP-NACMURALA-GLU-LIG-RXN ATP_c + CPD0-1456_c + D-GLT_c --> ADP_c + PROTON_c + Pi_c + UDP-AA-GLUTAMATE_c\n",
      "\n",
      "249 UDPACYLGLCNACDEACETYL-RXN UDP-OHMYR-ACETYLGLUCOSAMINE_c + WATER_c --> ACET_c + UDP-OHMYR-GLUCOSAMINE_c\n",
      "249 UDPACYLGLCNACDEACETYL-RXN UDP-OHMYR-ACETYLGLUCOSAMINE_c + WATER_c --> ACET_c + UDP-OHMYR-GLUCOSAMINE_c\n",
      "\n",
      "250 IMIDPHOSDEHYD-RXN D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c --> IMIDAZOLE-ACETOL-P_c + WATER_c\n",
      "250 IMIDPHOSDEHYD-RXN D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c --> IMIDAZOLE-ACETOL-P_c + WATER_c\n",
      "\n",
      "251 PRIBFAICARPISOM-RXN PHOSPHORIBOSYL-FORMIMINO-AICAR-P_c <=> PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c\n",
      "251 PRIBFAICARPISOM-RXN PHOSPHORIBOSYL-FORMIMINO-AICAR-P_c --> PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c\n",
      "\n",
      "252 SERINE-O-ACETTRAN-RXN ACETYL-COA_c + SER_c <=> ACETYLSERINE_c + CO-A_c\n",
      "252 SERINE-O-ACETTRAN-RXN ACETYL-COA_c + SER_c --> ACETYLSERINE_c + CO-A_c\n",
      "\n",
      "253 ACSERLY-RXN ACETYLSERINE_c + HS_c --> ACET_c + CYS_c + PROTON_c\n",
      "253 ACSERLY-RXN ACETYLSERINE_c + HS_c <=> ACET_c + CYS_c + PROTON_c\n",
      "\n",
      "254 325-BISPHOSPHATE-NUCLEOTIDASE-RXN 3-5-ADP_c + WATER_c --> AMP_c + Pi_c\n",
      "254 325-BISPHOSPHATE-NUCLEOTIDASE-RXN 3-5-ADP_c + WATER_c --> AMP_c + Pi_c\n",
      "\n",
      "255 ASPARTATE-SEMIALDEHYDE-DEHYDROGENASE-RXN L-ASPARTATE-SEMIALDEHYDE_c + NADP_c + Pi_c <=> L-BETA-ASPARTYL-P_c + NADPH_c + PROTON_c\n",
      "255 ASPARTATE-SEMIALDEHYDE-DEHYDROGENASE-RXN L-ASPARTATE-SEMIALDEHYDE_c + NADP_c + Pi_c <=> L-BETA-ASPARTYL-P_c + NADPH_c + PROTON_c\n",
      "\n",
      "256 DIAMINOPIMDECARB-RXN MESO-DIAMINOPIMELATE_c + PROTON_c --> CARBON-DIOXIDE_c + LYS_c\n",
      "256 DIAMINOPIMDECARB-RXN MESO-DIAMINOPIMELATE_c + PROTON_c --> CARBON-DIOXIDE_c + LYS_c\n",
      "\n",
      "257 S-ADENMETSYN-RXN ATP_c + MET_c + WATER_c --> PPI_c + Pi_c + S-ADENOSYLMETHIONINE_c\n",
      "257 S-ADENMETSYN-RXN ATP_c + MET_c + WATER_c --> PPI_c + Pi_c + S-ADENOSYLMETHIONINE_c\n",
      "\n",
      "258 THI-P-KIN-RXN ATP_c + THIAMINE-P_c --> ADP_c + THIAMINE-PYROPHOSPHATE_c\n",
      "258 THI-P-KIN-RXN ATP_c + THIAMINE-P_c --> ADP_c + THIAMINE-PYROPHOSPHATE_c\n",
      "\n",
      "259 GKI-RXN ATP_c + GLYCERATE_c --> 2-PG_c + ADP_c + PROTON_c\n",
      "259 GKI-RXN ATP_c + GLYCERATE_c --> 2-PG_c + ADP_c + PROTON_c\n",
      "\n",
      "260 GLUCOKIN-RXN ATP_c + Glucopyranose_c --> ADP_c + D-glucopyranose-6-phosphate_c + PROTON_c\n",
      "260 GLUCOKIN-RXN ATP_c + Glucopyranose_c --> ADP_c + D-glucopyranose-6-phosphate_c + PROTON_c\n",
      "\n",
      "261 MALIC-NADP-RXN MAL_c + NADP_c --> CARBON-DIOXIDE_c + NADPH_c + PYRUVATE_c\n",
      "261 MALIC-NADP-RXN MAL_c + NADP_c --> CARBON-DIOXIDE_c + NADPH_c + PYRUVATE_c\n",
      "\n",
      "262 ARGSUCCINLYA-RXN L-ARGININO-SUCCINATE_c <=> ARG_c + FUM_c\n",
      "262 ARGSUCCINLYA-RXN L-ARGININO-SUCCINATE_c <=> ARG_c + FUM_c\n",
      "\n",
      "263 CARBPSYN-RXN 2.0 ATP_c + GLN_c + H2CO3_c + WATER_c --> 2.0 ADP_c + CARBAMOYL-P_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "263 CARBPSYN-RXN 2.0 ATP_c + GLN_c + HCO3_c + WATER_c --> 2.0 ADP_c + CARBAMOYL-P_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "264 GLUTKIN-RXN ATP_c + GLT_c --> ADP_c + L-GLUTAMATE-5-P_c\n",
      "264 GLUTKIN-RXN ATP_c + GLT_c --> ADP_c + L-GLUTAMATE-5-P_c\n",
      "\n",
      "265 2.3.1.180-RXN ACETYL-COA_c + MALONYL-ACP_c + PROTON_c --> Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c + CO-A_c\n",
      "265 2.3.1.180-RXN ACETYL-COA_c + MALONYL-ACP_c + PROTON_c --> Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c + CO-A_c\n",
      "\n",
      "266 SUPEROX-DISMUT-RXN 2.0 PROTON_c + 2.0 SUPER-OXIDE_c --> HYDROGEN-PEROXIDE_c + OXYGEN-MOLECULE_c\n",
      "266 SUPEROX-DISMUT-RXN 2.0 PROTON_c + 2.0 SUPER-OXIDE_c --> HYDROGEN-PEROXIDE_c + OXYGEN-MOLECULE_c\n",
      "\n",
      "267 ADENOSYLHOMOCYSTEINE-NUCLEOSIDASE-RXN ADENOSYL-HOMO-CYS_c + WATER_c --> ADENINE_c + CPD-564_c\n",
      "267 ADENOSYLHOMOCYSTEINE-NUCLEOSIDASE-RXN ADENOSYL-HOMO-CYS_c + WATER_c --> ADENINE_c + CPD-564_c\n",
      "\n",
      "268 METHYLTHIOADENOSINE-NUCLEOSIDASE-RXN 5-METHYLTHIOADENOSINE_c + WATER_c --> ADENINE_c + CPD-560_c\n",
      "268 METHYLTHIOADENOSINE-NUCLEOSIDASE-RXN 5-METHYLTHIOADENOSINE_c + WATER_c --> ADENINE_c + CPD-560_c\n",
      "\n",
      "269 THIOSULFATE-SULFURTRANSFERASE-RXN HCN_c + S2O3_c --> HSCN_c + PROTON_c + SO3_c\n",
      "269 THIOSULFATE-SULFURTRANSFERASE-RXN HCN_c + S2O3_c --> HSCN_c + 2.0 PROTON_c + SO3_c\n",
      "\n",
      "270 DETHIOBIOTIN-SYN-RXN ATP_c + CARBON-DIOXIDE_c + DIAMINONONANOATE_c --> ADP_c + DETHIOBIOTIN_c + 3.0 PROTON_c + Pi_c\n",
      "270 DETHIOBIOTIN-SYN-RXN ATP_c + CARBON-DIOXIDE_c + DIAMINONONANOATE_c --> ADP_c + DETHIOBIOTIN_c + 3.0 PROTON_c + Pi_c\n",
      "\n",
      "271 CYTDEAM-RXN CYTOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + URACIL_c\n",
      "271 CYTDEAM-RXN CYTOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + URACIL_c\n",
      "\n",
      "272 L-GLN-FRUCT-6-P-AMINOTRANS-RXN FRUCTOSE-6P_c + GLN_c --> D-GLUCOSAMINE-6-P_c + GLT_c\n",
      "272 L-GLN-FRUCT-6-P-AMINOTRANS-RXN FRUCTOSE-6P_c + GLN_c <=> D-GLUCOSAMINE-6-P_c + GLT_c\n",
      "\n",
      "273 GLUTCYSLIG-RXN ATP_c + CYS_c + GLT_c --> ADP_c + L-GAMMA-GLUTAMYLCYSTEINE_c + PROTON_c + Pi_c\n",
      "273 GLUTCYSLIG-RXN ATP_c + CYS_c + GLT_c --> ADP_c + L-GAMMA-GLUTAMYLCYSTEINE_c + PROTON_c + Pi_c\n",
      "\n",
      "274 GLUTAMIN-RXN GLN_c + WATER_c --> AMMONIUM_c + GLT_c\n",
      "274 GLUTAMIN-RXN GLN_c + WATER_c --> AMMONIUM_c + GLT_c\n",
      "\n",
      "275 PRPPAMIDOTRANS-RXN GLN_c + PRPP_c + WATER_c --> 5-P-BETA-D-RIBOSYL-AMINE_c + GLT_c + PPI_c\n",
      "275 PRPPAMIDOTRANS-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + GLT_c + PPI_c <=> GLN_c + PRPP_c + WATER_c\n",
      "\n",
      "276 ORNCARBAMTRANSFER-RXN CARBAMOYL-P_c + L-ORNITHINE_c <=> L-CITRULLINE_c + PROTON_c + Pi_c\n",
      "276 ORNCARBAMTRANSFER-RXN CARBAMOYL-P_c + L-ORNITHINE_c <=> L-CITRULLINE_c + PROTON_c + Pi_c\n",
      "\n",
      "277 GART-RXN 10-FORMYL-THF_c + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE_c + PROTON_c + THF_c\n",
      "277 GART-RXN 10-FORMYL-THF_c + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE_c + PROTON_c + THF_c\n",
      "\n",
      "278 METHYLISOCITRATE-LYASE-RXN CPD-618_c <=> PYRUVATE_c + SUC_c\n",
      "278 METHYLISOCITRATE-LYASE-RXN CPD-618_c <=> PYRUVATE_c + SUC_c\n",
      "\n",
      "279 SUCCGLUALDDEHYD-RXN CPD-822_c + NAD_c + WATER_c --> N2-SUCCINYLGLUTAMATE_c + NADH_c + 2.0 PROTON_c\n",
      "279 SUCCGLUALDDEHYD-RXN CPD-822_c + NAD_c + WATER_c --> N2-SUCCINYLGLUTAMATE_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "280 PORPHOBILSYNTH-RXN 2.0 5-AMINO-LEVULINATE_c --> PORPHOBILINOGEN_c + PROTON_c + 2.0 WATER_c\n",
      "280 PORPHOBILSYNTH-RXN 2.0 5-AMINO-LEVULINATE_c --> PORPHOBILINOGEN_c + PROTON_c + 2.0 WATER_c\n",
      "\n",
      "281 GLYRIBONUCSYN-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + ATP_c + GLY_c <=> 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c + ADP_c + PROTON_c + Pi_c\n",
      "281 GLYRIBONUCSYN-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + ATP_c + GLY_c --> 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c + ADP_c + PROTON_c + Pi_c\n",
      "\n",
      "282 ETHAMLY-RXN ETHANOL-AMINE_c --> ACETALD_c + AMMONIUM_c\n",
      "282 ETHAMLY-RXN ETHANOL-AMINE_c --> ACETALD_c + AMMONIUM_c\n",
      "\n",
      "283 RXN0-5408 D-MYO-INOSITOL-1-MONOPHOSPHATE_c + WATER_c --> MYO-INOSITOL_c + Pi_c\n",
      "283 RXN0-5408 D-MYO-INOSITOL-1-MONOPHOSPHATE_c + WATER_c --> MYO-INOSITOL_c + Pi_c\n",
      "\n",
      "284 GLYCEROL-2-PHOSPHATASE-RXN CPD-536_c + WATER_c --> GLYCEROL_c + Pi_c\n",
      "284 GLYCEROL-2-PHOSPHATASE-RXN CPD-536_c + WATER_c --> GLYCEROL_c + Pi_c\n",
      "\n",
      "285 2-METHYLCITRATE-SYNTHASE-RXN OXALACETIC_ACID_c + PROPIONYL-COA_c + WATER_c --> CO-A_c + CPD-622_c + PROTON_c\n",
      "285 2-METHYLCITRATE-SYNTHASE-RXN OXALACETIC_ACID_c + PROPIONYL-COA_c + WATER_c <=> CO-A_c + CPD-622_c + PROTON_c\n",
      "\n",
      "286 ATPASE-RXN ATP_c + WATER_c --> ADP_c + PROTON_c + Pi_c\n",
      "286 ATPASE-RXN ATP_c + WATER_c --> ADP_c + PROTON_c + Pi_c\n",
      "\n",
      "287 RXN0-5462 GTP_c + WATER_c --> GDP_c + PROTON_c + Pi_c\n",
      "287 RXN0-5462 GTP_c + WATER_c --> GDP_c + PROTON_c + Pi_c\n",
      "\n",
      "288 GLU6PDEHYDROG-RXN D-glucopyranose-6-phosphate_c + NADP_c <=> D-6-P-GLUCONO-DELTA-LACTONE_c + NADPH_c + PROTON_c\n",
      "288 GLU6PDEHYDROG-RXN D-glucopyranose-6-phosphate_c + NADP_c --> D-6-P-GLUCONO-DELTA-LACTONE_c + NADPH_c + PROTON_c\n",
      "\n",
      "289 UGD-RXN CPD-12575_c + 2.0 NAD_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-GLUCURONATE_c\n",
      "289 UGD-RXN CPD-12575_c + 2.0 NAD_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-GLUCURONATE_c\n",
      "\n",
      "290 UDPMANNACADEHYDROG-RXN 2.0 NAD_c + UDP-MANNAC_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-MANNACA_c\n",
      "290 UDPMANNACADEHYDROG-RXN 2.0 NAD_c + UDP-MANNAC_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-MANNACA_c\n",
      "\n",
      "291 GLUTDECARBOX-RXN GLT_c + PROTON_c --> 4-AMINO-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "291 GLUTDECARBOX-RXN GLT_c + PROTON_c --> 4-AMINO-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "292 CYSTATHIONINE-BETA-LYASE-RXN L-CYSTATHIONINE_c + WATER_c --> AMMONIUM_c + HOMO-CYS_c + PYRUVATE_c\n",
      "292 CYSTATHIONINE-BETA-LYASE-RXN L-CYSTATHIONINE_c + WATER_c --> AMMONIUM_c + HOMO-CYS_c + PYRUVATE_c\n",
      "\n",
      "293 RXN-12565 ACETYL-COA_c + BUTYRYL-COA_c <=> CO-A_c + K-HEXANOYL-COA_c\n",
      "293 RXN-12565 ACETYL-COA_c + BUTYRYL-COA_c <=> CO-A_c + K-HEXANOYL-COA_c\n",
      "\n",
      "294 RXN-13617 ACETYL-COA_c + CPD-196_c <=> CO-A_c + CPD0-2123_c\n",
      "294 RXN-13617 CO-A_c + CPD0-2123_c --> ACETYL-COA_c + CPD-196_c\n",
      "\n",
      "295 ACETYL-COA-ACETYLTRANSFER-RXN 2.0 ACETYL-COA_c <=> ACETOACETYL-COA_c + CO-A_c\n",
      "295 ACETYL-COA-ACETYLTRANSFER-RXN 2.0 ACETYL-COA_c <=> ACETOACETYL-COA_c + CO-A_c\n",
      "\n",
      "296 RXN0-5114 3-P-SERINE_c + WATER_c --> Pi_c + SER_c\n",
      "296 RXN0-5114 3-P-SERINE_c + WATER_c --> Pi_c + SER_c\n",
      "\n",
      "297 RXN-16701 CPD-381_c + NAD_c <=> 2-KETOGLUTARATE_c + NADH_c + PROTON_c\n",
      "297 RXN-16701 CPD-381_c + NAD_c <=> 2-KETOGLUTARATE_c + NADH_c + PROTON_c\n",
      "\n",
      "298 PGLYCDEHYDROG-RXN G3P_c + NAD_c --> 3-P-HYDROXYPYRUVATE_c + NADH_c + PROTON_c\n",
      "298 PGLYCDEHYDROG-RXN G3P_c + NAD_c <=> 3-P-HYDROXYPYRUVATE_c + NADH_c + PROTON_c\n",
      "\n",
      "299 GTP-CYCLOHYDRO-I-RXN GTP_c + WATER_c --> DIHYDRONEOPTERIN-P3_c + FORMATE_c + PROTON_c\n",
      "299 GTP-CYCLOHYDRO-I-RXN GTP_c + WATER_c --> DIHYDRONEOPTERIN-P3_c + FORMATE_c + PROTON_c\n",
      "\n",
      "300 DIHYDROFOLATESYNTH-RXN 7-8-DIHYDROPTEROATE_c + ATP_c + GLT_c --> ADP_c + DIHYDROFOLATE_c + PROTON_c + Pi_c\n",
      "300 DIHYDROFOLATESYNTH-RXN 7-8-DIHYDROPTEROATE_c + ATP_c + GLT_c --> ADP_c + DIHYDROFOLATE_c + PROTON_c + Pi_c\n",
      "\n",
      "301 SAICARSYN-RXN ATP_c + L-ASPARTATE_c + PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c --> ADP_c + P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c + PROTON_c + Pi_c\n",
      "301 SAICARSYN-RXN ATP_c + L-ASPARTATE_c + PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c --> ADP_c + P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c + PROTON_c + Pi_c\n",
      "\n",
      "302 RXN0-5289 NADH_c + PROTON_c + TARTRONATE-S-ALD_c <=> GLYCERATE_c + NAD_c\n",
      "302 RXN0-5289 GLYCERATE_c + NAD_c <=> NADH_c + PROTON_c + TARTRONATE-S-ALD_c\n",
      "\n",
      "303 H2PTEROATESYNTH-RXN DIHYDROPTERIN-CH2OH-PP_c + P-AMINO-BENZOATE_c --> 7-8-DIHYDROPTEROATE_c + PPI_c\n",
      "303 H2PTEROATESYNTH-RXN DIHYDROPTERIN-CH2OH-PP_c + P-AMINO-BENZOATE_c --> 7-8-DIHYDROPTEROATE_c + PPI_c\n",
      "\n",
      "304 METHYLENETHFDEHYDROG-NADP-RXN METHYLENE-THF_c + NADP_c <=> 5-10-METHENYL-THF_c + NADPH_c\n",
      "304 METHYLENETHFDEHYDROG-NADP-RXN METHYLENE-THF_c + NADP_c <=> 5-10-METHENYL-THF_c + NADPH_c\n",
      "\n",
      "305 METHENYLTHFCYCLOHYDRO-RXN 5-10-METHENYL-THF_c + WATER_c <=> 10-FORMYL-THF_c + PROTON_c\n",
      "305 METHENYLTHFCYCLOHYDRO-RXN 5-10-METHENYL-THF_c + WATER_c <=> 10-FORMYL-THF_c + PROTON_c\n",
      "\n",
      "306 RXN0-3942 CPD-9000_c + WATER_c --> 4-AMINO-BUTYRATE_c + GLT_c\n",
      "306 RXN0-3942 CPD-9000_c + WATER_c --> 4-AMINO-BUTYRATE_c + GLT_c\n",
      "\n",
      "307 RXN-6622 CYS-GLY_c + WATER_c --> CYS_c + GLY_c\n",
      "307 RXN-6622 CYS-GLY_c + WATER_c --> CYS_c + GLY_c\n",
      "\n",
      "308 RXN0-6981 CPD0-1445_c + WATER_c --> GLT_c + L-ALPHA-ALANINE_c\n",
      "308 RXN0-6981 CPD0-1445_c + WATER_c --> GLT_c + L-ALPHA-ALANINE_c\n",
      "\n",
      "309 RXN0-1483 4.0 FE+2_c + OXYGEN-MOLECULE_c + 4.0 PROTON_c --> 4.0 FE+3_c + 2.0 WATER_c\n",
      "309 RXN0-1483 4.0 FE+2_c + OXYGEN-MOLECULE_c + 4.0 PROTON_c --> 4.0 FE+3_c + 2.0 WATER_c\n",
      "\n",
      "310 BTUR2-RXN ATP_c + COBINAMIDE_c + PROTON_c <=> ADENOSYLCOBINAMIDE_c + P3I_c\n",
      "310 BTUR2-RXN 2.0 ATP_c + 2.0 CPD-20903_c + Reduced-Flavoproteins_c --> 2.0 ADENOSYLCOBINAMIDE_c + Oxidized-Flavoproteins_c + 2.0 P3I_c + 2.0 PROTON_c\n",
      "\n",
      "311 PPGPPSYN-RXN GUANOSINE-5DP-3DP_c + WATER_c --> GDP_c + PPI_c\n",
      "311 PPGPPSYN-RXN GUANOSINE-5DP-3DP_c + WATER_c --> GDP_c + PPI_c + PROTON_c\n",
      "\n",
      "312 RXN0-6427 GDP-TP_c + WATER_c --> GTP_c + PPI_c\n",
      "312 RXN0-6427 GDP-TP_c + WATER_c --> GTP_c + PPI_c + PROTON_c\n",
      "\n",
      "313 RXN0-1603 WATER_c + XTP_c --> PPI_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "313 RXN0-1603 WATER_c + XTP_c --> PPI_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "314 RXN0-6382 ITP_c + WATER_c --> IMP_c + PPI_c + PROTON_c\n",
      "314 RXN0-6382 ITP_c + WATER_c --> IMP_c + PPI_c + PROTON_c\n",
      "\n",
      "315 RXN0-1602 DITP_c + WATER_c --> DIMP_c + PPI_c + PROTON_c\n",
      "315 RXN0-1602 DITP_c + WATER_c --> DIMP_c + PPI_c + PROTON_c\n",
      "\n",
      "316 RXN-15744 LACTALD_c + NADPH_c + PROTON_c --> NADP_c + PROPANE-1-2-DIOL_c\n",
      "316 RXN-15744 LACTALD_c + NADPH_c + PROTON_c --> NADP_c + PROPANE-1-2-DIOL_c\n",
      "\n",
      "317 RXN-15743 CPD-358_c + NADPH_c + PROTON_c --> CPD-8891_c + NADP_c\n",
      "317 RXN-15743 CPD-358_c + NADPH_c + PROTON_c --> CPD-8891_c + NADP_c\n",
      "\n",
      "318 GLUCONOKIN-RXN ATP_c + GLUCONATE_c --> ADP_c + CPD-2961_c + PROTON_c\n",
      "318 GLUCONOKIN-RXN ATP_c + GLUCONATE_c --> ADP_c + CPD-2961_c + PROTON_c\n",
      "\n",
      "319 OHMETHYLBILANESYN-RXN 4.0 PORPHOBILINOGEN_c + WATER_c --> 4.0 AMMONIUM_c + HYDROXYMETHYLBILANE_c\n",
      "319 OHMETHYLBILANESYN-RXN 4.0 PORPHOBILINOGEN_c + WATER_c --> 4.0 AMMONIUM_c + HYDROXYMETHYLBILANE_c\n",
      "\n",
      "320 2.7.7.60-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + CTP_c + PROTON_c --> 4-CYTIDINE-5-DIPHOSPHO-2-C_c + PPI_c\n",
      "320 2.7.7.60-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + CTP_c + PROTON_c --> 4-CYTIDINE-5-DIPHOSPHO-2-C_c + PPI_c\n",
      "\n",
      "321 GLUTDEHYD-RXN GLT_c + NADP_c + WATER_c <=> 2-KETOGLUTARATE_c + AMMONIUM_c + NADPH_c + PROTON_c\n",
      "321 GLUTDEHYD-RXN GLT_c + NADP_c + WATER_c <=> 2-KETOGLUTARATE_c + AMMONIUM_c + NADPH_c + PROTON_c\n",
      "\n",
      "322 GLUTAMATESYN-RXN 2-KETOGLUTARATE_c + GLN_c + NADPH_c + PROTON_c --> 2.0 GLT_c + NADP_c\n",
      "322 GLUTAMATESYN-RXN 2-KETOGLUTARATE_c + GLN_c + NADPH_c + PROTON_c --> 2.0 GLT_c + NADP_c\n",
      "\n",
      "323 GLYOXII-RXN S-LACTOYL-GLUTATHIONE_c + WATER_c --> D-LACTATE_c + GLUTATHIONE_c + PROTON_c\n",
      "323 GLYOXII-RXN S-LACTOYL-GLUTATHIONE_c + WATER_c --> D-LACTATE_c + GLUTATHIONE_c + PROTON_c\n",
      "\n",
      "324 ERYTH4PDEHYDROG-RXN ERYTHROSE-4P_c + NAD_c + WATER_c <=> ERYTHRONATE-4P_c + NADH_c + 2.0 PROTON_c\n",
      "324 ERYTH4PDEHYDROG-RXN ERYTHROSE-4P_c + NAD_c + WATER_c --> ERYTHRONATE-4P_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "325 GLUTAMIDOTRANS-RXN GLN_c + PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c --> AICAR_c + D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c + GLT_c + PROTON_c\n",
      "325 GLUTAMIDOTRANS-RXN GLN_c + PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c --> AICAR_c + D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c + GLT_c + PROTON_c\n",
      "\n",
      "326 DUTP-PYROP-RXN DUTP_c + WATER_c --> DUMP_c + PPI_c + PROTON_c\n",
      "326 DUTP-PYROP-RXN DUTP_c + WATER_c --> DUMP_c + PPI_c + PROTON_c\n",
      "\n",
      "327 THREDEHYD-RXN THR_c --> 2-OXOBUTANOATE_c + AMMONIUM_c\n",
      "327 THREDEHYD-RXN THR_c --> 2-OXOBUTANOATE_c + AMMONIUM_c\n",
      "\n",
      "328 2.7.9.3-RXN ATP_c + SE-2_c + WATER_c --> AMP_c + Pi_c + SEPO3_c\n",
      "328 2.7.9.3-RXN ATP_c + SE-2_c + WATER_c --> AMP_c + Pi_c + SEPO3_c\n",
      "\n",
      "329 SHIKIMATE-KINASE-RXN ATP_c + SHIKIMATE_c --> ADP_c + PROTON_c + SHIKIMATE-5P_c\n",
      "329 SHIKIMATE-KINASE-RXN ATP_c + SHIKIMATE_c <=> ADP_c + PROTON_c + SHIKIMATE-5P_c\n",
      "\n",
      "330 SUCCARGDIHYDRO-RXN CPD-421_c + 2.0 PROTON_c + 2.0 WATER_c --> 2.0 AMMONIUM_c + CARBON-DIOXIDE_c + N2-SUCCINYLORNITHINE_c\n",
      "330 SUCCARGDIHYDRO-RXN CPD-421_c + 2.0 PROTON_c + 2.0 WATER_c --> 2.0 AMMONIUM_c + CARBON-DIOXIDE_c + N2-SUCCINYLORNITHINE_c\n",
      "\n",
      "331 DAPASYN-RXN 8-AMINO-7-OXONONANOATE_c + S-ADENOSYLMETHIONINE_c <=> DIAMINONONANOATE_c + S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c\n",
      "331 DAPASYN-RXN 8-AMINO-7-OXONONANOATE_c + S-ADENOSYLMETHIONINE_c <=> DIAMINONONANOATE_c + S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c\n",
      "\n",
      "332 N-ACETYLTRANSFER-RXN ACETYL-COA_c + GLT_c --> ACETYL-GLU_c + CO-A_c + PROTON_c\n",
      "332 N-ACETYLTRANSFER-RXN ACETYL-COA_c + GLT_c <=> ACETYL-GLU_c + CO-A_c + PROTON_c\n",
      "\n",
      "333 RXN0-5038 CAMP_c + WATER_c --> AMP_c + PROTON_c\n",
      "333 RXN0-5038 CAMP_c + WATER_c --> AMP_c + PROTON_c\n",
      "\n",
      "334 DALADALALIG-RXN ATP_c + 2.0 D-ALANINE_c <=> ADP_c + D-ALA-D-ALA_c + PROTON_c + Pi_c\n",
      "334 DALADALALIG-RXN ATP_c + 2.0 D-ALANINE_c --> ADP_c + D-ALA-D-ALA_c + PROTON_c + Pi_c\n",
      "\n",
      "335 SUCCCOASYN-RXN ATP_c + CO-A_c + SUC_c <=> ADP_c + Pi_c + SUC-COA_c\n",
      "335 SUCCCOASYN-RXN ATP_c + CO-A_c + SUC_c <=> ADP_c + Pi_c + SUC-COA_c\n",
      "\n",
      "336 PANTOATE-BETA-ALANINE-LIG-RXN ATP_c + B-ALANINE_c + L-PANTOATE_c --> AMP_c + PANTOTHENATE_c + PPI_c + PROTON_c\n",
      "336 PANTOATE-BETA-ALANINE-LIG-RXN ATP_c + B-ALANINE_c + L-PANTOATE_c --> AMP_c + PANTOTHENATE_c + PPI_c + PROTON_c\n",
      "\n",
      "337 3-DEHYDROQUINATE-SYNTHASE-RXN 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c --> DEHYDROQUINATE_c + Pi_c\n",
      "337 3-DEHYDROQUINATE-SYNTHASE-RXN 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c --> DEHYDROQUINATE_c + Pi_c\n",
      "\n",
      "338 3PGAREARR-RXN 2-PG_c <=> G3P_c\n",
      "338 3PGAREARR-RXN 2-PG_c <=> G3P_c\n",
      "\n",
      "339 TRIOSEPISOMERIZATION-RXN DIHYDROXY-ACETONE-PHOSPHATE_c <=> GAP_c\n",
      "339 TRIOSEPISOMERIZATION-RXN GAP_c <=> DIHYDROXY-ACETONE-PHOSPHATE_c\n",
      "\n",
      "340 ERYTHRON4PDEHYDROG-RXN ERYTHRONATE-4P_c + NAD_c <=> 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + NADH_c + PROTON_c\n",
      "340 ERYTHRON4PDEHYDROG-RXN ERYTHRONATE-4P_c + NAD_c --> 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + NADH_c + PROTON_c\n",
      "\n",
      "341 RXN-8974 MALONATE-S-ALD_c + NADPH_c + PROTON_c --> 3-HYDROXY-PROPIONATE_c + NADP_c\n",
      "341 RXN-8974 MALONATE-S-ALD_c + NADPH_c + PROTON_c --> 3-HYDROXY-PROPIONATE_c + NADP_c\n",
      "\n",
      "342 GLUTATHIONE-SYN-RXN ATP_c + GLY_c + L-GAMMA-GLUTAMYLCYSTEINE_c --> ADP_c + GLUTATHIONE_c + PROTON_c + Pi_c\n",
      "342 GLUTATHIONE-SYN-RXN ATP_c + GLY_c + L-GAMMA-GLUTAMYLCYSTEINE_c --> ADP_c + GLUTATHIONE_c + PROTON_c + Pi_c\n",
      "\n",
      "343 RXN0-1863 5-BETA-L-THREO-PENTAPYRANOSYL-4-ULOSE-_c + GLT_c <=> 2-KETOGLUTARATE_c + UDP-4-AMINO-4-DEOXY-L-ARABINOSE_c\n",
      "343 RXN0-1863 2-KETOGLUTARATE_c + UDP-4-AMINO-4-DEOXY-L-ARABINOSE_c <=> 5-BETA-L-THREO-PENTAPYRANOSYL-4-ULOSE-_c + GLT_c\n",
      "\n",
      "344 RXN0-4281 METHYL-GLYOXAL_c + NADPH_c + PROTON_c --> ACETOL_c + NADP_c\n",
      "344 RXN0-4281 METHYL-GLYOXAL_c + NADPH_c + PROTON_c --> ACETOL_c + NADP_c\n",
      "\n",
      "345 RXN0-5061 ADP-L-GLYCERO-D-MANNO-HEPTOSE_c + CPD0-929_c --> ADP_c + CPD0-930_c + PROTON_c\n",
      "345 RXN0-5061 ADP-L-GLYCERO-D-MANNO-HEPTOSE_c + CPD0-929_c --> ADP_c + CPD0-930_c + PROTON_c\n",
      "\n",
      "346 HEMN-RXN COPROPORPHYRINOGEN_III_c + 2.0 S-ADENOSYLMETHIONINE_c --> 2.0 CARBON-DIOXIDE_c + 2.0 CH33ADO_c + 2.0 MET_c + PROTOPORPHYRINOGEN_c\n",
      "346 HEMN-RXN COPROPORPHYRINOGEN_III_c + 2.0 S-ADENOSYLMETHIONINE_c --> 2.0 CARBON-DIOXIDE_c + 2.0 CH33ADO_c + 2.0 MET_c + PROTOPORPHYRINOGEN_c\n",
      "\n",
      "347 DLACTDEHYDROGNAD-RXN D-LACTATE_c + NAD_c <=> NADH_c + PROTON_c + PYRUVATE_c\n",
      "347 DLACTDEHYDROGNAD-RXN NADH_c + PROTON_c + PYRUVATE_c --> D-LACTATE_c + NAD_c\n",
      "\n",
      "348 RXN-9537 R-3-hydroxymyristoyl-ACPs_c --> Tetradec-2-enoyl-ACPs_c + WATER_c\n",
      "348 RXN-9537 R-3-hydroxymyristoyl-ACPs_c --> Tetradec-2-enoyl-ACPs_c + WATER_c\n",
      "\n",
      "349 4.2.1.59-RXN 3-Hydroxy-octanoyl-ACPs_c --> 2-Octenoyl-ACPs_c + WATER_c\n",
      "349 4.2.1.59-RXN 3-Hydroxy-octanoyl-ACPs_c --> 2-Octenoyl-ACPs_c + WATER_c\n",
      "\n",
      "350 RXN0-305 OH-PYR_c <=> TARTRONATE-S-ALD_c\n",
      "350 RXN0-305 OH-PYR_c <=> TARTRONATE-S-ALD_c\n",
      "\n",
      "351 ADOMET-DMK-METHYLTRANSFER-RXN CPD-12115_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + PROTON_c + REDUCED-MENAQUINONE_c\n",
      "351 ADOMET-DMK-METHYLTRANSFER-RXN CPD-12115_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + PROTON_c + REDUCED-MENAQUINONE_c\n",
      "\n",
      "352 S-FORMYLGLUTATHIONE-HYDROLASE-RXN CPD-548_c + WATER_c --> FORMATE_c + GLUTATHIONE_c + PROTON_c\n",
      "352 S-FORMYLGLUTATHIONE-HYDROLASE-RXN CPD-548_c + WATER_c --> FORMATE_c + GLUTATHIONE_c + PROTON_c\n",
      "\n",
      "353 DMK-RXN DIHYDROXYNAPHTHOATE_c + OCTAPRENYL-DIPHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + CPD-12115_c + PPI_c\n",
      "353 DMK-RXN DIHYDROXYNAPHTHOATE_c + OCTAPRENYL-DIPHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + CPD-12115_c + PPI_c\n",
      "\n",
      "354 RXN-11477 3-Hydroxyglutaryl-ACP-methyl-ester_c --> Enoylglutaryl-ACP-methyl-esters_c + WATER_c\n",
      "354 RXN-11477 3-Hydroxyglutaryl-ACP-methyl-ester_c --> Enoylglutaryl-ACP-methyl-esters_c + WATER_c\n",
      "\n",
      "355 RXN-11481 3-hydroxypimeloyl-ACP-methyl-esters_c --> Enoylpimeloyl-ACP-methyl-esters_c + WATER_c\n",
      "355 RXN-11481 3-hydroxypimeloyl-ACP-methyl-esters_c --> Enoylpimeloyl-ACP-methyl-esters_c + WATER_c\n",
      "\n",
      "356 ACONITATEDEHYDR-RXN CIT_c <=> CIS-ACONITATE_c + WATER_c\n",
      "356 ACONITATEDEHYDR-RXN CIT_c <=> CIS-ACONITATE_c + WATER_c\n",
      "\n",
      "357 4.2.1.99-RXN CPD-1136_c + WATER_c <=> CPD-618_c\n",
      "357 4.2.1.99-RXN CPD-618_c <=> CPD-1136_c + WATER_c\n",
      "\n",
      "358 RXN-12093 AMMONIUM_c + ATP_c + CPD-13043_c --> 7-CYANO-7-DEAZAGUANINE_c + ADP_c + PROTON_c + Pi_c + WATER_c\n",
      "358 RXN-12093 AMMONIUM_c + ATP_c + CPD-13043_c --> 7-CYANO-7-DEAZAGUANINE_c + ADP_c + PROTON_c + Pi_c + WATER_c\n",
      "\n",
      "359 RXN0-4022 7-CYANO-7-DEAZAGUANINE_c + 2.0 NADPH_c + 3.0 PROTON_c --> 7-AMINOMETHYL-7-DEAZAGUANINE_c + 2.0 NADP_c\n",
      "359 RXN0-4022 7-CYANO-7-DEAZAGUANINE_c + 2.0 NADPH_c + 3.0 PROTON_c --> 7-AMINOMETHYL-7-DEAZAGUANINE_c + 2.0 NADP_c\n",
      "\n",
      "360 RXN-3641 3-KETO-ADIPYL-COA_c + CO-A_c --> ACETYL-COA_c + SUC-COA_c\n",
      "360 RXN-3641 3-KETO-ADIPYL-COA_c + CO-A_c --> ACETYL-COA_c + SUC-COA_c\n",
      "\n",
      "361 RXN0-2044 3-HYDROXYADIPYL-COA_c + NAD_c <=> 3-KETO-ADIPYL-COA_c + NADH_c + PROTON_c\n",
      "361 RXN0-2044 3-HYDROXYADIPYL-COA_c + NAD_c <=> 3-KETO-ADIPYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "362 RXN-11484 L-ALPHA-ALANINE_c + Pimeloyl-ACPs_c --> 8-AMINO-7-OXONONANOATE_c + ACP_c + CARBON-DIOXIDE_c\n",
      "362 RXN-11484 L-ALPHA-ALANINE_c + PROTON_c + Pimeloyl-ACPs_c --> 8-AMINO-7-OXONONANOATE_c + ACP_c + CARBON-DIOXIDE_c\n",
      "\n",
      "363 RXN-11483 Pimeloyl-ACP-methyl-esters_c + WATER_c --> METOH_c + Pimeloyl-ACPs_c\n",
      "363 RXN-11483 Pimeloyl-ACP-methyl-esters_c + WATER_c --> METOH_c + PROTON_c + Pimeloyl-ACPs_c\n",
      "\n",
      "364 RXN0-5208 2.0 ATP_c + PROTON_c --> ADENOSYL-P4_c + PPI_c\n",
      "364 RXN0-5208 2.0 ATP_c + PROTON_c --> ADENOSYL-P4_c + PPI_c\n",
      "\n",
      "365 RXN0-5364 CPD-381_c + OXYGEN-MOLECULE_c --> 2-KETOGLUTARATE_c + HYDROGEN-PEROXIDE_c\n",
      "365 RXN0-5364 CPD-381_c + OXYGEN-MOLECULE_c --> 2-KETOGLUTARATE_c + HYDROGEN-PEROXIDE_c\n",
      "\n",
      "366 RXN0-7104 CPD-207_c + WATER_c --> CO-A_c + PHENYLACETATE_c + PROTON_c\n",
      "366 RXN0-7104 CPD-207_c + WATER_c --> CO-A_c + PHENYLACETATE_c + PROTON_c\n",
      "\n",
      "367 RXN0-6562 3-P-HYDROXYPYRUVATE_c + WATER_c --> OH-PYR_c + Pi_c\n",
      "367 RXN0-6562 3-P-HYDROXYPYRUVATE_c + WATER_c --> OH-PYR_c + Pi_c\n",
      "\n",
      "368 2-OCTAPRENYLPHENOL-HYDROX-RXN 2-OCTAPRENYLPHENOL_c + NADPH_c + OXYGEN-MOLECULE_c + PROTON_c --> 2-OCTAPRENYL-6-HYDROXYPHENOL_c + NADP_c + WATER_c\n",
      "368 2-OCTAPRENYLPHENOL-HYDROX-RXN 2-OCTAPRENYLPHENOL_c + NADPH_c + OXYGEN-MOLECULE_c + PROTON_c --> 2-OCTAPRENYL-6-HYDROXYPHENOL_c + NADP_c + WATER_c\n",
      "\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "meta_model_ids = [rxn.id for rxn in halo_meta.reactions]\n",
    "for rxn in halo_iJO_meta.reactions:\n",
    "    if rxn.id in meta_model_ids:\n",
    "        rxn1 = halo_meta.reactions.get_by_id(rxn.id)\n",
    "        print(k,rxn.id,rxn.reaction)\n",
    "        print(k,rxn1.id,rxn1.reaction)\n",
    "        print('')\n",
    "        k += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. Combine two draft models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_subset(lst1,lst2):\n",
    "    # check if elements in lst1 are all in lst2\n",
    "    is_ = True\n",
    "    for item in lst1:\n",
    "        if item not in lst2: is_ = False\n",
    "    return is_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TD01GL002722 and TD01GL002723\n",
      "TD01GL001634 or TD01GL002723\n",
      "\n",
      "TD01GL002722 and TD01GL002723\n",
      "TD01GL001635 or TD01GL002722\n",
      "\n",
      "TD01GL001227 and TD01GL002214\n",
      "TD01GL001226 or TD01GL001227 or TD01GL002214 or TD01GL002215\n",
      "\n",
      "TD01GL003202 and TD01GL003203\n",
      "TD01GL003202 or TD01GL003203 or TD01GL003289\n",
      "\n",
      "TD01GL003202 and TD01GL003203\n",
      "TD01GL003202 or TD01GL003203 or TD01GL003289\n",
      "\n",
      "TD01GL000278 and TD01GL001910\n",
      "TD01GL000277 or TD01GL000278\n",
      "\n",
      "( TD01GL001243 and TD01GL001660 ) or ( TD01GL001130 and TD01GL001243 )\n",
      "TD01GL001130 or TD01GL001660 or TD01GL002558\n",
      "\n",
      "TD01GL000350 and TD01GL000351\n",
      "TD01GL000039 or TD01GL000350 or TD01GL000351\n",
      "\n",
      "TD01GL001060 and TD01GL003062 and TD01GL003063\n",
      "TD01GL001060 or TD01GL001380 or TD01GL003062\n",
      "\n",
      "TD01GL002722 and TD01GL002723\n",
      "TD01GL001634 or TD01GL001635\n",
      "\n",
      "For above reactions, use the one from ecoli at the moment.\n"
     ]
    }
   ],
   "source": [
    "# add reactions in metacyc draft model to halo_iJO_meta model\n",
    "combined_rxns = dict()\n",
    "for rxn in halo_iJO_meta.reactions: combined_rxns[rxn.id] = rxn\n",
    "for rxn in halo_meta.reactions:\n",
    "    eco_rxn = combined_rxns.get(rxn.id)\n",
    "    if eco_rxn is None: combined_rxns[rxn.id] = rxn\n",
    "    else: \n",
    "        # if the rxn exist in both models, combine their gr rule \n",
    "        gr_eco = eco_rxn.gene_reaction_rule\n",
    "        gr_meta = rxn.gene_reaction_rule\n",
    "        if gr_eco == gr_meta: continue\n",
    "            \n",
    "        ## case 1: if 'and' not in both gr rules. Combine all genes with 'or'.\n",
    "        if ' and ' not in gr_eco and ' and ' not in gr_meta:\n",
    "            genes = set(parse_gr(gr_eco)+parse_gr(gr_meta))\n",
    "            new_gr = ''\n",
    "            for gene in genes: new_gr += gene + ' or '\n",
    "            new_gr = new_gr[:-4]\n",
    "            \n",
    "            eco_rxn.gene_reaction_rule = new_gr\n",
    "        \n",
    "        ## case 2: if two grs only differ in 'and' 'or' while with the same genes. use the one from eco li\n",
    "        elif parse_gr(gr_eco) == parse_gr(gr_meta): pass\n",
    "        \n",
    "        ## case 3: if genes in metacyc model is a subset of eco model, use the one from eco model\n",
    "        elif is_subset(parse_gr(gr_meta),parse_gr(gr_eco)): pass\n",
    "        else: \n",
    "            print(gr_eco)\n",
    "            print(gr_meta)\n",
    "            #print(new_gr)\n",
    "            print('')\n",
    "print('For above reactions, use the one from ecoli at the moment.')\n",
    "            \n",
    "\n",
    "\n",
    "halo_meta_c = cobra.Model('combined_metacyc_iJO1366')\n",
    "halo_meta_c.add_reactions(combined_rxns.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1271\n",
      "Number of metabolits: 1641\n",
      "Number of compartments: 2\n",
      "Number of genes: 1230\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1836\n",
      "Number of metabolits: 1979\n",
      "Number of compartments: 2\n",
      "Number of genes: 1270\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.6 compare the combinded model and the universal model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 3-DEHYDROQUINATE-DEHYDRATASE-RXN DEHYDROQUINATE_c --> 3-DEHYDRO-SHIKIMATE_c + WATER_c\n",
      "1 3-DEHYDROQUINATE-DEHYDRATASE-RXN DEHYDROQUINATE_c <=> 3-DEHYDRO-SHIKIMATE_c + WATER_c\n",
      "\n",
      "2 TAGAALDOL-RXN TAGATOSE-1-6-DIPHOSPHATE_c <=> DIHYDROXY-ACETONE-PHOSPHATE_c + GAP_c\n",
      "2 TAGAALDOL-RXN TAGATOSE-1-6-DIPHOSPHATE_c --> DIHYDROXY-ACETONE-PHOSPHATE_c + GAP_c\n",
      "\n",
      "3 HISTPRATPHYD-RXN PHOSPHORIBOSYL-ATP_c + WATER_c --> PHOSPHORIBOSYL-AMP_c + PPI_c + PROTON_c\n",
      "3 HISTPRATPHYD-RXN PHOSPHORIBOSYL-ATP_c + WATER_c --> PHOSPHORIBOSYL-AMP_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "4 ASPARTATEKIN-RXN ATP_c + L-ASPARTATE_c <=> ADP_c + L-BETA-ASPARTYL-P_c\n",
      "4 ASPARTATEKIN-RXN ATP_c + L-ASPARTATE_c --> ADP_c + L-BETA-ASPARTYL-P_c\n",
      "\n",
      "5 RXN0-5111 GLYCOLLATE_e <=> GLYCOLLATE_c\n",
      "5 RXN0-5111 GLYCOLLATE_e + NA+_e <=> GLYCOLLATE_c + NA+_c\n",
      "\n",
      "6 RXN0-1981 ACET_e <=> ACET_c\n",
      "6 RXN0-1981 ACET_e + NA+_e <=> ACET_c + NA+_c\n",
      "\n",
      "7 RXN-9615 AMMONIUM_e <=> AMMONIUM_c\n",
      "7 RXN-9615 AMMONIUM_e --> AMMONIUM_c\n",
      "\n",
      "8 RXN0-1924 CYS_e <=> CYS_c\n",
      "8 RXN0-1924 CYS_c --> CYS_e\n",
      "\n",
      "9 TRANS-RXN0-569 LYS_e <=> LYS_c\n",
      "9 TRANS-RXN0-569 LYS_c --> LYS_e\n",
      "\n",
      "10 TRANS-RXN-137 NITRITE_e <=> NITRITE_c\n",
      "10 TRANS-RXN-137 NITRITE_c --> NITRITE_e\n",
      "\n",
      "11 ACETYLGLUTKIN-RXN ACETYL-GLU_c + ATP_c --> ADP_c + N-ACETYL-GLUTAMYL-P_c\n",
      "11 ACETYLGLUTKIN-RXN ACETYL-GLU_c + ATP_c <=> ADP_c + N-ACETYL-GLUTAMYL-P_c\n",
      "\n",
      "12 TRANS-RXN0-281 3-KETOBUTYRATE_e <=> 3-KETOBUTYRATE_c\n",
      "12 TRANS-RXN0-281 3-KETOBUTYRATE_e --> 3-KETOBUTYRATE_c\n",
      "\n",
      "13 RXN0-1923 ACETYLSERINE_e <=> ACETYLSERINE_c\n",
      "13 RXN0-1923 ACETYLSERINE_c --> ACETYLSERINE_e\n",
      "\n",
      "14 TRANS-RXN0-240 BIOTIN_e <=> BIOTIN_c\n",
      "14 TRANS-RXN0-240 BIOTIN_e --> BIOTIN_c\n",
      "\n",
      "15 TRANS-RXN0-445 CHITOBIOSE_e <=> CHITOBIOSE_c\n",
      "15 TRANS-RXN0-445 CHITOBIOSE_e --> CHITOBIOSE_c\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# there are some reactions with the same id while with different equations\n",
    "# \n",
    "k = 1\n",
    "meta_model_ids = [rxn.id for rxn in halo_meta_c.reactions]\n",
    "for rxn in universal.reactions:\n",
    "    if rxn.id in meta_model_ids:\n",
    "        rxn1 = halo_meta_c.reactions.get_by_id(rxn.id)\n",
    "        if rxn.reaction != rxn1.reaction:\n",
    "            print(k,rxn.id,rxn.reaction)\n",
    "            print(k,rxn1.id,rxn1.reaction)\n",
    "            print('')\n",
    "            k += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if there are some reactions with the same id while with different equations\n",
    "# \n",
    "k = 1\n",
    "meta_model_ids = [rxn.id for rxn in halo_iJO_meta.reactions]\n",
    "for rxn in universal.reactions:\n",
    "    if rxn.id in meta_model_ids:\n",
    "        rxn1 = halo_meta_c.reactions.get_by_id(rxn.id)\n",
    "        if rxn.reaction != rxn1.reaction:\n",
    "            print(k,rxn.id,rxn.reaction)\n",
    "            print(k,rxn1.id,rxn1.reaction)\n",
    "            print('')\n",
    "            k += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add exchange reactions for all metabolites in medium. Use _e for extracellular compartment. Set all these exchange reversible\n",
    "\n",
    "* For those metabolites that doesn't have any transporters, add reversible transport reactions\n",
    "* use metacyc id for all metabolites. if no metacyc id, then use bigg.metabolite id\n",
    "* use 'Exchange_xx' for exhange reactions. xx is a metabolite id\n",
    "* use 'Transport_xx' for transport reactions. xx is a metabolite id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exchange_CPD0-2422  <=> CPD0-2422_e\n",
      "\n",
      "Exchange_CL-  <=> CL-_e\n",
      "\n",
      "Exchange_Glucopyranose  <=> Glucopyranose_e\n",
      "\n",
      "Exchange_UREA  <=> UREA_e\n",
      "\n",
      "Exchange_MG+2  <=> MG+2_e\n",
      "MG+2_c c\n",
      "Transport_MG+2 MG+2_e <=> MG+2_c\n",
      "\n",
      "Exchange_HSO4  <=> HSO4_e\n",
      "\n",
      "Exchange_CPD-16459  <=> CPD-16459_e\n",
      "\n",
      "Exchange_FE+3  <=> FE+3_e\n",
      "FE+3_c c\n",
      "Transport_FE+3 FE+3_e <=> FE+3_c\n",
      "\n",
      "Exchange_AMMONIA  <=> AMMONIA_e\n",
      "\n",
      "Exchange_CIT  <=> CIT_e\n",
      "\n",
      "Exchange_FE+2  <=> FE+2_e\n",
      "\n",
      "Exchange_CA+2  <=> CA+2_e\n",
      "CA+2_c c\n",
      "Transport_CA+2 CA+2_e <=> CA+2_c\n",
      "\n",
      "Exchange_ZN+2  <=> ZN+2_e\n",
      "\n",
      "Exchange_MN+2  <=> MN+2_e\n",
      "Transport_MN+2 MN+2_e <=> MN+2_c\n",
      "\n",
      "Exchange_CO+2  <=> CO+2_e\n",
      "CO+2_c c\n",
      "Transport_CO+2 CO+2_e <=> CO+2_c\n",
      "\n",
      "Exchange_K+  <=> K+_e\n",
      "\n",
      "Exchange_CU+2  <=> CU+2_e\n",
      "CU+2_c c\n",
      "Transport_CU+2 CU+2_e <=> CU+2_c\n",
      "\n",
      "Exchange_NI+2  <=> NI+2_e\n",
      "\n",
      "Exchange_CPD-3  <=> CPD-3_e\n",
      "\n",
      "Exchange_OXYGEN-MOLECULE  <=> OXYGEN-MOLECULE_e\n",
      "\n",
      "Exchange_L-ALPHA-ALANINE  <=> L-ALPHA-ALANINE_e\n",
      "\n",
      "Exchange_CYS  <=> CYS_e\n",
      "\n",
      "Exchange_L-ASPARTATE  <=> L-ASPARTATE_e\n",
      "\n",
      "Exchange_GLT  <=> GLT_e\n",
      "\n",
      "Exchange_PHE  <=> PHE_e\n",
      "\n",
      "Exchange_GLY  <=> GLY_e\n",
      "\n",
      "Exchange_HIS  <=> HIS_e\n",
      "\n",
      "Exchange_ILE  <=> ILE_e\n",
      "\n",
      "Exchange_LYS  <=> LYS_e\n",
      "\n",
      "Exchange_LEU  <=> LEU_e\n",
      "\n",
      "Exchange_MET  <=> MET_e\n",
      "\n",
      "Exchange_ASN  <=> ASN_e\n",
      "\n",
      "Exchange_PRO  <=> PRO_e\n",
      "\n",
      "Exchange_GLN  <=> GLN_e\n",
      "\n",
      "Exchange_ARG  <=> ARG_e\n",
      "\n",
      "Exchange_SER  <=> SER_e\n",
      "\n",
      "Exchange_THR  <=> THR_e\n",
      "\n",
      "Exchange_VAL  <=> VAL_e\n",
      "\n",
      "Exchange_TRP  <=> TRP_e\n",
      "\n",
      "Exchange_TYR  <=> TYR_e\n",
      "\n",
      "Exchange_ADENINE  <=> ADENINE_e\n",
      "\n",
      "Exchange_CPD-15990  <=> CPD-15990_e\n",
      "\n",
      "Exchange_L-LACTATE  <=> L-LACTATE_e\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#####Last time stopped here\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('medium')\n",
    "\n",
    "mnx2meta = dict()\n",
    "for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "    cont = line.split()\n",
    "    mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "all_rxns = halo_meta_c.reactions\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    mnx = sh.cell(i,6).value\n",
    "    bigg = sh.cell(i,2).value\n",
    "    name = sh.cell(i,0).value\n",
    "    \n",
    "    met_ids = mnx2meta.get(mnx,None)\n",
    "    if met_ids is not None:\n",
    "        met_id = None\n",
    "        for item in met_ids:\n",
    "            try: \n",
    "                halo_meta_c.metabolites.get_by_id(item)\n",
    "                met_id = item\n",
    "                break\n",
    "            except: None\n",
    "        if met_id is None: met_id = met_ids[0]\n",
    "            \n",
    "    else: met_id = bigg\n",
    "    \n",
    "    # create exchange reaction\n",
    "    met_e = cobra.Metabolite(met_id+'_e')\n",
    "    met_e.name = name\n",
    "    met_e.compartment = 'e'\n",
    "    exg_rxn = cobra.Reaction('Exchange_'+met_id)\n",
    "    exg_rxn.upper_bound = 1000\n",
    "    exg_rxn.lower_bound = -1000\n",
    "    exg_rxn.add_metabolites({met_e:1})\n",
    "    #halo_meta_c.add_reaction(exg_rxn) \n",
    "    all_rxns.append(exg_rxn)\n",
    "    print(exg_rxn.id,exg_rxn.reaction)\n",
    "    \n",
    "    \n",
    "    # create transport reaction for some ions which transporter were not found\n",
    "    if met_id in ['CA+2','CO+2','CU+2','FE+3','MG+2','MN+2']:\n",
    "        try: \n",
    "            met_c = halo_meta_c.metabolites.get_by_id(met_id+'_c')\n",
    "            print(met_c,met_c.compartment)\n",
    "        except:\n",
    "            met_c = cobra.Metabolite(met_id+'_c')\n",
    "            met_c.name = name\n",
    "            met_c.compartment = 'c'\n",
    "\n",
    "        tran_rxn = cobra.Reaction('Transport_'+met_id)\n",
    "        tran_rxn.upper_bound = 1000\n",
    "        tran_rxn.lower_bound = -1000\n",
    "        tran_rxn.add_metabolites({met_e:-1,met_c:1})\n",
    "        all_rxns.append(tran_rxn)\n",
    "        print(tran_rxn.id,tran_rxn.reaction)\n",
    "    \n",
    "    \n",
    "    # add exchange and transport reactions\n",
    "    \n",
    "    #halo_meta_c.add_reaction(tran_rxn)# ,\n",
    "    \n",
    "    #\n",
    "    print('')\n",
    "    \n",
    "# if we use halo_meta_c.add_reactions([]) to add transport and exchange reactions, there would be a key error in \n",
    "# optlang costrant variant when use halo_meta_c.copy().\n",
    "halo_meta_c = cobra.Model(halo_meta_c.id)\n",
    "halo_meta_c.add_reactions(all_rxns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Add exchange reactions for all secreted metabolites. Use _e for extracellular compartment. Set all these exchange irreversible \n",
    "\n",
    "* use metacyc id for all metabolites. if no metacyc id, then use bigg.metabolite id\n",
    "* use 'Exchange_xx' for exhange reactions. xx is a metabolite id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exchange_ac ac_e --> \n",
      "\n",
      "Exchange_pyr pyr_e --> \n",
      "\n",
      "Exchange_co2 co2_e --> \n",
      "\n",
      "Exchange_h2o h2o_e --> \n",
      "\n"
     ]
    }
   ],
   "source": [
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('secretion')\n",
    "\n",
    "mnx2meta = dict()\n",
    "for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "    cont = line.split()\n",
    "    mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "all_rxns = halo_meta_c.reactions\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    mnx = sh.cell(i,6).value\n",
    "    bigg = sh.cell(i,2).value\n",
    "    name = sh.cell(i,0).value\n",
    "    \n",
    "    met_ids = mnx2meta.get(mnx,None)\n",
    "    if met_ids is not None:\n",
    "        met_id = None\n",
    "        for item in met_ids:\n",
    "            try: \n",
    "                halo_meta_c.metabolites.get_by_id(item)\n",
    "                met_id = item\n",
    "                break\n",
    "            except: None\n",
    "        if met_id is None: met_id = met_ids[0]\n",
    "            \n",
    "    else: met_id = bigg\n",
    "    \n",
    "    # create exchange reaction\n",
    "    met_e = cobra.Metabolite(met_id+'_e')\n",
    "    met_e.name = name\n",
    "    met_e.compartment = 'e'\n",
    "    exg_rxn = cobra.Reaction('Exchange_'+met_id)\n",
    "    exg_rxn.upper_bound = 1000\n",
    "    exg_rxn.lower_bound = 0\n",
    "    exg_rxn.add_metabolites({met_e:-1})\n",
    "    #halo_meta_c.add_reaction(exg_rxn) \n",
    "    all_rxns.append(exg_rxn)\n",
    "    \n",
    "    \n",
    "    # create transport reaction\n",
    "    '''\n",
    "    try: \n",
    "        met_c = halo_meta_c.metabolites.get_by_id(met_id+'_c')\n",
    "        print(met_c,met_c.compartment)\n",
    "    except:\n",
    "        met_c = cobra.Metabolite(met_id+'_c')\n",
    "        met_c.name = name\n",
    "        met_c.compartment = 'c'\n",
    "    \n",
    "    tran_rxn = cobra.Reaction('Transport_'+met_id)\n",
    "    tran_rxn.upper_bound = 1000\n",
    "    tran_rxn.lower_bound = -1000\n",
    "    tran_rxn.add_metabolites({met_e:-1,met_c:1})\n",
    "    all_rxns.append(tran_rxn)\n",
    "    '''\n",
    "    # add exchange and transport reactions\n",
    "    \n",
    "    #halo_meta_c.add_reaction(tran_rxn)# ,\n",
    "    print(exg_rxn.id,exg_rxn.reaction)\n",
    "    #print(tran_rxn.id,tran_rxn.reaction)\n",
    "    print('')\n",
    "    \n",
    "# if we use halo_meta_c.add_reactions([]) to add transport and exchange reactions, there would be a key error in \n",
    "# optlang costrant variant when use halo_meta_c.copy().\n",
    "halo_meta_c = cobra.Model(halo_meta_c.id)\n",
    "halo_meta_c.add_reactions(all_rxns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1889\n",
      "Number of metabolits: 2005\n",
      "Number of compartments: 2\n",
      "Number of genes: 1270\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add protein, dna, rna, ions, lipid synthesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_halo_met_c(bigg_id,mnx_id,mnx2meta,halomodel):\n",
    "    # check metacyc id\n",
    "    meta_ids = mnx2meta.get(mnx_id,None)\n",
    "    met = None\n",
    "    if meta_ids is not None:\n",
    "        # check which id is in model\n",
    "        for item in meta_ids:\n",
    "            try: \n",
    "                met = halomodel.metabolites.get_by_id(item+'_c')\n",
    "                break\n",
    "            except: met = cobra.Metabolite(item+'_c')\n",
    "    else:\n",
    "        item = bigg_id+'_c'\n",
    "        try: met = halomodel.metabolites.get_by_id(item)\n",
    "        except: met = cobra.Metabolite(item)\n",
    "    return met "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_if_model_can_produce(met,halomodel):\n",
    "    # add sink reaction\n",
    "    rxn_id = 'tmp'\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halomodel.add_reaction(rxn_sink)\n",
    "    halomodel.objective = 'tmp'\n",
    "    f = halomodel.optimize().objective_value\n",
    "    halomodel.remove_reactions([rxn_sink])\n",
    "    print('maximal flux:',met.id,f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_general_rxn(sheet_name,gen_met_name,gen_rxn_name,halomodel):\n",
    "    # add reaction for protein synthesis. 20 aas -> protein\n",
    "    book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "    sh = book.sheet_by_name(sheet_name)\n",
    "\n",
    "    mnx2meta = dict()\n",
    "    for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "        cont = line.split()\n",
    "        mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "    coeffs = dict()\n",
    "    for i in range(sh.nrows-1):\n",
    "        i += 1\n",
    "        mnx_id = sh.cell(i,9).value\n",
    "        bigg_id = sh.cell(i,4).value\n",
    "        met = find_halo_met_c(bigg_id,mnx_id,mnx2meta,halomodel)\n",
    "        coeffs[met] = sh.cell(i,3).value\n",
    "    met = cobra.Metabolite('{}_c'.format(gen_met_name))\n",
    "    met.name = gen_met_name\n",
    "    met.compartment = 'c'\n",
    "    coeffs[met] = 1\n",
    "\n",
    "    rxn = cobra.Reaction(gen_rxn_name)\n",
    "    rxn.add_metabolites(coeffs)\n",
    "    halomodel.add_reaction(rxn)\n",
    "    print(rxn.reaction)\n",
    "    test_if_model_can_produce(met,halomodel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.295792 ARG_c + 0.241055 ASN_c + 0.09158 CYS_c + 0.26316 GLN_c + 0.26316 GLT_c + 0.612638 GLY_c + 0.094738 HIS_c + 0.290529 ILE_c + 0.513689 L-ALPHA-ALANINE_c + 0.241055 L-ASPARTATE_c + 0.450531 LEU_c + 0.343161 LYS_c + 0.153686 MET_c + 0.185265 PHE_c + 0.221055 PRO_c + 0.215792 SER_c + 0.253687 THR_c + 0.056843 TRP_c + 0.137896 TYR_c + 0.423162 VAL_c --> protein_c\n",
      "maximal flux: protein_c 0.0\n"
     ]
    }
   ],
   "source": [
    "add_general_rxn('protein','protein','Protein_synthesis',halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.026166 DATP_c + 0.027017 DCTP_c + 0.027017 DGTP_c + 0.026166 TTP_c --> dna_c\n",
      "maximal flux: dna_c 0.0\n"
     ]
    }
   ],
   "source": [
    "add_general_rxn('dna','dna','DNA_synthesis',halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.175 ATP_c + 0.133508 CTP_c + 0.215096 GTP_c + 0.144104 UTP_c --> rna_c\n",
      "maximal flux: rna_c 0.0\n"
     ]
    }
   ],
   "source": [
    "add_general_rxn('rna','rna','RNA_synthesis',halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.013013 AMMONIUM_c + 0.005205 CA+2_c + 0.005205 CL-_c + 2.5e-05 CO+2_c + 7e-06 CPD-3_c + 0.000709 CU+2_c + 0.006715 FE+2_c + 0.007808 FE+3_c + 0.195193 K+_c + 0.008675 MG+2_c + 0.000691 MN+2_c + 0.000323 NI+2_c + 0.004338 SULFATE_c + 0.000341 ZN+2_c --> ions_c\n",
      "maximal flux: ions_c 0.0\n"
     ]
    }
   ],
   "source": [
    "add_general_rxn('ions','ions','ions_pool',halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.063814 CPD-12819_c + 0.075214 CPD-17086_c + 0.019456 KDO2-LIPID-IVA_c --> lipids_c\n",
      "maximal flux: lipids_c 0.0\n"
     ]
    }
   ],
   "source": [
    "add_general_rxn('lipids','lipids','lipids_synthesis',halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if model can produce the metabolites in biomass one by one\n",
    "Match metabolites based on\n",
    "* metacyc id\n",
    "* bigg id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_yeast_extraction(model,upper_bound):\n",
    "    # given the evidence that TD01 strain can growth wihout supplyment of amino acids, \n",
    "    # the strain should be able to produce 20 amino acids by itself.\n",
    "    # Block the uptake of 20 amino acids\n",
    "    amino_acids = ['CYS_c','L-ASPARTATE_c','GLT_c','PHE_c','GLY_c',\n",
    "                  'HIS_c','ILE_c','LYS_c','LEU_c','MET_c','ASN_c',\n",
    "                  'PRO_c','GLN_c','ARG_c','SER_c','THR_c','VAL_c',\n",
    "                  'TRP_c','L-ALPHA-ALANINE_c','TYR_c','L-LACTATE','CPD-15990', 'ADENINE']\n",
    "    for met_id in amino_acids:\n",
    "        exg_rxn = model.reactions.get_by_id('Exchange_'+met_id.replace('_c',''))\n",
    "        exg_rxn.upper_bound = upper_bound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def check_if_model_can_produce_mets_in_biomass_one_by_one(model):\n",
    "    book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "    sheets = ['protein','dna','rna','ions','lipids','others']\n",
    "\n",
    "    mnx2meta = dict()\n",
    "    for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "        cont = line.split()\n",
    "        mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "    for name in sheets:\n",
    "        sh = book.sheet_by_name(name)\n",
    "        print(name)\n",
    "        for i in range(sh.nrows-1):\n",
    "            i += 1\n",
    "            mnx_id = sh.cell(i,9).value\n",
    "            bigg_id = sh.cell(i,4).value\n",
    "            met = find_halo_met_c(bigg_id,mnx_id,mnx2meta,model)\n",
    "            if met is None: print(bigg_id+'_c',0)\n",
    "            else:\n",
    "                # add sink reaction\n",
    "                rxn_id = 'tmp'\n",
    "                rxn_sink = cobra.Reaction(rxn_id)\n",
    "                rxn_sink.add_metabolites({met:-1})\n",
    "                model.add_reaction(rxn_sink)\n",
    "                model.objective = 'tmp'\n",
    "                f = model.optimize().objective_value\n",
    "                model.remove_reactions([rxn_sink])\n",
    "                print(met.id,met.name,f)\n",
    "        print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein\n",
      "L-ALPHA-ALANINE_c L-alanine 0.0\n",
      "ARG_c L-arginine 0.0\n",
      "ASN_c L-asparagine 0.0\n",
      "L-ASPARTATE_c L-aspartate 0.0\n",
      "CYS_c L-cysteine 0.0\n",
      "GLN_c L-glutamine 0.0\n",
      "GLT_c L-glutamate 0.0\n",
      "GLY_c glycine 0.0\n",
      "HIS_c L-histidine 0.0\n",
      "ILE_c L-isoleucine 0.0\n",
      "LEU_c L-leucine 0.0\n",
      "LYS_c L-lysine 0.0\n",
      "MET_c L-methionine 0.0\n",
      "PHE_c L-phenylalanine 0.0\n",
      "PRO_c L-proline 0.0\n",
      "SER_c L-serine 0.0\n",
      "THR_c L-threonine 0.0\n",
      "TRP_c L-tryptophan 0.0\n",
      "TYR_c L-tyrosine 0.0\n",
      "VAL_c L-valine 0.0\n",
      "\n",
      "dna\n",
      "DATP_c dATP 0.0\n",
      "DCTP_c dCTP 0.0\n",
      "DGTP_c dGTP 0.0\n",
      "TTP_c dTTP 0.0\n",
      "\n",
      "rna\n",
      "ATP_c ATP 0.0\n",
      "CTP_c CTP 0.0\n",
      "GTP_c GTP 0.0\n",
      "UTP_c UTP 0.0\n",
      "\n",
      "ions\n",
      "CO+2_c Co2+ 1000.0\n",
      "CU+2_c Cu2+ 1000.0\n",
      "FE+2_c Fe2+ 1000.0\n",
      "FE+3_c Fe3+ 1000.0\n",
      "K+_c  1000.0\n",
      "MG+2_c Mg2+ 1000.0\n",
      "MN+2_c Manganese 1000.0\n",
      "CPD-3_c molybdate 0.0\n",
      "NI+2_c  0.0\n",
      "SULFATE_c sulfate 0.0\n",
      "ZN+2_c  1000.0\n",
      "CA+2_c  1000.0\n",
      "CL-_c chloride 1000.0\n",
      "AMMONIUM_c ammonium 0.0\n",
      "\n",
      "lipids\n",
      "KDO2-LIPID-IVA_c alpha-Kdo-(2->4)-alpha-Kdo-(2->6)-lipid IVA 0.0\n",
      "CPD-12819_c Phosphatidylethanolamine (dihexadecanoyl, n-C16:0) 0.0\n",
      "CPD-17086_c Phosphatidylethanolamine (dihexadec-9enoyl, n-C16:1) 0.0\n",
      "\n",
      "others\n",
      "PROTON_c H+ 250.0\n",
      "10-FORMYL-THF_c 10-FORMYL-THF 0.0\n",
      "2fe2s_c [2Fe-2S] iron-sulfur cluster 0.0\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c 3-(all-trans-octaprenyl)benzene-1,2-diol 0.0\n",
      "CPD-7_c [4Fe-4S] iron-sulfur cluster 0.0\n",
      "S-ADENOSYLMETHIONINE_c S-ADENOSYLMETHIONINE 0.0\n",
      "BIOTIN_c biotin 0.0\n",
      "CO-A_c coenzyme A 0.0\n",
      "FAD_c FAD 0.0\n",
      "WATER_c H2O 0.0\n",
      "METHYLENE-THF_c METHYLENE-THF 0.0\n",
      "CPD0-2278_c Two disacharide linked murein units, pentapeptide crosslinked tetrapeptide (A2pm->D-ala) (middle of chain) 0.0\n",
      "NAD_c NAD+ 0.0\n",
      "NADP_c NADP+ 0.0\n",
      "PROTOHEME_c ferroheme b 0.0\n",
      "PYRIDOXAL_PHOSPHATE_c pyridoxal 5'-phosphate 0.0\n",
      "RIBOFLAVIN_c riboflavin 0.0\n",
      "SIROHEME_c siroheme 0.0\n",
      "SUC-COA_c succinyl-CoA 0.0\n",
      "THF_c THF 0.0\n",
      "THIAMINE-PYROPHOSPHATE_c thiamine diphosphate 0.0\n",
      "CPD-9649_c Undecaprenyl diphosphate 0.0\n",
      "ADP_c ADP 0.0\n",
      "Pi_c phosphate 0.0\n",
      "PPI_c diphosphate 0.0\n",
      "ATP_c ATP 0.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "check_if_model_can_produce_mets_in_biomass_one_by_one(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 9. Gap-filling with iML1515"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_gap_filling(halo_met,halmodel,universal,**kwargs):\n",
    " \n",
    "    rxn_id = 'tmp'\n",
    "    try: \n",
    "        rxn_sink = halmodel.reactions.get_by_id(rxn_id)\n",
    "        halmodel.remove_reactions([rxn_sink])\n",
    "    except: None\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halmodel.add_reaction(rxn_sink)\n",
    "    halmodel.objective = 'tmp'\n",
    "    f1 = halmodel.optimize().objective_value\n",
    "    \n",
    "    added_genes = dict()# {'rxn_id':'genes'}\n",
    "    print(halo_met.id)\n",
    "    print('Before gap-filling:',f1)\n",
    "    if f1 != 0: print('No need for gap-filling')\n",
    "    else:\n",
    "        try: \n",
    "            solution = cobra.flux_analysis.gapfill(halmodel, universal,**kwargs)\n",
    "            all_rxns = dict()\n",
    "            rxns_to_add = list()\n",
    "            for rxn in solution[0]:\n",
    "                try: halmodel.reactions.get_by_id(rxn.id)\n",
    "                except: \n",
    "                    rxn_clean = rxn.copy()\n",
    "                    added_genes[rxn_clean.id] = rxn_clean.gene_reaction_rule\n",
    "                    \n",
    "                    # add _missing tag for those genes\n",
    "                    new_gr = rxn_clean.gene_reaction_rule\n",
    "                    for gene in parse_gr(rxn_clean.gene_reaction_rule):\n",
    "                        new_gr = new_gr.replace(gene,gene+'_missing')\n",
    "        \n",
    "                    rxn_clean.gene_reaction_rule = new_gr\n",
    "                    rxns_to_add.append(rxn_clean)\n",
    "            halmodel.add_reactions(rxns_to_add)\n",
    "            f2 = halmodel.optimize().objective_value\n",
    "            print('After gap-filling:',f2)\n",
    "            print('Added reactions:')\n",
    "            for rxn in rxns_to_add:print(rxn.id,rxn.reaction)\n",
    "        except: print('Gap-filling failed for',halo_met.id)\n",
    "\n",
    "    halmodel.remove_reactions([rxn_sink])\n",
    "    print('\\n\\n')\n",
    "    return added_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "def try_gap_filling(halo_met,halmodel,universal, **kwargs):\n",
    "    # glycogen_c\n",
    "    solution = [[]]\n",
    "    rxn_id = 'tmp'\n",
    "    try: \n",
    "        rxn_sink = halmodel.reactions.get_by_id(rxn_id)\n",
    "        halmodel.remove_reactions([rxn_sink])\n",
    "        halomodel.repair()\n",
    "    except: None\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halmodel.add_reaction(rxn_sink)\n",
    "    halmodel.objective = 'tmp'\n",
    "    optres = halmodel.optimize()\n",
    "    f1 = optres.objective_value\n",
    "    print('Before gap-filling:',f1)\n",
    "    if f1 != 0: \n",
    "        print('No need for gap-filling')\n",
    "        solution = [['No need for gap-filling']]\n",
    "        \n",
    "    else:\n",
    "        try: \n",
    "            solution = cobra.flux_analysis.gapfill(halmodel, universal,**kwargs)\n",
    "            halmodel.add_reactions(solution[0])\n",
    "            f2 = halmodel.optimize().objective_value\n",
    "            print('After gap-filling:',f2)\n",
    "            print('Added reactions:')\n",
    "            for rxn in solution[0]:print(rxn.id,rxn.reaction)\n",
    "            halmodel.remove_reactions(solution[0])\n",
    "        except: \n",
    "            print('Gap-filling failed for',halo_met.id)\n",
    "            solution = [['Failed']]\n",
    "    halmodel.remove_reactions([rxn_sink])\n",
    "    print('\\n\\n')\n",
    "    return solution"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# try to perform gap-filling for every component in biomass, get a list of reactions to be added.\n",
    "# output: \n",
    "# >met_id_in_eco, met_name, met_id_in_halo\n",
    "# rxn_id gr rule\n",
    "# \n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "fhand = open('../Results/gap_filled_reactions_try.txt','w')\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary.xlsx')\n",
    "sh = book.sheet_by_name('BiomassComparison')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    rxns = try_gap_filling(met,halo_meta_c,universal)[0]\n",
    "    \n",
    "    #if rxns[0] == 'Failed': \n",
    "    #    rxns = try_gap_filling(met,halo_meta_c,universal,demand_reactions=True)[0]\n",
    "    \n",
    "    fhand.write('>{0} {1} {2}\\n'.format(eco_met.id,eco_met.name,met.id))\n",
    "    if rxns[0] not in ['Failed','No need for gap-filling']:\n",
    "        for rxn in rxns:\n",
    "            fhand.write('{0}\\t{1}\\t{2}\\t{3}\\n'.format(rxn.id,rxn.name,rxn.reaction,rxn.gene_reaction_rule))\n",
    "    else: \n",
    "        fhand.write('{0}\\n'.format(rxns[0]))\n",
    "    fhand.write('\\n')\n",
    "    print('')\n",
    "fhand.close()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# block free uptake of lactate and TREHALOSE. Perform gap-filling with glucose as sole carbon source\n",
    "set_yeast_extraction(halo_meta_c,1000)\n",
    "# halo_meta_c.reactions.get_by_id('Exchange_L-LACTATE').upper_bound=0\n",
    "# halo_meta_c.reactions.get_by_id('Exchange_CPD-15990').upper_bound=0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Block uptake of 20 amino acids,  do gap-filling with iJO1366 as universal model. Do gap-filling for metabolites in order of (1) DNA , RNA, Protein, lipids, Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "fhand = open('../Results/gap_filled_reactions_v1_without_missing_genes.txt','w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_gap_fill_res(fhand,added_genes,met):\n",
    "    fhand.write('>{0} {1}\\n'.format(met.id,met.name))\n",
    "    if len(added_genes)==0: fhand.write('None')\n",
    "    else:\n",
    "        for k,v in added_genes.items(): fhand.write('{0},{1}\\n'.format(k,v))\n",
    "    fhand.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dna_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpizh_5l3_.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2010 rows, 3790 columns, 17824 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpud8xmt7j.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n",
      "After gap-filling: 96.08565305719085\n",
      "Added reactions:\n",
      "EX_pi_e Pi_e <=> \n",
      "EX_h_e PROTON_e <=> \n",
      "GLCtex_copy1 Glucopyranose_e --> Glucopyranose_c\n",
      "PItex Pi_e <=> Pi_c\n",
      "UREAtex UREA_e <=> UREA_c\n",
      "CITtex CIT_e <=> CIT_c\n",
      "\n",
      "\n",
      "\n",
      "rna_c\n",
      "Before gap-filling: 16.813968049268396\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "protein_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpf0wtxmpz.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2010 rows, 3802 columns, 17844 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmprtgd8g5l.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 7.646731749572431\n",
      "Added reactions:\n",
      "ASNS2 AMMONIUM_c + ATP_c + L-ASPARTATE_c --> AMP_c + ASN_c + PPI_c + PROTON_c\n",
      "HSK ATP_c + HOMO-SER_c --> ADP_c + O-PHOSPHO-L-HOMOSERINE_c + PROTON_c\n",
      "G5SADs L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c --> L-DELTA1-PYRROLINE_5-CARBOXYLATE_c + PROTON_c + WATER_c\n",
      "PPND NAD_c + PREPHENATE_c --> CARBON-DIOXIDE_c + NADH_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "ADSK APS_c + ATP_c --> ADP_c + PAPS_c + PROTON_c\n",
      "HSDy HOMO-SER_c + NADP_c <=> L-ASPARTATE-SEMIALDEHYDE_c + NADPH_c + PROTON_c\n",
      "THDPS DELTA1-PIPERIDEINE-2-6-DICARBOXYLATE_c + SUC-COA_c + WATER_c --> CO-A_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c\n",
      "EX_so4_e SULFATE_e <=> \n",
      "GLCtex_copy2 Glucopyranose_e <=> Glucopyranose_c\n",
      "\n",
      "\n",
      "\n",
      "lipids_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpawcqqws9.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2010 rows, 3820 columns, 17922 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp0gc4v07e.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_pi_e' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 113.89436024740593\n",
      "Added reactions:\n",
      "O2tex OXYGEN-MOLECULE_e <=> OXYGEN-MOLECULE_c\n",
      "EX_h2o_e WATER_e <=> \n",
      "PSSA161 SER_c + cdpdhdec9eg_c --> CMP_c + PROTON_c + ps161_c\n",
      "AGPAT160 1hdecg3p_c + palmACP_c --> ACP_c + CPD0-1422_c\n",
      "PSSA160 SER_c + cdpdhdecg_c --> CMP_c + CPD-12817_c + PROTON_c\n",
      "AGPAT161 1hdec9eg3p_c + hdeACP_c --> ACP_c + pa161_c\n",
      "\n",
      "\n",
      "\n",
      "PROTON_c\n",
      "PROTON_c\n",
      "Before gap-filling: 1000.0\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "10-FORMYL-THF_c\n",
      "10-FORMYL-THF_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpyisaohgd.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2010 rows, 3832 columns, 17964 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmppxgp9vye.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy1' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 25.078369905956123\n",
      "Added reactions:\n",
      "DNMPPA DIHYDRONEOPTERIN-P_c + WATER_c --> CPD-11770_c + Pi_c\n",
      "DNTPPA DIHYDRONEOPTERIN-P3_c + WATER_c --> DIHYDRONEOPTERIN-P_c + PPI_c + PROTON_c\n",
      "EX_nh4_e AMMONIUM_e <=> \n",
      "4HTHRK ATP_c + CPD0-2189_c --> 4-PHOSPHONOOXY-THREONINE_c + ADP_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2fe2s_c\n",
      "2fe2s_c\n",
      "Before gap-filling: 56.20155038759691\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c\n",
      "Before gap-filling: 17.595307917888594\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD-7_c\n",
      "CPD-7_c\n",
      "Before gap-filling: 28.68447082096933\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "S-ADENOSYLMETHIONINE_c\n",
      "S-ADENOSYLMETHIONINE_c\n",
      "Before gap-filling: 30.71672354948808\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "BIOTIN_c\n",
      "BIOTIN_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpgaib7gmk.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2011 rows, 3840 columns, 17994 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpfi8arh7j.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'HSDy' since it already exists.\n",
      "Ignoring reaction 'EX_so4_e' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'O2tex' since it already exists.\n",
      "Ignoring reaction 'EX_h2o_e' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 16.643550624133116\n",
      "Added reactions:\n",
      "DM_amob_c S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c --> \n",
      "EX_meoh_e ALCOHOL-GROUP_e --> \n",
      "DM_5drib_c CPD0-2167_c --> \n",
      "MEOHtex ALCOHOL-GROUP_e <=> METOH_c\n",
      "AI2K ATP_c + DIHYDROXYPENTANEDIONE_c --> ADP_c + PROTON_c + dhptdp_c\n",
      "PAI2I dhptdp_c --> dhptdd_c\n",
      "PAI2T CO-A_c + dhptdd_c --> ACETYL-COA_c + DIHYDROXY-ACETONE-PHOSPHATE_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CO-A_c\n",
      "CO-A_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpy8qjt627.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2014 rows, 3854 columns, 18026 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpgkfudm8f.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_pi_e' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy1' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'EX_so4_e' since it already exists.\n",
      "Ignoring reaction 'O2tex' since it already exists.\n",
      "Ignoring reaction 'EX_h2o_e' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 22.547914317925574\n",
      "Added reactions:\n",
      "PNTK ATP_c + PANTOTHENATE_c --> 4-P-PANTOTHENATE_c + ADP_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "FAD_c\n",
      "FAD_c\n",
      "Before gap-filling: 17.91936286709807\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "WATER_c\n",
      "WATER_c\n",
      "Before gap-filling: 1000.0\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "METHYLENE-THF_c\n",
      "METHYLENE-THF_c\n",
      "Before gap-filling: 25.078369905956116\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD0-2278_c\n",
      "CPD0-2278_c\n",
      "Before gap-filling: 18.33180568285976\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NAD_c\n",
      "NAD_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpsqhf1dyj.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2014 rows, 3856 columns, 18036 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpcisnfyvm.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_pi_e' since it already exists.\n",
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy1' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'EX_nh4_e' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 27.252081756245275\n",
      "Added reactions:\n",
      "NADS1 AMMONIUM_c + ATP_c + DEAMIDO-NAD_c --> AMP_c + NAD_c + PPI_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NADP_c\n",
      "NADP_c\n",
      "Before gap-filling: 26.011560693641684\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PROTOHEME_c\n",
      "PROTOHEME_c\n",
      "Before gap-filling: 25.247971145175832\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PYRIDOXAL_PHOSPHATE_c\n",
      "PYRIDOXAL_PHOSPHATE_c\n",
      "Before gap-filling: 128.5714285714286\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "RIBOFLAVIN_c\n",
      "RIBOFLAVIN_c\n",
      "Before gap-filling: 33.388981636059945\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SIROHEME_c\n",
      "SIROHEME_c\n",
      "Before gap-filling: 24.904877205119348\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUC-COA_c\n",
      "SUC-COA_c\n",
      "Before gap-filling: 21.76278563656146\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "THF_c\n",
      "THF_c\n",
      "Before gap-filling: 26.092628832354855\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "THIAMINE-PYROPHOSPHATE_c\n",
      "THIAMINE-PYROPHOSPHATE_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp_4sl8poa.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2014 rows, 3858 columns, 18050 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpenpc8dl1.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1503 rows, 4450 columns, 16736 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_pi_e' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'EX_so4_e' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'O2tex' since it already exists.\n",
      "Ignoring reaction 'EX_h2o_e' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 28.865979381443296\n",
      "Added reactions:\n",
      "THZPSN3 ATP_c + CPD-12279_c + DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c + iscssh_c --> AMP_c + CARBON-DIOXIDE_c + NADP_c + PPI_c + THZ-P_c + 2.0 WATER_c + iscs_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD-9649_c\n",
      "CPD-9649_c\n",
      "Before gap-filling: 14.634146341463406\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ADP_c\n",
      "ADP_c\n",
      "Before gap-filling: 44.17177914110425\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Pi_c\n",
      "Pi_c\n",
      "Before gap-filling: 1000.0\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PPI_c\n",
      "PPI_c\n",
      "Before gap-filling: 499.99999999999994\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ATP_c\n",
      "ATP_c\n",
      "Before gap-filling: 40.723981900452515\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "\n",
    "# DNA\n",
    "met = halo_meta_c.metabolites.dna_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# RNA \n",
    "met = halo_meta_c.metabolites.rna_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# Protein\n",
    "met = halo_meta_c.metabolites.protein_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# lipids\n",
    "met = halo_meta_c.metabolites.lipids_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# Others\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('others')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "    write_gap_fill_res(fhand,added_genes,met)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1929\n",
      "Number of metabolits: 2014\n",
      "Number of compartments: 2\n",
      "Number of genes: 1298\n",
      "Number of missing genes: 28\n",
      "Number of reactions with missing genes: 26\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update missing genes added by gap-filling. By find homologues in H. TD01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1929\n",
      "Number of metabolits: 2014\n",
      "Number of compartments: 2\n",
      "Number of genes: 1273\n",
      "Number of missing genes: 3\n",
      "Number of reactions with missing genes: 1\n"
     ]
    }
   ],
   "source": [
    "halo_meta_c = update_missing_genes(halo_meta_c,1e-20,report=True,coverage=45)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### For those reactions with missing genes, remove its gr rule and put a note as : \n",
    "{'source':'gapfilled from iJO1366, genes were not found in H.TD01'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THZPSN3 TD01GL000663 and TD01GL001192 and b0423_missing and b3990_missing and b4407_missing\n",
      "\n",
      "Number of reactions: 1929\n",
      "Number of metabolits: 2014\n",
      "Number of compartments: 2\n",
      "Number of genes: 1270\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "updated_rxns = list()\n",
    "for rxn in halo_meta_c.reactions:\n",
    "    if '_missing' in rxn.gene_reaction_rule:\n",
    "        print(rxn.id,rxn.gene_reaction_rule)\n",
    "        rxn.gene_reaction_rule = ''\n",
    "        rxn.notes['source']='gapfilled from iJO1366, genes were not found in H.TD01'\n",
    "    updated_rxns.append(rxn)\n",
    "\n",
    "print('')\n",
    "halo_meta_c = cobra.Model('model_v1')\n",
    "halo_meta_c.add_reactions(updated_rxns)\n",
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary.xlsx')\n",
    "sh = book.sheet_by_name('protein')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "    #write_gap_fill_res(fhand,added_genes,met)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Biomass verions 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.000223 10-FORMYL-THF_c + 0.000223 2-OCTAPRENYL-6-HYDROXYPHENOL_c + 2.6e-05 2fe2s_c + 75.37723 ATP_c + 2e-06 BIOTIN_c + 0.000576 CO-A_c + 0.00026 CPD-7_c + 5.5e-05 CPD-9649_c + 0.013894 CPD0-2278_c + 0.000223 FAD_c + 0.000223 METHYLENE-THF_c + 0.000447 NADP_c + 0.001831 NAD_c + 0.000223 PROTOHEME_c + 0.000223 PYRIDOXAL_PHOSPHATE_c + 0.000223 RIBOFLAVIN_c + 0.000223 S-ADENOSYLMETHIONINE_c + 0.000223 SIROHEME_c + 9.8e-05 SUC-COA_c + 0.000223 THF_c + 0.000223 THIAMINE-PYROPHOSPHATE_c + 70.028756 WATER_c + dna_c + ions_c + lipids_c + protein_c + rna_c --> 75.37723 ADP_c + 0.773903 PPI_c + 75.37723 PROTON_c + 75.37323 Pi_c\n"
     ]
    }
   ],
   "source": [
    "biomass_rxn = cobra.Reaction('Biomass_v1')\n",
    "coeffs = {halo_meta_c.metabolites.protein_c:-1,\n",
    "          halo_meta_c.metabolites.dna_c:-1,\n",
    "          halo_meta_c.metabolites.rna_c:-1,\n",
    "          halo_meta_c.metabolites.ions_c:-1,\n",
    "          halo_meta_c.metabolites.lipids_c:-1}\n",
    "# Others\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('others')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    coeff = sh.cell(i,3).value\n",
    "    coeffs[met] = coeff\n",
    "biomass_rxn.add_metabolites(coeffs)\n",
    "print(biomass_rxn.reaction)\n",
    "halo_meta_c.add_reactions([biomass_rxn])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "solution = cobra.flux_analysis.gapfill(halo_meta_c, universal,demand_reactions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "halo_meta_c.reactions.Exchange_Glucopyranose.upper_bound = 5\n",
    "halo_meta_c.reactions.get_by_id('Exchange_L-LACTATE').upper_bound=10\n",
    "halo_meta_c.reactions.get_by_id('Exchange_CPD-15990').upper_bound=10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 5.748 at 0x1328929198>\n"
     ]
    }
   ],
   "source": [
    "#print(halo_meta_c.reactions.Exchange_Glucopyranose.upper_bound)\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "print(halo_meta_c.optimize())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 6.187 at 0x1328929128>\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,1000)\n",
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "print(halo_meta_c.optimize())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save model and filled reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein\n",
      "L-ALPHA-ALANINE_c L-alanine 336.66666666666663\n",
      "ARG_c L-arginine 98.47826086956522\n",
      "ASN_c L-asparagine 264.04761904761915\n",
      "L-ASPARTATE_c L-aspartate 1000.0\n",
      "CYS_c L-cysteine 104.13793103448285\n",
      "GLN_c L-glutamine 252.5\n",
      "GLT_c L-glutamate 336.6666666666667\n",
      "GLY_c glycine 838.333333333333\n",
      "HIS_c L-histidine 7.5\n",
      "ILE_c L-isoleucine 113.33333333333331\n",
      "LEU_c L-leucine 112.22222222222221\n",
      "LYS_c L-lysine 117.6923076923077\n",
      "MET_c L-methionine 92.72727272727273\n",
      "PHE_c L-phenylalanine 15.0\n",
      "PRO_c L-proline 180.0\n",
      "SER_c L-serine 336.6666666666667\n",
      "THR_c L-threonine 278.1818181818182\n",
      "TRP_c L-tryptophan 5.0\n",
      "TYR_c L-tyrosine 15.0\n",
      "VAL_c L-valine 144.28571428571428\n",
      "\n",
      "dna\n",
      "DATP_c dATP 7.5\n",
      "DCTP_c dCTP 7.5\n",
      "DGTP_c dGTP 7.500000000000005\n",
      "TTP_c dTTP 7.5\n",
      "\n",
      "rna\n",
      "ATP_c ATP 7.5000000000000036\n",
      "CTP_c CTP 7.5\n",
      "GTP_c GTP 7.5\n",
      "UTP_c UTP 7.5\n",
      "\n",
      "ions\n",
      "CO+2_c Co2+ 1000.0\n",
      "CU+2_c Cu2+ 1000.0\n",
      "FE+2_c Fe2+ 1000.0\n",
      "FE+3_c Fe3+ 1000.0\n",
      "K+_c  1000.0\n",
      "MG+2_c Mg2+ 1000.0\n",
      "MN+2_c Manganese 1000.0\n",
      "CPD-3_c molybdate 750.0\n",
      "NI+2_c  750.0\n",
      "SULFATE_c sulfate 750.0000000000003\n",
      "ZN+2_c  1000.0\n",
      "CA+2_c  1000.0\n",
      "CL-_c chloride 1000.0\n",
      "AMMONIUM_c ammonium 1000.0\n",
      "\n",
      "lipids\n",
      "KDO2-LIPID-IVA_c alpha-Kdo-(2->4)-alpha-Kdo-(2->6)-lipid IVA 1.5\n",
      "CPD-12819_c Phosphatidylethanolamine (dihexadecanoyl, n-C16:0) 15.53299492385787\n",
      "CPD-17086_c Phosphatidylethanolamine (dihexadec-9enoyl, n-C16:1) 16.19047619047619\n",
      "\n",
      "others\n",
      "PROTON_c H+ 1000.0\n",
      "10-FORMYL-THF_c 10-FORMYL-THF 7.5000000000000675\n",
      "2fe2s_c [2Fe-2S] iron-sulfur cluster 55.666666666666764\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c 3-(all-trans-octaprenyl)benzene-1,2-diol 12.8\n",
      "CPD-7_c [4Fe-4S] iron-sulfur cluster 28.412098298676746\n",
      "S-ADENOSYLMETHIONINE_c S-ADENOSYLMETHIONINE 7.5\n",
      "BIOTIN_c biotin 2.5\n",
      "CO-A_c coenzyme A 7.5\n",
      "FAD_c FAD 1.875\n",
      "WATER_c H2O 1000.0\n",
      "METHYLENE-THF_c METHYLENE-THF 7.5\n",
      "CPD0-2278_c Two disacharide linked murein units, pentapeptide crosslinked tetrapeptide (A2pm->D-ala) (middle of chain) 1.25\n",
      "NAD_c NAD+ 3.75\n",
      "NADP_c NADP+ 3.7500000000000036\n",
      "PROTOHEME_c ferroheme b 22.936708860759495\n",
      "PYRIDOXAL_PHOSPHATE_c pyridoxal 5'-phosphate 15.0\n",
      "RIBOFLAVIN_c riboflavin 2.5\n",
      "SIROHEME_c siroheme 3.75\n",
      "SUC-COA_c succinyl-CoA 7.5\n",
      "THF_c THF 7.5\n",
      "THIAMINE-PYROPHOSPHATE_c thiamine diphosphate 7.5\n",
      "CPD-9649_c Undecaprenyl diphosphate 10.534979423868315\n",
      "ADP_c ADP 7.5\n",
      "Pi_c phosphate 1000.0\n",
      "PPI_c diphosphate 499.9999999999999\n",
      "ATP_c ATP 7.5000000000000036\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "check_if_model_can_produce_mets_in_biomass_one_by_one(halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in halo_meta_c.metabolites: \n",
    "    try: met.charge = int(met.charge)\n",
    "    except: pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.write_sbml_model(halo_meta_c,\n",
    "        '../../../ModelFiles/xml/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.xml')\n",
    "fhand.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_matlab_model(halo_meta_c,\n",
    "'../../../ModelFiles/mat/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_json_model(halo_meta_c,\n",
    "        '../../../ModelFiles/json/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cobra.io.save_yaml_model(halo_meta_c,\n",
    "        '../Results/combine_halo_meta100_iJO1366_gap_fill_with_core_biomass_without_missing_genes.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 1930\n",
      "Number of metabolits: 2014\n",
      "Number of compartments: 2\n",
      "Number of genes: 1270\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 2712\n",
      "Number of metabolits: 1877\n",
      "Number of compartments: 3\n",
      "Number of genes: 1516\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "report_model_status(eco)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare reactions: model_with/without missing genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "model_with_missing_genes = cobra.io.read_sbml_model('../Results/combine_halo_meta100_iJO1366_gap_fill_with_core_biomass_with_missing_genes.xml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "k = 0\n",
    "missing_reactions = 0\n",
    "for rxn in model_with_missing_genes.reactions:\n",
    "    if '_missing' in rxn.gene_reaction_rule:\n",
    "        missing_reactions += 1\n",
    "        try: \n",
    "            halo_meta_c.reactions.get_by_id(rxn.id)\n",
    "            k += 1\n",
    "        except: None\n",
    "print('Number of reactions with missing genes that were found by gap-filling:',k,'/',missing_reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
