{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction of HaloGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tools import *\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. compare draft models from template and metacyc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 DEOXYRIBOSE-P-ALD-RXN DEOXY-RIBOSE-5P_c --> ACETALD_c + GAP_c\n",
      "1 DEOXYRIBOSE-P-ALD-RXN DEOXY-RIBOSE-5P_c <=> ACETALD_c + GAP_c\n",
      "\n",
      "2 RXN-6321 5-FORMYL-THF_c + PROTON_c --> 5-10-METHENYL-THF_c + WATER_c\n",
      "2 RXN-6321 5-10-METHENYL-THF_c + WATER_c --> 5-FORMYL-THF_c + PROTON_c\n",
      "\n",
      "3 RXN0-5240 D-ALANINE_c + PYRIDOXAL_PHOSPHATE_c --> PYRIDOXAMINE-5P_c + PYRUVATE_c\n",
      "3 RXN0-5240 D-ALANINE_c + PYRIDOXAL_PHOSPHATE_c <=> PYRIDOXAMINE-5P_c + PYRUVATE_c\n",
      "\n",
      "4 THREONINE-ALDOLASE-RXN THR_c --> ACETALD_c + GLY_c\n",
      "4 THREONINE-ALDOLASE-RXN THR_c <=> ACETALD_c + GLY_c\n",
      "\n",
      "5 GLYOHMETRANS-RXN SER_c + THF_c <=> GLY_c + METHYLENE-THF_c + WATER_c\n",
      "5 GLYOHMETRANS-RXN SER_c + THF_c <=> GLY_c + METHYLENE-THF_c + WATER_c\n",
      "\n",
      "6 RXN0-6563 CPD0-2189_c <=> GLYCOLALDEHYDE_c + GLY_c\n",
      "6 RXN0-6563 CPD0-2189_c <=> GLYCOLALDEHYDE_c + GLY_c\n",
      "\n",
      "7 RXN-13403 2.0 S-ADENOSYLMETHIONINE_c + UROPORPHYRINOGEN-III_c --> 2.0 ADENOSYL-HOMO-CYS_c + DIHYDROSIROHYDROCHLORIN_c + PROTON_c\n",
      "7 RXN-13403 2.0 S-ADENOSYLMETHIONINE_c + UROPORPHYRINOGEN-III_c --> 2.0 ADENOSYL-HOMO-CYS_c + DIHYDROSIROHYDROCHLORIN_c + PROTON_c\n",
      "\n",
      "8 SIROHEME-FERROCHELAT-RXN FE+2_c + SIROHYDROCHLORIN_c --> 3.0 PROTON_c + SIROHEME_c\n",
      "8 SIROHEME-FERROCHELAT-RXN FE+2_c + SIROHYDROCHLORIN_c --> 2.0 PROTON_c + SIROHEME_c\n",
      "\n",
      "9 DIMETHUROPORDEHYDROG-RXN DIHYDROSIROHYDROCHLORIN_c + NAD_c --> NADH_c + PROTON_c + SIROHYDROCHLORIN_c\n",
      "9 DIMETHUROPORDEHYDROG-RXN DIHYDROSIROHYDROCHLORIN_c + NAD_c --> NADH_c + 2.0 PROTON_c + SIROHYDROCHLORIN_c\n",
      "\n",
      "10 RXN0-1461 COPROPORPHYRINOGEN_III_c + OXYGEN-MOLECULE_c + 2.0 PROTON_c --> 2.0 CARBON-DIOXIDE_c + PROTOPORPHYRINOGEN_c + 2.0 WATER_c\n",
      "10 RXN0-1461 COPROPORPHYRINOGEN_III_c + OXYGEN-MOLECULE_c + 2.0 PROTON_c --> 2.0 CARBON-DIOXIDE_c + PROTOPORPHYRINOGEN_c + 2.0 WATER_c\n",
      "\n",
      "11 GLUTATHIONE-REDUCT-NADPH-RXN NADPH_c + OXIDIZED-GLUTATHIONE_c + PROTON_c <=> 2.0 GLUTATHIONE_c + NADP_c\n",
      "11 GLUTATHIONE-REDUCT-NADPH-RXN NADPH_c + OXIDIZED-GLUTATHIONE_c + PROTON_c --> 2.0 GLUTATHIONE_c + NADP_c\n",
      "\n",
      "12 RXN0-2381 INDOLE-3-GLYCEROL-P_c --> GAP_c + INDOLE_c\n",
      "12 RXN0-2381 INDOLE-3-GLYCEROL-P_c <=> GAP_c + INDOLE_c\n",
      "\n",
      "13 RXN0-2382 INDOLE_c + SER_c --> TRP_c + WATER_c\n",
      "13 RXN0-2382 INDOLE_c + SER_c --> TRP_c + WATER_c\n",
      "\n",
      "14 TRYPSYN-RXN INDOLE-3-GLYCEROL-P_c + SER_c --> GAP_c + TRP_c + WATER_c\n",
      "14 TRYPSYN-RXN INDOLE-3-GLYCEROL-P_c + SER_c --> GAP_c + TRP_c + WATER_c\n",
      "\n",
      "15 TYROSINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + TYR_c <=> GLT_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "15 TYROSINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + TYR_c <=> GLT_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "\n",
      "16 BRANCHED-CHAINAMINOTRANSFERVAL-RXN 2-KETOGLUTARATE_c + VAL_c <=> 2-KETO-ISOVALERATE_c + GLT_c\n",
      "16 BRANCHED-CHAINAMINOTRANSFERVAL-RXN 2-KETOGLUTARATE_c + VAL_c <=> 2-KETO-ISOVALERATE_c + GLT_c\n",
      "\n",
      "17 BRANCHED-CHAINAMINOTRANSFERILEU-RXN 2-KETOGLUTARATE_c + ILE_c <=> 2-KETO-3-METHYL-VALERATE_c + GLT_c\n",
      "17 BRANCHED-CHAINAMINOTRANSFERILEU-RXN 2-KETOGLUTARATE_c + ILE_c <=> 2-KETO-3-METHYL-VALERATE_c + GLT_c\n",
      "\n",
      "18 BRANCHED-CHAINAMINOTRANSFERLEU-RXN 2K-4CH3-PENTANOATE_c + GLT_c --> 2-KETOGLUTARATE_c + LEU_c\n",
      "18 BRANCHED-CHAINAMINOTRANSFERLEU-RXN 2-KETOGLUTARATE_c + LEU_c <=> 2K-4CH3-PENTANOATE_c + GLT_c\n",
      "\n",
      "19 SHIKIMATE-5-DEHYDROGENASE-RXN 3-DEHYDRO-SHIKIMATE_c + NADPH_c + PROTON_c <=> NADP_c + SHIKIMATE_c\n",
      "19 SHIKIMATE-5-DEHYDROGENASE-RXN 3-DEHYDRO-SHIKIMATE_c + NADPH_c + PROTON_c --> NADP_c + SHIKIMATE_c\n",
      "\n",
      "20 GLUTSEMIALDEHYDROG-RXN L-GLUTAMATE-5-P_c + NADPH_c + PROTON_c --> L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c + NADP_c + Pi_c\n",
      "20 GLUTSEMIALDEHYDROG-RXN L-GLUTAMATE-5-P_c + NADPH_c + PROTON_c --> L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c + NADP_c + Pi_c\n",
      "\n",
      "21 RIBOFLAVINSYNDEAM-RXN DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + PROTON_c + WATER_c --> AMMONIUM_c + CPD-602_c\n",
      "21 RIBOFLAVINSYNDEAM-RXN DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + PROTON_c + WATER_c --> AMMONIUM_c + CPD-602_c\n",
      "\n",
      "22 RIBOFLAVINSYNREDUC-RXN CPD-602_c + NADPH_c + PROTON_c --> CPD-1086_c + NADP_c\n",
      "22 RIBOFLAVINSYNREDUC-RXN CPD-602_c + NADPH_c + PROTON_c --> CPD-1086_c + NADP_c\n",
      "\n",
      "23 RIBOFLAVINKIN-RXN ATP_c + RIBOFLAVIN_c --> ADP_c + FMN_c + PROTON_c\n",
      "23 RIBOFLAVINKIN-RXN ATP_c + RIBOFLAVIN_c --> ADP_c + FMN_c + PROTON_c\n",
      "\n",
      "24 FADSYN-RXN ATP_c + FMN_c + PROTON_c --> FAD_c + PPI_c\n",
      "24 FADSYN-RXN ATP_c + FMN_c + PROTON_c --> FAD_c + PPI_c\n",
      "\n",
      "25 RIBOPHOSPHAT-RXN CPD-1086_c + WATER_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + Pi_c\n",
      "25 RIBOPHOSPHAT-RXN CPD-1086_c + WATER_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + Pi_c\n",
      "\n",
      "26 RXN0-5187 FMN_c + WATER_c --> Pi_c + RIBOFLAVIN_c\n",
      "26 RXN0-5187 FMN_c + WATER_c --> Pi_c + RIBOFLAVIN_c\n",
      "\n",
      "27 RIBOFLAVIN-SYN-RXN 2.0 DIMETHYL-D-RIBITYL-LUMAZINE_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + RIBOFLAVIN_c\n",
      "27 RIBOFLAVIN-SYN-RXN 2.0 DIMETHYL-D-RIBITYL-LUMAZINE_c + PROTON_c --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE_c + RIBOFLAVIN_c\n",
      "\n",
      "28 GPPSYN-RXN CPD-4211_c + DELTA3-ISOPENTENYL-PP_c --> GERANYL-PP_c + PPI_c\n",
      "28 GPPSYN-RXN CPD-4211_c + DELTA3-ISOPENTENYL-PP_c --> GERANYL-PP_c + PPI_c\n",
      "\n",
      "29 FPPSYN-RXN DELTA3-ISOPENTENYL-PP_c + GERANYL-PP_c --> FARNESYL-PP_c + PPI_c\n",
      "29 FPPSYN-RXN DELTA3-ISOPENTENYL-PP_c + GERANYL-PP_c --> FARNESYL-PP_c + PPI_c\n",
      "\n",
      "30 UROGENIIISYN-RXN HYDROXYMETHYLBILANE_c --> UROPORPHYRINOGEN-III_c + WATER_c\n",
      "30 UROGENIIISYN-RXN HYDROXYMETHYLBILANE_c --> UROPORPHYRINOGEN-III_c + WATER_c\n",
      "\n",
      "31 UROGENDECARBOX-RXN 4.0 PROTON_c + UROPORPHYRINOGEN-III_c --> 4.0 CARBON-DIOXIDE_c + COPROPORPHYRINOGEN_III_c\n",
      "31 UROGENDECARBOX-RXN 4.0 PROTON_c + UROPORPHYRINOGEN-III_c --> 4.0 CARBON-DIOXIDE_c + COPROPORPHYRINOGEN_III_c\n",
      "\n",
      "32 URA-PHOSPH-RXN DEOXYURIDINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + URACIL_c\n",
      "32 URA-PHOSPH-RXN DEOXYURIDINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + URACIL_c\n",
      "\n",
      "33 THYM-PHOSPH-RXN Pi_c + THYMIDINE_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + THYMINE_c\n",
      "33 THYM-PHOSPH-RXN Pi_c + THYMIDINE_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + THYMINE_c\n",
      "\n",
      "34 RXN0-5224 CARBON-DIOXIDE_c + WATER_c <=> HCO3_c + PROTON_c\n",
      "34 RXN0-5224 HCO3_c + PROTON_c <=> CARBON-DIOXIDE_c + WATER_c\n",
      "\n",
      "35 AICARTRANSFORM-RXN 10-FORMYL-THF_c + AICAR_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c + THF_c\n",
      "35 AICARTRANSFORM-RXN 10-FORMYL-THF_c + AICAR_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c + THF_c\n",
      "\n",
      "36 IMPCYCLOHYDROLASE-RXN IMP_c + WATER_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c\n",
      "36 IMPCYCLOHYDROLASE-RXN IMP_c + WATER_c <=> PHOSPHORIBOSYL-FORMAMIDO-CARBOXAMIDE_c\n",
      "\n",
      "37 IMP-DEHYDROG-RXN IMP_c + NAD_c + WATER_c --> NADH_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "37 IMP-DEHYDROG-RXN IMP_c + NAD_c + WATER_c <=> NADH_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "38 RXN-11667 CPD-12173_c <=> CROTONYL-COA_c + WATER_c\n",
      "38 RXN-11667 S-3-HYDROXYBUTANOYL-COA_c <=> CROTONYL-COA_c + WATER_c\n",
      "\n",
      "39 RXN-12490 CPD0-2123_c + NADH_c + PROTON_c <=> CPD0-2244_c + NAD_c\n",
      "39 RXN-12490 CPD0-2244_c + NAD_c --> CPD0-2123_c + NADH_c + PROTON_c\n",
      "\n",
      "40 RXN-11662 ACETOACETYL-COA_c + NADH_c + PROTON_c <=> CPD-12173_c + NAD_c\n",
      "40 RXN-11662 NAD_c + S-3-HYDROXYBUTANOYL-COA_c <=> ACETOACETYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "41 RXN-13616 CPD0-2244_c <=> T2-DECENOYL-COA_c + WATER_c\n",
      "41 RXN-13616 T2-DECENOYL-COA_c + WATER_c --> CPD0-2244_c\n",
      "\n",
      "42 THI-P-SYN-RXN AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c + PROTON_c + THZ-P_c --> PPI_c + THIAMINE-P_c\n",
      "42 THI-P-SYN-RXN AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c + PROTON_c + THZ-P_c --> PPI_c + THIAMINE-P_c\n",
      "\n",
      "43 HYPOXANPRIBOSYLTRAN-RXN HYPOXANTHINE_c + PRPP_c --> IMP_c + PPI_c\n",
      "43 HYPOXANPRIBOSYLTRAN-RXN HYPOXANTHINE_c + PRPP_c --> IMP_c + PPI_c\n",
      "\n",
      "44 GUANPRIBOSYLTRAN-RXN GUANINE_c + PRPP_c --> GMP_c + PPI_c\n",
      "44 GUANPRIBOSYLTRAN-RXN GUANINE_c + PRPP_c --> GMP_c + PPI_c\n",
      "\n",
      "45 XANPRIBOSYLTRAN-RXN PRPP_c + XANTHINE_c --> PPI_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "45 XANPRIBOSYLTRAN-RXN PRPP_c + XANTHINE_c --> PPI_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "46 DADPKIN-RXN ATP_c + DADP_c <=> ADP_c + DATP_c\n",
      "46 DADPKIN-RXN ATP_c + DADP_c --> ADP_c + DATP_c\n",
      "\n",
      "47 ADENYL-KIN-RXN AMP_c + ATP_c <=> 2.0 ADP_c\n",
      "47 ADENYL-KIN-RXN AMP_c + ATP_c <=> 2.0 ADP_c\n",
      "\n",
      "48 DTDPKIN-RXN ATP_c + TDP_c <=> ADP_c + TTP_c\n",
      "48 DTDPKIN-RXN ATP_c + TDP_c --> ADP_c + TTP_c\n",
      "\n",
      "49 CDPKIN-RXN ATP_c + CDP_c <=> ADP_c + CTP_c\n",
      "49 CDPKIN-RXN ATP_c + CDP_c --> ADP_c + CTP_c\n",
      "\n",
      "50 GDPKIN-RXN ATP_c + GDP_c <=> ADP_c + GTP_c\n",
      "50 GDPKIN-RXN ATP_c + GDP_c --> ADP_c + GTP_c\n",
      "\n",
      "51 DGDPKIN-RXN ATP_c + DGDP_c <=> ADP_c + DGTP_c\n",
      "51 DGDPKIN-RXN ATP_c + DGDP_c --> ADP_c + DGTP_c\n",
      "\n",
      "52 UDPKIN-RXN ATP_c + UDP_c <=> ADP_c + UTP_c\n",
      "52 UDPKIN-RXN ATP_c + UDP_c --> ADP_c + UTP_c\n",
      "\n",
      "53 DUDPKIN-RXN ATP_c + DUDP_c <=> ADP_c + DUTP_c\n",
      "53 DUDPKIN-RXN ATP_c + DUDP_c --> ADP_c + DUTP_c\n",
      "\n",
      "54 DCDPKIN-RXN ATP_c + DCDP_c <=> ADP_c + DCTP_c\n",
      "54 DCDPKIN-RXN ATP_c + DCDP_c --> ADP_c + DCTP_c\n",
      "\n",
      "55 RXN-14074 AMP_c + GTP_c <=> ADP_c + GDP_c\n",
      "55 RXN-14074 AMP_c + GTP_c <=> ADP_c + GDP_c\n",
      "\n",
      "56 DIHYDROOROT-RXN DI-H-OROTATE_c + WATER_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c\n",
      "56 DIHYDROOROT-RXN DI-H-OROTATE_c + WATER_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c\n",
      "\n",
      "57 OROTPDECARB-RXN OROTIDINE-5-PHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + UMP_c\n",
      "57 OROTPDECARB-RXN OROTIDINE-5-PHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + UMP_c\n",
      "\n",
      "58 CITSYN-RXN ACETYL-COA_c + OXALACETIC_ACID_c + WATER_c --> CIT_c + CO-A_c + PROTON_c\n",
      "58 CITSYN-RXN ACETYL-COA_c + OXALACETIC_ACID_c + WATER_c --> CIT_c + CO-A_c + PROTON_c\n",
      "\n",
      "59 ACETALD-DEHYDROG-RXN ACETALD_c + CO-A_c + NAD_c <=> ACETYL-COA_c + NADH_c + PROTON_c\n",
      "59 ACETALD-DEHYDROG-RXN ACETALD_c + CO-A_c + NAD_c <=> ACETYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "60 INORGPYROPHOSPHAT-RXN PPI_c + WATER_c --> PROTON_c + 2.0 Pi_c\n",
      "60 INORGPYROPHOSPHAT-RXN PPI_c + WATER_c --> PROTON_c + 2.0 Pi_c\n",
      "\n",
      "61 TRIPHOSPHATASE-RXN P3I_c + WATER_c --> PPI_c + PROTON_c + Pi_c\n",
      "61 TRIPHOSPHATASE-RXN P3I_c + WATER_c --> PPI_c + Pi_c\n",
      "\n",
      "62 XMPXAN-RXN WATER_p + XANTHOSINE-5-PHOSPHATE_p --> Pi_p + XANTHOSINE_p\n",
      "62 XMPXAN-RXN WATER_c + XANTHOSINE-5-PHOSPHATE_c --> Pi_c + XANTHOSINE_c\n",
      "\n",
      "63 AMP-DEPHOSPHORYLATION-RXN AMP_p + WATER_p --> ADENOSINE_p + Pi_p\n",
      "63 AMP-DEPHOSPHORYLATION-RXN AMP_c + WATER_c --> ADENOSINE_c + Pi_c\n",
      "\n",
      "64 RXN-7609 GMP_p + WATER_p --> GUANOSINE_p + Pi_p\n",
      "64 RXN-7609 GMP_c + WATER_c --> GUANOSINE_c + Pi_c\n",
      "\n",
      "65 RXN-14142 DGMP_c + WATER_c --> DEOXYGUANOSINE_c + Pi_c\n",
      "65 RXN-14142 DGMP_c + WATER_c --> DEOXYGUANOSINE_c + Pi_c\n",
      "\n",
      "66 RXN-7607 IMP_c + WATER_c --> INOSINE_c + Pi_c\n",
      "66 RXN-7607 IMP_c + WATER_c --> INOSINE_c + Pi_c\n",
      "\n",
      "67 PEPCARBOXYKIN-RXN ATP_c + OXALACETIC_ACID_c --> ADP_c + CARBON-DIOXIDE_c + PHOSPHO-ENOL-PYRUVATE_c\n",
      "67 PEPCARBOXYKIN-RXN ATP_c + OXALACETIC_ACID_c --> ADP_c + CARBON-DIOXIDE_c + PHOSPHO-ENOL-PYRUVATE_c\n",
      "\n",
      "68 1.1.1.39-RXN MAL_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "68 1.1.1.39-RXN MAL_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "\n",
      "69 ALARACECAT-RXN L-ALPHA-ALANINE_c <=> D-ALANINE_c\n",
      "69 ALARACECAT-RXN L-ALPHA-ALANINE_c <=> D-ALANINE_c\n",
      "\n",
      "70 ALANINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + L-ALPHA-ALANINE_c <=> GLT_c + PYRUVATE_c\n",
      "70 ALANINE-AMINOTRANSFERASE-RXN 2-KETOGLUTARATE_c + L-ALPHA-ALANINE_c <=> GLT_c + PYRUVATE_c\n",
      "\n",
      "71 XYLULOKIN-RXN ATP_c + D-XYLULOSE_c --> ADP_c + PROTON_c + XYLULOSE-5-PHOSPHATE_c\n",
      "71 XYLULOKIN-RXN ATP_c + D-XYLULOSE_c --> ADP_c + PROTON_c + XYLULOSE-5-PHOSPHATE_c\n",
      "\n",
      "72 GLYCEROL-KIN-RXN ATP_c + GLYCEROL_c --> ADP_c + GLYCEROL-3P_c + PROTON_c\n",
      "72 GLYCEROL-KIN-RXN ATP_c + GLYCEROL_c <=> ADP_c + GLYCEROL-3P_c + PROTON_c\n",
      "\n",
      "73 D-PPENTOMUT-RXN DEOXY-D-RIBOSE-1-PHOSPHATE_c <=> DEOXY-RIBOSE-5P_c\n",
      "73 D-PPENTOMUT-RXN DEOXY-D-RIBOSE-1-PHOSPHATE_c <=> DEOXY-RIBOSE-5P_c\n",
      "\n",
      "74 PEPDEPHOS-RXN ADP_c + PHOSPHO-ENOL-PYRUVATE_c + PROTON_c --> ATP_c + PYRUVATE_c\n",
      "74 PEPDEPHOS-RXN ADP_c + PHOSPHO-ENOL-PYRUVATE_c + PROTON_c --> ATP_c + PYRUVATE_c\n",
      "\n",
      "75 DARAB5PISOM-RXN RIBULOSE-5P_c <=> ARABINOSE-5P_c\n",
      "75 DARAB5PISOM-RXN ARABINOSE-5P_c <=> RIBULOSE-5P_c\n",
      "\n",
      "76 RXN-8991 2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE_c <=> CPD-9451_c + WATER_c\n",
      "76 RXN-8991 2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE_c <=> CPD-9451_c + WATER_c\n",
      "\n",
      "77 3-ISOPROPYLMALISOM-RXN CPD-9451_c + WATER_c <=> 3-CARBOXY-3-HYDROXY-ISOCAPROATE_c\n",
      "77 3-ISOPROPYLMALISOM-RXN CPD-9451_c + WATER_c <=> 3-CARBOXY-3-HYDROXY-ISOCAPROATE_c\n",
      "\n",
      "78 ACETOLACTSYN-RXN PROTON_c + 2.0 PYRUVATE_c --> 2-ACETO-LACTATE_c + CARBON-DIOXIDE_c\n",
      "78 ACETOLACTSYN-RXN PROTON_c + 2.0 PYRUVATE_c --> 2-ACETO-LACTATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "79 ACETOOHBUTSYN-RXN 2-OXOBUTANOATE_c + PROTON_c + PYRUVATE_c --> 2-ACETO-2-HYDROXY-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "79 ACETOOHBUTSYN-RXN 2-OXOBUTANOATE_c + PROTON_c + PYRUVATE_c --> 2-ACETO-2-HYDROXY-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "80 2.5.1.19-RXN PHOSPHO-ENOL-PYRUVATE_c + SHIKIMATE-5P_c <=> 3-ENOLPYRUVYL-SHIKIMATE-5P_c + Pi_c\n",
      "80 2.5.1.19-RXN PHOSPHO-ENOL-PYRUVATE_c + SHIKIMATE-5P_c <=> 3-ENOLPYRUVYL-SHIKIMATE-5P_c + Pi_c\n",
      "\n",
      "81 ANTHRANSYN-RXN CHORISMATE_c + GLN_c --> ANTHRANILATE_c + GLT_c + PROTON_c + PYRUVATE_c\n",
      "81 ANTHRANSYN-RXN CHORISMATE_c + GLN_c <=> ANTHRANILATE_c + GLT_c + PROTON_c + PYRUVATE_c\n",
      "\n",
      "82 PRTRANS-RXN ANTHRANILATE_c + PRPP_c --> N-5-PHOSPHORIBOSYL-ANTHRANILATE_c + PPI_c\n",
      "82 PRTRANS-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c + PPI_c <=> ANTHRANILATE_c + PRPP_c\n",
      "\n",
      "83 CHORISMATEMUT-RXN CHORISMATE_c --> PREPHENATE_c\n",
      "83 CHORISMATEMUT-RXN CHORISMATE_c <=> PREPHENATE_c\n",
      "\n",
      "84 PREPHENATEDEHYDRAT-RXN PREPHENATE_c + PROTON_c --> CARBON-DIOXIDE_c + PHENYL-PYRUVATE_c + WATER_c\n",
      "84 PREPHENATEDEHYDRAT-RXN PREPHENATE_c + PROTON_c <=> CARBON-DIOXIDE_c + PHENYL-PYRUVATE_c + WATER_c\n",
      "\n",
      "85 CHORISMATE-SYNTHASE-RXN 3-ENOLPYRUVYL-SHIKIMATE-5P_c --> CHORISMATE_c + Pi_c\n",
      "85 CHORISMATE-SYNTHASE-RXN 3-ENOLPYRUVYL-SHIKIMATE-5P_c --> CHORISMATE_c + Pi_c\n",
      "\n",
      "86 IGPSYN-RXN CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c + PROTON_c --> CARBON-DIOXIDE_c + INDOLE-3-GLYCEROL-P_c + WATER_c\n",
      "86 IGPSYN-RXN CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c + PROTON_c --> CARBON-DIOXIDE_c + INDOLE-3-GLYCEROL-P_c + WATER_c\n",
      "\n",
      "87 PRAISOM-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c --> CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c\n",
      "87 PRAISOM-RXN N-5-PHOSPHORIBOSYL-ANTHRANILATE_c --> CARBOXYPHENYLAMINO-DEOXYRIBULOSE-P_c\n",
      "\n",
      "88 2PGADEHYDRAT-RXN 2-PG_c <=> PHOSPHO-ENOL-PYRUVATE_c + WATER_c\n",
      "88 2PGADEHYDRAT-RXN 2-PG_c <=> PHOSPHO-ENOL-PYRUVATE_c + WATER_c\n",
      "\n",
      "89 METHGLYSYN-RXN DIHYDROXY-ACETONE-PHOSPHATE_c --> METHYL-GLYOXAL_c + Pi_c\n",
      "89 METHGLYSYN-RXN DIHYDROXY-ACETONE-PHOSPHATE_c --> METHYL-GLYOXAL_c + Pi_c\n",
      "\n",
      "90 PEPSYNTH-RXN ATP_c + PYRUVATE_c + WATER_c --> AMP_c + PHOSPHO-ENOL-PYRUVATE_c + 2.0 PROTON_c + Pi_c\n",
      "90 PEPSYNTH-RXN ATP_c + PYRUVATE_c + WATER_c <=> AMP_c + PHOSPHO-ENOL-PYRUVATE_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "91 PGLUCISOM-RXN D-glucopyranose-6-phosphate_c <=> FRUCTOSE-6P_c\n",
      "91 PGLUCISOM-RXN D-glucopyranose-6-phosphate_c <=> FRUCTOSE-6P_c\n",
      "\n",
      "92 PHOSGLYPHOS-RXN ATP_c + G3P_c <=> ADP_c + DPG_c\n",
      "92 PHOSGLYPHOS-RXN ATP_c + G3P_c <=> ADP_c + DPG_c\n",
      "\n",
      "93 OXALODECARB-RXN OXALACETIC_ACID_c + PROTON_c --> CARBON-DIOXIDE_c + PYRUVATE_c\n",
      "93 OXALODECARB-RXN OXALACETIC_ACID_c + PROTON_c --> CARBON-DIOXIDE_c + PYRUVATE_c\n",
      "\n",
      "94 KDPGALDOL-RXN 2-KETO-3-DEOXY-6-P-GLUCONATE_c --> GAP_c + PYRUVATE_c\n",
      "94 KDPGALDOL-RXN 2-KETO-3-DEOXY-6-P-GLUCONATE_c --> GAP_c + PYRUVATE_c\n",
      "\n",
      "95 RIBULP3EPIM-RXN RIBULOSE-5P_c <=> XYLULOSE-5-PHOSPHATE_c\n",
      "95 RIBULP3EPIM-RXN RIBULOSE-5P_c <=> XYLULOSE-5-PHOSPHATE_c\n",
      "\n",
      "96 TRANSALDOL-RXN D-SEDOHEPTULOSE-7-P_c + GAP_c <=> ERYTHROSE-4P_c + FRUCTOSE-6P_c\n",
      "96 TRANSALDOL-RXN D-SEDOHEPTULOSE-7-P_c + GAP_c <=> ERYTHROSE-4P_c + FRUCTOSE-6P_c\n",
      "\n",
      "97 2TRANSKETO-RXN ERYTHROSE-4P_c + XYLULOSE-5-PHOSPHATE_c <=> FRUCTOSE-6P_c + GAP_c\n",
      "97 2TRANSKETO-RXN ERYTHROSE-4P_c + XYLULOSE-5-PHOSPHATE_c <=> FRUCTOSE-6P_c + GAP_c\n",
      "\n",
      "98 ALCOHOL-DEHYDROG-RXN ETOH_c + NAD_c <=> ACETALD_c + NADH_c + PROTON_c\n",
      "98 ALCOHOL-DEHYDROG-RXN ETOH_c + NAD_c <=> ACETALD_c + NADH_c + PROTON_c\n",
      "\n",
      "99 2.7.7.1-RXN ATP_c + NICOTINAMIDE_NUCLEOTIDE_c + PROTON_c --> NAD_c + PPI_c\n",
      "99 2.7.7.1-RXN ATP_c + NICOTINAMIDE_NUCLEOTIDE_c + PROTON_c --> NAD_c + PPI_c\n",
      "\n",
      "100 NICONUCADENYLYLTRAN-RXN ATP_c + NICOTINATE_NUCLEOTIDE_c + PROTON_c <=> DEAMIDO-NAD_c + PPI_c\n",
      "100 NICONUCADENYLYLTRAN-RXN ATP_c + NICOTINATE_NUCLEOTIDE_c + PROTON_c --> DEAMIDO-NAD_c + PPI_c\n",
      "\n",
      "101 PYRIMSYN3-RXN AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + ATP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c\n",
      "101 PYRIMSYN3-RXN AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + ATP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP_c\n",
      "\n",
      "102 OHMETPYRKIN-RXN ATP_c + HMP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + PROTON_c\n",
      "102 OHMETPYRKIN-RXN ATP_c + HMP_c --> ADP_c + AMINO-HYDROXYMETHYL-METHYL-PYR-P_c + PROTON_c\n",
      "\n",
      "103 DEOXYINOPHOSPHOR-RXN DEOXYINOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + HYPOXANTHINE_c\n",
      "103 DEOXYINOPHOSPHOR-RXN DEOXYINOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + HYPOXANTHINE_c\n",
      "\n",
      "104 INOPHOSPHOR-RXN INOSINE_c + Pi_c <=> HYPOXANTHINE_c + RIBOSE-1P_c\n",
      "104 INOPHOSPHOR-RXN INOSINE_c + Pi_c <=> HYPOXANTHINE_c + RIBOSE-1P_c\n",
      "\n",
      "105 ADENPHOSPHOR-RXN ADENOSINE_c + Pi_c <=> ADENINE_c + RIBOSE-1P_c\n",
      "105 ADENPHOSPHOR-RXN ADENOSINE_c + Pi_c <=> ADENINE_c + RIBOSE-1P_c\n",
      "\n",
      "106 RXN0-5199 GUANOSINE_c + Pi_c <=> GUANINE_c + RIBOSE-1P_c\n",
      "106 RXN0-5199 GUANOSINE_c + Pi_c <=> GUANINE_c + RIBOSE-1P_c\n",
      "\n",
      "107 DEOXYGUANPHOSPHOR-RXN DEOXYGUANOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + GUANINE_c\n",
      "107 DEOXYGUANPHOSPHOR-RXN DEOXYGUANOSINE_c + Pi_c <=> DEOXY-D-RIBOSE-1-PHOSPHATE_c + GUANINE_c\n",
      "\n",
      "108 DEOXYADENPHOSPHOR-RXN DEOXYADENOSINE_c + Pi_c <=> ADENINE_c + DEOXY-D-RIBOSE-1-PHOSPHATE_c\n",
      "108 DEOXYADENPHOSPHOR-RXN DEOXYADENOSINE_c + Pi_c <=> ADENINE_c + DEOXY-D-RIBOSE-1-PHOSPHATE_c\n",
      "\n",
      "109 OROPRIBTRANS-RXN OROTIDINE-5-PHOSPHATE_c + PPI_c <=> OROTATE_c + PRPP_c\n",
      "109 OROPRIBTRANS-RXN OROTIDINE-5-PHOSPHATE_c + PPI_c <=> OROTATE_c + PRPP_c\n",
      "\n",
      "110 DIHYDROXYISOVALDEHYDRAT-RXN CPD-13357_c --> 2-KETO-ISOVALERATE_c + WATER_c\n",
      "110 DIHYDROXYISOVALDEHYDRAT-RXN CPD-13357_c --> 2-KETO-ISOVALERATE_c + WATER_c\n",
      "\n",
      "111 DIHYDROXYMETVALDEHYDRAT-RXN 1-KETO-2-METHYLVALERATE_c --> 2-KETO-3-METHYL-VALERATE_c + WATER_c\n",
      "111 DIHYDROXYMETVALDEHYDRAT-RXN 1-KETO-2-METHYLVALERATE_c --> 2-KETO-3-METHYL-VALERATE_c + WATER_c\n",
      "\n",
      "112 PHOSACETYLTRANS-RXN ACETYL-COA_c + Pi_c <=> ACETYL-P_c + CO-A_c\n",
      "112 PHOSACETYLTRANS-RXN ACETYL-COA_c + Pi_c <=> ACETYL-P_c + CO-A_c\n",
      "\n",
      "113 PTAALT-RXN PROPIONYL-COA_c + Pi_c --> CO-A_c + PROPIONYL-P_c\n",
      "113 PTAALT-RXN PROPIONYL-COA_c + Pi_c <=> CO-A_c + PROPIONYL-P_c\n",
      "\n",
      "114 ACETATEKIN-RXN ACET_c + ATP_c <=> ACETYL-P_c + ADP_c\n",
      "114 ACETATEKIN-RXN ACET_c + ATP_c <=> ACETYL-P_c + ADP_c\n",
      "\n",
      "115 KDOTRANS-RXN CMP-KDO_c + LIPID-IV-A_c --> CMP_c + KDO-LIPID-IVA_c + PROTON_c\n",
      "115 KDOTRANS-RXN CMP-KDO_c + LIPID-IV-A_c --> CMP_c + KDO-LIPID-IVA_c + PROTON_c\n",
      "\n",
      "116 KDOTRANS2-RXN CMP-KDO_c + KDO-LIPID-IVA_c --> CMP_c + KDO2-LIPID-IVA_c + PROTON_c\n",
      "116 KDOTRANS2-RXN CMP-KDO_c + KDO-LIPID-IVA_c --> CMP_c + KDO2-LIPID-IVA_c + PROTON_c\n",
      "\n",
      "117 LIPIDXSYNTHESIS-RXN OH-MYRISTOYL_c + WATER_c --> BISOHMYR-GLUCOSAMINYL-1P_c + 2.0 PROTON_c + UMP_c\n",
      "117 LIPIDXSYNTHESIS-RXN OH-MYRISTOYL_c + WATER_c --> BISOHMYR-GLUCOSAMINYL-1P_c + 2.0 PROTON_c + UMP_c\n",
      "\n",
      "118 LIPIDADISACCHARIDESYNTH-RXN BISOHMYR-GLUCOSAMINYL-1P_c + OH-MYRISTOYL_c --> BISOHMYR-GLC_c + PROTON_c + UDP_c\n",
      "118 LIPIDADISACCHARIDESYNTH-RXN BISOHMYR-GLUCOSAMINYL-1P_c + OH-MYRISTOYL_c --> BISOHMYR-GLC_c + PROTON_c + UDP_c\n",
      "\n",
      "119 TETRAACYLDISACC4KIN-RXN ATP_c + BISOHMYR-GLC_c --> ADP_c + LIPID-IV-A_c + PROTON_c\n",
      "119 TETRAACYLDISACC4KIN-RXN ATP_c + BISOHMYR-GLC_c --> ADP_c + LIPID-IV-A_c + PROTON_c\n",
      "\n",
      "120 RXN0-5107 TTP_c + WATER_c --> PPI_c + PROTON_c + TMP_c\n",
      "120 RXN0-5107 TTP_c + WATER_c --> PPI_c + PROTON_c + TMP_c\n",
      "\n",
      "121 RXN-14139 UTP_c + WATER_c --> PPI_c + PROTON_c + UMP_c\n",
      "121 RXN-14139 UTP_c + WATER_c --> PPI_c + PROTON_c + UMP_c\n",
      "\n",
      "122 ATP-PYROPHOSPHATASE-RXN ATP_c + WATER_c --> AMP_c + PPI_c + PROTON_c\n",
      "122 ATP-PYROPHOSPHATASE-RXN ATP_c + WATER_c --> AMP_c + PPI_c + PROTON_c\n",
      "\n",
      "123 1.1.1.83-RXN CPD-660_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "123 1.1.1.83-RXN CPD-660_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + PYRUVATE_c\n",
      "\n",
      "124 RXN0-743 PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c <=> CPD0-181_c\n",
      "124 RXN0-743 CPD0-181_c --> PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c\n",
      "\n",
      "125 AMPSYN-RXN ADENYLOSUCC_c <=> AMP_c + FUM_c\n",
      "125 AMPSYN-RXN ADENYLOSUCC_c <=> AMP_c + FUM_c\n",
      "\n",
      "126 AICARSYN-RXN P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c <=> AICAR_c + FUM_c\n",
      "126 AICARSYN-RXN P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c <=> AICAR_c + FUM_c\n",
      "\n",
      "127 1.8.4.14-RXN CPD-8989_c + Red-Thioredoxin_c --> MET_c + Ox-Thioredoxin_c + WATER_c\n",
      "127 1.8.4.14-RXN CPD-8990_c + Red-Thioredoxin_c --> MET_c + Ox-Thioredoxin_c + WATER_c\n",
      "\n",
      "128 THIOREDOXIN-REDUCT-NADPH-RXN NADPH_c + Ox-Thioredoxin_c + PROTON_c --> NADP_c + Red-Thioredoxin_c\n",
      "128 THIOREDOXIN-REDUCT-NADPH-RXN NADPH_c + Ox-Thioredoxin_c + PROTON_c --> NADP_c + Red-Thioredoxin_c\n",
      "\n",
      "129 1.8.4.8-RXN PAPS_c + Red-Thioredoxin_c --> 3-5-ADP_c + Ox-Thioredoxin_c + 2.0 PROTON_c + SO3_c\n",
      "129 1.8.4.8-RXN 3-5-ADP_c + Ox-Thioredoxin_c + 2.0 PROTON_c + SO3_c <=> PAPS_c + Red-Thioredoxin_c\n",
      "\n",
      "130 CDPREDUCT-RXN CDP_c + Red-Thioredoxin_c --> DCDP_c + Ox-Thioredoxin_c + WATER_c\n",
      "130 CDPREDUCT-RXN CDP_c + Red-Thioredoxin_c --> DCDP_c + Ox-Thioredoxin_c + WATER_c\n",
      "\n",
      "131 THYMIDYLATESYN-RXN DUMP_c + METHYLENE-THF_c --> DIHYDROFOLATE_c + TMP_c\n",
      "131 THYMIDYLATESYN-RXN DUMP_c + METHYLENE-THF_c --> DIHYDROFOLATE_c + TMP_c\n",
      "\n",
      "132 P-PANTOCYSLIG-RXN 4-P-PANTOTHENATE_c + CTP_c + CYS_c --> CMP_c + PPI_c + PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c\n",
      "132 P-PANTOCYSLIG-RXN 4-P-PANTOTHENATE_c + CTP_c + CYS_c --> CMP_c + PPI_c + PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c\n",
      "\n",
      "133 P-PANTOCYSDECARB-RXN PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c --> CARBON-DIOXIDE_c + PANTETHEINE-P_c\n",
      "133 P-PANTOCYSDECARB-RXN PROTON_c + R-4-PHOSPHOPANTOTHENOYL-L-CYSTEINE_c --> CARBON-DIOXIDE_c + PANTETHEINE-P_c\n",
      "\n",
      "134 FORMYLTHFDEFORMYL-RXN 10-FORMYL-THF_c + WATER_c --> FORMATE_c + PROTON_c + THF_c\n",
      "134 FORMYLTHFDEFORMYL-RXN 10-FORMYL-THF_c + WATER_c --> FORMATE_c + PROTON_c + THF_c\n",
      "\n",
      "135 PGLUCONDEHYDRAT-RXN CPD-2961_c --> 2-KETO-3-DEOXY-6-P-GLUCONATE_c + WATER_c\n",
      "135 PGLUCONDEHYDRAT-RXN CPD-2961_c --> 2-KETO-3-DEOXY-6-P-GLUCONATE_c + WATER_c\n",
      "\n",
      "136 MALATE-DEH-RXN MAL_c + NAD_c <=> NADH_c + OXALACETIC_ACID_c + PROTON_c\n",
      "136 MALATE-DEH-RXN MAL_c + NAD_c <=> NADH_c + OXALACETIC_ACID_c + PROTON_c\n",
      "\n",
      "137 FUMHYDR-RXN FUM_c + WATER_c <=> MAL_c\n",
      "137 FUMHYDR-RXN MAL_c <=> FUM_c + WATER_c\n",
      "\n",
      "138 D--TARTRATE-DEHYDRATASE-RXN D-TARTRATE_c --> OXALACETIC_ACID_c + WATER_c\n",
      "138 D--TARTRATE-DEHYDRATASE-RXN D-TARTRATE_c <=> OXALACETIC_ACID_c + WATER_c\n",
      "\n",
      "139 PNPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXINE-5P_c --> HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "139 PNPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXINE-5P_c --> HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "\n",
      "140 PMPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXAMINE-5P_c + WATER_c --> AMMONIUM_c + HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "140 PMPOXI-RXN OXYGEN-MOLECULE_c + PYRIDOXAMINE-5P_c + WATER_c --> AMMONIUM_c + HYDROGEN-PEROXIDE_c + PYRIDOXAL_PHOSPHATE_c\n",
      "\n",
      "141 PSERTRANSAMPYR-RXN 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + GLT_c <=> 2-KETOGLUTARATE_c + 4-PHOSPHONOOXY-THREONINE_c\n",
      "141 PSERTRANSAMPYR-RXN 2-KETOGLUTARATE_c + 4-PHOSPHONOOXY-THREONINE_c <=> 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + GLT_c\n",
      "\n",
      "142 SUCCINYLDIAMINOPIMTRANS-RXN 2-KETOGLUTARATE_c + N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c <=> GLT_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c\n",
      "142 SUCCINYLDIAMINOPIMTRANS-RXN GLT_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c --> 2-KETOGLUTARATE_c + N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c\n",
      "\n",
      "143 PSERTRANSAM-RXN 3-P-HYDROXYPYRUVATE_c + GLT_c --> 2-KETOGLUTARATE_c + 3-P-SERINE_c\n",
      "143 PSERTRANSAM-RXN 2-KETOGLUTARATE_c + 3-P-SERINE_c <=> 3-P-HYDROXYPYRUVATE_c + GLT_c\n",
      "\n",
      "144 HEMEOSYN-RXN FARNESYL-PP_c + PROTOHEME_c + WATER_c --> CPD-17063_c + PPI_c\n",
      "144 HEMEOSYN-RXN FARNESYL-PP_c + PROTOHEME_c + WATER_c --> CPD-17063_c + PPI_c\n",
      "\n",
      "145 AGMATIN-RXN AGMATHINE_c + WATER_c --> PUTRESCINE_c + UREA_c\n",
      "145 AGMATIN-RXN AGMATHINE_c + WATER_c --> PUTRESCINE_c + UREA_c\n",
      "\n",
      "146 ARGDECARBOX-RXN ARG_p + PROTON_p --> AGMATHINE_p + CARBON-DIOXIDE_p\n",
      "146 ARGDECARBOX-RXN ARG_c + PROTON_c <=> AGMATHINE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "147 ARGSUCCINSYN-RXN ATP_c + L-ASPARTATE_c + L-CITRULLINE_c --> AMP_c + L-ARGININO-SUCCINATE_c + PPI_c + PROTON_c\n",
      "147 ARGSUCCINSYN-RXN ATP_c + L-ASPARTATE_c + L-CITRULLINE_c --> AMP_c + L-ARGININO-SUCCINATE_c + PPI_c + PROTON_c\n",
      "\n",
      "148 SPERMIDINESYN-RXN PUTRESCINE_c + S-ADENOSYLMETHIONINAMINE_c --> 5-METHYLTHIOADENOSINE_c + PROTON_c + SPERMIDINE_c\n",
      "148 SPERMIDINESYN-RXN PUTRESCINE_c + S-ADENOSYLMETHIONINAMINE_c <=> 5-METHYLTHIOADENOSINE_c + PROTON_c + SPERMIDINE_c\n",
      "\n",
      "149 GSAAMINOTRANS-RXN GLUTAMATE-1-SEMIALDEHYDE_c <=> 5-AMINO-LEVULINATE_c\n",
      "149 GSAAMINOTRANS-RXN GLUTAMATE-1-SEMIALDEHYDE_c --> 5-AMINO-LEVULINATE_c\n",
      "\n",
      "150 CPM-KDOSYNTH-RXN CTP_c + KDO_c --> CMP-KDO_c + PPI_c\n",
      "150 CPM-KDOSYNTH-RXN CTP_c + KDO_c --> CMP-KDO_c + PPI_c\n",
      "\n",
      "151 5.4.2.10-RXN GLUCOSAMINE-1P_c <=> D-GLUCOSAMINE-6-P_c\n",
      "151 5.4.2.10-RXN D-GLUCOSAMINE-6-P_c <=> GLUCOSAMINE-1P_c\n",
      "\n",
      "152 UDPNACETYLGLUCOSAMENOLPYRTRANS-RXN PHOSPHO-ENOL-PYRUVATE_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> Pi_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c\n",
      "152 UDPNACETYLGLUCOSAMENOLPYRTRANS-RXN PHOSPHO-ENOL-PYRUVATE_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> Pi_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c\n",
      "\n",
      "153 GLUTRACE-RXN D-GLT_c <=> GLT_c\n",
      "153 GLUTRACE-RXN GLT_c <=> D-GLT_c\n",
      "\n",
      "154 UDPGLCNACEPIM-RXN UDP-N-ACETYL-D-GLUCOSAMINE_c <=> UDP-MANNAC_c\n",
      "154 UDPGLCNACEPIM-RXN UDP-N-ACETYL-D-GLUCOSAMINE_c <=> UDP-MANNAC_c\n",
      "\n",
      "155 MALONYL-ACPDECARBOX-RXN MALONYL-ACP_c + PROTON_c --> ACETYL-ACP_c + CARBON-DIOXIDE_c\n",
      "155 MALONYL-ACPDECARBOX-RXN MALONYL-ACP_c + PROTON_c --> ACETYL-ACP_c + CARBON-DIOXIDE_c\n",
      "\n",
      "156 3-OXOACYL-ACP-SYNTH-BASE-RXN ACETYL-ACP_c + MALONYL-ACP_c + PROTON_c --> ACP_c + Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c\n",
      "156 3-OXOACYL-ACP-SYNTH-BASE-RXN ACETYL-ACP_c + MALONYL-ACP_c + PROTON_c --> ACP_c + Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c\n",
      "\n",
      "157 2-OCTAPRENYL-6-OHPHENOL-METHY-RXN 2-OCTAPRENYL-6-HYDROXYPHENOL_c + S-ADENOSYLMETHIONINE_c --> 2-OCTAPRENYL-6-METHOXYPHENOL_c + ADENOSYL-HOMO-CYS_c + PROTON_c\n",
      "157 2-OCTAPRENYL-6-OHPHENOL-METHY-RXN 2-OCTAPRENYL-6-HYDROXYPHENOL_c + S-ADENOSYLMETHIONINE_c --> 2-OCTAPRENYL-6-METHOXYPHENOL_c + ADENOSYL-HOMO-CYS_c + PROTON_c\n",
      "\n",
      "158 DHHB-METHYLTRANSFER-RXN OCTAPRENYL-METHYL-OH-METHOXY-BENZQ_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + CPD-9956_c + PROTON_c\n",
      "158 DHHB-METHYLTRANSFER-RXN OCTAPRENYL-METHYL-OH-METHOXY-BENZQ_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + CPD-9956_c + PROTON_c\n",
      "\n",
      "159 H2PTERIDINEPYROPHOSPHOKIN-RXN AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE_c + ATP_c --> AMP_c + DIHYDROPTERIN-CH2OH-PP_c + PROTON_c\n",
      "159 H2PTERIDINEPYROPHOSPHOKIN-RXN AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE_c + ATP_c --> AMP_c + DIHYDROPTERIN-CH2OH-PP_c + PROTON_c\n",
      "\n",
      "160 PANTEPADENYLYLTRAN-RXN ATP_c + PANTETHEINE-P_c + PROTON_c --> DEPHOSPHO-COA_c + PPI_c\n",
      "160 PANTEPADENYLYLTRAN-RXN ATP_c + PANTETHEINE-P_c + PROTON_c --> DEPHOSPHO-COA_c + PPI_c\n",
      "\n",
      "161 DXPREDISOM-RXN DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c --> 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + NADP_c\n",
      "161 DXPREDISOM-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + NADP_c <=> DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c\n",
      "\n",
      "162 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PROTON_c --> 2-OCTAPRENYLPHENOL_c + CARBON-DIOXIDE_c\n",
      "162 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PROTON_c --> 2-OCTAPRENYLPHENOL_c + CARBON-DIOXIDE_c\n",
      "\n",
      "163 GABATRANSAM-RXN 2-KETOGLUTARATE_c + 4-AMINO-BUTYRATE_c --> GLT_c + SUCC-S-ALD_c\n",
      "163 GABATRANSAM-RXN 2-KETOGLUTARATE_c + 4-AMINO-BUTYRATE_c <=> GLT_c + SUCC-S-ALD_c\n",
      "\n",
      "164 ACETYLORNTRANSAM-RXN 2-KETOGLUTARATE_c + N-ALPHA-ACETYLORNITHINE_c <=> CPD-469_c + GLT_c\n",
      "164 ACETYLORNTRANSAM-RXN 2-KETOGLUTARATE_c + N-ALPHA-ACETYLORNITHINE_c <=> CPD-469_c + GLT_c\n",
      "\n",
      "165 SUCCSEMIALDDEHYDROG-RXN NADP_c + SUCC-S-ALD_c + WATER_c --> NADPH_c + 2.0 PROTON_c + SUC_c\n",
      "165 SUCCSEMIALDDEHYDROG-RXN NADP_c + SUCC-S-ALD_c + WATER_c --> NADPH_c + 2.0 PROTON_c + SUC_c\n",
      "\n",
      "166 SUCCINATE-SEMIALDEHYDE-DEHYDROGENASE-RXN NAD_c + SUCC-S-ALD_c + WATER_c --> NADH_c + 2.0 PROTON_c + SUC_c\n",
      "166 SUCCINATE-SEMIALDEHYDE-DEHYDROGENASE-RXN NAD_c + SUCC-S-ALD_c + WATER_c --> NADH_c + 2.0 PROTON_c + SUC_c\n",
      "\n",
      "167 4.3.1.17-RXN SER_c --> AMMONIUM_c + PYRUVATE_c\n",
      "167 4.3.1.17-RXN SER_c --> AMMONIUM_c + PYRUVATE_c\n",
      "\n",
      "168 ACETOOHBUTREDUCTOISOM-RXN 2-ACETO-2-HYDROXY-BUTYRATE_c + NADPH_c + PROTON_c <=> 1-KETO-2-METHYLVALERATE_c + NADP_c\n",
      "168 ACETOOHBUTREDUCTOISOM-RXN 2-ACETO-2-HYDROXY-BUTYRATE_c + NADPH_c + PROTON_c --> 1-KETO-2-METHYLVALERATE_c + NADP_c\n",
      "\n",
      "169 ACETOLACTREDUCTOISOM-RXN CPD-13357_c + NADP_c <=> 2-ACETO-LACTATE_c + NADPH_c + PROTON_c\n",
      "169 ACETOLACTREDUCTOISOM-RXN 2-ACETO-LACTATE_c + NADPH_c + PROTON_c --> CPD-13357_c + NADP_c\n",
      "\n",
      "170 NAD-KIN-RXN ATP_c + NAD_c --> ADP_c + NADP_c + PROTON_c\n",
      "170 NAD-KIN-RXN ATP_c + NAD_c --> ADP_c + NADP_c + PROTON_c\n",
      "\n",
      "171 GLYOXI-RXN GLUTATHIONE_c + METHYL-GLYOXAL_c --> S-LACTOYL-GLUTATHIONE_c\n",
      "171 GLYOXI-RXN S-LACTOYL-GLUTATHIONE_c <=> GLUTATHIONE_c + METHYL-GLYOXAL_c\n",
      "\n",
      "172 3.1.3.74-RXN PYRIDOXAL_PHOSPHATE_c + WATER_c --> PYRIDOXAL_c + Pi_c\n",
      "172 3.1.3.74-RXN PYRIDOXAL_PHOSPHATE_c + WATER_c --> PYRIDOXAL_c + Pi_c\n",
      "\n",
      "173 DXS-RXN GAP_c + PROTON_c + PYRUVATE_c --> CARBON-DIOXIDE_c + DEOXYXYLULOSE-5P_c\n",
      "173 DXS-RXN GAP_c + PROTON_c + PYRUVATE_c --> CARBON-DIOXIDE_c + DEOXYXYLULOSE-5P_c\n",
      "\n",
      "174 2.7.1.148-RXN 4-CYTIDINE-5-DIPHOSPHO-2-C_c + ATP_c --> 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c + ADP_c + PROTON_c\n",
      "174 2.7.1.148-RXN 4-CYTIDINE-5-DIPHOSPHO-2-C_c + ATP_c --> 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c + ADP_c + PROTON_c\n",
      "\n",
      "175 RXN0-302 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c --> 2C-METH-D-ERYTHRITOL-CYCLODIPHOSPHATE_c + CMP_c\n",
      "175 RXN0-302 2-PHOSPHO-4-CYTIDINE-5-DIPHOSPHO-2-C-MET_c --> 2C-METH-D-ERYTHRITOL-CYCLODIPHOSPHATE_c + CMP_c\n",
      "\n",
      "176 PHOSMANMUT-RXN MANNOSE-1P_c <=> CPD-15979_c\n",
      "176 PHOSMANMUT-RXN MANNOSE-1P_c <=> CPD-15979_c\n",
      "\n",
      "177 PYRUVDEH-RXN CO-A_c + NAD_c + PYRUVATE_c --> ACETYL-COA_c + CARBON-DIOXIDE_c + NADH_c\n",
      "177 PYRUVDEH-RXN CO-A_c + NAD_c + PYRUVATE_c --> ACETYL-COA_c + CARBON-DIOXIDE_c + NADH_c\n",
      "\n",
      "178 2OXOGLUTARATEDEH-RXN 2-KETOGLUTARATE_c + CO-A_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + SUC-COA_c\n",
      "178 2OXOGLUTARATEDEH-RXN 2-KETOGLUTARATE_c + CO-A_c + NAD_c --> CARBON-DIOXIDE_c + NADH_c + SUC-COA_c\n",
      "\n",
      "179 MMUM-RXN CPD-397_c + HOMO-CYS_c --> 2.0 MET_c + PROTON_c\n",
      "179 MMUM-RXN CPD-397_c + HOMO-CYS_c --> 2.0 MET_c + PROTON_c\n",
      "\n",
      "180 TRE6PHYDRO-RXN TREHALOSE-6P_c + WATER_c --> D-glucopyranose-6-phosphate_c + Glucopyranose_c\n",
      "180 TRE6PHYDRO-RXN TREHALOSE-6P_c + WATER_c --> D-glucopyranose-6-phosphate_c + Glucopyranose_c\n",
      "\n",
      "181 GTP-CYCLOHYDRO-II-RXN GTP_c + 3.0 WATER_c --> DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + FORMATE_c + PPI_c + 2.0 PROTON_c\n",
      "181 GTP-CYCLOHYDRO-II-RXN GTP_c + 3.0 WATER_c --> DIAMINO-OH-PHOSPHORIBOSYLAMINO-PYR_c + FORMATE_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "182 MANNONOXIDOREDUCT-RXN D-MANNONATE_c + NAD_c <=> FRUCTURONATE_c + NADH_c + PROTON_c\n",
      "182 MANNONOXIDOREDUCT-RXN D-MANNONATE_c + NAD_c <=> FRUCTURONATE_c + NADH_c + PROTON_c\n",
      "\n",
      "183 GLYOCARBOLIG-RXN 2.0 GLYOX_c + PROTON_c --> CARBON-DIOXIDE_c + TARTRONATE-S-ALD_c\n",
      "183 GLYOCARBOLIG-RXN 2.0 GLYOX_c + PROTON_c --> CARBON-DIOXIDE_c + TARTRONATE-S-ALD_c\n",
      "\n",
      "184 QUINOPRIBOTRANS-RXN 2.0 PROTON_c + PRPP_c + QUINOLINATE_c --> CARBON-DIOXIDE_c + NICOTINATE_NUCLEOTIDE_c + PPI_c\n",
      "184 QUINOPRIBOTRANS-RXN 2.0 PROTON_c + PRPP_c + QUINOLINATE_c --> CARBON-DIOXIDE_c + NICOTINATE_NUCLEOTIDE_c + PPI_c\n",
      "\n",
      "185 MALONYL-COA-ACP-TRANSACYL-RXN ACP_c + MALONYL-COA_c <=> CO-A_c + MALONYL-ACP_c\n",
      "185 MALONYL-COA-ACP-TRANSACYL-RXN ACP_c + MALONYL-COA_c --> CO-A_c + MALONYL-ACP_c\n",
      "\n",
      "186 UDPNACETYLGLUCOSAMACYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-N-ACETYL-D-GLUCOSAMINE_c <=> ACP_c + UDP-OHMYR-ACETYLGLUCOSAMINE_c\n",
      "186 UDPNACETYLGLUCOSAMACYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-N-ACETYL-D-GLUCOSAMINE_c --> ACP_c + UDP-OHMYR-ACETYLGLUCOSAMINE_c\n",
      "\n",
      "187 SUCCDIAMINOPIMDESUCC-RXN N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c + WATER_c --> LL-DIAMINOPIMELATE_c + SUC_c\n",
      "187 SUCCDIAMINOPIMDESUCC-RXN N-SUCCINYLLL-2-6-DIAMINOPIMELATE_c + WATER_c --> LL-DIAMINOPIMELATE_c + SUC_c\n",
      "\n",
      "188 ADENPRIBOSYLTRAN-RXN ADENINE_c + PRPP_c --> AMP_c + PPI_c\n",
      "188 ADENPRIBOSYLTRAN-RXN ADENINE_c + PRPP_c --> AMP_c + PPI_c\n",
      "\n",
      "189 RXN-8992 5.0 DELTA3-ISOPENTENYL-PP_c + FARNESYL-PP_c --> OCTAPRENYL-DIPHOSPHATE_c + 5.0 PPI_c\n",
      "189 RXN-8992 5.0 DELTA3-ISOPENTENYL-PP_c + FARNESYL-PP_c --> OCTAPRENYL-DIPHOSPHATE_c + 5.0 PPI_c\n",
      "\n",
      "190 CATAL-RXN 2.0 HYDROGEN-PEROXIDE_c --> OXYGEN-MOLECULE_c + 2.0 WATER_c\n",
      "190 CATAL-RXN 2.0 HYDROGEN-PEROXIDE_c --> OXYGEN-MOLECULE_c + 2.0 WATER_c\n",
      "\n",
      "191 RXN-10819 ATP_c + CO-A_c + PHENYLACETATE_c --> AMP_c + CPD-207_c + PPI_c\n",
      "191 RXN-10819 ATP_c + CO-A_c + PHENYLACETATE_c --> AMP_c + CPD-207_c + PPI_c\n",
      "\n",
      "192 3.6.1.41-RXN ADENOSYL-P4_c + WATER_c --> 2.0 ADP_c + 2.0 PROTON_c\n",
      "192 3.6.1.41-RXN ADENOSYL-P4_c + WATER_c --> 2.0 ADP_c + 2.0 PROTON_c\n",
      "\n",
      "193 RXN0-4342 ATP_c + D-BETA-D-HEPTOSE-1-P_c + PROTON_c --> ADP-D-GLYCERO-D-MANNO-HEPTOSE_c + PPI_c\n",
      "193 RXN0-4342 ATP_c + D-BETA-D-HEPTOSE-1-P_c + PROTON_c --> ADP-D-GLYCERO-D-MANNO-HEPTOSE_c + PPI_c\n",
      "\n",
      "194 RXN-12002 ATP_c + UMP_c <=> ADP_c + UDP_c\n",
      "194 RXN-12002 ATP_c + UMP_c --> ADP_c + UDP_c\n",
      "\n",
      "195 RXN-7913 ATP_c + DCMP_c <=> ADP_c + DCDP_c\n",
      "195 RXN-7913 ATP_c + DCMP_c --> ADP_c + DCDP_c\n",
      "\n",
      "196 RXN-11832 ATP_c + CMP_c <=> ADP_c + CDP_c\n",
      "196 RXN-11832 ATP_c + CMP_c <=> ADP_c + CDP_c\n",
      "\n",
      "197 GMKALT-RXN ATP_c + DGMP_c <=> ADP_c + DGDP_c\n",
      "197 GMKALT-RXN ATP_c + DGMP_c <=> ADP_c + DGDP_c\n",
      "\n",
      "198 GUANYL-KIN-RXN ATP_c + GMP_c <=> ADP_c + GDP_c\n",
      "198 GUANYL-KIN-RXN ATP_c + GMP_c <=> ADP_c + GDP_c\n",
      "\n",
      "199 ACETOACETYL-COA-TRANSFER-RXN 3-KETOBUTYRATE_c + ACETYL-COA_c --> ACETOACETYL-COA_c + ACET_c\n",
      "199 ACETOACETYL-COA-TRANSFER-RXN 3-KETOBUTYRATE_c + ACETYL-COA_c <=> ACETOACETYL-COA_c + ACET_c\n",
      "\n",
      "200 DTDPGLUCOSEPP-RXN GLC-1-P_c + PROTON_c + TTP_c --> DTDP-D-GLUCOSE_c + PPI_c\n",
      "200 DTDPGLUCOSEPP-RXN GLC-1-P_c + PROTON_c + TTP_c --> DTDP-D-GLUCOSE_c + PPI_c\n",
      "\n",
      "201 DTDPGLUCDEHYDRAT-RXN DTDP-D-GLUCOSE_c --> DTDP-DEOH-DEOXY-GLUCOSE_c + WATER_c\n",
      "201 DTDPGLUCDEHYDRAT-RXN DTDP-D-GLUCOSE_c --> DTDP-DEOH-DEOXY-GLUCOSE_c + WATER_c\n",
      "\n",
      "202 DTDPDEHYDRHAMEPIM-RXN DTDP-DEOH-DEOXY-GLUCOSE_c --> DTDP-DEOH-DEOXY-MANNOSE_c\n",
      "202 DTDPDEHYDRHAMEPIM-RXN DTDP-DEOH-DEOXY-GLUCOSE_c --> DTDP-DEOH-DEOXY-MANNOSE_c\n",
      "\n",
      "203 DTDPDEHYRHAMREDUCT-RXN DTDP-DEOH-DEOXY-MANNOSE_c + NADPH_c + PROTON_c --> DTDP-RHAMNOSE_c + NADP_c\n",
      "203 DTDPDEHYRHAMREDUCT-RXN DTDP-RHAMNOSE_c + NADP_c <=> DTDP-DEOH-DEOXY-MANNOSE_c + NADPH_c + PROTON_c\n",
      "\n",
      "204 UDPNACETYLMURAMATEDEHYDROG-RXN NADPH_c + PROTON_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c --> NADP_c + UDP-N-ACETYLMURAMATE_c\n",
      "204 UDPNACETYLMURAMATEDEHYDROG-RXN NADPH_c + PROTON_c + UDP-ACETYL-CARBOXYVINYL-GLUCOSAMINE_c --> NADP_c + UDP-N-ACETYLMURAMATE_c\n",
      "\n",
      "205 RXN0-4301 D-SEDOHEPTULOSE-7-P_c --> D-ALPHABETA-D-HEPTOSE-7-PHOSPHATE_c\n",
      "205 RXN0-4301 D-SEDOHEPTULOSE-7-P_c --> D-ALPHABETA-D-HEPTOSE-7-PHOSPHATE_c\n",
      "\n",
      "206 RXN-9510 FMN_c + NADH_c + PROTON_c --> FMNH2_c + NAD_c\n",
      "206 RXN-9510 FMN_c + NADH_c + 2.0 PROTON_c --> FMNH2_c + NAD_c\n",
      "\n",
      "207 RXN-12444 FMN_c + NADPH_c + PROTON_c --> FMNH2_c + NADP_c\n",
      "207 RXN-12444 FMN_c + NADPH_c + 2.0 PROTON_c --> FMNH2_c + NADP_c\n",
      "\n",
      "208 GDPPYPHOSKIN-RXN ATP_c + GDP_c --> AMP_c + GUANOSINE-5DP-3DP_c + PROTON_c\n",
      "208 GDPPYPHOSKIN-RXN ATP_c + GDP_c --> AMP_c + GUANOSINE-5DP-3DP_c\n",
      "\n",
      "209 GTPPYPHOSKIN-RXN ATP_c + GTP_c --> AMP_c + GDP-TP_c + PROTON_c\n",
      "209 GTPPYPHOSKIN-RXN ATP_c + GTP_c --> AMP_c + GDP-TP_c\n",
      "\n",
      "210 PROTOHEMEFERROCHELAT-RXN FE+2_c + PROTOPORPHYRIN_IX_c --> PROTOHEME_c + 2.0 PROTON_c\n",
      "210 PROTOHEMEFERROCHELAT-RXN PROTOHEME_c + 2.0 PROTON_c <=> FE+2_c + PROTOPORPHYRIN_IX_c\n",
      "\n",
      "211 4OHBENZOATE-OCTAPRENYLTRANSFER-RXN 4-hydroxybenzoate_c + OCTAPRENYL-DIPHOSPHATE_c --> 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PPI_c\n",
      "211 4OHBENZOATE-OCTAPRENYLTRANSFER-RXN 4-hydroxybenzoate_c + OCTAPRENYL-DIPHOSPHATE_c --> 3-OCTAPRENYL-4-HYDROXYBENZOATE_c + PPI_c\n",
      "\n",
      "212 DEPHOSPHOCOAKIN-RXN ATP_c + DEPHOSPHO-COA_c --> ADP_c + CO-A_c + PROTON_c\n",
      "212 DEPHOSPHOCOAKIN-RXN ATP_c + DEPHOSPHO-COA_c --> ADP_c + CO-A_c + PROTON_c\n",
      "\n",
      "213 THIAZOLSYN3-RXN ATP_c + THZ_c --> ADP_c + PROTON_c + THZ-P_c\n",
      "213 THIAZOLSYN3-RXN ATP_c + THZ_c --> ADP_c + PROTON_c + THZ-P_c\n",
      "\n",
      "214 KDO-8PPHOSPHAT-RXN KDO-8P_c + WATER_c --> KDO_c + Pi_c\n",
      "214 KDO-8PPHOSPHAT-RXN KDO-8P_c + WATER_c --> KDO_c + Pi_c\n",
      "\n",
      "215 ARGININE-N-SUCCINYLTRANSFERASE-RXN ARG_c + SUC-COA_c --> CO-A_c + CPD-421_c + PROTON_c\n",
      "215 ARGININE-N-SUCCINYLTRANSFERASE-RXN ARG_c + SUC-COA_c --> CO-A_c + CPD-421_c + PROTON_c\n",
      "\n",
      "216 SUCCGLUDESUCC-RXN N2-SUCCINYLGLUTAMATE_c + WATER_c --> GLT_c + SUC_c\n",
      "216 SUCCGLUDESUCC-RXN N2-SUCCINYLGLUTAMATE_c + WATER_c --> GLT_c + SUC_c\n",
      "\n",
      "217 GUANINE-DEAMINASE-RXN GUANINE_c + PROTON_c + WATER_c --> AMMONIUM_c + XANTHINE_c\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "217 GUANINE-DEAMINASE-RXN GUANINE_c + PROTON_c + WATER_c --> AMMONIUM_c + XANTHINE_c\n",
      "\n",
      "218 DTMPKI-RXN ATP_c + TMP_c <=> ADP_c + TDP_c\n",
      "218 DTMPKI-RXN ATP_c + TMP_c --> ADP_c + TDP_c\n",
      "\n",
      "219 PHOSPHOGLUCMUT-RXN GLC-1-P_c <=> D-glucopyranose-6-phosphate_c\n",
      "219 PHOSPHOGLUCMUT-RXN GLC-1-P_c <=> D-glucopyranose-6-phosphate_c\n",
      "\n",
      "220 RXN0-5185 CPD-2961_c + WATER_c --> GLUCONATE_c + Pi_c\n",
      "220 RXN0-5185 CPD-2961_c + WATER_c --> GLUCONATE_c + Pi_c\n",
      "\n",
      "221 RIBOSYLHOMOCYSTEINASE-RXN CPD-564_c --> DIHYDROXYPENTANEDIONE_c + HOMO-CYS_c\n",
      "221 RIBOSYLHOMOCYSTEINASE-RXN CPD-564_c --> DIHYDROXYPENTANEDIONE_c + HOMO-CYS_c\n",
      "\n",
      "222 BADH-RXN BETAINE_ALDEHYDE_c + NAD_c + WATER_c --> BETAINE_c + NADH_c + 2.0 PROTON_c\n",
      "222 BADH-RXN BETAINE_ALDEHYDE_c + NAD_c + WATER_c <=> BETAINE_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "223 PYRNUTRANSHYDROGEN-RXN NADPH_c + NAD_c --> NADH_c + NADP_c\n",
      "223 PYRNUTRANSHYDROGEN-RXN NADPH_c + NAD_c <=> NADH_c + NADP_c\n",
      "\n",
      "224 GMP-SYN-GLUT-RXN ATP_c + GLN_c + WATER_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GLT_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "224 GMP-SYN-GLUT-RXN ATP_c + GLN_c + WATER_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GLT_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "225 GMP-SYN-NH3-RXN AMMONIUM_c + ATP_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "225 GMP-SYN-NH3-RXN AMMONIUM_c + ATP_c + XANTHOSINE-5-PHOSPHATE_c --> AMP_c + GMP_c + PPI_c + 2.0 PROTON_c\n",
      "\n",
      "226 PABASYN-RXN CHORISMATE_c + GLN_c --> 4-AMINO-4-DEOXYCHORISMATE_c + GLT_c\n",
      "226 PABASYN-RXN CHORISMATE_c + GLN_c <=> 4-AMINO-4-DEOXYCHORISMATE_c + GLT_c\n",
      "\n",
      "227 ADCLY-RXN 4-AMINO-4-DEOXYCHORISMATE_c --> P-AMINO-BENZOATE_c + PROTON_c + PYRUVATE_c\n",
      "227 ADCLY-RXN 4-AMINO-4-DEOXYCHORISMATE_c --> P-AMINO-BENZOATE_c + PROTON_c + PYRUVATE_c\n",
      "\n",
      "228 URACIL-PRIBOSYLTRANS-RXN PRPP_c + URACIL_c --> PPI_c + UMP_c\n",
      "228 URACIL-PRIBOSYLTRANS-RXN PRPP_c + URACIL_c --> PPI_c + UMP_c\n",
      "\n",
      "229 CTPSYN-RXN ATP_c + GLN_c + UTP_c + WATER_c --> ADP_c + CTP_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "229 CTPSYN-RXN ATP_c + GLN_c + UTP_c + WATER_c --> ADP_c + CTP_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "230 5.1.3.20-RXN ADP-D-GLYCERO-D-MANNO-HEPTOSE_c --> ADP-L-GLYCERO-D-MANNO-HEPTOSE_c\n",
      "230 5.1.3.20-RXN ADP-D-GLYCERO-D-MANNO-HEPTOSE_c --> ADP-L-GLYCERO-D-MANNO-HEPTOSE_c\n",
      "\n",
      "231 RXN0-300 NADPH_c + OH-PYR_c + PROTON_c --> GLYCERATE_c + NADP_c\n",
      "231 RXN0-300 NADPH_c + OH-PYR_c + PROTON_c --> GLYCERATE_c + NADP_c\n",
      "\n",
      "232 GLYOXYLATE-REDUCTASE-NADP+-RXN GLYOX_c + NADPH_c + PROTON_c --> GLYCOLLATE_c + NADP_c\n",
      "232 GLYOXYLATE-REDUCTASE-NADP+-RXN GLYOX_c + NADPH_c + PROTON_c --> GLYCOLLATE_c + NADP_c\n",
      "\n",
      "233 ACETYLORNDEACET-RXN N-ALPHA-ACETYLORNITHINE_c + WATER_c --> ACET_c + L-ORNITHINE_c\n",
      "233 ACETYLORNDEACET-RXN N-ALPHA-ACETYLORNITHINE_c + WATER_c <=> ACET_c + L-ORNITHINE_c\n",
      "\n",
      "234 ADDALT-RXN DEOXYADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + DEOXYINOSINE_c\n",
      "234 ADDALT-RXN DEOXYADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + DEOXYINOSINE_c\n",
      "\n",
      "235 ADENODEAMIN-RXN ADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + INOSINE_c\n",
      "235 ADENODEAMIN-RXN ADENOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + INOSINE_c\n",
      "\n",
      "236 ADENYLOSUCCINATE-SYNTHASE-RXN GTP_c + IMP_c + L-ASPARTATE_c --> ADENYLOSUCC_c + GDP_c + 2.0 PROTON_c + Pi_c\n",
      "236 ADENYLOSUCCINATE-SYNTHASE-RXN GTP_c + IMP_c + L-ASPARTATE_c --> ADENYLOSUCC_c + GDP_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "237 ASPARAGHYD-RXN ASN_p + WATER_p --> AMMONIUM_p + L-ASPARTATE_p\n",
      "237 ASPARAGHYD-RXN ASN_c + WATER_c --> AMMONIUM_c + L-ASPARTATE_c\n",
      "\n",
      "238 ASPCARBTRANS-RXN CARBAMOYL-P_c + L-ASPARTATE_c --> CARBAMYUL-L-ASPARTATE_c + PROTON_c + Pi_c\n",
      "238 ASPCARBTRANS-RXN CARBAMOYL-P_c + L-ASPARTATE_c <=> CARBAMYUL-L-ASPARTATE_c + PROTON_c + Pi_c\n",
      "\n",
      "239 DIAMINOPIMEPIM-RXN LL-DIAMINOPIMELATE_c <=> MESO-DIAMINOPIMELATE_c\n",
      "239 DIAMINOPIMEPIM-RXN LL-DIAMINOPIMELATE_c <=> MESO-DIAMINOPIMELATE_c\n",
      "\n",
      "240 SUCCORNTRANSAM-RXN 2-KETOGLUTARATE_c + N2-SUCCINYLORNITHINE_c --> CPD-822_c + GLT_c\n",
      "240 SUCCORNTRANSAM-RXN 2-KETOGLUTARATE_c + N2-SUCCINYLORNITHINE_c <=> CPD-822_c + GLT_c\n",
      "\n",
      "241 THRESYN-RXN O-PHOSPHO-L-HOMOSERINE_c + WATER_c --> Pi_c + THR_c\n",
      "241 THRESYN-RXN O-PHOSPHO-L-HOMOSERINE_c + WATER_c --> Pi_c + THR_c\n",
      "\n",
      "242 DAHPSYN-RXN ERYTHROSE-4P_c + PHOSPHO-ENOL-PYRUVATE_c + WATER_c --> 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c + Pi_c\n",
      "242 DAHPSYN-RXN ERYTHROSE-4P_c + PHOSPHO-ENOL-PYRUVATE_c + WATER_c <=> 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c + Pi_c\n",
      "\n",
      "243 NMNAMIDOHYDRO-RXN NICOTINAMIDE_NUCLEOTIDE_c + WATER_c --> AMMONIUM_c + NICOTINATE_NUCLEOTIDE_c\n",
      "243 NMNAMIDOHYDRO-RXN NICOTINAMIDE_NUCLEOTIDE_c + WATER_c --> AMMONIUM_c + NICOTINATE_NUCLEOTIDE_c\n",
      "\n",
      "244 SPERMACTRAN-RXN ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-568_c + PROTON_c\n",
      "244 SPERMACTRAN-RXN ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-568_c + PROTON_c\n",
      "\n",
      "245 RXN0-7165 ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-3462_c + PROTON_c\n",
      "245 RXN0-7165 ACETYL-COA_c + SPERMIDINE_c --> CO-A_c + CPD-3462_c + PROTON_c\n",
      "\n",
      "246 MALSYN-RXN ACETYL-COA_c + GLYOX_c + WATER_c --> CO-A_c + MAL_c + PROTON_c\n",
      "246 MALSYN-RXN ACETYL-COA_c + GLYOX_c + WATER_c --> CO-A_c + MAL_c + PROTON_c\n",
      "\n",
      "247 GLUC1PURIDYLTRANS-RXN GLC-1-P_c + PROTON_c + UTP_c --> CPD-12575_c + PPI_c\n",
      "247 GLUC1PURIDYLTRANS-RXN GLC-1-P_c + PROTON_c + UTP_c <=> CPD-12575_c + PPI_c\n",
      "\n",
      "248 UDPHYDROXYMYRGLUCOSAMNACETYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-OHMYR-GLUCOSAMINE_c --> ACP_c + OH-MYRISTOYL_c + PROTON_c\n",
      "248 UDPHYDROXYMYRGLUCOSAMNACETYLTRANS-RXN R-3-hydroxymyristoyl-ACPs_c + UDP-OHMYR-GLUCOSAMINE_c --> ACP_c + OH-MYRISTOYL_c + PROTON_c\n",
      "\n",
      "249 UDP-NACMUR-ALA-LIG-RXN ATP_c + L-ALPHA-ALANINE_c + UDP-N-ACETYLMURAMATE_c --> ADP_c + CPD0-1456_c + PROTON_c + Pi_c\n",
      "249 UDP-NACMUR-ALA-LIG-RXN ATP_c + L-ALPHA-ALANINE_c + UDP-N-ACETYLMURAMATE_c --> ADP_c + CPD0-1456_c + PROTON_c + Pi_c\n",
      "\n",
      "250 UDP-NACMURALA-GLU-LIG-RXN ATP_c + CPD0-1456_c + D-GLT_c --> ADP_c + PROTON_c + Pi_c + UDP-AA-GLUTAMATE_c\n",
      "250 UDP-NACMURALA-GLU-LIG-RXN ATP_c + CPD0-1456_c + D-GLT_c --> ADP_c + PROTON_c + Pi_c + UDP-AA-GLUTAMATE_c\n",
      "\n",
      "251 UDPACYLGLCNACDEACETYL-RXN UDP-OHMYR-ACETYLGLUCOSAMINE_c + WATER_c --> ACET_c + UDP-OHMYR-GLUCOSAMINE_c\n",
      "251 UDPACYLGLCNACDEACETYL-RXN UDP-OHMYR-ACETYLGLUCOSAMINE_c + WATER_c --> ACET_c + UDP-OHMYR-GLUCOSAMINE_c\n",
      "\n",
      "252 IMIDPHOSDEHYD-RXN D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c --> IMIDAZOLE-ACETOL-P_c + WATER_c\n",
      "252 IMIDPHOSDEHYD-RXN D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c --> IMIDAZOLE-ACETOL-P_c + WATER_c\n",
      "\n",
      "253 PRIBFAICARPISOM-RXN PHOSPHORIBOSYL-FORMIMINO-AICAR-P_c <=> PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c\n",
      "253 PRIBFAICARPISOM-RXN PHOSPHORIBOSYL-FORMIMINO-AICAR-P_c --> PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c\n",
      "\n",
      "254 SERINE-O-ACETTRAN-RXN ACETYL-COA_c + SER_c <=> ACETYLSERINE_c + CO-A_c\n",
      "254 SERINE-O-ACETTRAN-RXN ACETYL-COA_c + SER_c --> ACETYLSERINE_c + CO-A_c\n",
      "\n",
      "255 ACSERLY-RXN ACETYLSERINE_c + HS_c --> ACET_c + CYS_c + PROTON_c\n",
      "255 ACSERLY-RXN ACETYLSERINE_c + HS_c <=> ACET_c + CYS_c + PROTON_c\n",
      "\n",
      "256 325-BISPHOSPHATE-NUCLEOTIDASE-RXN 3-5-ADP_c + WATER_c --> AMP_c + Pi_c\n",
      "256 325-BISPHOSPHATE-NUCLEOTIDASE-RXN 3-5-ADP_c + WATER_c --> AMP_c + Pi_c\n",
      "\n",
      "257 ASPARTATE-SEMIALDEHYDE-DEHYDROGENASE-RXN L-ASPARTATE-SEMIALDEHYDE_c + NADP_c + Pi_c <=> L-BETA-ASPARTYL-P_c + NADPH_c + PROTON_c\n",
      "257 ASPARTATE-SEMIALDEHYDE-DEHYDROGENASE-RXN L-ASPARTATE-SEMIALDEHYDE_c + NADP_c + Pi_c <=> L-BETA-ASPARTYL-P_c + NADPH_c + PROTON_c\n",
      "\n",
      "258 DIAMINOPIMDECARB-RXN MESO-DIAMINOPIMELATE_c + PROTON_c --> CARBON-DIOXIDE_c + LYS_c\n",
      "258 DIAMINOPIMDECARB-RXN MESO-DIAMINOPIMELATE_c + PROTON_c --> CARBON-DIOXIDE_c + LYS_c\n",
      "\n",
      "259 S-ADENMETSYN-RXN ATP_c + MET_c + WATER_c --> PPI_c + Pi_c + S-ADENOSYLMETHIONINE_c\n",
      "259 S-ADENMETSYN-RXN ATP_c + MET_c + WATER_c --> PPI_c + Pi_c + S-ADENOSYLMETHIONINE_c\n",
      "\n",
      "260 THI-P-KIN-RXN ATP_c + THIAMINE-P_c --> ADP_c + THIAMINE-PYROPHOSPHATE_c\n",
      "260 THI-P-KIN-RXN ATP_c + THIAMINE-P_c --> ADP_c + THIAMINE-PYROPHOSPHATE_c\n",
      "\n",
      "261 GKI-RXN ATP_c + GLYCERATE_c --> 2-PG_c + ADP_c + PROTON_c\n",
      "261 GKI-RXN ATP_c + GLYCERATE_c --> 2-PG_c + ADP_c + PROTON_c\n",
      "\n",
      "262 GLUCOKIN-RXN ATP_c + Glucopyranose_c --> ADP_c + D-glucopyranose-6-phosphate_c + PROTON_c\n",
      "262 GLUCOKIN-RXN ATP_c + Glucopyranose_c --> ADP_c + D-glucopyranose-6-phosphate_c + PROTON_c\n",
      "\n",
      "263 MALIC-NADP-RXN MAL_c + NADP_c --> CARBON-DIOXIDE_c + NADPH_c + PYRUVATE_c\n",
      "263 MALIC-NADP-RXN MAL_c + NADP_c --> CARBON-DIOXIDE_c + NADPH_c + PYRUVATE_c\n",
      "\n",
      "264 ARGSUCCINLYA-RXN L-ARGININO-SUCCINATE_c <=> ARG_c + FUM_c\n",
      "264 ARGSUCCINLYA-RXN L-ARGININO-SUCCINATE_c <=> ARG_c + FUM_c\n",
      "\n",
      "265 CARBPSYN-RXN 2.0 ATP_c + GLN_c + HCO3_c + WATER_c --> 2.0 ADP_c + CARBAMOYL-P_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "265 CARBPSYN-RXN 2.0 ATP_c + GLN_c + HCO3_c + WATER_c --> 2.0 ADP_c + CARBAMOYL-P_c + GLT_c + 2.0 PROTON_c + Pi_c\n",
      "\n",
      "266 GLUTKIN-RXN ATP_c + GLT_c --> ADP_c + L-GLUTAMATE-5-P_c\n",
      "266 GLUTKIN-RXN ATP_c + GLT_c --> ADP_c + L-GLUTAMATE-5-P_c\n",
      "\n",
      "267 2.3.1.180-RXN ACETYL-COA_c + MALONYL-ACP_c + PROTON_c --> Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c + CO-A_c\n",
      "267 2.3.1.180-RXN ACETYL-COA_c + MALONYL-ACP_c + PROTON_c --> Acetoacetyl-ACPs_c + CARBON-DIOXIDE_c + CO-A_c\n",
      "\n",
      "268 SUPEROX-DISMUT-RXN 2.0 PROTON_p + 2.0 SUPER-OXIDE_p --> HYDROGEN-PEROXIDE_p + OXYGEN-MOLECULE_p\n",
      "268 SUPEROX-DISMUT-RXN 2.0 PROTON_c + 2.0 SUPER-OXIDE_c --> HYDROGEN-PEROXIDE_c + OXYGEN-MOLECULE_c\n",
      "\n",
      "269 ADENOSYLHOMOCYSTEINE-NUCLEOSIDASE-RXN ADENOSYL-HOMO-CYS_c + WATER_c --> ADENINE_c + CPD-564_c\n",
      "269 ADENOSYLHOMOCYSTEINE-NUCLEOSIDASE-RXN ADENOSYL-HOMO-CYS_c + WATER_c --> ADENINE_c + CPD-564_c\n",
      "\n",
      "270 METHYLTHIOADENOSINE-NUCLEOSIDASE-RXN 5-METHYLTHIOADENOSINE_c + WATER_c --> ADENINE_c + CPD-560_c\n",
      "270 METHYLTHIOADENOSINE-NUCLEOSIDASE-RXN 5-METHYLTHIOADENOSINE_c + WATER_c --> ADENINE_c + CPD-560_c\n",
      "\n",
      "271 THIOSULFATE-SULFURTRANSFERASE-RXN CPD-13584_p + S2O3_p --> CPD-19512_p + H2SO3_p + PROTON_p\n",
      "271 THIOSULFATE-SULFURTRANSFERASE-RXN HCN_c + S2O3_c --> HSCN_c + 2.0 PROTON_c + SO3_c\n",
      "\n",
      "272 DETHIOBIOTIN-SYN-RXN ATP_c + CARBON-DIOXIDE_c + DIAMINONONANOATE_c --> ADP_c + DETHIOBIOTIN_c + 3.0 PROTON_c + Pi_c\n",
      "272 DETHIOBIOTIN-SYN-RXN ATP_c + CARBON-DIOXIDE_c + DIAMINONONANOATE_c --> ADP_c + DETHIOBIOTIN_c + 3.0 PROTON_c + Pi_c\n",
      "\n",
      "273 DCTP-DEAM-RXN DCTP_c + PROTON_c + WATER_c --> AMMONIUM_c + DUTP_c\n",
      "273 DCTP-DEAM-RXN DCTP_c + PROTON_c + WATER_c --> AMMONIUM_c + DUTP_c\n",
      "\n",
      "274 CYTDEAM-RXN CYTOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + URACIL_c\n",
      "274 CYTDEAM-RXN CYTOSINE_c + PROTON_c + WATER_c --> AMMONIUM_c + URACIL_c\n",
      "\n",
      "275 L-GLN-FRUCT-6-P-AMINOTRANS-RXN FRUCTOSE-6P_c + GLN_c --> D-GLUCOSAMINE-6-P_c + GLT_c\n",
      "275 L-GLN-FRUCT-6-P-AMINOTRANS-RXN FRUCTOSE-6P_c + GLN_c <=> D-GLUCOSAMINE-6-P_c + GLT_c\n",
      "\n",
      "276 GLUTCYSLIG-RXN ATP_c + CYS_c + GLT_c --> ADP_c + L-GAMMA-GLUTAMYLCYSTEINE_c + PROTON_c + Pi_c\n",
      "276 GLUTCYSLIG-RXN ATP_c + CYS_c + GLT_c --> ADP_c + L-GAMMA-GLUTAMYLCYSTEINE_c + PROTON_c + Pi_c\n",
      "\n",
      "277 GLUTAMIN-RXN GLN_p + WATER_p --> AMMONIUM_p + GLT_p\n",
      "277 GLUTAMIN-RXN GLN_c + WATER_c --> AMMONIUM_c + GLT_c\n",
      "\n",
      "278 PRPPAMIDOTRANS-RXN GLN_c + PRPP_c + WATER_c --> 5-P-BETA-D-RIBOSYL-AMINE_c + GLT_c + PPI_c\n",
      "278 PRPPAMIDOTRANS-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + GLT_c + PPI_c <=> GLN_c + PRPP_c + WATER_c\n",
      "\n",
      "279 RXN-13854 3.0 NADH_c + NITRITE_c + 5.0 PROTON_c --> AMMONIUM_c + 3.0 NAD_c + 2.0 WATER_c\n",
      "279 RXN-13854 3.0 NADH_c + NITRITE_c + 5.0 PROTON_c --> AMMONIUM_c + 3.0 NAD_c + 2.0 WATER_c\n",
      "\n",
      "280 ORNCARBAMTRANSFER-RXN CARBAMOYL-P_c + L-ORNITHINE_c <=> L-CITRULLINE_c + PROTON_c + Pi_c\n",
      "280 ORNCARBAMTRANSFER-RXN CARBAMOYL-P_c + L-ORNITHINE_c <=> L-CITRULLINE_c + PROTON_c + Pi_c\n",
      "\n",
      "281 GART-RXN 10-FORMYL-THF_c + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE_c + PROTON_c + THF_c\n",
      "281 GART-RXN 10-FORMYL-THF_c + 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c <=> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE_c + PROTON_c + THF_c\n",
      "\n",
      "282 METHYLISOCITRATE-LYASE-RXN CPD-618_c <=> PYRUVATE_c + SUC_c\n",
      "282 METHYLISOCITRATE-LYASE-RXN CPD-618_c <=> PYRUVATE_c + SUC_c\n",
      "\n",
      "283 SUCCGLUALDDEHYD-RXN CPD-822_c + NAD_c + WATER_c --> N2-SUCCINYLGLUTAMATE_c + NADH_c + 2.0 PROTON_c\n",
      "283 SUCCGLUALDDEHYD-RXN CPD-822_c + NAD_c + WATER_c --> N2-SUCCINYLGLUTAMATE_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "284 PORPHOBILSYNTH-RXN 2.0 5-AMINO-LEVULINATE_c --> PORPHOBILINOGEN_c + PROTON_c + 2.0 WATER_c\n",
      "284 PORPHOBILSYNTH-RXN 2.0 5-AMINO-LEVULINATE_c --> PORPHOBILINOGEN_c + PROTON_c + 2.0 WATER_c\n",
      "\n",
      "285 GLYRIBONUCSYN-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + ATP_c + GLY_c <=> 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c + ADP_c + PROTON_c + Pi_c\n",
      "285 GLYRIBONUCSYN-RXN 5-P-BETA-D-RIBOSYL-AMINE_c + ATP_c + GLY_c --> 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c + ADP_c + PROTON_c + Pi_c\n",
      "\n",
      "286 ETHAMLY-RXN ETHANOL-AMINE_c --> ACETALD_c + AMMONIUM_c\n",
      "286 ETHAMLY-RXN ETHANOL-AMINE_c --> ACETALD_c + AMMONIUM_c\n",
      "\n",
      "287 RXN0-5408 D-MYO-INOSITOL-1-MONOPHOSPHATE_c + WATER_c --> MYO-INOSITOL_c + Pi_c\n",
      "287 RXN0-5408 D-MYO-INOSITOL-1-MONOPHOSPHATE_c + WATER_c --> MYO-INOSITOL_c + Pi_c\n",
      "\n",
      "288 GLYCEROL-2-PHOSPHATASE-RXN CPD-536_c + WATER_c --> GLYCEROL_c + Pi_c\n",
      "288 GLYCEROL-2-PHOSPHATASE-RXN CPD-536_c + WATER_c --> GLYCEROL_c + Pi_c\n",
      "\n",
      "289 2-METHYLCITRATE-SYNTHASE-RXN OXALACETIC_ACID_c + PROPIONYL-COA_c + WATER_c --> CO-A_c + CPD-622_c + PROTON_c\n",
      "289 2-METHYLCITRATE-SYNTHASE-RXN OXALACETIC_ACID_c + PROPIONYL-COA_c + WATER_c <=> CO-A_c + CPD-622_c + PROTON_c\n",
      "\n",
      "290 ATPASE-RXN ATP_c + WATER_c --> ADP_c + PROTON_c + Pi_c\n",
      "290 ATPASE-RXN ATP_c + WATER_c --> ADP_c + PROTON_c + Pi_c\n",
      "\n",
      "291 RXN0-5462 GTP_p + WATER_p --> GDP_p + PROTON_p + Pi_p\n",
      "291 RXN0-5462 GTP_c + WATER_c --> GDP_c + PROTON_c + Pi_c\n",
      "\n",
      "292 GLU6PDEHYDROG-RXN D-glucopyranose-6-phosphate_c + NADP_c <=> D-6-P-GLUCONO-DELTA-LACTONE_c + NADPH_c + PROTON_c\n",
      "292 GLU6PDEHYDROG-RXN D-glucopyranose-6-phosphate_c + NADP_c --> D-6-P-GLUCONO-DELTA-LACTONE_c + NADPH_c + PROTON_c\n",
      "\n",
      "293 UGD-RXN CPD-12575_c + 2.0 NAD_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-GLUCURONATE_c\n",
      "293 UGD-RXN CPD-12575_c + 2.0 NAD_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-GLUCURONATE_c\n",
      "\n",
      "294 UDPMANNACADEHYDROG-RXN 2.0 NAD_c + UDP-MANNAC_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-MANNACA_c\n",
      "294 UDPMANNACADEHYDROG-RXN 2.0 NAD_c + UDP-MANNAC_c + WATER_c --> 2.0 NADH_c + 3.0 PROTON_c + UDP-MANNACA_c\n",
      "\n",
      "295 GLUTDECARBOX-RXN GLT_c + PROTON_c --> 4-AMINO-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "295 GLUTDECARBOX-RXN GLT_c + PROTON_c --> 4-AMINO-BUTYRATE_c + CARBON-DIOXIDE_c\n",
      "\n",
      "296 CYSTATHIONINE-BETA-LYASE-RXN L-CYSTATHIONINE_c + WATER_c --> AMMONIUM_c + HOMO-CYS_c + PYRUVATE_c\n",
      "296 CYSTATHIONINE-BETA-LYASE-RXN L-CYSTATHIONINE_c + WATER_c --> AMMONIUM_c + HOMO-CYS_c + PYRUVATE_c\n",
      "\n",
      "297 RXN-12565 ACETYL-COA_c + BUTYRYL-COA_c <=> CO-A_c + K-HEXANOYL-COA_c\n",
      "297 RXN-12565 ACETYL-COA_c + BUTYRYL-COA_c <=> CO-A_c + K-HEXANOYL-COA_c\n",
      "\n",
      "298 RXN-13617 ACETYL-COA_c + CPD-196_c <=> CO-A_c + CPD0-2123_c\n",
      "298 RXN-13617 CO-A_c + CPD0-2123_c --> ACETYL-COA_c + CPD-196_c\n",
      "\n",
      "299 ACETYL-COA-ACETYLTRANSFER-RXN 2.0 ACETYL-COA_c <=> ACETOACETYL-COA_c + CO-A_c\n",
      "299 ACETYL-COA-ACETYLTRANSFER-RXN 2.0 ACETYL-COA_c <=> ACETOACETYL-COA_c + CO-A_c\n",
      "\n",
      "300 RXN0-5114 3-P-SERINE_p + WATER_p --> Pi_p + SER_p\n",
      "300 RXN0-5114 3-P-SERINE_c + WATER_c --> Pi_c + SER_c\n",
      "\n",
      "301 RXN-16701 CPD-381_c + NAD_c <=> 2-KETOGLUTARATE_c + NADH_c + PROTON_c\n",
      "301 RXN-16701 CPD-381_c + NAD_c <=> 2-KETOGLUTARATE_c + NADH_c + PROTON_c\n",
      "\n",
      "302 PGLYCDEHYDROG-RXN G3P_c + NAD_c --> 3-P-HYDROXYPYRUVATE_c + NADH_c + PROTON_c\n",
      "302 PGLYCDEHYDROG-RXN G3P_c + NAD_c <=> 3-P-HYDROXYPYRUVATE_c + NADH_c + PROTON_c\n",
      "\n",
      "303 GTP-CYCLOHYDRO-I-RXN GTP_c + WATER_c --> DIHYDRONEOPTERIN-P3_c + FORMATE_c + PROTON_c\n",
      "303 GTP-CYCLOHYDRO-I-RXN GTP_c + WATER_c --> DIHYDRONEOPTERIN-P3_c + FORMATE_c + PROTON_c\n",
      "\n",
      "304 DIHYDROFOLATESYNTH-RXN 7-8-DIHYDROPTEROATE_c + ATP_c + GLT_c --> ADP_c + DIHYDROFOLATE_c + PROTON_c + Pi_c\n",
      "304 DIHYDROFOLATESYNTH-RXN 7-8-DIHYDROPTEROATE_c + ATP_c + GLT_c --> ADP_c + DIHYDROFOLATE_c + PROTON_c + Pi_c\n",
      "\n",
      "305 SAICARSYN-RXN ATP_c + L-ASPARTATE_c + PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c --> ADP_c + P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c + PROTON_c + Pi_c\n",
      "305 SAICARSYN-RXN ATP_c + L-ASPARTATE_c + PHOSPHORIBOSYL-CARBOXY-AMINOIMIDAZOLE_c --> ADP_c + P-RIBOSYL-4-SUCCCARB-AMINOIMIDAZOLE_c + PROTON_c + Pi_c\n",
      "\n",
      "306 RXN0-5289 NADH_c + PROTON_c + TARTRONATE-S-ALD_c <=> GLYCERATE_c + NAD_c\n",
      "306 RXN0-5289 GLYCERATE_c + NAD_c <=> NADH_c + PROTON_c + TARTRONATE-S-ALD_c\n",
      "\n",
      "307 H2PTEROATESYNTH-RXN DIHYDROPTERIN-CH2OH-PP_c + P-AMINO-BENZOATE_c --> 7-8-DIHYDROPTEROATE_c + PPI_c\n",
      "307 H2PTEROATESYNTH-RXN DIHYDROPTERIN-CH2OH-PP_c + P-AMINO-BENZOATE_c --> 7-8-DIHYDROPTEROATE_c + PPI_c\n",
      "\n",
      "308 METHYLENETHFDEHYDROG-NADP-RXN METHYLENE-THF_c + NADP_c <=> 5-10-METHENYL-THF_c + NADPH_c\n",
      "308 METHYLENETHFDEHYDROG-NADP-RXN METHYLENE-THF_c + NADP_c <=> 5-10-METHENYL-THF_c + NADPH_c\n",
      "\n",
      "309 METHENYLTHFCYCLOHYDRO-RXN 5-10-METHENYL-THF_c + WATER_c <=> 10-FORMYL-THF_c + PROTON_c\n",
      "309 METHENYLTHFCYCLOHYDRO-RXN 5-10-METHENYL-THF_c + WATER_c <=> 10-FORMYL-THF_c + PROTON_c\n",
      "\n",
      "310 RXN0-3942 CPD-9000_c + WATER_c --> 4-AMINO-BUTYRATE_c + GLT_c\n",
      "310 RXN0-3942 CPD-9000_c + WATER_c --> 4-AMINO-BUTYRATE_c + GLT_c\n",
      "\n",
      "311 RXN-6622 CYS-GLY_c + WATER_c --> CYS_c + GLY_c\n",
      "311 RXN-6622 CYS-GLY_c + WATER_c --> CYS_c + GLY_c\n",
      "\n",
      "312 RXN0-6981 CPD0-1445_c + WATER_c --> GLT_c + L-ALPHA-ALANINE_c\n",
      "312 RXN0-6981 CPD0-1445_c + WATER_c --> GLT_c + L-ALPHA-ALANINE_c\n",
      "\n",
      "313 RXN0-1483 4.0 FE+2_p + OXYGEN-MOLECULE_p + 4.0 PROTON_p --> 4.0 FE+3_p + 2.0 WATER_p\n",
      "313 RXN0-1483 4.0 FE+2_c + OXYGEN-MOLECULE_c + 4.0 PROTON_c --> 4.0 FE+3_c + 2.0 WATER_c\n",
      "\n",
      "314 BTUR2-RXN ATP_c + COBINAMIDE_c + PROTON_c <=> ADENOSYLCOBINAMIDE_c + P3I_c\n",
      "314 BTUR2-RXN 2.0 ATP_c + 2.0 CPD-20903_c + Reduced-Flavoproteins_c --> 2.0 ADENOSYLCOBINAMIDE_c + Oxidized-Flavoproteins_c + 2.0 P3I_c + 2.0 PROTON_c\n",
      "\n",
      "315 PPGPPSYN-RXN GUANOSINE-5DP-3DP_c + WATER_c --> GDP_c + PPI_c\n",
      "315 PPGPPSYN-RXN GUANOSINE-5DP-3DP_c + WATER_c --> GDP_c + PPI_c + PROTON_c\n",
      "\n",
      "316 RXN0-6427 GDP-TP_c + WATER_c --> GTP_c + PPI_c\n",
      "316 RXN0-6427 GDP-TP_c + WATER_c --> GTP_c + PPI_c + PROTON_c\n",
      "\n",
      "317 RXN0-6382 ITP_c + WATER_c --> IMP_c + PPI_c + PROTON_c\n",
      "317 RXN0-6382 ITP_c + WATER_c --> IMP_c + PPI_c + PROTON_c\n",
      "\n",
      "318 RXN0-1602 DITP_c + WATER_c --> DIMP_c + PPI_c + PROTON_c\n",
      "318 RXN0-1602 DITP_c + WATER_c --> DIMP_c + PPI_c + PROTON_c\n",
      "\n",
      "319 RXN0-1603 WATER_c + XTP_c --> PPI_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "319 RXN0-1603 WATER_c + XTP_c --> PPI_c + PROTON_c + XANTHOSINE-5-PHOSPHATE_c\n",
      "\n",
      "320 RXN-15744 LACTALD_c + NADPH_c + PROTON_c --> NADP_c + PROPANE-1-2-DIOL_c\n",
      "320 RXN-15744 LACTALD_c + NADPH_c + PROTON_c --> NADP_c + PROPANE-1-2-DIOL_c\n",
      "\n",
      "321 RXN-15743 CPD-358_c + NADPH_c + PROTON_c --> CPD-8891_c + NADP_c\n",
      "321 RXN-15743 CPD-358_c + NADPH_c + PROTON_c --> CPD-8891_c + NADP_c\n",
      "\n",
      "322 GLUCONOKIN-RXN ATP_c + GLUCONATE_c --> ADP_c + CPD-2961_c + PROTON_c\n",
      "322 GLUCONOKIN-RXN ATP_c + GLUCONATE_c --> ADP_c + CPD-2961_c + PROTON_c\n",
      "\n",
      "323 OHMETHYLBILANESYN-RXN 4.0 PORPHOBILINOGEN_c + WATER_c --> 4.0 AMMONIUM_c + HYDROXYMETHYLBILANE_c\n",
      "323 OHMETHYLBILANESYN-RXN 4.0 PORPHOBILINOGEN_c + WATER_c --> 4.0 AMMONIUM_c + HYDROXYMETHYLBILANE_c\n",
      "\n",
      "324 2.7.7.60-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + CTP_c + PROTON_c --> 4-CYTIDINE-5-DIPHOSPHO-2-C_c + PPI_c\n",
      "324 2.7.7.60-RXN 2-C-METHYL-D-ERYTHRITOL-4-PHOSPHATE_c + CTP_c + PROTON_c --> 4-CYTIDINE-5-DIPHOSPHO-2-C_c + PPI_c\n",
      "\n",
      "325 GLUTAMATESYN-RXN 2-KETOGLUTARATE_c + GLN_c + NADPH_c + PROTON_c --> 2.0 GLT_c + NADP_c\n",
      "325 GLUTAMATESYN-RXN 2-KETOGLUTARATE_c + GLN_c + NADPH_c + PROTON_c --> 2.0 GLT_c + NADP_c\n",
      "\n",
      "326 GLUTDEHYD-RXN GLT_c + NADP_c + WATER_c <=> 2-KETOGLUTARATE_c + AMMONIUM_c + NADPH_c + PROTON_c\n",
      "326 GLUTDEHYD-RXN GLT_c + NADP_c + WATER_c <=> 2-KETOGLUTARATE_c + AMMONIUM_c + NADPH_c + PROTON_c\n",
      "\n",
      "327 GLYOXII-RXN S-LACTOYL-GLUTATHIONE_c + WATER_c --> D-LACTATE_c + GLUTATHIONE_c + PROTON_c\n",
      "327 GLYOXII-RXN S-LACTOYL-GLUTATHIONE_c + WATER_c --> D-LACTATE_c + GLUTATHIONE_c + PROTON_c\n",
      "\n",
      "328 ERYTH4PDEHYDROG-RXN ERYTHROSE-4P_c + NAD_c + WATER_c <=> ERYTHRONATE-4P_c + NADH_c + 2.0 PROTON_c\n",
      "328 ERYTH4PDEHYDROG-RXN ERYTHROSE-4P_c + NAD_c + WATER_c --> ERYTHRONATE-4P_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "329 GLUTAMIDOTRANS-RXN GLN_c + PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c --> AICAR_c + D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c + GLT_c + PROTON_c\n",
      "329 GLUTAMIDOTRANS-RXN GLN_c + PHOSPHORIBULOSYL-FORMIMINO-AICAR-P_c --> AICAR_c + D-ERYTHRO-IMIDAZOLE-GLYCEROL-P_c + GLT_c + PROTON_c\n",
      "\n",
      "330 DUTP-PYROP-RXN DUTP_c + WATER_c --> DUMP_c + PPI_c + PROTON_c\n",
      "330 DUTP-PYROP-RXN DUTP_c + WATER_c --> DUMP_c + PPI_c + PROTON_c\n",
      "\n",
      "331 THREDEHYD-RXN THR_c --> 2-OXOBUTANOATE_c + AMMONIUM_c\n",
      "331 THREDEHYD-RXN THR_c --> 2-OXOBUTANOATE_c + AMMONIUM_c\n",
      "\n",
      "332 2.7.9.3-RXN ATP_c + SE-2_c + WATER_c --> AMP_c + Pi_c + SEPO3_c\n",
      "332 2.7.9.3-RXN ATP_c + SE-2_c + WATER_c --> AMP_c + Pi_c + SEPO3_c\n",
      "\n",
      "333 SHIKIMATE-KINASE-RXN ATP_c + SHIKIMATE_c --> ADP_c + PROTON_c + SHIKIMATE-5P_c\n",
      "333 SHIKIMATE-KINASE-RXN ATP_c + SHIKIMATE_c --> ADP_c + PROTON_c + SHIKIMATE-5P_c\n",
      "\n",
      "334 SUCCARGDIHYDRO-RXN CPD-421_c + 2.0 PROTON_c + 2.0 WATER_c --> 2.0 AMMONIUM_c + CARBON-DIOXIDE_c + N2-SUCCINYLORNITHINE_c\n",
      "334 SUCCARGDIHYDRO-RXN CPD-421_c + 2.0 PROTON_c + 2.0 WATER_c --> 2.0 AMMONIUM_c + CARBON-DIOXIDE_c + N2-SUCCINYLORNITHINE_c\n",
      "\n",
      "335 DAPASYN-RXN 8-AMINO-7-OXONONANOATE_c + S-ADENOSYLMETHIONINE_c <=> DIAMINONONANOATE_c + S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c\n",
      "335 DAPASYN-RXN 8-AMINO-7-OXONONANOATE_c + S-ADENOSYLMETHIONINE_c <=> DIAMINONONANOATE_c + S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c\n",
      "\n",
      "336 N-ACETYLTRANSFER-RXN ACETYL-COA_c + GLT_c --> ACETYL-GLU_c + CO-A_c + PROTON_c\n",
      "336 N-ACETYLTRANSFER-RXN ACETYL-COA_c + GLT_c <=> ACETYL-GLU_c + CO-A_c + PROTON_c\n",
      "\n",
      "337 RXN0-5038 CAMP_c + WATER_c --> AMP_c + PROTON_c\n",
      "337 RXN0-5038 CAMP_c + WATER_c --> AMP_c + PROTON_c\n",
      "\n",
      "338 DALADALALIG-RXN ATP_c + 2.0 D-ALANINE_c <=> ADP_c + D-ALA-D-ALA_c + PROTON_c + Pi_c\n",
      "338 DALADALALIG-RXN ATP_c + 2.0 D-ALANINE_c --> ADP_c + D-ALA-D-ALA_c + PROTON_c + Pi_c\n",
      "\n",
      "339 SUCCCOASYN-RXN ATP_c + CO-A_c + SUC_c <=> ADP_c + Pi_c + SUC-COA_c\n",
      "339 SUCCCOASYN-RXN ATP_c + CO-A_c + SUC_c <=> ADP_c + Pi_c + SUC-COA_c\n",
      "\n",
      "340 PANTOATE-BETA-ALANINE-LIG-RXN ATP_c + B-ALANINE_c + L-PANTOATE_c --> AMP_c + PANTOTHENATE_c + PPI_c + PROTON_c\n",
      "340 PANTOATE-BETA-ALANINE-LIG-RXN ATP_c + B-ALANINE_c + L-PANTOATE_c --> AMP_c + PANTOTHENATE_c + PPI_c + PROTON_c\n",
      "\n",
      "341 3-DEHYDROQUINATE-SYNTHASE-RXN 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c --> DEHYDROQUINATE_c + Pi_c\n",
      "341 3-DEHYDROQUINATE-SYNTHASE-RXN 3-DEOXY-D-ARABINO-HEPTULOSONATE-7-P_c --> DEHYDROQUINATE_c + Pi_c\n",
      "\n",
      "342 3PGAREARR-RXN 2-PG_c <=> G3P_c\n",
      "342 3PGAREARR-RXN 2-PG_c <=> G3P_c\n",
      "\n",
      "343 TRIOSEPISOMERIZATION-RXN DIHYDROXY-ACETONE-PHOSPHATE_c <=> GAP_c\n",
      "343 TRIOSEPISOMERIZATION-RXN GAP_c <=> DIHYDROXY-ACETONE-PHOSPHATE_c\n",
      "\n",
      "344 ERYTHRON4PDEHYDROG-RXN ERYTHRONATE-4P_c + NAD_c <=> 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + NADH_c + PROTON_c\n",
      "344 ERYTHRON4PDEHYDROG-RXN ERYTHRONATE-4P_c + NAD_c --> 3OH-4P-OH-ALPHA-KETOBUTYRATE_c + NADH_c + PROTON_c\n",
      "\n",
      "345 RXN-8974 MALONATE-S-ALD_c + NADPH_c + PROTON_c --> 3-HYDROXY-PROPIONATE_c + NADP_c\n",
      "345 RXN-8974 MALONATE-S-ALD_c + NADPH_c + PROTON_c --> 3-HYDROXY-PROPIONATE_c + NADP_c\n",
      "\n",
      "346 GLUTATHIONE-SYN-RXN ATP_c + GLY_c + L-GAMMA-GLUTAMYLCYSTEINE_c --> ADP_c + GLUTATHIONE_c + PROTON_c + Pi_c\n",
      "346 GLUTATHIONE-SYN-RXN ATP_c + GLY_c + L-GAMMA-GLUTAMYLCYSTEINE_c --> ADP_c + GLUTATHIONE_c + PROTON_c + Pi_c\n",
      "\n",
      "347 RXN0-1863 5-BETA-L-THREO-PENTAPYRANOSYL-4-ULOSE-_c + GLT_c <=> 2-KETOGLUTARATE_c + UDP-4-AMINO-4-DEOXY-L-ARABINOSE_c\n",
      "347 RXN0-1863 2-KETOGLUTARATE_c + UDP-4-AMINO-4-DEOXY-L-ARABINOSE_c <=> 5-BETA-L-THREO-PENTAPYRANOSYL-4-ULOSE-_c + GLT_c\n",
      "\n",
      "348 RXN0-4281 METHYL-GLYOXAL_c + NADPH_c + PROTON_c --> ACETOL_c + NADP_c\n",
      "348 RXN0-4281 METHYL-GLYOXAL_c + NADPH_c + PROTON_c --> ACETOL_c + NADP_c\n",
      "\n",
      "349 RXN0-5061 ADP-L-GLYCERO-D-MANNO-HEPTOSE_c + CPD0-929_c --> ADP_c + CPD0-930_c + PROTON_c\n",
      "349 RXN0-5061 ADP-L-GLYCERO-D-MANNO-HEPTOSE_c + CPD0-929_c --> ADP_c + CPD0-930_c + PROTON_c\n",
      "\n",
      "350 HEMN-RXN COPROPORPHYRINOGEN_III_c + 2.0 S-ADENOSYLMETHIONINE_c --> 2.0 CARBON-DIOXIDE_c + 2.0 CH33ADO_c + 2.0 MET_c + PROTOPORPHYRINOGEN_c\n",
      "350 HEMN-RXN COPROPORPHYRINOGEN_III_c + 2.0 S-ADENOSYLMETHIONINE_c --> 2.0 CARBON-DIOXIDE_c + 2.0 CH33ADO_c + 2.0 MET_c + PROTOPORPHYRINOGEN_c\n",
      "\n",
      "351 DLACTDEHYDROGNAD-RXN D-LACTATE_c + NAD_c <=> NADH_c + PROTON_c + PYRUVATE_c\n",
      "351 DLACTDEHYDROGNAD-RXN NADH_c + PROTON_c + PYRUVATE_c --> D-LACTATE_c + NAD_c\n",
      "\n",
      "352 RXN-9537 R-3-hydroxymyristoyl-ACPs_c --> Tetradec-2-enoyl-ACPs_c + WATER_c\n",
      "352 RXN-9537 R-3-hydroxymyristoyl-ACPs_c --> Tetradec-2-enoyl-ACPs_c + WATER_c\n",
      "\n",
      "353 4.2.1.59-RXN 3-Hydroxy-octanoyl-ACPs_c --> 2-Octenoyl-ACPs_c + WATER_c\n",
      "353 4.2.1.59-RXN 3-Hydroxy-octanoyl-ACPs_c --> 2-Octenoyl-ACPs_c + WATER_c\n",
      "\n",
      "354 RXN0-305 OH-PYR_c <=> TARTRONATE-S-ALD_c\n",
      "354 RXN0-305 OH-PYR_c <=> TARTRONATE-S-ALD_c\n",
      "\n",
      "355 ADOMET-DMK-METHYLTRANSFER-RXN CPD-12115_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + PROTON_c + REDUCED-MENAQUINONE_c\n",
      "355 ADOMET-DMK-METHYLTRANSFER-RXN CPD-12115_c + S-ADENOSYLMETHIONINE_c --> ADENOSYL-HOMO-CYS_c + PROTON_c + REDUCED-MENAQUINONE_c\n",
      "\n",
      "356 S-FORMYLGLUTATHIONE-HYDROLASE-RXN CPD-548_c + WATER_c --> FORMATE_c + GLUTATHIONE_c + PROTON_c\n",
      "356 S-FORMYLGLUTATHIONE-HYDROLASE-RXN CPD-548_c + WATER_c --> FORMATE_c + GLUTATHIONE_c + PROTON_c\n",
      "\n",
      "357 DMK-RXN DIHYDROXYNAPHTHOATE_c + OCTAPRENYL-DIPHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + CPD-12115_c + PPI_c\n",
      "357 DMK-RXN DIHYDROXYNAPHTHOATE_c + OCTAPRENYL-DIPHOSPHATE_c + PROTON_c --> CARBON-DIOXIDE_c + CPD-12115_c + PPI_c\n",
      "\n",
      "358 RXN-11477 3-Hydroxyglutaryl-ACP-methyl-ester_c --> Enoylglutaryl-ACP-methyl-esters_c + WATER_c\n",
      "358 RXN-11477 3-Hydroxyglutaryl-ACP-methyl-ester_c --> Enoylglutaryl-ACP-methyl-esters_c + WATER_c\n",
      "\n",
      "359 RXN-11481 3-hydroxypimeloyl-ACP-methyl-esters_c --> Enoylpimeloyl-ACP-methyl-esters_c + WATER_c\n",
      "359 RXN-11481 3-hydroxypimeloyl-ACP-methyl-esters_c --> Enoylpimeloyl-ACP-methyl-esters_c + WATER_c\n",
      "\n",
      "360 ACONITATEDEHYDR-RXN CIT_c <=> CIS-ACONITATE_c + WATER_c\n",
      "360 ACONITATEDEHYDR-RXN CIT_c <=> CIS-ACONITATE_c + WATER_c\n",
      "\n",
      "361 4.2.1.99-RXN CPD-1136_c + WATER_c <=> CPD-618_c\n",
      "361 4.2.1.99-RXN CPD-618_c <=> CPD-1136_c + WATER_c\n",
      "\n",
      "362 RXN-12093 AMMONIUM_c + ATP_c + CPD-13043_c --> 7-CYANO-7-DEAZAGUANINE_c + ADP_c + PROTON_c + Pi_c + WATER_c\n",
      "362 RXN-12093 AMMONIUM_c + ATP_c + CPD-13043_c --> 7-CYANO-7-DEAZAGUANINE_c + ADP_c + PROTON_c + Pi_c + WATER_c\n",
      "\n",
      "363 RXN0-4022 7-CYANO-7-DEAZAGUANINE_c + 2.0 NADPH_c + 3.0 PROTON_c --> 7-AMINOMETHYL-7-DEAZAGUANINE_c + 2.0 NADP_c\n",
      "363 RXN0-4022 7-CYANO-7-DEAZAGUANINE_c + 2.0 NADPH_c + 3.0 PROTON_c --> 7-AMINOMETHYL-7-DEAZAGUANINE_c + 2.0 NADP_c\n",
      "\n",
      "364 RXN-3641 3-KETO-ADIPYL-COA_c + CO-A_c --> ACETYL-COA_c + SUC-COA_c\n",
      "364 RXN-3641 3-KETO-ADIPYL-COA_c + CO-A_c --> ACETYL-COA_c + SUC-COA_c\n",
      "\n",
      "365 RXN0-2044 3-HYDROXYADIPYL-COA_c + NAD_c <=> 3-KETO-ADIPYL-COA_c + NADH_c + PROTON_c\n",
      "365 RXN0-2044 3-HYDROXYADIPYL-COA_c + NAD_c <=> 3-KETO-ADIPYL-COA_c + NADH_c + PROTON_c\n",
      "\n",
      "366 RXN-11484 L-ALPHA-ALANINE_c + Pimeloyl-ACPs_c --> 8-AMINO-7-OXONONANOATE_c + ACP_c + CARBON-DIOXIDE_c\n",
      "366 RXN-11484 L-ALPHA-ALANINE_c + PROTON_c + Pimeloyl-ACPs_c --> 8-AMINO-7-OXONONANOATE_c + ACP_c + CARBON-DIOXIDE_c\n",
      "\n",
      "367 RXN-11483 Pimeloyl-ACP-methyl-esters_c + WATER_c --> METOH_c + Pimeloyl-ACPs_c\n",
      "367 RXN-11483 Pimeloyl-ACP-methyl-esters_c + WATER_c --> METOH_c + PROTON_c + Pimeloyl-ACPs_c\n",
      "\n",
      "368 RXN0-5208 2.0 ATP_c + PROTON_c --> ADENOSYL-P4_c + PPI_c\n",
      "368 RXN0-5208 2.0 ATP_c + PROTON_c --> ADENOSYL-P4_c + PPI_c\n",
      "\n",
      "369 RXN0-5364 CPD-381_c + OXYGEN-MOLECULE_c --> 2-KETOGLUTARATE_c + HYDROGEN-PEROXIDE_c\n",
      "369 RXN0-5364 CPD-381_c + OXYGEN-MOLECULE_c --> 2-KETOGLUTARATE_c + HYDROGEN-PEROXIDE_c\n",
      "\n",
      "370 RXN0-7104 CPD-207_c + WATER_c --> CO-A_c + PHENYLACETATE_c + PROTON_c\n",
      "370 RXN0-7104 CPD-207_c + WATER_c --> CO-A_c + PHENYLACETATE_c + PROTON_c\n",
      "\n",
      "371 RXN0-6562 3-P-HYDROXYPYRUVATE_c + WATER_c --> OH-PYR_c + Pi_c\n",
      "371 RXN0-6562 3-P-HYDROXYPYRUVATE_c + WATER_c --> OH-PYR_c + Pi_c\n",
      "\n",
      "372 2-OCTAPRENYLPHENOL-HYDROX-RXN 2-OCTAPRENYLPHENOL_c + NADPH_c + OXYGEN-MOLECULE_c + PROTON_c --> 2-OCTAPRENYL-6-HYDROXYPHENOL_c + NADP_c + WATER_c\n",
      "372 2-OCTAPRENYLPHENOL-HYDROX-RXN 2-OCTAPRENYLPHENOL_c + NADPH_c + OXYGEN-MOLECULE_c + PROTON_c --> 2-OCTAPRENYL-6-HYDROXYPHENOL_c + NADP_c + WATER_c\n",
      "\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "meta_model_ids = [rxn.id for rxn in halo_meta.reactions]\n",
    "for rxn in halo_iJO_meta.reactions:\n",
    "    if rxn.id in meta_model_ids:\n",
    "        rxn1 = halo_meta.reactions.get_by_id(rxn.id)\n",
    "        print(k,rxn.id,rxn.reaction)\n",
    "        print(k,rxn1.id,rxn1.reaction)\n",
    "        print('')\n",
    "        k += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate the reversibility of reactions based on iML1515 model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_iML1515_reactions_with_metacyc_ids():\n",
    "    model = cobra.io.read_sbml_model('../../../ComplementaryData/iML1515.xml')\n",
    "    \n",
    "    met_rxns = dict()\n",
    "    for rxn in model.reactions:\n",
    "        # ['META:3.6.3.10-RXN', 'META:3.6.3.12-RXN', 'META:TRANS-RXN-2']\n",
    "        metcyc = rxn.annotation.get('biocyc') \n",
    "        if metcyc is None: continue\n",
    "        elif type(metcyc) is list:\n",
    "            for item in metcyc: met_rxns[item.replace('META:','')] = rxn\n",
    "        else: met_rxns[metcyc.replace('META:','')] = rxn\n",
    "    del model\n",
    "    return met_rxns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_reversible(rxn):\n",
    "    if rxn.lower_bound <0 and rxn.upper_bound >0: return 1\n",
    "    else: return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_bound(model,rxn_id,lb=None,ub=None):\n",
    "    rxn = model.reactions.get_by_id(rxn_id)\n",
    "    if lb is not None: rxn.lower_bound = lb\n",
    "    if ub is not None: rxn.upper_bound = ub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "def diff_in_reversibility(halo_model,eco_rxns_met):\n",
    "    for rxn in halo_model.reactions:\n",
    "        eco_rxn = eco_rxns_met.get(rxn.id)\n",
    "        if eco_rxn is None: continue\n",
    "\n",
    "        if is_reversible(eco_rxn)*is_reversible(rxn)>0: continue\n",
    "\n",
    "        if is_reversible(eco_rxn)>0 and is_reversible(rxn)<0:\n",
    "            rxn.lower_bound = -1000\n",
    "            rxn.upper_bound = 1000\n",
    "            continue\n",
    "\n",
    "        print(rxn.id)\n",
    "        print(rxn.reaction)\n",
    "        print(eco_rxn.reaction)\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "984\n"
     ]
    }
   ],
   "source": [
    "eco_rxns_met =  load_iML1515_reactions_with_metacyc_ids()\n",
    "print(len(eco_rxns_met))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3-DEHYDROQUINATE-DEHYDRATASE-RXN\n",
      "DEHYDROQUINATE_c <=> 3-DEHYDRO-SHIKIMATE_c + WATER_c\n",
      "3dhq_c --> 3dhsk_c + h2o_c\n",
      "\n",
      "ACETYLGLUTKIN-RXN\n",
      "ACETYL-GLU_c + ATP_c <=> ADP_c + N-ACETYL-GLUTAMYL-P_c\n",
      "acglu_c + atp_c --> acg5p_c + adp_c\n",
      "\n",
      "TRANS-RXN-42\n",
      "K+_p + PROTON_c <=> K+_c + PROTON_p\n",
      "h_p + k_c --> h_c + k_p\n",
      "\n",
      "TRANS-RXN-242\n",
      "HOMO-SER_p + PROTON_c <=> HOMO-SER_c + PROTON_p\n",
      "h_p + hom__L_c --> h_c + hom__L_p\n",
      "\n",
      "TRANS-RXN-122A\n",
      "L-ASPARTATE_p + 2 PROTON_p <=> L-ASPARTATE_c + 2 PROTON_c\n",
      "asp__L_p + 2.0 h_p --> asp__L_c + 2.0 h_c\n",
      "\n",
      "TRANS-RXN0-0244\n",
      "PROTON_c + THR_p <=> PROTON_p + THR_c\n",
      "h_p + thr__L_c --> h_c + thr__L_p\n",
      "\n",
      "RXN-5076\n",
      "PROTON_p + XANTHINE_p <=> PROTON_c + XANTHINE_c\n",
      "h_p + xan_p --> h_c + xan_c\n",
      "\n",
      "RXN0-4083\n",
      "NA+_p + SER_p <=> NA+_c + SER_c\n",
      "na1_p + ser__L_p --> na1_c + ser__L_c\n",
      "\n",
      "TRANS-RXN-132\n",
      "PROTON_p + URACIL_p <=> PROTON_c + URACIL_c\n",
      "h_p + ura_p --> h_c + ura_c\n",
      "\n",
      "RXN0-6\n",
      "FE+2_p + PROTON_c <=> FE+2_c + PROTON_p\n",
      "fe2_c + h_p --> fe2_p + h_c\n",
      "\n",
      "TRANS-RXN-99\n",
      "CHOLINE_p + PROTON_p <=> CHOLINE_c + PROTON_c\n",
      "chol_p + h_p --> chol_c + h_c\n",
      "\n",
      "TRANS-RXN-40\n",
      "L-ARABINOSE_p + PROTON_c <=> L-ARABINOSE_c + PROTON_p\n",
      "arab__L_c + h_p --> arab__L_p + h_c\n",
      "\n",
      "TRANS-RXN-3\n",
      "K+_p + PROTON_p <=> K+_c + PROTON_c\n",
      "h_p + k_p --> h_c + k_c\n",
      "\n",
      "TRANS-RXN-118\n",
      "NA+_p + PRO_p <=> NA+_c + PRO_c\n",
      "na1_p + pro__L_p --> na1_c + pro__L_c\n",
      "\n",
      "TRANS-RXN-121\n",
      "2 PROTON_p + SUC_p <=> 2 PROTON_c + SUC_c\n",
      "2.0 h_p + succ_p --> 2.0 h_c + succ_c\n",
      "\n"
     ]
    }
   ],
   "source": [
    " diff_in_reversibility(halo_meta_c,eco_rxns_met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually modify the reversibility of above reactions\n",
    "lb_to_be_zero = ['3-DEHYDROQUINATE-DEHYDRATASE-RXN',\n",
    "                 'ACETYLGLUTKIN-RXN',\n",
    "                 'TRANS-RXN-122A',\n",
    "                 'RXN-5076',\n",
    "                 'RXN0-4083',\n",
    "                 'TRANS-RXN-132',\n",
    "                 'TRANS-RXN-99',\n",
    "                 'TRANS-RXN-3',\n",
    "                 'TRANS-RXN-118',\n",
    "                 'TRANS-RXN-121'\n",
    "                ]\n",
    "ub_to_be_zero = ['TRANS-RXN-42',\n",
    "                 'TRANS-RXN-242',\n",
    "                 'TRANS-RXN0-0244',\n",
    "                 'RXN0-6',\n",
    "                 'TRANS-RXN-40']\n",
    "for rxn_id in lb_to_be_zero: set_bound(halo_meta_c,rxn_id,lb=0)\n",
    "for rxn_id in ub_to_be_zero: set_bound(halo_meta_c,rxn_id,ub=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    " diff_in_reversibility(halo_meta_c,eco_rxns_met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.295792 ARG_c + 0.241055 ASN_c + 0.09158 CYS_c + 0.26316 GLN_c + 0.26316 GLT_c + 0.612638 GLY_c + 0.094738 HIS_c + 0.290529 ILE_c + 0.513689 L-ALPHA-ALANINE_c + 0.241055 L-ASPARTATE_c + 0.450531 LEU_c + 0.343161 LYS_c + 0.153686 MET_c + 0.185265 PHE_c + 0.221055 PRO_c + 0.215792 SER_c + 0.253687 THR_c + 0.056843 TRP_c + 0.137896 TYR_c + 0.423162 VAL_c --> protein_c\n",
      "maximal flux: protein_c 0.0\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.026166 DATP_c + 0.027017 DCTP_c + 0.027017 DGTP_c + 0.026166 TTP_c --> dna_c\n",
      "maximal flux: dna_c 0.0\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.175 ATP_c + 0.133508 CTP_c + 0.215096 GTP_c + 0.144104 UTP_c --> rna_c\n",
      "maximal flux: rna_c 0.0\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.013013 AMMONIUM_c + 0.005205 CA+2_c + 0.005205 CL-_c + 2.5e-05 CO+2_c + 7e-06 CPD-3_c + 0.000709 CU+2_c + 0.006715 FE+2_c + 0.007808 FE+3_c + 0.195193 K+_c + 0.008675 MG+2_c + 0.000691 MN+2_c + 0.000323 NI+2_c + 0.004338 SULFATE_c + 0.000341 ZN+2_c --> ions_c\n",
      "maximal flux: ions_c 0.0\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.063814 CPD-12819_c + 0.075214 CPD-17086_c + 0.019456 KDO2-LIPID-IVA_c --> lipids_c\n",
      "maximal flux: lipids_c 0.0\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if model can produce the metabolites in biomass one by one\n",
    "Match metabolites based on\n",
    "* metacyc id\n",
    "* bigg id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_yeast_extraction(model,upper_bound):\n",
    "    # given the evidence that TD01 strain can growth wihout supplyment of amino acids, \n",
    "    # the strain should be able to produce 20 amino acids by itself.\n",
    "    # Block the uptake of 20 amino acids\n",
    "    amino_acids = ['CYS_c','L-ASPARTATE_c','GLT_c','PHE_c','GLY_c',\n",
    "                  'HIS_c','ILE_c','LYS_c','LEU_c','MET_c','ASN_c',\n",
    "                  'PRO_c','GLN_c','ARG_c','SER_c','THR_c','VAL_c',\n",
    "                  'TRP_c','L-ALPHA-ALANINE_c','TYR_c','L-LACTATE','CPD-15990', 'ADENINE']\n",
    "    for met_id in amino_acids:\n",
    "        exg_rxn = model.reactions.get_by_id('Exchange_'+met_id.replace('_c',''))\n",
    "        exg_rxn.upper_bound = upper_bound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def check_if_model_can_produce_mets_in_biomass_one_by_one(model):\n",
    "    book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "    sheets = ['protein','dna','rna','ions','lipids','others']\n",
    "\n",
    "    mnx2meta = dict()\n",
    "    for line in open('../../../ComplementaryData/MNX2metacyc.tsv'):\n",
    "        cont = line.split()\n",
    "        mnx2meta[cont[0]] = mnx2meta.get(cont[0],[]) + [cont[1]]\n",
    "\n",
    "    for name in sheets:\n",
    "        sh = book.sheet_by_name(name)\n",
    "        print(name)\n",
    "        for i in range(sh.nrows-1):\n",
    "            i += 1\n",
    "            mnx_id = sh.cell(i,9).value\n",
    "            bigg_id = sh.cell(i,4).value\n",
    "            met = find_halo_met_c(bigg_id,mnx_id,mnx2meta,model)\n",
    "            if met is None: print(bigg_id+'_c',0)\n",
    "            else:\n",
    "                # add sink reaction\n",
    "                rxn_id = 'tmp'\n",
    "                rxn_sink = cobra.Reaction(rxn_id)\n",
    "                rxn_sink.add_metabolites({met:-1})\n",
    "                model.add_reaction(rxn_sink)\n",
    "                model.objective = 'tmp'\n",
    "                f = model.optimize().objective_value\n",
    "                model.remove_reactions([rxn_sink])\n",
    "                print(met.id,met.name,f)\n",
    "        print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein\n",
      "L-ALPHA-ALANINE_c L-alanine 0.0\n",
      "ARG_c L-arginine 0.0\n",
      "ASN_c L-asparagine 0.0\n",
      "L-ASPARTATE_c L-aspartate 0.0\n",
      "CYS_c L-cysteine 0.0\n",
      "GLN_c L-glutamine 0.0\n",
      "GLT_c L-glutamate 0.0\n",
      "GLY_c glycine 0.0\n",
      "HIS_c L-histidine 0.0\n",
      "ILE_c L-isoleucine 0.0\n",
      "LEU_c L-leucine 0.0\n",
      "LYS_c L-lysine 0.0\n",
      "MET_c L-methionine 0.0\n",
      "PHE_c L-phenylalanine 0.0\n",
      "PRO_c L-proline 0.0\n",
      "SER_c L-serine 0.0\n",
      "THR_c L-threonine 0.0\n",
      "TRP_c L-tryptophan 0.0\n",
      "TYR_c L-tyrosine 0.0\n",
      "VAL_c L-valine 0.0\n",
      "\n",
      "dna\n",
      "DATP_c dATP 0.0\n",
      "DCTP_c dCTP 0.0\n",
      "DGTP_c dGTP 0.0\n",
      "TTP_c dTTP 0.0\n",
      "\n",
      "rna\n",
      "ATP_c ATP 0.0\n",
      "CTP_c CTP 0.0\n",
      "GTP_c GTP 0.0\n",
      "UTP_c UTP 0.0\n",
      "\n",
      "ions\n",
      "CO+2_c Co2+ 1000.0\n",
      "CU+2_c Cu2+ 1000.0\n",
      "FE+2_c Fe2+ 0.0\n",
      "FE+3_c Fe3+ 1000.0\n",
      "K+_c  0.0\n",
      "MG+2_c Mg2+ 1000.0\n",
      "MN+2_c Manganese 1000.0\n",
      "CPD-3_c molybdate 0.0\n",
      "NI+2_c  0.0\n",
      "SULFATE_c sulfate 0.0\n",
      "ZN+2_c  1000.0\n",
      "CA+2_c  1000.0\n",
      "CL-_c chloride 1000.0\n",
      "AMMONIUM_c ammonium 0.0\n",
      "\n",
      "lipids\n",
      "KDO2-LIPID-IVA_c alpha-Kdo-(2->4)-alpha-Kdo-(2->6)-lipid IVA (E. coli) 0.0\n",
      "CPD-12819_c Phosphatidylethanolamine (dihexadecanoyl, n-C16:0) 0.0\n",
      "CPD-17086_c Phosphatidylethanolamine (dihexadec-9enoyl, n-C16:1) 0.0\n",
      "\n",
      "others\n",
      "PROTON_c H+ 0.0\n",
      "10-FORMYL-THF_c 10-FORMYL-THF 0.0\n",
      "2fe2s_c [2Fe-2S] iron-sulfur cluster 0.0\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c 3-(all-trans-octaprenyl)benzene-1,2-diol 0.0\n",
      "CPD-7_c [4Fe-4S] iron-sulfur cluster 0.0\n",
      "S-ADENOSYLMETHIONINE_c S-adenosyl-L-methionine 0.0\n",
      "BIOTIN_c biotin 0.0\n",
      "CO-A_c coenzyme A 0.0\n",
      "FAD_c FAD 0.0\n",
      "WATER_c H2O 0.0\n",
      "METHYLENE-THF_c METHYLENE-THF 0.0\n",
      "CPD0-2278_c  0.0\n",
      "NAD_c NAD+ 0.0\n",
      "NADP_c NADP+ 0.0\n",
      "PROTOHEME_c protoheme 0.0\n",
      "PYRIDOXAL_PHOSPHATE_c pyridoxal 5'-phosphate 0.0\n",
      "RIBOFLAVIN_c riboflavin 0.0\n",
      "SIROHEME_c siroheme 0.0\n",
      "SUC-COA_c succinyl-CoA 0.0\n",
      "THF_c THF 0.0\n",
      "THIAMINE-PYROPHOSPHATE_c thiamine diphosphate 0.0\n",
      "CPD-9649_c Undecaprenyl diphosphate 0.0\n",
      "ADP_c ADP 0.0\n",
      "Pi_c phosphate 0.0\n",
      "PPI_c diphosphate 0.0\n",
      "ATP_c ATP 0.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "check_if_model_can_produce_mets_in_biomass_one_by_one(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 9. Gap-filling with iML1515"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_gap_filling(halo_met,halmodel,universal,**kwargs):\n",
    " \n",
    "    rxn_id = 'tmp'\n",
    "    try: \n",
    "        rxn_sink = halmodel.reactions.get_by_id(rxn_id)\n",
    "        halmodel.remove_reactions([rxn_sink])\n",
    "    except: None\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halmodel.add_reaction(rxn_sink)\n",
    "    halmodel.objective = 'tmp'\n",
    "    f1 = halmodel.optimize().objective_value\n",
    "    \n",
    "    added_genes = dict()# {'rxn_id':'genes'}\n",
    "    print(halo_met.id)\n",
    "    print('Before gap-filling:',f1)\n",
    "    if f1 != 0: print('No need for gap-filling')\n",
    "    else:\n",
    "        try: \n",
    "            with universal:\n",
    "                solution = cobra.flux_analysis.gapfill(halmodel, universal,**kwargs)\n",
    "            all_rxns = dict()\n",
    "            rxns_to_add = list()\n",
    "            for rxn in solution[0]:\n",
    "                try: halmodel.reactions.get_by_id(rxn.id)\n",
    "                except: \n",
    "                    rxn_clean = rxn.copy()\n",
    "                    added_genes[rxn_clean.id] = rxn_clean.gene_reaction_rule\n",
    "                    \n",
    "                    # add _missing tag for those genes\n",
    "                    new_gr = rxn_clean.gene_reaction_rule\n",
    "                    for gene in parse_gr(rxn_clean.gene_reaction_rule):\n",
    "                        new_gr = new_gr.replace(gene,gene+'_missing')\n",
    "        \n",
    "                    rxn_clean.gene_reaction_rule = new_gr\n",
    "                    rxns_to_add.append(rxn_clean)\n",
    "            halmodel.add_reactions(rxns_to_add)\n",
    "            f2 = halmodel.optimize().objective_value\n",
    "            print('After gap-filling:',f2)\n",
    "            print('Added reactions:')\n",
    "            for rxn in rxns_to_add:print(rxn.id,rxn.reaction)\n",
    "        except: print('Gap-filling failed for',halo_met.id)\n",
    "\n",
    "    halmodel.remove_reactions([rxn_sink])\n",
    "    halmodel.repair()\n",
    "    print('\\n\\n')\n",
    "    return added_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "def do_gap_filling_combine(halo_met,halmodel,universal,all_rxns,**kwargs):\n",
    "    # add filled reactions in the end, not after each step\n",
    "    rxn_id = 'tmp'\n",
    "    try: \n",
    "        rxn_sink = halmodel.reactions.get_by_id(rxn_id)\n",
    "        halmodel.remove_reactions([rxn_sink])\n",
    "    except: None\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halmodel.add_reaction(rxn_sink)\n",
    "    halmodel.objective = 'tmp'\n",
    "    f1 = halmodel.optimize().objective_value\n",
    "    \n",
    "    added_genes = dict()# {'rxn_id':'genes'}\n",
    "    print(halo_met.id)\n",
    "    print('Before gap-filling:',f1)\n",
    "    if f1 != 0: print('No need for gap-filling')\n",
    "    else:\n",
    "        try: \n",
    "            with universal:\n",
    "                solution = cobra.flux_analysis.gapfill(halmodel, universal,**kwargs)\n",
    "            rxns_to_add = list()\n",
    "            for rxn in solution[0]:\n",
    "                try: halmodel.reactions.get_by_id(rxn.id)\n",
    "                except: \n",
    "                    rxn_clean = rxn.copy()\n",
    "                    added_genes[rxn_clean.id] = rxn_clean.gene_reaction_rule\n",
    "                    \n",
    "                    # add _missing tag for those genes\n",
    "                    new_gr = rxn_clean.gene_reaction_rule\n",
    "                    for gene in parse_gr(rxn_clean.gene_reaction_rule):\n",
    "                        new_gr = new_gr.replace(gene,gene+'_missing')\n",
    "        \n",
    "                    rxn_clean.gene_reaction_rule = new_gr\n",
    "                    rxns_to_add.append(rxn_clean)\n",
    "                    all_rxns[rxn_clean.id] = rxn_clean\n",
    "                    \n",
    "            #halmodel.add_reactions(rxns_to_add)\n",
    "            #f2 = halmodel.optimize().objective_value\n",
    "            #print('After gap-filling:',f2)\n",
    "            print('Added reactions:')\n",
    "            for rxn in rxns_to_add:print(rxn.id,rxn.reaction)\n",
    "        except: print('Gap-filling failed for',halo_met.id)\n",
    "\n",
    "    halmodel.remove_reactions([rxn_sink])\n",
    "    halmodel.repair()\n",
    "    print('\\n\\n')\n",
    "    return added_genes,all_rxns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def try_gap_filling(halo_met,halmodel,universal, **kwargs):\n",
    "    # glycogen_c\n",
    "    \n",
    "    solution = [[]]\n",
    "    rxn_id = 'tmp'\n",
    "    try: \n",
    "        rxn_sink = halmodel.reactions.get_by_id(rxn_id)\n",
    "        halmodel.remove_reactions([rxn_sink])\n",
    "        halomodel.repair()\n",
    "    except: None\n",
    "    rxn_sink = cobra.Reaction(rxn_id)\n",
    "    rxn_sink.add_metabolites({met:-1})\n",
    "    halmodel.add_reaction(rxn_sink)\n",
    "    halmodel.objective = 'tmp'\n",
    "    optres = halmodel.optimize()\n",
    "    f1 = optres.objective_value\n",
    "    print('Before gap-filling:',f1)\n",
    "    if f1 != 0: \n",
    "        print('No need for gap-filling')\n",
    "        solution = [['No need for gap-filling']]\n",
    "        \n",
    "    else:\n",
    "        try: \n",
    "            solution = cobra.flux_analysis.gapfill(halmodel, universal,**kwargs)\n",
    "            halmodel.add_reactions(solution[0])\n",
    "            f2 = halmodel.optimize().objective_value\n",
    "            print('After gap-filling:',f2)\n",
    "            print('Added reactions:')\n",
    "            for rxn in solution[0]:print(rxn.id,rxn.reaction)\n",
    "            halmodel.remove_reactions(solution[0])\n",
    "        except: \n",
    "            print('Gap-filling failed for',halo_met.id)\n",
    "            solution = [['Failed']]\n",
    "    halmodel.remove_reactions([rxn_sink])\n",
    "    print('\\n\\n')\n",
    "    return solution"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# try to perform gap-filling for every component in biomass, get a list of reactions to be added.\n",
    "# output: \n",
    "# >met_id_in_eco, met_name, met_id_in_halo\n",
    "# rxn_id gr rule\n",
    "# \n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "fhand = open('../Results/gap_filled_reactions_try.txt','w')\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('BiomassComparison')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    rxns = try_gap_filling(met,halo_meta_c,universal)[0]\n",
    "    \n",
    "    #if rxns[0] == 'Failed': \n",
    "    #    rxns = try_gap_filling(met,halo_meta_c,universal,demand_reactions=True)[0]\n",
    "    \n",
    "    fhand.write('>{0} {1} {2}\\n'.format(eco_met.id,eco_met.name,met.id))\n",
    "    if rxns[0] not in ['Failed','No need for gap-filling']:\n",
    "        for rxn in rxns:\n",
    "            fhand.write('{0}\\t{1}\\t{2}\\t{3}\\n'.format(rxn.id,rxn.name,rxn.reaction,rxn.gene_reaction_rule))\n",
    "    else: \n",
    "        fhand.write('{0}\\n'.format(rxns[0]))\n",
    "    fhand.write('\\n')\n",
    "    print('')\n",
    "fhand.close()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# block free uptake of lactate and TREHALOSE. Perform gap-filling with glucose as sole carbon source\n",
    "set_yeast_extraction(halo_meta_c,1000)\n",
    "# halo_meta_c.reactions.get_by_id('Exchange_L-LACTATE').upper_bound=0\n",
    "# halo_meta_c.reactions.get_by_id('Exchange_CPD-15990').upper_bound=0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Block uptake of 20 amino acids,  do gap-filling with iJO1366 as universal model. Do gap-filling for metabolites in order of (1) DNA , RNA, Protein, lipids, Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "fhand = open('../Results/gap_filled_reactions_v1_without_missing_genes.txt','w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_gap_fill_res(fhand,added_genes,met):\n",
    "    fhand.write('>{0} {1}\\n'.format(met.id,met.name))\n",
    "    if len(added_genes)==0: fhand.write('None')\n",
    "    else:\n",
    "        for k,v in added_genes.items(): fhand.write('{0},{1}\\n'.format(k,v))\n",
    "    fhand.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dna_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpvregvza4.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2213 rows, 4008 columns, 18884 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp0vprkb4i.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n",
      "After gap-filling: 38.93863862420134\n",
      "Added reactions:\n",
      "EX_h_e PROTON_e <=> \n",
      "FTHFLi ATP_c + FORMATE_c + THF_c --> 10-FORMYL-THF_c + ADP_c + Pi_c\n",
      "Htex PROTON_e <=> PROTON_p\n",
      "GLCtex_copy1 Glucopyranose_e --> Glucopyranose_p\n",
      "PItex CPD-16459_e <=> Pi_p\n",
      "FE2tex FE+2_e <=> FE+2_p\n",
      "UREAtex UREA_e <=> UREA_p\n",
      "CITtex CIT_e <-- CIT_p\n",
      "GLUt2rpp GLT_p + PROTON_p <=> GLT_c + PROTON_c\n",
      "UREAtpp UREA_p <=> UREA_c\n",
      "CITt3pp CIT_c + PROTON_p --> CIT_p + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "rna_c\n",
      "Before gap-filling: 6.70011100029226\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "protein_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmptbz0rt_d.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2216 rows, 4030 columns, 18942 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpgs3arj4f.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 6.792163186719584\n",
      "Added reactions:\n",
      "ASNS2 AMMONIUM_c + ATP_c + L-ASPARTATE_c --> AMP_c + ASN_c + PPI_c + PROTON_c\n",
      "HSK ATP_c + HOMO-SER_c --> ADP_c + O-PHOSPHO-L-HOMOSERINE_c + PROTON_c\n",
      "G5SADs L-GLUTAMATE_GAMMA-SEMIALDEHYDE_c --> L-DELTA1-PYRROLINE_5-CARBOXYLATE_c + PROTON_c + WATER_c\n",
      "PPND NAD_c + PREPHENATE_c --> CARBON-DIOXIDE_c + NADH_c + P-HYDROXY-PHENYLPYRUVATE_c\n",
      "ADSK APS_c + ATP_c --> ADP_c + PAPS_c + PROTON_c\n",
      "HSDy HOMO-SER_c + NADP_c <=> L-ASPARTATE-SEMIALDEHYDE_c + NADPH_c + PROTON_c\n",
      "SO4tex HSO4_e <=> SULFATE_p\n",
      "NO2t2rpp NITRITE_p + PROTON_p <=> NITRITE_c + PROTON_c\n",
      "THDPS DELTA1-PIPERIDEINE-2-6-DICARBOXYLATE_c + SUC-COA_c + WATER_c --> CO-A_c + N-SUCCINYL-2-AMINO-6-KETOPIMELATE_c\n",
      "GLCtex_copy2 Glucopyranose_e <=> Glucopyranose_p\n",
      "\n",
      "\n",
      "\n",
      "lipids_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpabn4zewd.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2216 rows, 4050 columns, 19030 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp4gf8_zsw.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'FTHFLi' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 24.59910591105682\n",
      "Added reactions:\n",
      "GLUt4pp GLT_p + NA+_p --> GLT_c + NA+_c\n",
      "PSSA161 SER_c + cdpdhdec9eg_c --> CMP_c + PROTON_c + ps161_c\n",
      "PSSA160 SER_c + cdpdhdecg_c --> CMP_c + CPD-12817_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "PROTON_c\n",
      "PROTON_c\n",
      "Before gap-filling: 1000.0\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "10-FORMYL-THF_c\n",
      "10-FORMYL-THF_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp1pqasgd3.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2216 rows, 4056 columns, 19058 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpfwu1y7a0.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'FTHFLi' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'GLUt4pp' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 8.54700854700855\n",
      "Added reactions:\n",
      "DNMPPA DIHYDRONEOPTERIN-P_c + WATER_c --> CPD-11770_c + Pi_c\n",
      "DNTPPA DIHYDRONEOPTERIN-P3_c + WATER_c --> DIHYDRONEOPTERIN-P_c + PPI_c + PROTON_c\n",
      "GCALDD GLYCOLALDEHYDE_c + NAD_c + WATER_c --> GLYCOLLATE_c + NADH_c + 2.0 PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2fe2s_c\n",
      "2fe2s_c\n",
      "Before gap-filling: 17.85714285714285\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpk3ipdpk9.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2217 rows, 4062 columns, 19088 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp3szv8u3l.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy1' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 3.030303030303028\n",
      "Added reactions:\n",
      "NADH18pp DEMETHYLMENAQUINONE_c + NADH_c + 4.0 PROTON_c --> CPD-12115_c + NAD_c + 3.0 PROTON_p\n",
      "OPHHX3 2-OCTAPRENYLPHENOL_c + 2.0 ATP_c + NAD_c + 3.0 WATER_c --> 2-OCTAPRENYL-6-HYDROXYPHENOL_c + 2.0 ADP_c + NADH_c + 3.0 PROTON_c + 2.0 Pi_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD-7_c\n",
      "CPD-7_c\n",
      "Before gap-filling: 8.928571428571434\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "S-ADENOSYLMETHIONINE_c\n",
      "S-ADENOSYLMETHIONINE_c\n",
      "Before gap-filling: 9.090909090909095\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "BIOTIN_c\n",
      "BIOTIN_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmphxub9uqi.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2217 rows, 4066 columns, 19118 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpz0g6lrv4.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'HSDy' since it already exists.\n",
      "Ignoring reaction 'SO4tex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 5.1546391752577305\n",
      "Added reactions:\n",
      "DM_amob_c S-ADENOSYL-4-METHYLTHIO-2-OXOBUTANOATE_c --> \n",
      "GART 5-PHOSPHO-RIBOSYL-GLYCINEAMIDE_c + ATP_c + FORMATE_c --> 5-P-RIBOSYL-N-FORMYLGLYCINEAMIDE_c + ADP_c + PROTON_c + Pi_c\n",
      "ACACt2pp 3-KETOBUTYRATE_p + PROTON_p <=> 3-KETOBUTYRATE_c + PROTON_c\n",
      "EX_meoh_e ALCOHOL-GROUP_e --> \n",
      "DM_5drib_c CPD0-2167_c --> \n",
      "MEOHtrpp ALCOHOL-GROUP_p <=> METOH_c\n",
      "MEOHtex ALCOHOL-GROUP_e <=> ALCOHOL-GROUP_p\n",
      "AI2K ATP_c + DIHYDROXYPENTANEDIONE_c --> ADP_c + PROTON_c + dhptdp_c\n",
      "PAI2I dhptdp_c --> dhptdd_c\n",
      "PAI2T CO-A_c + dhptdd_c --> ACETYL-COA_c + DIHYDROXY-ACETONE-PHOSPHATE_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CO-A_c\n",
      "CO-A_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp6grgbmow.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4086 columns, 19176 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpdiga32x5.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'FTHFLi' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'SO4tex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'GLUt4pp' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 5.988023952095702\n",
      "Added reactions:\n",
      "PNTK ATP_c + PANTOTHENATE_c --> 4-P-PANTOTHENATE_c + ADP_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "FAD_c\n",
      "FAD_c\n",
      "Before gap-filling: 5.7471264367816275\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "WATER_c\n",
      "WATER_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpgzyce877.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4088 columns, 19186 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpt2ywvpx9.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 500.0\n",
      "Added reactions:\n",
      "O2tex OXYGEN-MOLECULE_e <=> OXYGEN-MOLECULE_p\n",
      "FE3tex FE+3_e <=> FE+3_p\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "METHYLENE-THF_c\n",
      "METHYLENE-THF_c\n",
      "Before gap-filling: 8.06451612903225\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD0-2278_p\n",
      "CPD0-2278_p\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp3rnopyvs.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4092 columns, 19194 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmplf0r0fyv.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'THDPS' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 3.2154340836012993\n",
      "Added reactions:\n",
      "DALAt2pp D-ALANINE_p + PROTON_p --> D-ALANINE_c + PROTON_c\n",
      "FUMt1pp FUM_p <=> FUM_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NAD_c\n",
      "NAD_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpoc86ed2z.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4096 columns, 19206 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpkng2zrt9.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'FTHFLi' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'GLUt4pp' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 7.692307692307963\n",
      "Added reactions:\n",
      "NADS1 AMMONIUM_c + ATP_c + DEAMIDO-NAD_c --> AMP_c + NAD_c + PPI_c + PROTON_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NADP_c\n",
      "NADP_c\n",
      "Before gap-filling: 7.19424460431647\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PROTOHEME_c\n",
      "PROTOHEME_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpb65p2i9a.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4098 columns, 19220 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpaj5pfj12.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy1' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'GLUt4pp' since it already exists.\n",
      "Ignoring reaction 'DM_5drib_c' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 3.7593984962406037\n",
      "Added reactions:\n",
      "PPPGO3 3.0 FUM_c + PROTOPORPHYRINOGEN_c --> PROTOPORPHYRIN_IX_c + 3.0 SUC_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PYRIDOXAL_PHOSPHATE_c\n",
      "PYRIDOXAL_PHOSPHATE_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpxre1qg9z.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4100 columns, 19228 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpbpwc_vh3.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 19.230769230769198\n",
      "Added reactions:\n",
      "GLYCLTt2rpp GLYCOLLATE_p + PROTON_p <=> GLYCOLLATE_c + PROTON_c\n",
      "PDX5PO2 NAD_c + PYRIDOXINE-5P_c --> NADH_c + PROTON_c + PYRIDOXAL_PHOSPHATE_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "RIBOFLAVIN_c\n",
      "RIBOFLAVIN_c\n",
      "Before gap-filling: 10.526315789473678\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SIROHEME_c\n",
      "SIROHEME_c\n",
      "Before gap-filling: 4.310344827586206\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUC-COA_c\n",
      "SUC-COA_c\n",
      "Before gap-filling: 5.586592178770935\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "THF_c\n",
      "THF_c\n",
      "Before gap-filling: 8.695652173913045\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "THIAMINE-PYROPHOSPHATE_c\n",
      "THIAMINE-PYROPHOSPHATE_c\n",
      "Before gap-filling: 0.0\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmpnp4ohb3m.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2221 rows, 4104 columns, 19246 nonzeros\n",
      "Read LP format model from file /var/folders/dx/ghtq02dx2w307xx_5hncb421qny3_w/T/tmp65a1lioj.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1872 rows, 5380 columns, 20970 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_h_e' since it already exists.\n",
      "Ignoring reaction 'FTHFLi' since it already exists.\n",
      "Ignoring reaction 'Htex' since it already exists.\n",
      "Ignoring reaction 'PItex' since it already exists.\n",
      "Ignoring reaction 'FE2tex' since it already exists.\n",
      "Ignoring reaction 'UREAtex' since it already exists.\n",
      "Ignoring reaction 'CITtex' since it already exists.\n",
      "Ignoring reaction 'UREAtpp' since it already exists.\n",
      "Ignoring reaction 'CITt3pp' since it already exists.\n",
      "Ignoring reaction 'ADSK' since it already exists.\n",
      "Ignoring reaction 'SO4tex' since it already exists.\n",
      "Ignoring reaction 'GLCtex_copy2' since it already exists.\n",
      "Ignoring reaction 'NADH18pp' since it already exists.\n",
      "Ignoring reaction 'O2tex' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After gap-filling: 7.142857142857134\n",
      "Added reactions:\n",
      "O2tpp OXYGEN-MOLECULE_p <=> OXYGEN-MOLECULE_c\n",
      "THZPSN3 ATP_c + CPD-12279_c + DEOXYXYLULOSE-5P_c + NADPH_c + PROTON_c + iscssh_c --> AMP_c + CARBON-DIOXIDE_c + NADP_c + PPI_c + THZ-P_c + 2.0 WATER_c + iscs_c\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CPD-9649_c\n",
      "CPD-9649_c\n",
      "Before gap-filling: 2.409638554216865\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ADP_c\n",
      "ADP_c\n",
      "Before gap-filling: 14.084507042253584\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Pi_c\n",
      "Pi_c\n",
      "Before gap-filling: 1000.0\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PPI_c\n",
      "PPI_c\n",
      "Before gap-filling: 124.99999999999997\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ATP_c\n",
      "ATP_c\n",
      "Before gap-filling: 12.65822784810122\n",
      "No need for gap-filling\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "# ATP\n",
    "#met = halo_meta_c.metabolites.ATP_c\n",
    "#added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "#write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# DNA\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.dna_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# RNA \n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.rna_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# Protein\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.protein_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# lipids\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.lipids_c\n",
    "added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "\n",
    "# Others\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('others')\n",
    "for i in range(sh.nrows-1):\n",
    "    set_yeast_extraction(halo_meta_c,0)\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False,iterations=1)\n",
    "    write_gap_fill_res(fhand,added_genes,met)\n",
    "    print('')\n",
    "fhand.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "all_filled_rxns = {}\n",
    "# ATP\n",
    "#met = halo_meta_c.metabolites.ATP_c\n",
    "#added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "#write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# DNA\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.dna_c\n",
    "added_genes,all_filled_rxns = do_gap_filling_combine(met,halo_meta_c,universal,all_filled_rxns,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# RNA \n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.rna_c\n",
    "added_genes,all_filled_rxns = do_gap_filling_combine(met,halo_meta_c,universal,all_filled_rxns,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# Protein\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.protein_c\n",
    "added_genes,all_filled_rxns = do_gap_filling_combine(met,halo_meta_c,universal,all_filled_rxns,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "# lipids\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "met = halo_meta_c.metabolites.lipids_c\n",
    "added_genes,all_filled_rxns = do_gap_filling_combine(met,halo_meta_c,universal,all_filled_rxns,demand_reactions=False,iterations=1)\n",
    "write_gap_fill_res(fhand,added_genes,met)\n",
    "\n",
    "\n",
    "# Others\n",
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary_iML1515.xlsx')\n",
    "sh = book.sheet_by_name('others')\n",
    "for i in range(sh.nrows-1):\n",
    "    set_yeast_extraction(halo_meta_c,0)\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    added_genes,all_filled_rxns = do_gap_filling_combine(met,halo_meta_c,universal,all_filled_rxns,demand_reactions=False,iterations=1)\n",
    "    write_gap_fill_res(fhand,added_genes,met)\n",
    "    print('')\n",
    "fhand.close()\n",
    "halo_meta_c.add_reactions(list(all_filled_rxns.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 2053\n",
      "Number of metabolits: 2221\n",
      "Number of compartments: 3\n",
      "Number of genes: 1377\n",
      "Number of missing genes: 53\n",
      "Number of reactions with missing genes: 41\n"
     ]
    }
   ],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update missing genes added by gap-filling. By find homologues in H. TD01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions: 2053\n",
      "Number of metabolits: 2220\n",
      "Number of compartments: 3\n",
      "Number of genes: 1342\n",
      "Number of missing genes: 10\n",
      "Number of reactions with missing genes: 2\n"
     ]
    }
   ],
   "source": [
    "halo_meta_c = update_missing_genes(halo_meta_c,1e-3,report=True,coverage=45)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### For those reactions with missing genes, remove its gr rule and put a note as : \n",
    "{'source':'gapfilled from iJO1366, genes were not found in H.TD01'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NADH18pp TD01GL000419 and TD01GL002642 and TD01GL002645 and TD01GL002810 and b2279_missing and b2280_missing and b2281_missing and b2282_missing and b2283_missing and b2286_missing and b2287_missing and b2288_missing\n",
      "THZPSN3 TD01GL000663 and TD01GL001192 and TD01GL003669 and b0423_missing and b3990_missing\n",
      "\n",
      "Number of reactions: 2053\n",
      "Number of metabolits: 2220\n",
      "Number of compartments: 3\n",
      "Number of genes: 1330\n",
      "Number of missing genes: 0\n",
      "Number of reactions with missing genes: 0\n"
     ]
    }
   ],
   "source": [
    "updated_rxns = list()\n",
    "for rxn in halo_meta_c.reactions:\n",
    "    if '_missing' in rxn.gene_reaction_rule:\n",
    "        print(rxn.id,rxn.gene_reaction_rule)\n",
    "        rxn.gene_reaction_rule = ''\n",
    "        rxn.notes['source']='gapfilled from iJO1366, genes were not found in H.TD01'\n",
    "    updated_rxns.append(rxn)\n",
    "\n",
    "print('')\n",
    "halo_meta_c = cobra.Model('model_v1')\n",
    "halo_meta_c.add_reactions(updated_rxns)\n",
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "book = xlrd.open_workbook('../Results/core_biomass_medium_summary.xlsx')\n",
    "sh = book.sheet_by_name('protein')\n",
    "for i in range(sh.nrows-1):\n",
    "    i += 1\n",
    "    eco_met = eco.metabolites.get_by_id(sh.cell(i,0).value)\n",
    "    met = create_universal_met(eco_met,halo_meta_c,universal,mnx2meta)\n",
    "    print(met)\n",
    "    added_genes = do_gap_filling(met,halo_meta_c,universal,demand_reactions=False)\n",
    "    #write_gap_fill_res(fhand,added_genes,met)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Biomass verions 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.000223 10-FORMYL-THF_c + 0.000223 2-OCTAPRENYL-6-HYDROXYPHENOL_c + 2.6e-05 2fe2s_c + 75.37723 ATP_c + 2e-06 BIOTIN_c + 0.000576 CO-A_c + 0.00026 CPD-7_c + 5.5e-05 CPD-9649_c + 0.013894 CPD0-2278_p + 0.000223 FAD_c + 0.000223 METHYLENE-THF_c + 0.000447 NADP_c + 0.001831 NAD_c + 0.000223 PROTOHEME_c + 0.000223 PYRIDOXAL_PHOSPHATE_c + 0.000223 RIBOFLAVIN_c + 0.000223 S-ADENOSYLMETHIONINE_c + 0.000223 SIROHEME_c + 9.8e-05 SUC-COA_c + 0.000223 THF_c + 0.000223 THIAMINE-PYROPHOSPHATE_c + 70.028756 WATER_c + dna_c + ions_c + lipids_c + protein_c + rna_c --> 75.37723 ADP_c + 0.773903 PPI_c + 75.37723 PROTON_c + 75.37323 Pi_c\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "solution = cobra.flux_analysis.gapfill(halo_meta_c, universal,demand_reactions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "halo_meta_c.reactions.Exchange_Glucopyranose.upper_bound = 5\n",
    "halo_meta_c.reactions.get_by_id('Exchange_L-LACTATE').upper_bound=0\n",
    "halo_meta_c.reactions.get_by_id('Exchange_CPD-15990').upper_bound=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 0.000 at 0x13360997f0>\n"
     ]
    }
   ],
   "source": [
    "#print(halo_meta_c.reactions.Exchange_Glucopyranose.upper_bound)\n",
    "set_yeast_extraction(halo_meta_c,0)\n",
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "print(halo_meta_c.optimize())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 0.000 at 0x1336099a20>\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,1000)\n",
    "halo_meta_c.objective = 'Biomass_v1'\n",
    "print(halo_meta_c.optimize())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save model and filled reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein\n",
      "L-ALPHA-ALANINE_c L-alanine 3.5294117647058822\n",
      "ARG_c L-arginine 1.0714285714285723\n",
      "ASN_c L-asparagine 3.333333333333334\n",
      "L-ASPARTATE_c L-aspartate 8.571428571428573\n",
      "CYS_c L-cysteine 0.923076923076923\n",
      "GLN_c L-glutamine 1.8750000000000013\n",
      "GLT_c L-glutamate 2.857142857142861\n",
      "GLY_c glycine 20.0\n",
      "HIS_c L-histidine 1.0526315789473684\n",
      "ILE_c L-isoleucine 1.0169491525423715\n",
      "LEU_c L-leucine 1.0169491525423728\n",
      "LYS_c L-lysine 1.03448275862069\n",
      "MET_c L-methionine 0.6451612903225806\n",
      "PHE_c L-phenylalanine 0.5940594059405943\n",
      "PRO_c L-proline 1.3333333333333286\n",
      "SER_c L-serine 12.0\n",
      "THR_c L-threonine 3.1578947368421053\n",
      "TRP_c L-tryptophan 0.46874999999999956\n",
      "TYR_c L-tyrosine 0.6741573033707868\n",
      "VAL_c L-valine 1.3333333333333335\n",
      "\n",
      "dna\n",
      "DATP_c dATP 0.4444444444444464\n",
      "DCTP_c dCTP 0.5405405405405405\n",
      "DGTP_c dGTP 0.4878048780487809\n",
      "TTP_c dTTP 0.5263157894736843\n",
      "\n",
      "rna\n",
      "ATP_c ATP 0.48780487804878003\n",
      "CTP_c CTP 0.6060606060606069\n",
      "GTP_c GTP 0.5405405405405403\n",
      "UTP_c UTP 0.6818181818181819\n",
      "\n",
      "ions\n",
      "CO+2_c Co2+ 1000.0\n",
      "CU+2_c Cu2+ 1000.0\n",
      "FE+2_c Fe2+ 30.0\n",
      "FE+3_c Fe3+ 1000.0\n",
      "K+_c  0.0\n",
      "MG+2_c Mg2+ 1000.0\n",
      "MN+2_c Manganese 1000.0\n",
      "CPD-3_c molybdate 0.0\n",
      "NI+2_c  0.0\n",
      "SULFATE_c sulfate 1000.0\n",
      "ZN+2_c  1000.0\n",
      "CA+2_c  1000.0\n",
      "CL-_c chloride 1000.0\n",
      "AMMONIUM_c ammonium 20.0\n",
      "\n",
      "lipids\n",
      "KDO2-LIPID-IVA_c alpha-Kdo-(2->4)-alpha-Kdo-(2->6)-lipid IVA (E. coli) 0.07263922518159807\n",
      "CPD-12819_c Phosphatidylethanolamine (dihexadecanoyl, n-C16:0) 0.12793176972281442\n",
      "CPD-17086_c Phosphatidylethanolamine (dihexadec-9enoyl, n-C16:1) 0.1279317697228145\n",
      "\n",
      "others\n",
      "PROTON_c H+ 1000.0\n",
      "10-FORMYL-THF_c 10-FORMYL-THF 0.31746031746031766\n",
      "2fe2s_c [2Fe-2S] iron-sulfur cluster 0.6250000000000001\n",
      "2-OCTAPRENYL-6-HYDROXYPHENOL_c 3-(all-trans-octaprenyl)benzene-1,2-diol 0.09677419354838714\n",
      "CPD-7_c [4Fe-4S] iron-sulfur cluster 0.31249999999999956\n",
      "S-ADENOSYLMETHIONINE_c S-adenosyl-L-methionine 0.31249999999999983\n",
      "BIOTIN_c biotin 0.16901408450704217\n",
      "CO-A_c coenzyme A 0.2090592334494774\n",
      "FAD_c FAD 0.21052631578947256\n",
      "WATER_c H2O 510.0\n",
      "METHYLENE-THF_c METHYLENE-THF 0.2985074626865671\n",
      "CPD0-2278_c  0.0\n",
      "NAD_c NAD+ 0.27906976744186096\n",
      "NADP_c NADP+ 0.2643171806167377\n",
      "PROTOHEME_c protoheme 0.13392857142857142\n",
      "PYRIDOXAL_PHOSPHATE_c pyridoxal 5'-phosphate 0.6896551724137924\n",
      "RIBOFLAVIN_c riboflavin 0.37037037037037\n",
      "SIROHEME_c siroheme 0.15306122448979576\n",
      "SUC-COA_c succinyl-CoA 0.1954397394136814\n",
      "THF_c THF 0.3208556149732619\n",
      "THIAMINE-PYROPHOSPHATE_c thiamine diphosphate 0.25423728813559326\n",
      "CPD-9649_c Undecaprenyl diphosphate 0.07672634271099744\n",
      "ADP_c ADP 0.5405405405405368\n",
      "Pi_c phosphate 1000.0\n",
      "PPI_c diphosphate 5.0\n",
      "ATP_c ATP 0.48780487804877826\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set_yeast_extraction(halo_meta_c,0)\n",
    "check_if_model_can_produce_mets_in_biomass_one_by_one(halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in halo_meta_c.metabolites: \n",
    "    try: met.charge = int(met.charge)\n",
    "    except: pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "ecpy_path = '../../../ecpy/'\n",
    "sys.path.append(os.path.abspath(ecpy_path))\n",
    "import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid null reference in method 'Species_setCompartment', argument 2 of type 'std::string const &'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-141-792a42bcf90b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave_model_into_formats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhalo_meta_c\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'../../../ModelFiles/'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Halo_GEM_v1'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Documents/GitHub/Halo-GEM/ecpy/utils.py\u001b[0m in \u001b[0;36msave_model_into_formats\u001b[0;34m(model, model_dir, model_name)\u001b[0m\n\u001b[1;32m     60\u001b[0m     '''\n\u001b[1;32m     61\u001b[0m     \u001b[0mxml_model\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel_dir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'xml'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmodel_name\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'.xml'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m     \u001b[0mcobra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_sbml_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mxml_model\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Saved'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mxml_model\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/cobra/io/sbml.py\u001b[0m in \u001b[0;36mwrite_sbml_model\u001b[0;34m(cobra_model, filename, f_replace, **kwargs)\u001b[0m\n\u001b[1;32m    886\u001b[0m     \u001b[0mf_replace\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdict\u001b[0m \u001b[0mof\u001b[0m \u001b[0mreplacement\u001b[0m \u001b[0mfunctions\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mid\u001b[0m \u001b[0mreplacement\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    887\u001b[0m     \"\"\"\n\u001b[0;32m--> 888\u001b[0;31m     \u001b[0mdoc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_model_to_sbml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcobra_model\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf_replace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mf_replace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    889\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    890\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstring_types\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/cobra/io/sbml.py\u001b[0m in \u001b[0;36m_model_to_sbml\u001b[0;34m(cobra_model, f_replace, units)\u001b[0m\n\u001b[1;32m   1029\u001b[0m         \u001b[0mspecie\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetHasOnlySubstanceUnits\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1030\u001b[0m         \u001b[0mspecie\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetName\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1031\u001b[0;31m         \u001b[0mspecie\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetCompartment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompartment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1032\u001b[0m         \u001b[0ms_fbc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspecie\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetPlugin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"fbc\"\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: libsbml.FbcSpeciesPlugin\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1033\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmetabolite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcharge\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/libsbml/__init__.py\u001b[0m in \u001b[0;36msetCompartment\u001b[0;34m(self, sid)\u001b[0m\n\u001b[1;32m  25781\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  25782\u001b[0m         \"\"\"\n\u001b[0;32m> 25783\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_libsbml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSpecies_setCompartment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m  25784\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  25785\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid null reference in method 'Species_setCompartment', argument 2 of type 'std::string const &'"
     ]
    }
   ],
   "source": [
    "utils.save_model_into_formats(halo_meta_c,'../../../ModelFiles/','Halo_GEM_v1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cobra.io.write_sbml_model(halo_meta_c,\n",
    "        '../../../ModelFiles/xml/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.xml')\n",
    "fhand.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cobra.io.save_matlab_model(halo_meta_c,\n",
    "'../../../ModelFiles/mat/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.mat')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cobra.io.save_json_model(halo_meta_c,\n",
    "        '../../../ModelFiles/json/combine_halo_meta100_iML1515_gap_fill_with_core_biomass_without_missing_genes.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cobra.io.save_yaml_model(halo_meta_c,\n",
    "        '../Results/combine_halo_meta100_iJO1366_gap_fill_with_core_biomass_without_missing_genes.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report_model_status(halo_meta_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report_model_status(eco)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare reactions: model_with/without missing genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "model_with_missing_genes = cobra.io.read_sbml_model('../Results/combine_halo_meta100_iJO1366_gap_fill_with_core_biomass_with_missing_genes.xml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "k = 0\n",
    "missing_reactions = 0\n",
    "for rxn in model_with_missing_genes.reactions:\n",
    "    if '_missing' in rxn.gene_reaction_rule:\n",
    "        missing_reactions += 1\n",
    "        try: \n",
    "            halo_meta_c.reactions.get_by_id(rxn.id)\n",
    "            k += 1\n",
    "        except: None\n",
    "print('Number of reactions with missing genes that were found by gap-filling:',k,'/',missing_reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
